// koishi-plugin-awa-mc-ament/lib/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Config: () => Config,
  apply: () => apply,
  inject: () => inject,
  name: () => name,
  usage: () => usage
});
module.exports = __toCommonJS(src_exports);
var import_koishi = require("koishi");
var import_fs2 = require("fs");
var import_node_path2 = __toESM(require("node:path"));

// src/generate_image.ts
var import_node_fs = require("node:fs");
var import_promises = require("node:fs/promises");
var import_node_path = __toESM(require("node:path"));
var defaultTemplate = /* @__PURE__ */ __name(async (options) => `
<!DOCTYPE html>
<html>
<head>
    <style>
        @font-face {
            font-family: 'Minecraft';
            src: url(data:font/truetype;charset=utf-8;base64,${options.fontBase64}) format('truetype');
        }

        body {
            margin: 0;
            width: ${options.width}px;
            height: ${options.height}px;
            background: url(data:image/png;base64,${options.bgBase64}) no-repeat;
            background-size: 100% 100%;  
            background-color: transparent; /* 确保背景色为透明 */
            display: flex;
            flex-direction: row; /* 横向排列 */
            align-items: center; /* 垂直居中 */
            padding: 20px 30px 20px 30px;
            box-sizing: border-box;
        }

        .icon {
            width: 200px;
            height: 200px;
            margin-right: 40px;
            background-image: url(data:image/png;base64,${options.iconBase64}); /* 使用图片 URL */
            
            background-size: cover;

            background-position: center center; /* 居中裁剪 */
            overflow: hidden; /* 隐藏溢出部分 */
            position: relative; /* 为伪元素定位做准备 */
        }

        .text-container {
            flex: 1;
            color: #FFFFFF;
            font-family: 'Minecraft', sans-serif;
            text-shadow: 2px 2px #3F3F3F;
            display: flex;
            flex-direction: column;
            justify-content: space-between; /* 上下排列，确保 title 和 description 分开 */
            height: 81%; /* 确保内容撑满 */
            align-items: flex-end; /* 右对齐 */
        }

        .title {
            font-size: 72px;
            margin-bottom: 0;
            color: #FCFC00;
            align-self: flex-start; /* 右上角 */
        }

        .description {
            font-size: 72px;
            opacity: 0.8;
            align-self: flex-start; /* 右下角 */
        }
    </style>
</head>
<body>
    <div class="icon"></div>
    <div class="text-container">
        <div class="title">${options.title}</div>
        <div class="description">${options.description}</div>
    </div>
</body>
</html>
`, "defaultTemplate");
async function renderAmentImage(ctx, args) {
  const browserPage = await ctx.puppeteer.page();
  try {
    if (args.iconMode === "url") {
      await ctx.http.head(args.icon);
    }
    let arg_icon;
    if (args.iconMode === "path")
      arg_icon = `data:image/png;base64,${await (0, import_promises.readFile)(args.icon, "base64")}`;
    else if (args.iconMode === "base64")
      arg_icon = args.icon;
    else if (args.iconMode === "url")
      arg_icon = `url(${args.icon})`;
    const html = await defaultTemplate({
      title: args.title,
      description: args.description,
      iconBase64: args.icon,
      width: 1280,
      height: 256,
      bgBase64: args.bgBase64,
      fontBase64: args.fontBase64
    });
    (0, import_node_fs.writeFileSync)(import_node_path.default.join(__dirname, "tmp.html"), html);
    browserPage.on("console", (msg) => {
      ctx.logger.debug(`Puppeteer console: ${msg.text()}`);
    });
    browserPage.on("pageerror", (error) => {
      ctx.logger.error(`Puppeteer page error: ${error.message}`);
    });
    await browserPage.setContent(html);
    await browserPage.setViewport({ width: 1280, height: 256 });
    const res = await browserPage.screenshot({
      encoding: "base64",
      type: args.page_screenshotformat,
      omitBackground: true,
      fullPage: true,
      // quality 参数仅对 jpeg/webp 有效，png 不支持
      ...args.page_screenshotformat !== "png" && { quality: args.page_screenshotquality }
    });
    return res;
  } catch (e) {
    ctx.logger.error(`error: ${e}`);
  } finally {
  }
}
__name(renderAmentImage, "renderAmentImage");

// src/index.ts
var import_promises2 = require("fs/promises");

// src/utils.ts
var import_fs = require("fs");
var import_path = require("path");
async function validateFonts(ctx) {
  const assetsDir = (0, import_path.join)(__dirname, "..", "assets");
  if (!(0, import_fs.existsSync)(assetsDir)) {
    (0, import_fs.mkdirSync)(assetsDir, { recursive: true });
  }
  const fontConfigs = [
    {
      filename: "MinecraftAE.ttf",
      downloadUrl: "https://gitee.com/vincent-zyu/koishi-plugin-awa-mc-ament/releases/download/fonts/Minecraft_AE.ttf"
    }
  ];
  for (const fontConfig of fontConfigs) {
    const fontPath = (0, import_path.join)(assetsDir, fontConfig.filename);
    if (!(0, import_fs.existsSync)(fontPath)) {
      ctx.logger.info(`字体文件 ${fontConfig.filename} 不存在，开始下载...`);
      try {
        const response = await ctx.http.get(fontConfig.downloadUrl, { responseType: "arraybuffer" });
        const fontBuffer = Buffer.from(response);
        (0, import_fs.writeFileSync)(fontPath, fontBuffer);
        ctx.logger.info(`字体文件 ${fontConfig.filename} 下载完成`);
      } catch (error) {
        ctx.logger.error(`下载字体文件 ${fontConfig.filename} 失败: ${error.message}`);
      }
    } else {
      ctx.logger.debug(`字体文件 ${fontConfig.filename} 已存在`);
    }
  }
}
__name(validateFonts, "validateFonts");

// src/index.ts
var name = "koishi=plugin-qwq-mc-ament";
var inject = {
  required: ["puppeteer", "http", "i18n"]
};
var pkg = JSON.parse(
  (0, import_fs2.readFileSync)(import_node_path2.default.resolve(__dirname, "../package.json"), "utf-8")
);
var usage = `
<h1>🎮 Koishi 插件：awa-mc-ament - Minecraft 成就生成器</h1>
<h2>🎯 插件版本：v${pkg.version}</h2>
<p>插件使用问题 / Bug反馈 / 插件开发交流，欢迎加入QQ群：<b>259248174</b></p>
<p>📖 <a href="https://gitee.com/vincent-zyu/koishi-plugin-awa-mc-ament" target="_blank">【点我跳转到 Gitee 查看带图片的完整 README】</a></p>

<hr>

<h2>✨ 功能概述</h2>
<p>生成 Minecraft 风格的成就/进度图片，支持自定义标题、描述和图标。</p>

<hr>

<h2>🔍 图标获取优先级</h2>
<p>本插件会按照以下优先级自动选择图标来源：</p>
<ol>
  <li>🎮 <b>Minecraft 游戏图标</b>（需要启用<a href="https://gitee.com/vincent-zyu/fastapi-awa-fuzzy-search-backend" target="_blank">Pytorch后端服务</a>）</li>
  <li>💬 <b>引用消息的图片</b></li>
  <li>🖼️ <b>参数传入的图片</b></li>
  <li>👤 <b>@用户的头像</b></li>
  <li>🎲 <b>默认幸运方块图标</b>（fallback）</li>
</ol>

<hr>

<h2>📝 使用示例</h2>
<ul>
  <li>
    <code>ament -t 标题 -d 介绍 --mcicon 钻石</code><br>
    → 借助 PyTorch+FastAPI 后端，进行语义相似度检测，选出 Minecraft 图片文件作为 icon
  </li>
  <br>
  <li>
    <code>【先引用一条消息】 ament -t 标题 -d 介绍</code><br>
    → 使用引用消息的第一张图片作为 icon
  </li>
  <br>
  <li>
    <code>ament -t 标题 -d 介绍 --icon [图片]</code><br>
    → 使用传入的 icon 图片参数作为 icon
  </li>
  <br>
  <li>
    <code>ament -t 标题 -d 介绍 @某人</code><br>
    → 使用 session 消息中第一个艾特元素的用户头像作为 icon
  </li>
  <br>
  <li>
    <code>ament -t 标题 -d 介绍</code><br>
    → fallback 到默认准备好的幸运方块问号 icon
  </li>
</ul>

<hr>

<h2>⚙️ MC 图标后端（可选）</h2>
<p>如需使用 <code>--mcicon</code> 参数进行 Minecraft 游戏图标搜索，请：</p>
<ol>
  <li>启用配置项中的 "启用MC图标后端服务"</li>
  <li>自行部署 PyTorch+FastAPI 后端服务</li>
  <li>
    后端项目地址：<br>
    <a href="https://gitee.com/vincent-zyu/fastapi-awa-fuzzy-search-backend" target="_blank">
      【点我跳转】https://gitee.com/vincent-zyu/fastapi-awa-fuzzy-search-backend
    </a>
  </li>
</ol>

<hr>

<h3>📜 许可声明</h3>
<p>本插件为开源免费项目，基于 MIT 协议开放。欢迎修改、分发与二次开发。</p>
`;
var Config = import_koishi.Schema.intersect(
  [
    import_koishi.Schema.object(
      {
        banAtUserArg: import_koishi.Schema.boolean().default(false).experimental().description("🚫 是否禁止使用at用户作为成就图标来源 </br> <i> (qq官机得打开这个，因为官机必须艾特bot才能用指令...) </i> ")
      }
    ).description("⚙️ Args-参数相关"),
    import_koishi.Schema.object(
      {
        fontPath: import_koishi.Schema.string().default(import_node_path2.default.join(__dirname, "./../assets/MinecraftAE.ttf")).role("textarea", { rows: [2, 5] }).description("🔤 字体文件绝对路径"),
        bgPath: import_koishi.Schema.string().default(import_node_path2.default.join(__dirname, "./../assets/AdvancementMade_BG.png")).role("textarea", { rows: [2, 5] }).description("🖼️ 背景图绝对路径")
      }
    ).description("📦 Assets-静态资源资产相关"),
    import_koishi.Schema.object(
      {
        browserScreenshotquality: import_koishi.Schema.number().role("slider").min(0).max(100).step(1).default(60).description("📸 Puppeteer截图质量参数，图片压缩质量, 范围0-100"),
        browserScreenshotFormat: import_koishi.Schema.union([
          import_koishi.Schema.const("jpeg").description("JPEG - 有损压缩，文件小"),
          import_koishi.Schema.const("png").description("PNG - 无损压缩，支持透明"),
          import_koishi.Schema.const("webp").description("WebP - 现代格式，兼顾质量与体积")
        ]).default("jpeg").role("radio").description("🖼️ 截图输出格式")
      }
    ).description("🌐 PuppeteerConfig-浏览器配置相关"),
    import_koishi.Schema.object(
      {
        enableMciconBackend: import_koishi.Schema.boolean().default(false).experimental().description("🎮 <b>(可选)</b>是否启用MC图标后端服务。启用本选项 会给ament指令增加一个--icon参数，例如：--icon diamond <br/> <i> 需要自行部署一个PyTorch+FastAPI后端: https://gitee.com/vincent-zyu/fastapi-awa-fuzzy-search-backend </i>  "),
        mciconBackendAddres: import_koishi.Schema.string().default("http://192.168.31.233:8989").description("🔗 mc图标后端地址，完整URL（包含 http:// 或 https://）")
      }
    ).description("🎯 MCICON-后端服务相关"),
    import_koishi.Schema.object(
      {
        VerboseLoggerMode: import_koishi.Schema.boolean().default(false).description("🔍 是否开启详细输出")
      }
    ).description("🐛 DebugConfig-调试内容相关")
  ]
);
function apply(ctx, config) {
  validateFonts(ctx);
  const amentCommand = ctx.command(
    "ament",
    "生成MC风格的成就/进度图片\n	【注意图标获取的优先级】：Minecaft游戏图标 > 引用消息的图片 > 参数传入的图片 > at用户的头像 > 默认fallback幸运方块图标。【没说明白就去看source code】\n"
  ).option("arg0_title", "-t, --title <arg0_title:string> 成就标题", { fallback: "请输入标题" }).option("arg1_description", "-d, --description <arg1_description:string> 成就描述", { fallback: "请输入描述" }).option("arg2_icon", "-i, --icon <arg2_icon:image> 成就图标");
  if (config.enableMciconBackend) {
    amentCommand.option("arg3_mcicon", "-m, --mcicon <arg3_mcicon:string> Minecraft游戏图标搜索关键词");
  }
  amentCommand.action(
    async ({ session, options }) => {
      const fallback_img_path = import_node_path2.default.join(__dirname, "./../assets/fallback_icon.jpg");
      const fallback_base64_str = (0, import_fs2.readFileSync)(fallback_img_path).toString("base64");
      const fallback_base64_str_with_head = `data:image/jpeg;base64,${fallback_base64_str}`;
      let iconSource = "LUCKYBLOCK";
      const firstAtUser = extractAtUser(session.content);
      if (config.VerboseLoggerMode)
        ctx.logger.info("fitstAtUser = " + firstAtUser);
      if ("id" in firstAtUser && !config.banAtUserArg)
        iconSource = "ATUSER";
      if (options.arg2_icon)
        iconSource = "CMDARG";
      if (session.quote) {
        const firstImgUrl = await extractFirstImageUrl(session.quote.content);
        if (firstImgUrl !== "")
          iconSource = "QUOTEMSG";
      }
      if (options.arg3_mcicon)
        iconSource = "MCICON";
      let icon_format;
      let ament_icon_image_element;
      if (iconSource === "MCICON") {
        icon_format = "url";
        const bestItem = await getBestFuzzySearchRes(ctx, options.arg3_mcicon);
        ctx.logger.info(`options.arg3_mcicon = ${options.arg3_mcicon}`);
        if (bestItem.isSucceed === false) {
          await session.send(`获取Minecraft图标有问题哦, msg=${bestItem.res}, res=${bestItem.res}`);
          return;
        }
        ctx.logger.info(JSON.stringify(bestItem.res));
        ament_icon_image_element = `${config.mciconBackendAddres}/mcimg/${bestItem.res.name.toString().replace(/\\/g, "/")}`;
      } else if (iconSource === "QUOTEMSG") {
        icon_format = "url";
        ament_icon_image_element = await extractFirstImageUrl(session.quote.content);
      } else if (iconSource === "CMDARG") {
        icon_format = "url";
        ament_icon_image_element = options.arg2_icon.src;
      } else if (iconSource === "ATUSER") {
        icon_format = "url";
        const firstUserDict = extractAtUser(session.content);
        ament_icon_image_element = (await session.bot.getUser(firstUserDict["id"], session.event.guild.id)).avatar;
      } else if (iconSource === "LUCKYBLOCK") {
        icon_format = "base64";
        ament_icon_image_element = fallback_base64_str_with_head;
      }
      let args_msg = "🛠️[debug]\n";
      args_msg += `📝[options.arg0_title] = ${options.arg0_title}
`;
      args_msg += `📖[options.arg1_description] = ${options.arg1_description}
`;
      args_msg += `🎨[options.arg2_icon] = ${String(options.arg2_icon).slice(0.1)}
`;
      args_msg += `🖼️[options.arg3_mcicon] = ${options.arg3_mcicon}
`;
      args_msg += "\n";
      args_msg += `[iconSource] = ${iconSource}`;
      args_msg += `[icon_format] = ${icon_format}
`;
      args_msg += `[ament_icon_image_element](raw) = ${ament_icon_image_element.slice(0, 100)}
`;
      args_msg += "\n---------\n";
      args_msg += `🅰️[config.fontpath] = ${config.fontPath}
`;
      args_msg += `🌌[config.bgpath] = ${config.bgPath}
`;
      let ament_icon_base64;
      if (icon_format === "base64") {
        ament_icon_base64 = fallback_base64_str;
      } else if (icon_format === "url") {
        const ament_icon_buffer = await ctx.http.file(ament_icon_image_element);
        ament_icon_base64 = Buffer.from(ament_icon_buffer.data).toString("base64");
      }
      args_msg += `ament_icon_base64 = ${String(ament_icon_base64).slice(0, 50)}`;
      if (config.VerboseLoggerMode) {
        await session.send(
          args_msg + `
ament_icon_image_element(image) = ` + import_koishi.h.image(ament_icon_image_element)
        );
      }
      logInfo(args_msg);
      const font_base64 = await fileToBase64(config.fontPath);
      const bg_base64 = await fileToBase64(config.bgPath);
      const res = await renderAmentImage(
        ctx,
        {
          title: options.arg0_title,
          description: options.arg1_description,
          icon: ament_icon_base64,
          iconMode: "base64",
          width: 320,
          height: 64,
          // fontPath: path.join(ctx.baseDir, 'assets', 'Minecraft_AE.ttf'),
          // bgPath: path.join(ctx.baseDir, 'assets', 'AdvancementMade_BG.png')
          fontBase64: font_base64,
          bgBase64: bg_base64,
          page_screenshotquality: config.browserScreenshotquality,
          page_screenshotformat: config.browserScreenshotFormat
        }
      );
      const mimeType = config.browserScreenshotFormat === "png" ? "image/png" : config.browserScreenshotFormat === "webp" ? "image/webp" : "image/jpeg";
      await session.send(
        (0, import_koishi.h)(
          "image",
          { url: `data:${mimeType};base64,` + res }
        )
      );
    }
  );
  function logInfo(...args) {
    ctx.logger.info(...args);
  }
  __name(logInfo, "logInfo");
  const extractImageUrl = /* @__PURE__ */ __name((content) => {
    let urls = import_koishi.h.select(content, "img").map((item) => item.attrs.src);
    if (urls?.length > 0) {
      return urls;
    }
    urls = import_koishi.h.select(content, "mface").map((item) => item.attrs.url);
    return urls?.length > 0 ? urls : null;
  }, "extractImageUrl");
  async function fileToBase64(filePath) {
    try {
      const absolutePath = import_node_path2.default.isAbsolute(filePath) ? filePath : import_node_path2.default.resolve(__dirname, filePath);
      const buffer = await (0, import_promises2.readFile)(absolutePath);
      return buffer.toString("base64");
    } catch (error) {
      ctx.logger.error(`文件转换成base64失败: ${error.message}`);
      throw error;
    }
  }
  __name(fileToBase64, "fileToBase64");
  const extractFirstImageUrl = /* @__PURE__ */ __name(async (content) => {
    if (!content) {
      return "";
    }
    try {
      logInfo("extractFirstImageUrl content:", content);
      let elementContent = content;
      if (typeof content === "string") {
        elementContent = import_koishi.h.parse(`${content}`);
      }
      let url = "";
      const imgElements = import_koishi.h.select(elementContent, "img, image, mface");
      if (imgElements.length > 0) {
        const firstElement = imgElements[0];
        url = firstElement.attrs?.src || firstElement.attrs?.url || "";
      }
      logInfo("extractFirstImageUrl解析结果：", url);
      return url;
    } catch (error) {
      ctx.logger.error("extractFirstImageUrl error:", error);
      return "";
    }
  }, "extractFirstImageUrl");
  const extractAtUser = /* @__PURE__ */ __name((content) => {
    if (!content) {
      return "";
    }
    try {
      if (config.VerboseLoggerMode)
        logInfo("extractextractAtUser content:", content);
      let elementContent = content;
      if (typeof content === "string") {
        elementContent = import_koishi.h.parse(`${content}`);
      }
      let user = {};
      const atElements = import_koishi.h.select(elementContent, "at");
      if (atElements.length > 0) {
        const firstElement = atElements[0];
        user = firstElement.attrs;
      }
      if (config.VerboseLoggerMode)
        logInfo("extractAtUser解析结果: ", user);
      return user;
    } catch (error) {
      ctx.logger.error("extractFirstImageUrl error:", error);
      return "";
    }
  }, "extractAtUser");
  async function getBestFuzzySearchRes(session, keyword) {
    try {
      let url = `${config.mciconBackendAddres}/fuzzy_guess`;
      const params = new URLSearchParams();
      params.append("keyword", keyword);
      params.append("limit", "20");
      if (params.toString()) url += `?${params.toString()}`;
      ctx.logger.info(`getBestFuzzySearchRes(): url = ${url}`);
      const result = await ctx.http.get(url);
      ctx.logger.info(JSON.stringify(result).slice(0, 100));
      if (result.code !== 200) {
        await session.send(`错误, code!=200: ${result.message}`);
        return {
          isSucceed: false,
          msg: `error: code!==200`,
          res: result
        };
      }
      const data = result.data;
      const results = data.results;
      if (results.length === 0) {
        return {
          isSucceed: false,
          msg: `results[] is empty`,
          res: null
        };
      } else {
        return {
          isSucceed: true,
          msg: `succeed`,
          res: results[0]
        };
      }
    } catch (e) {
      ctx.logger.error(`error in getBestFuzzySearchRes(): ${e}`);
      return {
        isSucceed: false,
        msg: `error: ${e}`,
        res: null
      };
    }
  }
  __name(getBestFuzzySearchRes, "getBestFuzzySearchRes");
}
__name(apply, "apply");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Config,
  apply,
  inject,
  name,
  usage
});
