// koishi-plugin-mc-tools/lib/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Config: () => Config,
  apply: () => apply,
  inject: () => inject,
  name: () => name,
  usage: () => usage
});
module.exports = __toCommonJS(src_exports);
var import_koishi8 = require("koishi");

// src/tool/player.ts
var import_koishi = require("koishi");
var import_crypto = require("crypto");
function getOfflineUUID(username) {
  const data = `OfflinePlayer:${username}`;
  const hash = (0, import_crypto.createHash)("md5").update(data).digest();
  hash[6] = hash[6] & 15 | 48;
  hash[8] = hash[8] & 63 | 128;
  const hex = hash.toString("hex");
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20)}`;
}
__name(getOfflineUUID, "getOfflineUUID");
async function fetchPlayerProfile(ctx, username) {
  try {
    const playerData = await ctx.http.get(`https://api.mojang.com/users/profiles/minecraft/${username}`);
    if (!playerData) throw new Error(`不存在玩家 ${username}`);
    const profileData = await ctx.http.get(`https://sessionserver.mojang.com/session/minecraft/profile/${playerData.id}`);
    const texturesData = profileData.properties?.[0]?.value ? JSON.parse(Buffer.from(profileData.properties[0].value, "base64").toString()) : null;
    if (!texturesData?.textures?.SKIN) throw new Error(`无法获取玩家皮肤`);
    const profile = {
      name: playerData.name,
      uuid: playerData.id,
      uuidDashed: playerData.id.replace(/(\w{8})(\w{4})(\w{4})(\w{4})(\w{12})/, "$1-$2-$3-$4-$5"),
      skin: {
        url: texturesData.textures.SKIN.url,
        model: texturesData.textures.SKIN.metadata?.model || "classic"
      }
    };
    if (texturesData?.textures?.CAPE) profile.cape = { url: texturesData.textures.CAPE.url };
    return profile;
  } catch (error) {
    ctx.logger.error(`玩家信息获取失败: ${error.message}`, error);
    throw error;
  }
}
__name(fetchPlayerProfile, "fetchPlayerProfile");
async function renderWithPuppeteer(ctx, html, selector) {
  const page = await ctx.puppeteer.page();
  await page.setContent(html);
  await page.waitForFunction(
    (s) => document.querySelector(s) && Array.from(document.querySelectorAll("canvas")).every((c) => c.toDataURL() !== "data:,"),
    { timeout: 5e3 },
    selector
  );
  await new Promise((resolve) => setTimeout(resolve, 100));
  const screenshot = await (await page.$(selector)).screenshot({ encoding: "base64", omitBackground: true });
  await page.close();
  return screenshot;
}
__name(renderWithPuppeteer, "renderWithPuppeteer");
async function renderPlayerSkin(ctx, skinUrl, capeUrl, renderElytra = false, backgroundColor) {
  const viewportWidth = renderElytra ? 600 : capeUrl ? 400 : 360;
  const skinViewWidth = renderElytra ? 300 : capeUrl ? 200 : 180;
  const capeCode = capeUrl ? `
    await view.loadCape("${capeUrl}");
    ${renderElytra ? "view.playerObject.cape.visible = false; view.playerObject.elytra.visible = true;" : "view.playerObject.cape.visible = true; view.playerObject.elytra.visible = false;"}
  ` : "";
  const backgroundStyle = backgroundColor ? `background:${backgroundColor};` : `background:transparent;`;
  const html = `<html><head>
    <script src="https://unpkg.com/skinview3d@3.1.0/bundles/skinview3d.bundle.js"></script>
    <style>body{margin:0;${backgroundStyle}display:flex;justify-content:center;align-items:center}.container{display:flex;width:${viewportWidth}px;height:400px}.view{width:${skinViewWidth}px;height:400px}</style>
    </head><body><div class="container">
    <canvas id="view1" class="view"></canvas><canvas id="view2" class="view"></canvas></div>
    <script>(async()=>{
      const createViewer=(id,angle)=>{
        const v=new skinview3d.SkinViewer({canvas:document.getElementById(id),width:${skinViewWidth},height:400,preserveDrawingBuffer:true,fov:30,zoom:0.95});
        v.renderer.setClearColor(0x000000,0);v.playerObject.rotation.y=angle;v.animation=null;return v;
      };
      const views=[createViewer('view1',-Math.PI/5),createViewer('view2',Math.PI*4/5)];
      for(const view of views){
        await view.loadSkin("${skinUrl}");${capeCode}
        view.render();
      }
    })()</script></body></html>`;
  return renderWithPuppeteer(ctx, html, ".container");
}
__name(renderPlayerSkin, "renderPlayerSkin");
async function renderPlayerHead(ctx, skinUrl, backgroundColor) {
  const backgroundStyle = backgroundColor ? `background:${backgroundColor};` : `background:transparent;`;
  const html = `<html><head>
    <script src="https://unpkg.com/skinview3d@3.1.0/bundles/skinview3d.bundle.js"></script>
    <style>body{margin:0;${backgroundStyle}display:flex;justify-content:center;align-items:center}.container{width:400px;height:400px}</style>
    </head><body><div class="container"><canvas id="view" width="400" height="400"></canvas></div>
    <script>(async()=>{
      const viewer=new skinview3d.SkinViewer({canvas:document.getElementById('view'),width:400,height:400,preserveDrawingBuffer:true,fov:10,zoom:1.0});
      viewer.renderer.setClearColor(0x000000,0);
      await viewer.loadSkin("${skinUrl}");
      viewer.playerObject.rotation.x=0.05;
      viewer.playerObject.skin.head.scale.set(3.0,3.0,3.0);
      viewer.playerObject.skin.head.position.y=1.0;
      viewer.playerObject.scale.set(0.6,0.6,0.6);
      viewer.playerObject.position.y=-5;
      viewer.animation=null;
      viewer.render();
    })()</script></body></html>`;
  return renderWithPuppeteer(ctx, html, ".container");
}
__name(renderPlayerHead, "renderPlayerHead");
function registerPlayer(ctx, parent) {
  const player = parent.subcommand(".player <username>", "查询 Minecraft 玩家信息").action(async ({}, username) => {
    if (!username) return "请输入玩家用户名";
    const offlineUUID = getOfflineUUID(username);
    let profile = null;
    try {
      profile = await fetchPlayerProfile(ctx, username);
    } catch (e) {
    }
    const message = [import_koishi.h.text(`玩家: ${profile ? profile.name : username}`)];
    if (profile) {
      const modelType = profile.skin.model === "slim" ? "纤细" : "经典";
      message.push(import_koishi.h.text(` [${modelType}] `));
      if (profile.cape) message.push(import_koishi.h.text("(披风)"));
    }
    message.push(import_koishi.h.text(`
Offine UUID: ${offlineUUID}`));
    if (profile) message.push(import_koishi.h.text(`
Online UUID: ${profile.uuidDashed}`));
    message.push(
      import_koishi.h.text('\n使用 "/give @p minecraft:xxx" 获取玩家头颅'),
      import_koishi.h.text(`
[1.12-]skull 1 3 {SkullOwner:"${profile ? profile.name : username}"}`),
      import_koishi.h.text(`
[1.13+]player_head{SkullOwner:"${profile ? profile.name : username}"}`)
    );
    return (0, import_koishi.h)("message", message);
  });
  player.subcommand(".skin <username>", "获取玩家皮肤预览").option("elytra", "-e 显示鞘翅").option("cape", "-c 不显示披风").option("bg", "-b <color:string> 设置背景色(HEX)").action(async ({ options }, username) => {
    if (!username) return "请输入玩家用户名";
    try {
      const profile = await fetchPlayerProfile(ctx, username);
      const showCape = Boolean(!options.cape && profile.cape?.url);
      const showElytra = Boolean(options.elytra && profile.cape?.url);
      const skinImage = await renderPlayerSkin(ctx, profile.skin.url, showCape || showElytra ? profile.cape?.url : void 0, showElytra, options.bg);
      return import_koishi.h.image(`data:image/png;base64,${skinImage}`);
    } catch (error) {
      ctx.logger.error(`获取玩家皮肤预览失败: ${error.message}`, error);
      return `获取玩家皮肤失败: ${error.message}`;
    }
  });
  player.subcommand(".head <username>", "获取玩家大头娃娃").option("bg", "-b <color:string> 设置背景色(HEX)").action(async ({ options }, username) => {
    if (!username) return "请输入玩家用户名";
    try {
      const profile = await fetchPlayerProfile(ctx, username);
      return import_koishi.h.image(`data:image/png;base64,${await renderPlayerHead(ctx, profile.skin.url, options.bg)}`);
    } catch (error) {
      ctx.logger.error(`获取玩家大头娃娃失败: ${error.message}`, error);
      return `获取玩家皮肤失败: ${error.message}`;
    }
  });
  player.subcommand(".raw <username>", "获取玩家原始皮肤").action(async ({}, username) => {
    if (!username) return "请输入玩家用户名";
    try {
      const profile = await fetchPlayerProfile(ctx, username);
      return import_koishi.h.image(profile.skin.url);
    } catch (error) {
      ctx.logger.error(`获取玩家原始皮肤失败: ${error.message}`, error);
      return `获取玩家皮肤失败: ${error.message}`;
    }
  });
}
__name(registerPlayer, "registerPlayer");

// src/server/info.ts
var import_koishi2 = require("koishi");
var net = __toESM(require("net"));
var dgram = __toESM(require("dgram"));
var FORBIDDEN_PATTERNS = [
  /^localhost$/,
  /^127\./,
  /^0\.0\.0\.0$/,
  /^\[::\]/,
  /^::$/,
  /^\[::1\]/,
  /^10\./,
  /^(192\.168)\./,
  /^(172\.(1[6-9]|2[0-9]|3[0-1]))\./,
  /^(169\.254)\./,
  /^fe80:/,
  /^[fd]/,
  /^ff/
];
function validateServerAddress(input) {
  const lowerAddr = input.toLowerCase();
  if (FORBIDDEN_PATTERNS.some((pattern) => pattern.test(lowerAddr))) return null;
  const portPart = lowerAddr.includes(":") ? lowerAddr.substring(lowerAddr.lastIndexOf(":") + 1) : null;
  if (portPart) {
    const port = parseInt(portPart, 10);
    if (isNaN(port) || port < 1 || port > 65535) return null;
  }
  return input;
}
__name(validateServerAddress, "validateServerAddress");
async function pingServer(host, port, type) {
  const startTime = Date.now();
  return new Promise((resolve) => {
    if (type === "java") {
      const socket = new net.Socket();
      const onError = /* @__PURE__ */ __name(() => {
        socket.destroy();
        resolve(-1);
      }, "onError");
      socket.setTimeout(1e4);
      socket.on("connect", () => {
        socket.destroy();
        resolve(Date.now() - startTime);
      });
      socket.on("error", onError);
      socket.on("timeout", onError);
      socket.connect(port, host);
    } else {
      const client = dgram.createSocket("udp4");
      const timer = setTimeout(() => {
        client.close();
        resolve(-1);
      }, 1e4);
      const cleanup = /* @__PURE__ */ __name(() => {
        clearTimeout(timer);
        client.close();
      }, "cleanup");
      client.on("message", () => {
        cleanup();
        resolve(Date.now() - startTime);
      });
      client.on("error", () => {
        cleanup();
        resolve(-1);
      });
      const pingData = Buffer.from([1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 0, 254, 254, 254, 254, 253, 253, 253, 253, 18, 52, 86, 120]);
      client.send(pingData, port, host, (err) => {
        if (err) {
          cleanup();
          resolve(-1);
        }
      });
    }
  });
}
__name(pingServer, "pingServer");
function parseServerAddress(address, defaultPort) {
  const ipv6WithPortMatch = address.match(/^\[(.+)\]:(\d+)$/);
  if (ipv6WithPortMatch) return { host: ipv6WithPortMatch[1], port: parseInt(ipv6WithPortMatch[2], 10) };
  const ipv6Match = address.match(/^\[(.+)\]$/);
  if (ipv6Match) return { host: ipv6Match[1], port: defaultPort };
  if (address.split(":").length > 2 && !address.endsWith("]")) return { host: address, port: defaultPort };
  const lastColonIndex = address.lastIndexOf(":");
  if (lastColonIndex > -1) {
    const host = address.substring(0, lastColonIndex);
    const port = parseInt(address.substring(lastColonIndex + 1), 10);
    if (!isNaN(port)) return { host, port };
  }
  return { host: address, port: defaultPort };
}
__name(parseServerAddress, "parseServerAddress");
async function fetchServerStatus(server, forceType, config) {
  const serverType = forceType || "java";
  const defaultPort = serverType === "java" ? 25565 : 19132;
  const address = validateServerAddress(server);
  if (!address) {
    const { host: host2, port: port2 } = parseServerAddress(server, defaultPort);
    return { online: false, host: host2, port: port2, players: { online: null, max: null }, error: "无效地址" };
  }
  const { host, port } = parseServerAddress(address, defaultPort);
  const apiEndpoints = config?.serverApis?.filter((api) => api.type === serverType)?.map((api) => api.url) || [];
  const apiResults = await Promise.allSettled(
    apiEndpoints.map((apiUrl) => fetch(apiUrl.replace("${address}", address), { headers: { "User-Agent": "Koishi-MC-Info/1.0" } }).then((res) => res.ok ? res.json() : Promise.reject(`API 请求失败: ${res.status}`)))
  );
  const successfulData = apiResults.find((r) => r.status === "fulfilled")?.value;
  if (successfulData) {
    const status = normalizeApiResponse(successfulData, address, serverType);
    if (status.online) {
      status.ping = await pingServer(status.host, status.port, serverType);
      return status;
    }
  }
  return { online: false, host, port, players: { online: null, max: null }, error: "查询失败" };
}
__name(fetchServerStatus, "fetchServerStatus");
function normalizeApiResponse(data, address, serverType) {
  const [hostFromAddr, portStr] = address.split(":");
  const defaultPort = serverType === "java" ? 25565 : 19132;
  const portFromAddr = parseInt(portStr) || defaultPort;
  if (data.online === false || ["error", "offline"].includes(data.status?.toLowerCase())) {
    return { online: false, host: hostFromAddr, port: portFromAddr, players: { online: null, max: null }, error: data.error || data.description };
  }
  let finalHost = data.hostname || data.host || data.server || hostFromAddr;
  let finalPort = data.port ?? data.ipv6Port;
  if (finalPort == null) {
    const ipv6Match = finalHost.match(/^\[(.+)\]:(\d+)$/);
    const hostPortMatch = finalHost.lastIndexOf(":") > finalHost.indexOf(":") ? null : finalHost.match(/^([^:]+):(\d+)$/);
    const match = ipv6Match || hostPortMatch;
    if (match) {
      finalHost = match[1];
      finalPort = parseInt(match[2], 10);
    }
  }
  finalPort = finalPort ?? portFromAddr;
  const processListData = /* @__PURE__ */ __name((items) => Array.isArray(items) ? items.map((item) => typeof item === "string" ? { name: item } : item) : void 0, "processListData");
  const motdText = (() => {
    if (!data.motd) return data.description?.text || data.description;
    if (typeof data.motd === "string") return data.motd;
    if (typeof data.motd === "object") {
      const textArray = data.motd.clean || data.motd.raw;
      return Array.isArray(textArray) ? textArray.join("\n") : textArray;
    }
    return null;
  })();
  const playerList = data.players?.list || data.players?.sample?.map((p) => p.name) || (Array.isArray(data.players) ? data.players : data.player_list);
  return {
    online: true,
    host: finalHost,
    port: finalPort,
    ip_address: data.ip_address || data.ip,
    eula_blocked: data.eula_blocked,
    motd: motdText,
    version: {
      name_clean: data.version?.name_clean ?? data.version,
      name: data.version?.name ?? data.protocol?.name
    },
    players: {
      online: data.players?.online ?? data.players?.now,
      max: data.players?.max,
      list: playerList?.map((p) => (typeof p === "string" ? p : p.name) || "")
    },
    icon: data.icon || data.favicon,
    srv_record: data.srv_record || data.srv,
    mods: processListData(data.mods || data.modinfo?.modList),
    software: data.software,
    plugins: processListData(data.plugins),
    gamemode: data.gamemode,
    server_id: data.server_id,
    edition: data.edition || (serverType === "bedrock" ? "MCPE" : null)
  };
}
__name(normalizeApiResponse, "normalizeApiResponse");
function formatServerStatus(status, config) {
  if (!status.online) return status.error;
  const formatList = /* @__PURE__ */ __name((list, limit) => {
    if (!list?.length) return null;
    const limitedList = list.slice(0, limit || list.length);
    const text = limitedList.map((item) => item.version ? `${item.name}-${item.version}` : item.name).join(", ");
    return limit && limit < list.length ? `${text}...` : text;
  }, "formatList");
  const getValue = /* @__PURE__ */ __name((name2, limit) => {
    switch (name2) {
      case "ip":
        return status.ip_address;
      case "srv":
        return status.srv_record ? `${status.srv_record.host}:${status.srv_record.port}` : null;
      case "icon":
        return status.icon?.startsWith("data:image/png;base64,") ? import_koishi2.h.image(status.icon).toString() : null;
      case "motd":
        return status.motd;
      case "version":
        return status.version?.name_clean;
      case "online":
        return status.players.online?.toString();
      case "max":
        return status.players.max?.toString();
      case "ping":
        return status.ping != null && status.ping !== -1 ? `${status.ping}ms` : null;
      case "software":
        return status.software;
      case "edition":
        return status.edition === "MCPE" ? "基岩版" : status.edition === "MCEE" ? "教育版" : status.edition;
      case "gamemode":
        return status.gamemode;
      case "eulablock":
        return status.eula_blocked ? "是" : null;
      case "serverid":
        return status.server_id;
      case "playercount": {
        const count = status.players.list?.length;
        return count > 0 ? count.toString() : null;
      }
      case "plugincount": {
        const count = status.plugins?.length;
        return count > 0 ? count.toString() : null;
      }
      case "modcount": {
        const count = status.mods?.length;
        return count > 0 ? count.toString() : null;
      }
      case "playerlist":
        return formatList(status.players.list?.map((name3) => ({ name: name3 })), limit);
      case "pluginlist":
        return formatList(status.plugins, limit);
      case "modlist":
        return formatList(status.mods, limit);
      default:
        return null;
    }
  }, "getValue");
  return config.serverTemplate.split("\n").map((line) => {
    const placeholders = [...line.matchAll(/\{([^{}:]+)(?::(\d+))?\}/g)];
    if (placeholders.length > 0 && placeholders.every((p) => !getValue(p[1], p[2] ? parseInt(p[2]) : void 0))) return "";
    return line.replace(/\{([^{}:]+)(?::(\d+))?\}/g, (match, name2, limitStr) => getValue(name2, limitStr ? parseInt(limitStr) : void 0) ?? "");
  }).filter((line) => line.trim().length > 0).join("\n").replace(/\n{3,}/g, "\n\n").trim();
}
__name(formatServerStatus, "formatServerStatus");
function registerInfo(parent, config) {
  const commandAction = /* @__PURE__ */ __name(async (session, server, type) => {
    const targetServer = server || (config.serverMaps.find((m) => m.platform === session.platform && m.channelId === session.guildId)?.serverAddress ?? null);
    if (!targetServer) return "请提供服务器地址";
    const status = await fetchServerStatus(targetServer, type, config);
    return formatServerStatus(status, config);
  }, "commandAction");
  const mcinfo = parent.subcommand(".info [server]", "查询 Java 服务器").usage(`用法: mc.info [地址[:端口]]
查询 Java 版服务器的状态。`).action(async ({ session }, server) => commandAction(session, server, "java"));
  mcinfo.subcommand(".be [server]", "查询 Bedrock 服务器").usage(`用法: mc.info.be [地址[:端口]]
查询基岩版服务器的状态。`).action(async ({ session }, server) => commandAction(session, server, "bedrock"));
}
__name(registerInfo, "registerInfo");

// src/utils/fileManager.ts
var fs = __toESM(require("fs/promises"));
var path = __toESM(require("path"));
var FileManager = class {
  constructor(ctx) {
    this.ctx = ctx;
    this.dataDir = path.join(ctx.baseDir, "data");
  }
  static {
    __name(this, "FileManager");
  }
  dataDir;
  /**
   * 保存数据到JSON文件
   * @template T 数据类型
   * @param filename 文件名
   * @param data 数据对象
   * @returns 保存是否成功
   */
  async saveJson(filename, data) {
    try {
      const filePath = path.join(this.dataDir, filename);
      await fs.writeFile(filePath, JSON.stringify(data, null, 2), "utf8");
      return true;
    } catch (error) {
      this.ctx.logger.error(`保存文件失败 (${filename}): ${error.message}`);
      return false;
    }
  }
  /**
   * 从JSON文件读取数据
   * @template T 数据类型
   * @param filename 文件名
   * @param defaultValue 默认值，当文件不存在或读取失败时返回
   * @returns 读取到的数据或默认值
   */
  async loadJson(filename, defaultValue) {
    try {
      const filePath = path.join(this.dataDir, filename);
      try {
        const data = await fs.readFile(filePath, "utf8");
        return JSON.parse(data);
      } catch (error) {
        if (error.code === "ENOENT") {
          await this.saveJson(filename, defaultValue);
          return defaultValue;
        }
        throw error;
      }
    } catch (error) {
      this.ctx.logger.error(`读取文件失败 (${filename}): ${error.message}`);
      return defaultValue;
    }
  }
  /**
   * 获取白名单绑定数据
   * @returns 白名单绑定数据对象，如果不存在则返回空对象
   */
  async getWhitelistBindings() {
    return await this.loadJson("whitelist.json", {});
  }
  /**
   * 保存白名单绑定数据
   * @param bindings 绑定数据
   * @returns 保存是否成功
   */
  async saveWhitelistBindings(bindings) {
    return await this.saveJson("whitelist.json", bindings);
  }
};

// src/server/server.ts
var import_rcon_client = require("rcon-client");
async function executeRconCommand(command, serverConfig) {
  const [host, portStr] = (serverConfig.rconAddress || "").split(":");
  const port = parseInt(portStr || "");
  if (!serverConfig.rconPassword || !host || !portStr || isNaN(port)) {
    throw new Error(`服务器 #${serverConfig.id} RCON 配置错误`);
  }
  const rcon = await import_rcon_client.Rcon.connect({ host, port, password: serverConfig.rconPassword });
  try {
    const result = await rcon.send(command);
    return result;
  } finally {
    await rcon.end();
  }
}
__name(executeRconCommand, "executeRconCommand");
function findServer(config, serverId) {
  const rconServer = config.rconServers.find((s) => s.id === serverId);
  return {
    found: !!rconServer,
    id: serverId,
    displayName: `服务器 #${serverId}`,
    rconConfig: rconServer || null
  };
}
__name(findServer, "findServer");
function setupServerIdBefore({ session, options }, config) {
  if (!session) return "";
  if (!options.server) {
    const mapping = config.serverMaps.find((m) => m.platform === session.platform && m.channelId === session.guildId);
    if (!mapping) {
      return "该群组未配置对应服务器";
    }
    options.server = mapping.serverId;
  }
}
__name(setupServerIdBefore, "setupServerIdBefore");
function registerServer(ctx, parent, config) {
  const server = parent.subcommand(".server", "管理 Minecraft 服务器").usage("mc.server - 向 Minecraft 服务器内发送消息和执行命令");
  server.subcommand(".say <message:text>", "发送聊天消息").usage("mc.server.say <消息内容> - 发送消息到 Minecraft 服务器").option("server", "-s <serverId:number> 指定服务器 ID").before((argv) => setupServerIdBefore(argv, config)).action(async ({ session, options }, message) => {
    if (!session) return;
    if (!message) return "请输入要发送的消息";
    const serverId = options.server;
    const serverInfo = findServer(config, serverId);
    if (!serverInfo.found) return `未找到服务器 #${serverId}`;
    if (!serverInfo.rconConfig) return `服务器 #${serverId} 未配置 RCON`;
    const sender = session.username || session.userId;
    const command = `say ${sender}: ${message}`;
    try {
      await executeRconCommand(command, serverInfo.rconConfig);
      return `已执行命令 [#${serverId}]`;
    } catch (error) {
      return `命令执行失败 [#${serverId}] - ${error.message}`;
    }
  });
  server.subcommand(".run <command:text>", "执行命令").usage("mc.server.run <命令内容> - 执行指定 Minecraft 命令").option("server", "-s <serverId:number> 指定服务器 ID").before((argv) => setupServerIdBefore(argv, config)).action(async ({ session, options }, command) => {
    if (!session) return;
    if (!command) return "请输入要执行的命令";
    const serverId = options.server;
    const serverInfo = findServer(config, serverId);
    if (!serverInfo.found) return `未找到服务器 #${serverId}`;
    if (!serverInfo.rconConfig) return `服务器 #${serverId} 未配置 RCON`;
    try {
      const result = await executeRconCommand(command, serverInfo.rconConfig);
      return result ? `已执行命令 [#${serverId}]
${result}` : `已执行命令 [#${serverId}]`;
    } catch (error) {
      return `命令执行失败 [#${serverId}] - ${error.message}`;
    }
  });
  if (config.bindEnabled) {
    const fileManager = new FileManager(ctx);
    server.subcommand(".bind [username:string]", "白名单管理").usage("mc.server.bind [用户名] - 绑定或解绑 Minecraft 用户名").option("server", "-s <serverId:number> 指定服务器 ID").option("remove", "-r 解绑指定用户名").action(async ({ session, options }, username) => {
      if (!session || !session.userId) return;
      const bindings = await fileManager.getWhitelistBindings();
      const userId = session.userId;
      if (!username) {
        const userBindings = bindings[userId];
        if (!userBindings || Object.keys(userBindings).length === 0) return "未绑定任何用户名";
        const bindingList = Object.entries(userBindings).map(([name2, serverId2]) => `${name2} → 服务器#${serverId2}`).join("\n");
        return `已绑定的用户名：
${bindingList}`;
      }
      const serverId = options.server || config.serverMaps.find((m) => m.platform === session.platform && m.channelId === session.channelId)?.serverId;
      if (options.remove) {
        if (!bindings[userId]?.[username]) return `未找到绑定用户名 ${username}`;
        const boundServerId = bindings[userId][username];
        const serverInfo2 = findServer(config, boundServerId);
        if (!serverInfo2.found || !serverInfo2.rconConfig) return `服务器 #${boundServerId} 不存在或未配置RCON`;
        try {
          await executeRconCommand(`whitelist remove ${username}`, serverInfo2.rconConfig);
          delete bindings[userId][username];
          if (Object.keys(bindings[userId]).length === 0) delete bindings[userId];
          await fileManager.saveWhitelistBindings(bindings);
          return `已解绑用户名 ${username} [#${boundServerId}]`;
        } catch (error) {
          ctx.logger.warn(`白名单移除失败: ${error.message} [#${boundServerId}]`);
          return `白名单移除失败，未解除绑定: ${error.message} [#${boundServerId}]`;
        }
      }
      if (username.length < 3 || username.length > 16) return "无效的用户名";
      if (!serverId) return "该群组未配置对应服务器";
      const serverInfo = findServer(config, serverId);
      if (!serverInfo.found) return `未找到服务器 #${serverId}`;
      if (!serverInfo.rconConfig) return `服务器 #${serverId} 未配置 RCON`;
      for (const [uid, userBindings] of Object.entries(bindings)) {
        if (uid !== userId && username in userBindings) {
          return `用户名 ${username} 已被其他用户绑定到服务器 #${userBindings[username]}`;
        }
      }
      if (bindings[userId]?.[username] === serverId) return `已绑定用户名 ${username} 到服务器 #${serverId}`;
      try {
        await executeRconCommand(`whitelist add ${username}`, serverInfo.rconConfig);
        if (!bindings[userId]) bindings[userId] = {};
        bindings[userId][username] = serverId;
        await fileManager.saveWhitelistBindings(bindings);
        return `已绑定用户名 ${username} 到服务器 #${serverId}`;
      } catch (error) {
        return `白名单添加失败，未绑定: ${error.message} [#${serverId}]`;
      }
    });
  }
}
__name(registerServer, "registerServer");

// src/tool/ver.ts
async function getLatestVersion() {
  const apiUrl = "https://launchermeta.mojang.com/mc/game/version_manifest.json";
  const response = await fetch(apiUrl);
  if (!response.ok) throw new Error(`API 响应错误: ${response.status}`);
  const { latest, versions } = await response.json();
  const release = versions.find((v) => v.id === latest.release);
  const snapshot = versions.find((v) => v.id === latest.snapshot);
  return {
    release: { id: release.id, releaseTime: release.releaseTime },
    snapshot: { id: snapshot.id, releaseTime: snapshot.releaseTime }
  };
}
__name(getLatestVersion, "getLatestVersion");
async function sendUpdateNotification(ctx, targets, versionType, versionInfo) {
  const filteredTargets = targets.filter((t) => t.type === "both" || t.type === versionType);
  if (!filteredTargets.length) return;
  const typeName = versionType === "release" ? "正式版" : "快照版";
  const updateMsg = `Minecraft ${typeName}更新：${versionInfo.id}
发布时间: ${new Date(versionInfo.releaseTime).toLocaleString("zh-CN")}`;
  const broadcastChannels = filteredTargets.map((t) => `${t.platform}:${t.channelId}`);
  await ctx.broadcast(broadcastChannels, updateMsg);
}
__name(sendUpdateNotification, "sendUpdateNotification");
var sharedVersionCache = { release: { id: "", releaseTime: "" }, snapshot: { id: "", releaseTime: "" } };
function registerVer(mc) {
  mc.subcommand(".ver", "查询 Minecraft 最新版本").action(async () => {
    const formatVersionInfo = /* @__PURE__ */ __name((release, snapshot) => {
      const formatDate = /* @__PURE__ */ __name((date) => new Date(date).toLocaleDateString("zh-CN"), "formatDate");
      return `Minecraft 最新版本：
正式版: ${release.id}(${formatDate(release.releaseTime)})
快照版: ${snapshot.id}(${formatDate(snapshot.releaseTime)})`;
    }, "formatVersionInfo");
    try {
      const versions = await getLatestVersion();
      Object.assign(sharedVersionCache, versions);
      return formatVersionInfo(versions.release, versions.snapshot);
    } catch (error) {
      if (sharedVersionCache.release.id && sharedVersionCache.snapshot.id) return formatVersionInfo(sharedVersionCache.release, sharedVersionCache.snapshot);
      return "获取 Minecraft 版本信息失败";
    }
  });
}
__name(registerVer, "registerVer");
function regVerCheck(ctx, config) {
  const trackedVersion = { release: { id: "", releaseTime: "" }, snapshot: { id: "", releaseTime: "" } };
  const checkVersions = /* @__PURE__ */ __name(async () => {
    try {
      const latest = await getLatestVersion();
      Object.assign(sharedVersionCache, latest);
      const isFirstCheck = !trackedVersion.release.id;
      if (!isFirstCheck) {
        if (latest.release.id !== trackedVersion.release.id) await sendUpdateNotification(ctx, config.noticeTargets, "release", latest.release);
        if (latest.snapshot.id !== trackedVersion.snapshot.id && latest.snapshot.id !== latest.release.id) await sendUpdateNotification(ctx, config.noticeTargets, "snapshot", latest.snapshot);
      }
      Object.assign(trackedVersion, latest);
    } catch (error) {
      return;
    }
  }, "checkVersions");
  checkVersions();
  ctx.setInterval(checkVersions, config.updInterval * 6e4);
}
__name(regVerCheck, "regVerCheck");

// src/tool/status.ts
var servicesToCheck = {
  "Minecraft Net": "https://minecraft.net/",
  "Session": "http://session.minecraft.net/",
  "Textures": "http://textures.minecraft.net/",
  "Mojang API": "https://api.mojang.com/",
  "Account": "http://account.mojang.com/",
  "Session Server": "https://sessionserver.mojang.com/"
};
function formatStatusMessage(status) {
  const statusLines = Object.entries(status).map(([service, isOnline]) => {
    const symbol = isOnline ? "[√]" : "[×]";
    return `${symbol} ${service}`;
  });
  return ["Minecraft 服务状态:", ...statusLines].join("\n");
}
__name(formatStatusMessage, "formatStatusMessage");
async function checkServiceStatus(url) {
  try {
    const response = await fetch(url, { signal: AbortSignal.timeout(1e4), redirect: "follow" });
    return response.status < 500;
  } catch {
    return false;
  }
}
__name(checkServiceStatus, "checkServiceStatus");
async function getMinecraftStatus() {
  const statusEntries = await Promise.all(
    Object.entries(servicesToCheck).map(async ([name2, url]) => {
      const isOnline = await checkServiceStatus(url);
      return [name2, isOnline];
    })
  );
  return Object.fromEntries(statusEntries);
}
__name(getMinecraftStatus, "getMinecraftStatus");
function registerStatus(mc) {
  mc.subcommand(".status", "查询 Minecraft 服务状态").action(async ({}) => {
    try {
      const currentStatus = await getMinecraftStatus();
      return formatStatusMessage(currentStatus);
    } catch (error) {
      return "获取 Minecraft 服务状态失败";
    }
  });
}
__name(registerStatus, "registerStatus");
function regStatusCheck(ctx, config) {
  const targets = config.statusNoticeTargets;
  if (!targets?.length) return;
  let lastConfirmedState = true;
  let pendingState = null;
  let count = 0;
  const check = /* @__PURE__ */ __name(async () => {
    try {
      const current = await getMinecraftStatus();
      const onlineCount = Object.values(current).filter((v) => v).length;
      const totalCount = Object.keys(servicesToCheck).length;
      let currentState = null;
      if (onlineCount === totalCount) currentState = true;
      else if (onlineCount === 0) currentState = false;
      if (currentState === null || currentState === lastConfirmedState) {
        count = 0;
        pendingState = null;
        return;
      }
      if (currentState === pendingState) {
        count++;
      } else {
        pendingState = currentState;
        count = 1;
      }
      if (count >= 3) {
        const msg = currentState ? "Minecraft 服务恢复正常" : "Minecraft 服务全部宕机";
        const channels = targets.map((t) => `${t.platform}:${t.channelId}`);
        await ctx.broadcast(channels, msg);
        lastConfirmedState = currentState;
        count = 0;
        pendingState = null;
      }
    } catch (e) {
      ctx.logger.warn("检查 Minecraft 服务状态失败:", e);
    }
  }, "check");
  check();
  ctx.setInterval(check, config.statusUpdInterval * 6e4);
}
__name(regStatusCheck, "regStatusCheck");

// src/resource/render.ts
var import_koishi3 = require("koishi");
async function takeScreenshot(url, ctx, onImageCallback) {
  try {
    const browser = await ctx.puppeteer.browser;
    const context = await browser.createBrowserContext();
    const page = await context.newPage();
    try {
      await page.setRequestInterception(true);
      page.on("request", (request) => {
        const resourceType = request.resourceType();
        const requestUrl = request.url().toLowerCase();
        if (requestUrl.includes("at.alicdn.com") && (requestUrl.endsWith(".js") || requestUrl.includes("font_")) || requestUrl.includes("iconfont") && requestUrl.includes(".svg")) {
          request.continue();
        } else if (["image", "media", "font", "script"].includes(resourceType) && /\.(gif|analytics|tracking|ad|pixel)|\/ad(s|vert(ising)?)?\/|(pagead2\.googlesyndication|adservice\.google|amazon-adsystem|googletagmanager|scorecardresearch)\.com/.test(requestUrl)) {
          request.abort();
        } else {
          request.continue();
        }
      });
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 1e4 });
      await page.evaluate(() => new Promise((resolve) => {
        document.readyState === "complete" ? resolve(true) : window.addEventListener("load", () => resolve(true), { once: true });
        setTimeout(resolve, 5e3);
      }));
      await optimizePage(page, url);
      const contentBox = await getContentBox(page);
      return await captureScreenshots(page, contentBox, onImageCallback);
    } finally {
      await context.close();
    }
  } catch (e) {
    ctx.logger.error(`渲染截图失败: ${e.message}`, e);
    return null;
  }
}
__name(takeScreenshot, "takeScreenshot");
async function getContentBox(page) {
  const url = page.url();
  return await page.evaluate((currentUrl) => {
    const siteMap = {
      "mcmod.cn": [".item-row", ".post-row", ".class-text"],
      "modrinth.com": [".new-page.sidebar", ".markdown-body", ".project-description"],
      "minecraft.wiki": ["#bodyContent", "#content"]
    };
    const selectors = Object.entries(siteMap).find(([site]) => currentUrl.includes(site))?.[1] || [];
    selectors.push("main", ".content");
    for (const selector of selectors) {
      const element = document.querySelector(selector);
      if (element) {
        const rect = element.getBoundingClientRect();
        return {
          x: Math.max(0, rect.left),
          y: Math.max(0, rect.top),
          width: Math.min(rect.width, window.innerWidth),
          height: rect.height
        };
      }
    }
    return null;
  }, url);
}
__name(getContentBox, "getContentBox");
async function captureScreenshots(page, contentBox, onImageCallback) {
  if (!contentBox) return null;
  const maxHeight = 4096;
  const screenshotOpts = { type: "webp", quality: 80, optimizeForSpeed: true, omitBackground: true };
  if (contentBox.height <= maxHeight) {
    const image = await page.screenshot({ ...screenshotOpts, clip: contentBox });
    if (!image) return null;
    const imageElement = import_koishi3.h.image(image, "image/webp");
    if (onImageCallback) {
      await onImageCallback(imageElement);
      return null;
    }
    return imageElement;
  }
  const pageCount = Math.ceil(contentBox.height / maxHeight);
  const screenshots = await Promise.all(
    Array(pageCount).fill(0).map(async (_, i) => {
      const startY = contentBox.y + i * maxHeight;
      const height = Math.min(maxHeight, contentBox.height - i * maxHeight);
      const image = await page.screenshot({
        ...screenshotOpts,
        clip: { x: contentBox.x, y: startY, width: contentBox.width, height }
      });
      return { image, index: i };
    })
  );
  for (const { image, index } of screenshots.sort((a, b) => a.index - b.index)) {
    if (image) {
      const imageElement = import_koishi3.h.image(image, "image/webp");
      if (onImageCallback) {
        await onImageCallback(imageElement);
      } else if (index === 0) {
        return imageElement;
      }
    }
  }
  return null;
}
__name(captureScreenshots, "captureScreenshots");
async function optimizePage(page, url) {
  const siteType = url.includes("mcmod.cn") ? "mcmod" : url.includes("minecraft.wiki") ? "minecraft-wiki" : url.includes("modrinth.com") ? "modrinth" : "generic";
  await page.evaluate((type) => {
    const selectorsToRemove = [
      "footer",
      "header",
      "nav",
      ".ads-container",
      "ins.adsbygoogle",
      "iframe",
      "script",
      ...type === "mcmod" ? [".comment-ad", ".class-rating-submit"] : [],
      ...type === "minecraft-wiki" ? [
        ".mw-editsection",
        ".noprint",
        ".mw-indicators",
        "#siteNotice",
        "#mw-page-base",
        "#mw-head-base",
        ".wiki-nav",
        ".page-header",
        "#mw-head",
        "#mw-navigation",
        ".mcw-sidebar"
      ] : [],
      ...type === "modrinth" ? [
        ".notification-container",
        ".vue-notification-group",
        ".project-description + div",
        ".joined-buttons",
        ".donate-button",
        ".social-buttons",
        ".btn-group",
        ".sidebar-left",
        ".sidebar-right",
        ".header-wrapper"
      ] : []
    ];
    selectorsToRemove.forEach((selector) => {
      document.querySelectorAll(selector).forEach((el) => el?.remove());
    });
    if (type === "mcmod") {
      document.querySelectorAll(".uknowtoomuch").forEach((el) => {
        if (el.parentNode) {
          const newElement = document.createElement("span");
          newElement.textContent = el.textContent;
          el.parentNode.replaceChild(newElement, el);
        }
      });
    } else if (type === "minecraft-wiki") {
      document.querySelectorAll(".collapsible").forEach((el) => {
        el.classList.remove("collapsed");
        el.classList.add("expanded");
      });
    } else if (type === "modrinth") {
      document.querySelectorAll("details").forEach((detail) => detail.setAttribute("open", "true"));
      document.querySelectorAll("img").forEach((img) => {
        img.loading = "eager";
        if (img.dataset.src) {
          img.src = img.dataset.src;
          delete img.dataset.src;
        }
      });
    }
  }, siteType);
}
__name(optimizePage, "optimizePage");
async function renderOutput(session, content, url = null, ctx, config, screenshot = false) {
  if (config.useScreenshot && screenshot && url && ctx.puppeteer) {
    try {
      const screenshotResult = await takeScreenshot(url, ctx, async (image) => {
        await session.send(image);
      });
      return screenshotResult || "";
    } catch (error) {
      ctx.logger.error("截图失败", error);
    }
  }
  if (config.useForward && session.platform === "onebot") {
    try {
      const messages = content.map((item) => ({
        type: "node",
        data: {
          name: "MC Tools",
          uin: session.selfId,
          content: typeof item === "object" && item?.type === "img" ? `[CQ:image,file=${item.attrs?.src || ""}]` : item
        }
      }));
      const isGroup = session.subtype === "group";
      const target = isGroup ? session.guildId : session.userId;
      const method = isGroup ? "sendGroupForwardMsg" : "sendPrivateForwardMsg";
      await session.bot.internal[method](target, messages);
      return "";
    } catch (error) {
      ctx.logger.error("合并转发失败", error);
      if (!config.useFallback) return "";
    }
  }
  try {
    for (const item of content) await session.send(item);
    return "";
  } catch (error) {
    ctx.logger.error("消息发送失败", error);
  }
}
__name(renderOutput, "renderOutput");

// src/resource/modrinth.ts
var import_koishi7 = require("koishi");

// src/resource/curseforge.ts
var import_koishi4 = require("koishi");

// src/resource/download.ts
var CF_API_BASE = "https://api.curseforge.com/v1";
var MR_API_BASE = "https://api.modrinth.com/v2";
async function fetchAPI(ctx, url, options = {}) {
  try {
    return await ctx.http.get(url, options);
  } catch (error) {
    ctx.logger.error(`API请求失败: ${url}`, error);
    return null;
  }
}
__name(fetchAPI, "fetchAPI");
async function getModrinthVersions(ctx, projectId, options = {}) {
  const result = await fetchAPI(ctx, `${MR_API_BASE}/project/${projectId}/version`);
  if (!result) return [];
  return [...result].sort((a, b) => new Date(b.date_published).getTime() - new Date(a.date_published).getTime()).filter((v) => !options.version || v.game_versions?.includes(options.version)).filter((v) => !options.loader || v.loaders?.includes(options.loader));
}
__name(getModrinthVersions, "getModrinthVersions");
async function getCurseForgeFiles(ctx, modId, apiKey, options = {}, index = 0, pageSize = 50) {
  const params = { index };
  if (options.version) params.gameVersion = options.version;
  if (options.loader && options.loader in { forge: 1, fabric: 1, quilt: 1 }) params.modLoaderType = options.loader === "forge" ? 1 : options.loader === "fabric" ? 4 : 5;
  const response = await fetchAPI(ctx, `${CF_API_BASE}/mods/${modId}/files`, { headers: { "x-api-key": apiKey }, params });
  return { files: response?.data || [], pagination: response?.pagination || { index, pageSize, resultCount: 0, totalCount: 0 } };
}
__name(getCurseForgeFiles, "getCurseForgeFiles");
function formatFileSize(bytes) {
  const units = ["B", "KB", "MB", "GB"];
  let size = bytes, unitIndex = 0;
  while (size >= 1024 && unitIndex < units.length - 1) {
    size /= 1024;
    unitIndex++;
  }
  return `${size.toFixed(unitIndex > 0 ? 2 : 0)} ${units[unitIndex]}`;
}
__name(formatFileSize, "formatFileSize");
async function handleUserInput(session, input, allFiles, isLastPage) {
  if (!input || input.toLowerCase() === "c") return { action: "cancel" };
  if (input.toLowerCase() === "n") return isLastPage ? { action: "cancel" } : { action: "next" };
  const choice = parseInt(input);
  if (isNaN(choice) || choice < 1 || choice > allFiles.length) {
    await session.send(`请回复序号下载文件，输入n查看下页，输入c取消`);
    return handleUserInput(session, await session.prompt(6e4), allFiles, isLastPage);
  }
  return { action: "select", index: choice - 1 };
}
__name(handleUserInput, "handleUserInput");
function formatFileInfo(file, platform, globalIndex) {
  const index = globalIndex + 1;
  if (platform === "modrinth") {
    return `${index}. ${file.name} [${file.game_versions?.join(", ")}] [${file.loaders?.join(", ")}] (${formatFileSize(file.files[0].size)})`;
  } else {
    const loaders = (file.gameVersions || []).filter((v) => !/^\d+\.\d+(\.\d+)?$/.test(v) && v !== "Client").join(", ");
    const versions = (file.gameVersions || []).filter((v) => /^\d+\.\d+(\.\d+)?$/.test(v)).join(", ");
    return `${index}. ${file.displayName || file.fileName} [${versions}] [${loaders}] (${formatFileSize(file.fileLength)})`;
  }
}
__name(formatFileInfo, "formatFileInfo");
async function displayFileList(session, files, pageInfo, platform, ctx, config, startIndex) {
  const messages = [
    "请回复序号下载文件，输入n查看下页，输入c取消",
    ...files.map((file, i) => formatFileInfo(file, platform, startIndex + i)),
    pageInfo
  ];
  await renderOutput(session, messages, null, ctx, config, false);
}
__name(displayFileList, "displayFileList");
async function handleDownload(ctx, session, platform, project, config, options = {}) {
  try {
    let allFiles = [], currentIndex = 0, currentPage = 0, totalItems = 0;
    let hasMoreResults = true;
    let cfPagination = null;
    if (platform === "modrinth") {
      allFiles = await getModrinthVersions(ctx, project.project_id, options);
      totalItems = allFiles.length;
      hasMoreResults = false;
      if (!allFiles?.length) return "该项目未找到任何版本";
    }
    while (true) {
      if (platform === "curseforge" && currentPage * 10 >= allFiles.length && hasMoreResults) {
        const result2 = await getCurseForgeFiles(ctx, project.id, config.curseforgeEnabled, options, currentIndex);
        if (!result2.files?.length) {
          if (allFiles.length === 0) return "该项目未找到任何文件";
          hasMoreResults = false;
        } else {
          allFiles = [...allFiles, ...result2.files];
          currentIndex += result2.pagination.pageSize;
          cfPagination = result2.pagination;
          totalItems = cfPagination.totalCount;
          hasMoreResults = allFiles.length < cfPagination.totalCount;
        }
      }
      const startIndex = currentPage * 10;
      if (startIndex >= allFiles.length) return "已取消下载";
      const endIndex = Math.min(startIndex + 10, allFiles.length);
      const pageFiles = allFiles.slice(startIndex, endIndex);
      const totalPages = Math.ceil(totalItems / 10);
      const isLastPage = !hasMoreResults && endIndex >= allFiles.length;
      const pageInfo = `第 ${currentPage + 1}/${totalPages || "?"} 页${isLastPage ? "（最后一页）" : ""}`;
      await displayFileList(session, pageFiles, pageInfo, platform, ctx, config, startIndex);
      const input = await session.prompt(6e4);
      const result = await handleUserInput(session, input, allFiles, isLastPage);
      if (result.action === "cancel") {
        return "已取消下载";
      } else if (result.action === "next") {
        if (!isLastPage) currentPage++;
        else return "已取消下载";
      } else if (result.action === "select" && result.index !== void 0) {
        const selectedFile = allFiles[result.index];
        if (platform === "modrinth") {
          if (selectedFile.files.length > 1) {
            const fileMessages = [
              "请选择要下载的文件：",
              ...selectedFile.files.map((f, i) => `${i + 1}. ${f.filename} (${formatFileSize(f.size)}) [${f.primary ? "主要" : "次要"}]`)
            ];
            await renderOutput(session, fileMessages, null, ctx, config, false);
            const fileInput = await session.prompt(6e4);
            const fileIndex = parseInt(fileInput) - 1;
            if (isNaN(fileIndex) || fileIndex < 0 || fileIndex >= selectedFile.files.length) return "无效选择";
            await session.send(`[${selectedFile.files[fileIndex].filename}](${selectedFile.files[fileIndex].url})`);
          } else {
            await session.send(`[${selectedFile.files[0].filename}](${selectedFile.files[0].url})`);
          }
        } else {
          if (!selectedFile.downloadUrl) return "获取下载链接失败";
          await session.send(`[${selectedFile.fileName}](${selectedFile.downloadUrl})`);
        }
        return "";
      }
    }
  } catch (error) {
    ctx.logger.error(`下载处理失败:`, error);
    return "下载过程中出错";
  }
}
__name(handleDownload, "handleDownload");

// src/resource/curseforge.ts
var CF_API_BASE2 = "https://api.curseforge.com/v1";
async function searchCurseForgeProjects(ctx, keyword, api, options = {}) {
  try {
    if (!api) return { results: [], pagination: { totalCount: 0 } };
    const params = { gameId: 432, searchFilter: keyword, sortOrder: options["sortOrder"] || "desc" };
    const validParams = [
      "categoryId",
      "classId",
      "gameVersion",
      "modLoaderType",
      "gameVersionTypeId",
      "authorId",
      "primaryAuthorId",
      "slug",
      "categoryIds",
      "gameVersions",
      "modLoaderTypes",
      "sortField",
      "pageSize",
      "index"
    ];
    validParams.forEach((param) => {
      if (options[param] === void 0) return;
      if (Array.isArray(options[param])) {
        params[param] = options[param].join(",");
      } else if (typeof options[param] === "string" && (param === "categoryIds" || param === "gameVersions" || param === "modLoaderTypes")) {
        try {
          const parsed = JSON.parse(options[param]);
          params[param] = Array.isArray(parsed) ? parsed.join(",") : options[param];
        } catch {
          params[param] = options[param];
        }
      } else {
        params[param] = options[param];
      }
    });
    const response = await ctx.http.get(`${CF_API_BASE2}/mods/search`, { headers: { "x-api-key": api }, params });
    return { results: response.data || [], pagination: response.pagination || { totalCount: 0 } };
  } catch (error) {
    ctx.logger.error("CurseForge 搜索失败:", error);
    return { results: [], pagination: { totalCount: 0 } };
  }
}
__name(searchCurseForgeProjects, "searchCurseForgeProjects");
async function getCurseForgeProject(ctx, projectId, api) {
  try {
    if (!api) return null;
    const projectRes = await ctx.http.get(`${CF_API_BASE2}/mods/${projectId}`, { headers: { "x-api-key": api } });
    const project = projectRes.data;
    if (!project) return null;
    const formatDate = /* @__PURE__ */ __name((date) => new Date(date).toLocaleString(), "formatDate");
    const content = [
      project.logo.url && import_koishi4.h.image(project.logo.url),
      `[${project.name}]
${project.summary}`,
      // 基本信息
      [
        `分类: ${project.categories?.map((c) => c.name).join(", ")}`,
        `加载器: ${project.latestFilesIndexes?.map((f) => {
          return Object.entries(CF_MAPS.LOADER).find(([_, val]) => val === f.modLoader)?.[0] || f.modLoader;
        }).filter((v, i, a) => a.indexOf(v) === i).join(", ")}`,
        `支持版本: ${project.latestFilesIndexes?.map((f) => f.gameVersion).filter((v, i, a) => a.indexOf(v) === i).join(", ")}`,
        `作者: ${project.authors.map((a) => a.name).join(", ")}`,
        `更新于: ${formatDate(project.dateModified)}`,
        `下载量: ${project.downloadCount.toLocaleString()}`
      ].filter(Boolean).map((item) => `● ${item}`).join("\n")
    ].filter(Boolean);
    const links = [
      project.links?.websiteUrl && `官方网站: ${project.links.websiteUrl}`,
      project.links?.wikiUrl && `Wiki: ${project.links.wikiUrl}`,
      project.links?.issuesUrl && `问题追踪: ${project.links.issuesUrl}`,
      project.links?.sourceUrl && `源代码: ${project.links.sourceUrl}`
    ].filter(Boolean);
    if (links.length > 0) content.push(`相关链接：
${links.join("\n")}`);
    if (project.screenshots?.length) {
      content.push("图库：");
      project.screenshots.slice(0, 3).forEach((s) => {
        content.push(import_koishi4.h.image(s.url));
      });
    }
    content.push(`项目地址：${project.links.websiteUrl}`);
    return { content, url: project.links.websiteUrl };
  } catch (error) {
    ctx.logger.error("CurseForge 详情获取失败:", error);
    return null;
  }
}
__name(getCurseForgeProject, "getCurseForgeProject");
function registerCurseForge(ctx, mc, config) {
  mc.subcommand(".curseforge <keyword:string>", `查询 CurseForge 资源`).option("type", `-t <type:string> 资源类型(${Object.keys(CF_MAPS.TYPE).join("|")})`).option("version", "-v <version:string> 支持版本").option("loader", "-l <loader:string> 加载器").option("skip", "-k <count:number> 跳过结果数").option("shot", "-s 截图模式").option("download", "-d 下载模式").action(async ({ session, options }, keyword) => {
    if (!keyword) return "请输入关键词";
    if (!config.curseforgeEnabled) return "未配置 CurseForge API 密钥";
    try {
      const searchOptions = {
        categoryId: options.type ? CF_MAPS.TYPE[options.type] : void 0,
        gameVersion: options.version,
        index: Math.max(0, options.skip || 0),
        pageSize: 1,
        modLoaderType: options.loader ? CF_MAPS.LOADER[options.loader] : void 0
      };
      const { results } = await searchCurseForgeProjects(ctx, keyword, config.curseforgeEnabled, searchOptions);
      if (!results.length) return "未找到匹配的资源";
      if (options.download) return handleDownload(ctx, session, "curseforge", results[0], config, options);
      const projectInfo = await getCurseForgeProject(ctx, results[0].id, config.curseforgeEnabled);
      if (!projectInfo) return "获取详情失败";
      const result = await renderOutput(session, projectInfo.content, projectInfo.url, ctx, config, options.shot);
      return config.useForward && result === "" && !options.shot ? void 0 : result;
    } catch (error) {
      ctx.logger.error("CurseForge 查询失败:", error);
      return "查询时出错";
    }
  });
}
__name(registerCurseForge, "registerCurseForge");

// src/resource/mcmod.ts
var import_koishi5 = require("koishi");
function processMcmodIntroduction(body, paragraphLimit) {
  const result = [];
  body = body.replace(/\r\n/g, "\n").replace(/\n{3,}/g, "\n\n").trim();
  const images = [];
  let match;
  const imgRegex = /!\[((?:\[[^\]]*\]|[^\]])*)\]\(([^)]+)\)/g;
  while ((match = imgRegex.exec(body)) !== null) {
    images.push({ index: match.index, length: match[0].length, url: match[2], text: match[0] });
  }
  const processParagraphs = /* @__PURE__ */ __name((text, startBuffer = "") => {
    let buffer = startBuffer;
    for (const para of text.split("\n\n").filter((p) => p.trim())) {
      const isHeading = /^#{1,6}\s/.test(para.trim());
      if (isHeading && buffer || buffer && buffer.length + para.length + 2 > paragraphLimit) {
        result.push(buffer);
        buffer = para;
      } else {
        buffer = buffer ? `${buffer}
${para}` : para;
      }
    }
    return buffer;
  }, "processParagraphs");
  if (!images.length) {
    const finalBuffer = processParagraphs(body);
    if (finalBuffer) result.push(finalBuffer);
    return result;
  }
  let lastIndex = 0, contentBuffer = "";
  for (let i = 0; i < images.length; i++) {
    const image = images[i];
    if (image.index > lastIndex) {
      contentBuffer = processParagraphs(body.substring(lastIndex, image.index).trim(), contentBuffer);
      if (contentBuffer) {
        result.push(contentBuffer);
        contentBuffer = "";
      }
    }
    result.push(import_koishi5.h.image(image.url));
    lastIndex = image.index + image.length;
    if (i < images.length - 1) {
      const nextImage = images[i + 1];
      const textBetween = body.substring(lastIndex, nextImage.index).trim();
      if (textBetween && (textBetween.length < paragraphLimit / 2 || !textBetween.includes("\n\n"))) {
        result.push(textBetween);
        lastIndex = nextImage.index;
      }
    }
  }
  contentBuffer = processParagraphs(body.substring(lastIndex).trim(), contentBuffer);
  if (contentBuffer) result.push(contentBuffer);
  return result;
}
__name(processMcmodIntroduction, "processMcmodIntroduction");
var getMcmodApiBase = /* @__PURE__ */ __name((config) => typeof config?.mcmodEnabled === "string" && config.mcmodEnabled.trim() ? config.mcmodEnabled.trim().endsWith("/") ? config.mcmodEnabled.trim() : config.mcmodEnabled.trim() + "/" : "", "getMcmodApiBase");
var getMapValue = /* @__PURE__ */ __name((map, key, defaultValue = `未知(${key})`) => map[key] || defaultValue, "getMapValue");
async function searchMcmodProjects(ctx, keyword, options = {}, config = null) {
  try {
    const pageSize = 30;
    const page = options["page"] || (options["offset"] ? Math.floor(options["offset"] / pageSize) + 1 : 1);
    const params = {
      q: keyword,
      page,
      ...options["mold"] === 1 || options["mcmold"] ? { mold: 1 } : {},
      ...options["type"] && MCMOD_MAPS.FILTER[options["type"]] > 0 ? { filter: MCMOD_MAPS.FILTER[options["type"]] } : {}
    };
    const response = await ctx.http.get(`${getMcmodApiBase(config)}api/search`, { params });
    if (response?.results?.length) {
      return {
        results: response.results.map(({ id, name: name2, description, type, url, category }) => ({ id, name: name2, description, type, url, category })),
        pagination: {
          page: response.page || 1,
          total: response.total || 1,
          totalResults: response.totalResults || response.results.length,
          pageSize,
          offset: (response.page - 1) * pageSize
        }
      };
    }
    return { results: [], pagination: { page: 1, total: 0, totalResults: 0, pageSize, offset: 0 } };
  } catch (error) {
    ctx.logger.error("MCMOD 搜索失败:", error);
    return { results: [], pagination: { page: 1, total: 0, totalResults: 0, pageSize: 0, offset: 0 } };
  }
}
__name(searchMcmodProjects, "searchMcmodProjects");
async function getMcmodProject(ctx, project, config = null) {
  try {
    const response = await ctx.http.get(`${getMcmodApiBase(config)}api/class`, {
      params: { id: project.extra.id, others: false, community: project.community === true, relations: project.relations === true }
    });
    if (!response?.basicInfo) throw new Error("无法获取模组详情");
    const { basicInfo, compatibility, links, authors, resources, introduction, community, relations } = response;
    const modName = [basicInfo.shortName, basicInfo.name, basicInfo.englishName ? `[${basicInfo.englishName}]` : null].filter(Boolean).join(" ");
    let versionInfo = "未知";
    if (compatibility?.mcVersions) {
      versionInfo = ["forge", "fabric", "neoforge", "quilt", "behaviorPack", "dataPack"].map((platform) => compatibility.mcVersions[platform] && `${platform === "behaviorPack" ? "行为包" : platform === "dataPack" ? "数据包" : platform.charAt(0).toUpperCase() + platform.slice(1)}: ${compatibility.mcVersions[platform].join(", ")}`).filter(Boolean).join("\n● ");
    }
    const getModStatus = /* @__PURE__ */ __name((status) => {
      if (!status) return "未知";
      switch (status.status) {
        case "active":
          return "活跃";
        case "semi-abandoned":
          return "半弃坑";
        case "inactive":
          return "停更";
      }
    }, "getModStatus");
    const content = [
      basicInfo.img && import_koishi5.h.image(basicInfo.img),
      [
        modName,
        `状态: ${[getModStatus(basicInfo.status), basicInfo.status?.isOpenSource ? "开源" : "闭源"].join(", ")}`,
        `分类: ${basicInfo.categories?.map((id) => getMapValue(MCMOD_MAPS.MOD_CATEGORY, id)).join(", ") || "未知"}`,
        `标签: ${basicInfo.tags?.join(", ") || "无标签"}`,
        `作者: ${authors?.map((a) => `${a.name}${a.position ? ` (${a.position})` : ""}`).join(", ") || "未知"}`,
        `支持平台: ${compatibility?.platforms?.join(", ") || "未知"}`,
        `运作方式: ${compatibility?.apis?.join(", ") || "未知"}`,
        `运行环境: ${compatibility?.environment || "未知"}`,
        `支持版本:
● ${versionInfo}`,
        `Mod资料:
${resources.map((res) => `● ${getMapValue(MCMOD_MAPS.RESOURCE_TYPE, res.typeId)} (${res.count}条)`).join("\n")}`
      ].join("\n"),
      links?.length && `相关链接:
${links.slice(0, 10).map((link) => `● ${link.title}: ${link.url}`).join("\n")}`
    ].filter(Boolean);
    if (project.relations && relations?.length) {
      const relationItems = ["模组关系:"];
      for (const relation of relations) {
        if (relation.version) {
          relationItems.push(`[${relation.version}]版本：`);
          if (relation.dependencyMods?.length) {
            relationItems.push(`  依赖模组: ${relation.dependencyMods.slice(0, 10).map((mod) => `[${mod.name}](https://www.mcmod.cn/class/${mod.id}.html)`).join(", ")}`);
          }
          if (relation.relationMods?.length) {
            relationItems.push(`  关联模组: ${relation.relationMods.slice(0, 10).map((mod) => `[${mod.name}](https://www.mcmod.cn/class/${mod.id}.html)`).join(", ")}`);
          }
        }
      }
      if (relationItems.length > 1) content.push(relationItems.join("\n"));
    }
    if (community) {
      if (community.tutorials?.length) {
        content.push("Mod教程:");
        content.push(community.tutorials.slice(0, 10).map((t) => `● [${t.title}](https://www.mcmod.cn/post/${t.id}.html)`).join("\n"));
      }
      if (community.discussions?.length) {
        content.push("Mod讨论:");
        content.push(community.discussions.slice(0, 10).map((d) => `● [${d.title}](https://bbs.mcmod.cn/thread-${d.id}-1-1.html)`).join("\n"));
      }
    }
    if (introduction) {
      content.push(`详细介绍：`);
      const allIntroParts = processMcmodIntroduction(introduction, config.maxDescLength);
      const introParts = allIntroParts.slice(0, config.maxParagraphs);
      content.push(...introParts);
      if (introParts.length < allIntroParts.length) content.push("（更多内容请查看完整页面）");
    }
    content.push(`查看详情: ${project.url}`);
    return { content, url: project.url, icon: basicInfo.img || null };
  } catch (error) {
    ctx.logger.error("MCMOD 详情获取失败:", error);
    return null;
  }
}
__name(getMcmodProject, "getMcmodProject");
async function getMcmodModpack(ctx, modpack, config = null) {
  try {
    const response = await ctx.http.get(`${getMcmodApiBase(config)}api/modpack`, {
      params: { id: modpack.extra.id, others: false, community: modpack.community === true, relations: modpack.relations === true }
    });
    if (!response?.basicInfo) throw new Error("无法获取整合包详情");
    const { basicInfo, compatibility, links, authors, introduction, relations } = response;
    const packName = [basicInfo.shortName, basicInfo.name, basicInfo.englishName ? `[${basicInfo.englishName}]` : null].filter(Boolean).join(" ");
    const content = [
      basicInfo.img && import_koishi5.h.image(basicInfo.img),
      [
        packName,
        `分类: ${basicInfo.categories?.map((id) => getMapValue(MCMOD_MAPS.CATEGORY, id)).join(", ") || "未知"}`,
        `作者: ${authors?.map((a) => `${a.name}${a.position ? ` (${a.position})` : ""}`).join(", ") || "未知"}`,
        compatibility?.packType && `整合包类型: ${compatibility.packType}`,
        compatibility?.apiType?.length ? `运作方式: ${compatibility.apiType.join(", ")}` : null,
        compatibility?.packMethod?.length ? `打包方式: ${compatibility.packMethod.join(", ")}` : null,
        compatibility?.mcVersions ? `支持版本: ${Array.isArray(compatibility.mcVersions) ? compatibility.mcVersions.join(", ") : "未知"}` : null
      ].filter(Boolean).join("\n"),
      links?.length && `相关链接:
${links.slice(0, 10).map((link) => `● ${link.title}: ${link.url}`).join("\n")}`
    ].filter(Boolean);
    if (modpack.relations && relations?.mods?.length) {
      content.push("包含模组:");
      content.push(relations.mods.slice(0, 10).map((mod) => `● [${mod.name}](https://www.mcmod.cn/class/${mod.id}.html)${mod.version ? ` (${mod.version})` : ""}`).join("\n"));
    }
    if (modpack.community && relations?.tutorials?.length) {
      content.push("相关教程:");
      content.push(relations.tutorials.slice(0, 10).map((t) => `● [${t.title}](https://www.mcmod.cn/post/${t.id}.html)`).join("\n"));
    }
    if (introduction) {
      content.push(`详细介绍：`);
      const allIntroParts = processMcmodIntroduction(introduction, config.maxDescLength);
      const introParts = allIntroParts.slice(0, config.maxParagraphs);
      content.push(...introParts);
      if (introParts.length < allIntroParts.length) content.push("（更多内容请查看完整页面）");
    }
    content.push(`查看详情: ${modpack.url}`);
    return { content, url: modpack.url, icon: basicInfo.img || null };
  } catch (error) {
    ctx.logger.error("MCMOD 整合包获取失败:", error);
    return null;
  }
}
__name(getMcmodModpack, "getMcmodModpack");
async function getMcmodPost(ctx, post, config = null) {
  try {
    const response = await ctx.http.get(`${getMcmodApiBase(config)}api/post`, {
      params: { id: post.extra.id, others: post.others === true }
    });
    if (!response?.content) throw new Error("无法获取教程详情");
    const { title, content, author, metrics } = response;
    const postContent = [
      `教程：${title || post.name}`,
      post.others && author ? `作者：${author.name}` : null,
      post.others && metrics?.statistics ? `发布于：${new Date(metrics.statistics.createTime).toLocaleString("zh-CN")}` : null,
      post.others && metrics?.statistics?.viewCount ? `浏览量：${metrics.statistics.viewCount}` : null,
      `查看原文：${post.url}`
    ].filter(Boolean);
    if (content) {
      postContent.push(`
内容预览：`);
      const allContentParts = processMcmodIntroduction(content, config.maxDescLength);
      const contentParts = allContentParts.slice(0, config.maxParagraphs);
      postContent.push(...contentParts);
      if (contentParts.length < allContentParts.length) postContent.push("（更多内容请查看完整页面）");
    }
    return { content: postContent, url: post.url, icon: post.others && author ? author.avatar : null };
  } catch (error) {
    ctx.logger.error("MCMOD 教程获取失败:", error);
    return null;
  }
}
__name(getMcmodPost, "getMcmodPost");
async function getMcmodItem(ctx, item, config = null) {
  try {
    const response = await ctx.http.get(`${getMcmodApiBase(config)}api/item`, { params: { id: item.extra.id, others: item.others === true } });
    if (!response || !Array.isArray(response) || !response.length) throw new Error("无法获取物品详情");
    const itemContent = [];
    let iconUrl = null;
    for (let i = 0; i < response.length; i++) {
      const itemInfo = response[i];
      const { name: name2, englishName, icon, command, category, introduction, properties, recipes } = itemInfo;
      const itemElements = [
        icon && import_koishi5.h.image(icon),
        [
          `${name2}${englishName ? ` [${englishName}]` : ""}`,
          category && `分类：${category}`,
          ...properties?.length ? properties.map((prop) => `${prop.name}：${prop.value}`) : [],
          command && `获取：${command}`
        ].filter(Boolean).join("\n")
      ].filter(Boolean);
      itemContent.push(...itemElements);
      if (recipes?.length) {
        const recipesText = [`合成配方(${recipes.length}个)：`];
        for (let j = 0; j < recipes.length; j++) {
          const recipe = recipes[j];
          const recipeType = recipe.type || "未知类型";
          const materials = recipe.materials?.map((m) => {
            const materialName = m.name || "未知原料";
            return m.count > 1 ? `${materialName} x${m.count}` : materialName;
          }).join(", ") || "无原料信息";
          const result = recipe.result?.name || "未知产物";
          const resultCount = recipe.result?.count > 1 ? ` x${recipe.result.count}` : "";
          recipesText.push(`• 使用: ${recipeType}`);
          recipesText.push(`  原料: ${materials}`);
          recipesText.push(`  产物: ${result}${resultCount}`);
          if (recipe.notes) recipesText.push(`  备注: ${recipe.notes}`);
        }
        itemContent.push(recipesText.join("\n"));
      }
      if (introduction) {
        itemContent.push(`物品介绍：`);
        const allIntroParts = processMcmodIntroduction(introduction, config.maxDescLength);
        const introParts = allIntroParts.slice(0, config.maxParagraphs);
        itemContent.push(...introParts);
        if (introParts.length < allIntroParts.length) itemContent.push("（更多内容请查看完整页面）");
      }
      if (item.others && itemInfo.teams?.relatedItems?.length) {
        const relatedItems = [`相关内容：`];
        itemInfo.teams.relatedItems.slice(0, 10).forEach((related) => {
          relatedItems.push(`• [${related.name}](${related.url})`);
        });
        itemContent.push(relatedItems.join("\n"));
      }
    }
    itemContent.push(`查看详情: ${item.url}`);
    return { content: itemContent, url: item.url, icon: iconUrl };
  } catch (error) {
    ctx.logger.error("MCMOD 资料详情获取失败:", error);
    return null;
  }
}
__name(getMcmodItem, "getMcmodItem");
function registerMcmod(ctx, mc, config) {
  mc.subcommand(".mod <keyword:string>", "查询 MCMOD 百科").option("type", "-t <type:string> 资源类型").option("mold", "-m 启用复杂搜索").option("community", "-c 获取教程讨论").option("relations", "-r 获取模组关系").option("others", "-o 获取额外信息").option("page", "-p <page:number> 页码").option("shot", "-s 截图模式").action(async ({ session, options }, keyword) => {
    if (!keyword) return "请输入关键词";
    try {
      const projects = await searchMcmodProjects(ctx, keyword, { type: options.type, mold: options.mold ? 1 : 0, page: options.page }, config);
      if (!projects.results.length) return "未找到匹配的资源";
      const project = projects.results[0];
      let projectInfo;
      switch (project.type) {
        case "class":
          projectInfo = await getMcmodProject(ctx, {
            name: project.name,
            url: project.url,
            description: project.description,
            community: options.community,
            relations: options.relations,
            extra: { id: project.id, type: project.type, category: project.category }
          }, config);
          break;
        case "modpack":
          projectInfo = await getMcmodModpack(ctx, {
            name: project.name,
            url: project.url,
            description: project.description,
            community: options.community,
            relations: options.relations,
            extra: { id: project.id, type: project.type, category: project.category }
          }, config);
          break;
        case "post":
          projectInfo = await getMcmodPost(ctx, {
            name: project.name,
            url: project.url,
            description: project.description,
            others: options.others,
            extra: { id: project.id, type: project.type }
          }, config);
          break;
        case "item":
          projectInfo = await getMcmodItem(ctx, {
            name: project.name,
            url: project.url,
            description: project.description,
            others: options.others,
            extra: { id: project.id, type: project.type }
          }, config);
          break;
        default:
          projectInfo = {
            content: [
              `[${project.name}]`,
              project.description || "暂无描述",
              `类型: ${getMapValue(MCMOD_MAPS.TYPE, project.type, "未知")}`,
              `查看详情: ${project.url}`
            ].filter(Boolean),
            url: project.url,
            icon: null
          };
      }
      const result = await renderOutput(session, projectInfo.content, projectInfo.url, ctx, config, options.shot);
      return config.useForward && result === "" && !options.shot ? void 0 : result;
    } catch (error) {
      ctx.logger.error("MCMOD 查询失败:", error);
      return "查询时出错";
    }
  });
}
__name(registerMcmod, "registerMcmod");

// src/resource/mcwiki.ts
var import_koishi6 = require("koishi");
var WIKI_API_BASE = "https://zh.minecraft.wiki/api.php";
async function searchMcwikiPages(ctx, keyword, options = {}) {
  try {
    const { offset = 0, what } = options;
    const sroffset = Number(offset);
    const params = { action: "query", list: "search", srsearch: keyword, format: "json" };
    if (sroffset > 0) params.sroffset = sroffset;
    if (what) params.srwhat = what;
    const response = await ctx.http.get(WIKI_API_BASE, { params });
    const results = response.query?.search || [];
    const searchInfo = response.query?.searchinfo || {};
    const continueInfo = response.continue || {};
    return {
      results,
      pagination: {
        totalResults: searchInfo.totalhits || 0,
        offset: sroffset,
        nextOffset: continueInfo.sroffset ? Number(continueInfo.sroffset) : void 0,
        exhausted: results.length === 0 || !continueInfo.sroffset
      }
    };
  } catch (error) {
    ctx.logger.error("Minecraft Wiki 搜索失败:", error);
    return { results: [], pagination: { totalResults: 0, offset: Number(options.offset || 0), exhausted: true } };
  }
}
__name(searchMcwikiPages, "searchMcwikiPages");
async function getMcwikiPage(ctx, pageId, config) {
  try {
    const response = await ctx.http.get(WIKI_API_BASE, {
      params: {
        action: "query",
        pageids: pageId,
        prop: "info|extracts|images",
        inprop: "url",
        explaintext: true,
        format: "json"
      }
    });
    const page = response.query?.pages?.[pageId];
    if (!page) return null;
    let imageUrl = null;
    if (page.images?.length > 0) {
      try {
        const imgResponse = await ctx.http.get(WIKI_API_BASE, {
          params: {
            action: "query",
            titles: page.images[0].title,
            prop: "imageinfo",
            iiprop: "url",
            format: "json"
          }
        });
        const imgPages = imgResponse.query?.pages;
        imageUrl = imgPages?.[Object.keys(imgPages || {})[0]]?.imageinfo?.[0]?.url;
      } catch (error) {
        ctx.logger.error("Minecraft Wiki 图片获取失败:", error);
      }
    }
    const extractContent = page.extract ? processWikiExtract(page.extract, config.maxDescLength).slice(0, config.maxParagraphs) : [];
    if (page.extract && processWikiExtract(page.extract, config.maxDescLength).length > config.maxParagraphs) extractContent.push("（更多内容请查看完整页面）");
    const content = [
      imageUrl && import_koishi6.h.image(imageUrl),
      extractContent.length > 0 ? `[${page.title}]
${extractContent[0]}` : `[${page.title}]`,
      ...extractContent.slice(1),
      `完整页面: ${page.fullurl}`
    ].filter(Boolean);
    return { content, url: page.fullurl, icon: imageUrl };
  } catch (error) {
    ctx.logger.error("Minecraft Wiki 详情获取失败:", error);
    return null;
  }
}
__name(getMcwikiPage, "getMcwikiPage");
function splitIntoParagraphs(text, paragraphLimit) {
  const result = [];
  if (text.length <= paragraphLimit) {
    result.push(text);
    return result;
  }
  const sentenceBreaks = text.match(/[。！？\.!?]+/g);
  if (sentenceBreaks?.length > 5) {
    let subParagraph = "";
    for (const sentence of text.split(/(?<=[。！？\.!?]+)/)) {
      if ((subParagraph + sentence).length > paragraphLimit) {
        if (subParagraph.trim()) result.push(subParagraph.trim());
        subParagraph = sentence;
      } else {
        subParagraph += sentence;
      }
    }
    if (subParagraph.trim()) result.push(subParagraph.trim());
  } else {
    for (let j = 0; j < text.length; j += paragraphLimit) {
      result.push(text.substring(j, Math.min(j + paragraphLimit, text.length)).trim());
    }
  }
  return result;
}
__name(splitIntoParagraphs, "splitIntoParagraphs");
function processWikiExtract(extract, paragraphLimit) {
  const normalizedText = extract.replace(/\r\n/g, "\n").replace(/\n{3,}/g, "\n\n");
  const paragraphs = normalizedText.split("\n\n");
  const result = [];
  let currentSection = "";
  let lastSectionType = "none";
  for (const paragraph of paragraphs) {
    if (!paragraph.trim()) continue;
    const isHeading = /^==+ .+ ==+$/.test(paragraph);
    const isSubHeading = /^=== .+ ===$/.test(paragraph);
    const saveCurrentSection = /* @__PURE__ */ __name(() => {
      if (currentSection) {
        result.push(currentSection.trim());
        currentSection = "";
      }
    }, "saveCurrentSection");
    if (isHeading) {
      saveCurrentSection();
      result.push(paragraph);
      lastSectionType = "heading";
    } else if (isSubHeading) {
      saveCurrentSection();
      currentSection = paragraph;
      lastSectionType = "subheading";
    } else if (paragraph.length > paragraphLimit) {
      saveCurrentSection();
      result.push(...splitIntoParagraphs(paragraph, paragraphLimit));
      lastSectionType = "content";
    } else if (lastSectionType === "subheading") {
      currentSection += "\n" + paragraph;
      lastSectionType = "content";
    } else {
      saveCurrentSection();
      currentSection = paragraph;
      lastSectionType = "content";
    }
  }
  if (currentSection) result.push(currentSection.trim());
  return result;
}
__name(processWikiExtract, "processWikiExtract");
function registerMcwiki(ctx, mc, config) {
  mc.subcommand(".wiki <keyword:string>", "查询 Minecraft Wiki").option("skip", "-k <count:number> 跳过结果数").option("what", "-w <what:string> 搜索范围").option("exact", "-e 精确匹配").option("shot", "-s 截图模式").action(async ({ session, options }, keyword) => {
    if (!keyword) return "请输入关键词";
    try {
      const searchKey = options.exact ? `"${keyword}"` : keyword;
      const searchOptions = { offset: Math.max(0, options.skip || 0), what: options.what };
      const { results: pages, pagination } = await searchMcwikiPages(ctx, searchKey, searchOptions);
      if (!pages.length) return "未找到匹配的条目";
      const pageInfo = await getMcwikiPage(ctx, pages[0].pageid, config);
      if (!pageInfo) return "获取详情失败";
      const result = await renderOutput(session, pageInfo.content, pageInfo.url, ctx, config, options.shot);
      return config.useForward && result === "" && !options.shot ? void 0 : result;
    } catch (error) {
      ctx.logger.error("Minecraft Wiki 查询失败:", error);
      return "查询时出错";
    }
  });
}
__name(registerMcwiki, "registerMcwiki");

// src/resource/maps.ts
var PLATFORMS = {
  modrinth: {
    name: "Modrinth",
    /**
     * 搜索Modrinth项目
     * @param {Context} ctx - Koishi上下文
     * @param {string} keyword - 搜索关键词
     * @param {Config} config - 配置对象
     * @param {Object} options - 搜索选项
     * @returns {Promise<Array>} 搜索结果列表
     */
    search: /* @__PURE__ */ __name((ctx, keyword, config, options = {}) => searchModrinthProjects(ctx, keyword, options), "search"),
    /**
     * 获取Modrinth项目详情
     * @type {Function}
     */
    getDetail: getModrinthProject,
    /**
     * 转换Modrinth项目数据为统一格式
     * @param {Object} p - 项目数据
     * @returns {Object} 统一格式的项目数据
     */
    transform: /* @__PURE__ */ __name((p) => ({
      platform: "Modrinth",
      name: p.title,
      description: p.description,
      url: `https://modrinth.com/${p.project_type}/${p.slug}`,
      extra: { id: p.project_id, type: p.project_type, author: p.author, downloads: p.downloads }
    }), "transform"),
    /**
     * 检查配置是否启用该平台
     * @param {Config} config - 配置对象
     * @returns {boolean} 是否启用
     */
    checkConfig: /* @__PURE__ */ __name((config) => config.modrinthEnabled, "checkConfig")
  },
  curseforge: {
    name: "CurseForge",
    /**
     * 搜索CurseForge项目
     * @param {Context} ctx - Koishi上下文
     * @param {string} keyword - 搜索关键词
     * @param {Config} config - 配置对象
     * @param {Object} options - 搜索选项
     * @returns {Promise<Object>} 包含结果和分页信息的对象
     */
    search: /* @__PURE__ */ __name((ctx, keyword, config, options = {}) => searchCurseForgeProjects(ctx, keyword, config.curseforgeEnabled, options), "search"),
    /**
     * 获取CurseForge项目详情
     * @param {Context} ctx - Koishi上下文
     * @param {string} id - 项目ID
     * @param {Config} config - 配置对象
     * @returns {Promise<Object>} 项目详情
     */
    getDetail: /* @__PURE__ */ __name((ctx, id, config) => getCurseForgeProject(ctx, id, config.curseforgeEnabled), "getDetail"),
    /**
     * 转换CurseForge项目数据为统一格式
     * @param {Object} p - 项目数据
     * @returns {Object} 统一格式的项目数据
     */
    transform: /* @__PURE__ */ __name((p) => ({
      platform: "CurseForge",
      name: p.name,
      description: p.summary,
      url: p.links?.websiteUrl || "",
      extra: { id: p.id, author: p.authors.map((a) => a.name).join(", "), downloads: p.downloadCount, type: p.classId }
    }), "transform"),
    /**
     * 检查配置是否启用该平台
     * @param {Config} config - 配置对象
     * @returns {boolean} 是否启用
     */
    checkConfig: /* @__PURE__ */ __name((config) => config.curseforgeEnabled, "checkConfig")
  },
  mcmod: {
    name: "MCMOD",
    /**
     * 搜索MCMOD项目
     * @param {Context} ctx - Koishi上下文
     * @param {string} keyword - 搜索关键词
     * @param {Config} config - 配置对象
     * @param {Object} options - 搜索选项
     * @returns {Promise<Object>} 包含结果和分页信息的对象
     */
    search: /* @__PURE__ */ __name((ctx, keyword, config, options = {}) => searchMcmodProjects(ctx, keyword, options, config), "search"),
    /**
     * 获取MCMOD项目详情
     * @param {Context} ctx - Koishi上下文
     * @param {string} id - 项目ID
     * @param {Config} config - 配置对象
     * @returns {Promise<Object>} 项目详情
     */
    getDetail: /* @__PURE__ */ __name((ctx, project, config) => getMcmodProject(ctx, project, config), "getDetail"),
    /**
     * 转换MCMOD项目数据为统一格式
     * @param {Object} p - 项目数据
     * @returns {Object} 统一格式的项目数据
     */
    transform: /* @__PURE__ */ __name((p) => ({
      platform: "MCMOD",
      name: p.name,
      description: p.description,
      url: p.url,
      extra: { id: p.id, type: p.type, category: p.category }
    }), "transform"),
    /**
     * 检查配置是否启用该平台
     * @param {Config} config - 配置对象
     * @returns {boolean} 是否启用
     */
    checkConfig: /* @__PURE__ */ __name((config) => !!config.mcmodEnabled, "checkConfig")
  },
  mcwiki: {
    name: "Minecraft Wiki",
    /**
     * 搜索Minecraft Wiki页面
     * @param {Context} ctx - Koishi上下文
     * @param {string} keyword - 搜索关键词
     * @param {Config} config - 配置对象
     * @param {Object} options - 搜索选项
     * @returns {Promise<Array>} 搜索结果列表
     */
    search: /* @__PURE__ */ __name((ctx, keyword, config, options = {}) => searchMcwikiPages(ctx, keyword, options), "search"),
    /**
     * 获取Minecraft Wiki页面详情
     * @param {Context} ctx - Koishi上下文
     * @param {string} id - 页面ID
     * @param {Config} config - 配置对象
     * @returns {Promise<Object>} 页面详情
     */
    getDetail: getMcwikiPage,
    /**
     * 转换Minecraft Wiki页面数据为统一格式
     * @param {Object} p - 页面数据
     * @returns {Object} 统一格式的页面数据
     */
    transform: /* @__PURE__ */ __name((p) => ({
      platform: "Minecraft Wiki",
      name: p.title,
      description: p.snippet.replace(/<\/?[^>]+(>|$)/g, ""),
      url: `https://zh.minecraft.wiki/w/${encodeURIComponent(p.title)}`,
      extra: { id: p.pageid }
    }), "transform"),
    /**
     * 检查配置是否启用该平台
     * @param {Config} config - 配置对象
     * @returns {boolean} 是否启用
     */
    checkConfig: /* @__PURE__ */ __name((config) => config.mcwikiEnabled, "checkConfig")
  }
};
var CF_MAPS = {
  /** 资源类型映射 */
  TYPE: {
    "mod": 6,
    "resourcepack": 12,
    "world": 17,
    "plugin": 5,
    "modpack": 4471,
    "addon": 4559,
    "customization": 4546,
    "shader": 6552,
    "datapack": 6945
  },
  /** 加载器类型映射 */
  LOADER: {
    "any": 0,
    "forge": 1,
    "cauldron": 2,
    "liteloader": 3,
    "fabric": 4,
    "quilt": 5,
    "neoforge": 6
  },
  /** 发布类型映射 */
  RELEASE: { 1: "正式版", 2: "快照版", 3: "开发版" },
  /** 依赖关系映射 */
  RELATION: { 1: "必需", 2: "可选", 3: "不兼容", 4: "内置", 5: "工具" }
};
var STATUS_MAP = {
  compatibility: { required: "必需", optional: "可选", unsupported: "不支持" },
  type: { mod: "模组", modpack: "整合包", resourcepack: "资源包", shader: "着色器" }
};
var MCMOD_MAPS = {
  /** 搜索过滤器映射 */
  FILTER: {
    "mod": 1,
    // 模组
    "modpack": 2,
    // 整合包
    "item": 3,
    // 资料
    "post": 4,
    // 教程
    "author": 5,
    // 作者
    "user": 6,
    // 用户
    "community": 7
    // 社群
  },
  /** 资源类型映射 */
  RESOURCE_TYPE: {
    "1": "物品/方块",
    "2": "生物",
    "3": "附魔",
    "4": "世界生成",
    "5": "维度",
    "6": "BUFF/DEBUFF",
    "7": "多方块结构",
    "8": "自然生成",
    "9": "绑定热键",
    "10": "游戏设定",
    "11": "指令",
    "12": "API",
    "191": "编辑规范"
  },
  /** 整合包分类映射 */
  CATEGORY: {
    "1": "科技",
    "2": "魔法",
    "3": "冒险",
    "4": "建筑",
    "5": "地图",
    "6": "任务",
    "7": "硬核",
    "8": "休闲",
    "9": "大型",
    "10": "轻量",
    "11": "剧情",
    "12": "水槽",
    "13": "空岛",
    "14": "PvP",
    "15": "国创"
  },
  /** 模组分类映射 */
  MOD_CATEGORY: {
    "1": "科技",
    "2": "魔法",
    "3": "冒险",
    "4": "农业",
    "5": "装饰",
    "6": "安全",
    "7": "API",
    "8": "资源",
    "9": "世界",
    "10": "群系",
    "11": "生物",
    "12": "能源",
    "13": "存储",
    "14": "物流",
    "15": "道具",
    "16": "红石",
    "17": "食物",
    "18": "模型",
    "19": "指南",
    "20": "破坏",
    "21": "自定义",
    "22": "Meme",
    "23": "实用",
    "24": "脚本",
    "25": "中式",
    "26": "日式",
    "27": "西式",
    "28": "恐怖",
    "29": "建材",
    "30": "生存",
    "31": "指令",
    "32": "优化",
    "33": "国创",
    "34": "关卡",
    "35": "结构"
  },
  /** 资源类型映射 */
  TYPE: {
    "class": "模组",
    "modpack": "整合包",
    "item": "资料",
    "post": "教程",
    "author": "作者",
    "user": "用户",
    "community": "社群"
  }
};

// src/resource/modrinth.ts
var MR_API_BASE2 = "https://api.modrinth.com/v2";
function parseFacets(facetsStr) {
  if (!facetsStr) return [];
  try {
    return facetsStr.startsWith("[") && facetsStr.endsWith("]") ? JSON.parse(facetsStr) : facetsStr.split(",").map((facet) => {
      const parts = facet.trim().split(":");
      return parts.length >= 2 ? [parts.length === 2 ? `${parts[0]}:${parts[1]}` : `${parts[0]}${parts[1]}${parts[2]}`] : [facet.trim()];
    });
  } catch {
    return [];
  }
}
__name(parseFacets, "parseFacets");
async function searchModrinthProjects(ctx, keyword, options = {}) {
  try {
    const { facets, sort, offset, limit, ...otherOptions } = options;
    const params = {
      query: keyword,
      ...sort && { index: sort },
      ...offset !== void 0 && { offset },
      ...limit !== void 0 && { limit },
      ...Object.fromEntries(Object.entries(otherOptions).filter(([_, v]) => v !== void 0))
    };
    if (facets) params["facets"] = JSON.stringify(parseFacets(facets));
    const response = await ctx.http.get(`${MR_API_BASE2}/search`, { params });
    return response.hits || [];
  } catch (error) {
    ctx.logger.error("Modrinth 搜索失败:", error);
    return [];
  }
}
__name(searchModrinthProjects, "searchModrinthProjects");
async function getModrinthDependencies(ctx, projectId) {
  try {
    return await ctx.http.get(`${MR_API_BASE2}/project/${projectId}/dependencies`);
  } catch (error) {
    ctx.logger.error("Modrinth 依赖获取失败:", error);
    return null;
  }
}
__name(getModrinthDependencies, "getModrinthDependencies");
function splitIntoParagraphs2(text, paragraphLimit) {
  const result = [];
  if (text.length <= paragraphLimit) {
    result.push(text);
    return result;
  }
  const sentenceBreaks = text.match(/[。！？\.!?]+/g);
  if (sentenceBreaks?.length > 5) {
    let subParagraph = "";
    for (const sentence of text.split(/(?<=[。！？\.!?]+)/)) {
      if ((subParagraph + sentence).length > paragraphLimit) {
        if (subParagraph.trim()) result.push(subParagraph.trim());
        subParagraph = sentence;
      } else {
        subParagraph += sentence;
      }
    }
    if (subParagraph.trim()) result.push(subParagraph.trim());
  } else {
    for (let j = 0; j < text.length; j += paragraphLimit) {
      result.push(text.substring(j, Math.min(j + paragraphLimit, text.length)).trim());
    }
  }
  return result;
}
__name(splitIntoParagraphs2, "splitIntoParagraphs");
function processModrinthBody(body, paragraphLimit) {
  const images = [];
  body = body.replace(/<img\s+src="([^"]+)"(?:\s+alt="[^"]*")?\s*\/?>/g, (match, url) => {
    images.push(url);
    return `__IMAGE__${images.length - 1}__`;
  });
  body = body.replace(/<h[1-6][^>]*>(.*?)<\/h[1-6]>/g, "$1").replace(/<center>([\s\S]*?)<\/center>/g, "$1").replace(/<p[^>]*>([\s\S]*?)<\/p>/g, "$1\n\n").replace(/<[^>]*>/g, "").replace(/<!--[\s\S]*?-->/g, "");
  const normalizedText = body.replace(/\r\n/g, "\n").replace(/\n{3,}/g, "\n\n");
  const paragraphs = normalizedText.split("\n\n");
  const result = [];
  for (const paragraph of paragraphs) {
    if (!paragraph.trim()) continue;
    const imageMatch = paragraph.trim().match(/^__IMAGE__(\d+)__$/);
    if (imageMatch) {
      const imageIndex = parseInt(imageMatch[1]);
      if (!isNaN(imageIndex) && imageIndex >= 0 && imageIndex < images.length) result.push(`__IMAGE__${images[imageIndex]}__`);
      continue;
    }
    const cleanParagraph = paragraph.replace(/[#\-*]/g, "").replace(/\[([^\]]+)\]\(([^)]+)\)/g, "$1: $2").replace(/__IMAGE__\d+__/g, "");
    if (!cleanParagraph.trim()) continue;
    result.push(...splitIntoParagraphs2(cleanParagraph.trim(), paragraphLimit));
  }
  return result;
}
__name(processModrinthBody, "processModrinthBody");
async function getModrinthProject(ctx, projectId, config, showDependencies = false) {
  try {
    const project = await ctx.http.get(`${MR_API_BASE2}/project/${projectId}`);
    if (!project) return null;
    const projectUrl = `https://modrinth.com/${project.project_type}/${project.slug}`;
    const formatDate = /* @__PURE__ */ __name((date) => date ? new Date(date).toLocaleString() : "未知", "formatDate");
    const content = [
      project.icon_url && import_koishi7.h.image(project.icon_url),
      `[${project.title}]
${project.description}`,
      // 详细信息
      [
        `分类: ${project.categories?.join(", ")}`,
        project.additional_categories?.length ? `子分类: ${project.additional_categories.join(", ")}` : null,
        `类型: ${STATUS_MAP.type && STATUS_MAP.type[project.project_type] || project.project_type}`,
        `加载器: ${project.loaders?.join(", ")}`,
        `客户端: ${STATUS_MAP.compatibility && STATUS_MAP.compatibility[project.client_side] || project.client_side} 服务端: ${STATUS_MAP.compatibility && STATUS_MAP.compatibility[project.server_side] || project.server_side}`,
        `支持版本: ${project.game_versions?.join(", ")}`,
        `更新于: ${formatDate(project.updated)}`,
        `下载量: ${project.downloads?.toLocaleString()}`,
        `许可: ${project.license?.id}${project.license?.name ? ` (${project.license.name})` : ""}`
      ].filter(Boolean).map((item) => `● ${item}`).join("\n")
    ].filter(Boolean);
    if (showDependencies) {
      const dependencies = await getModrinthDependencies(ctx, projectId);
      if (dependencies) {
        if (dependencies.projects?.length > 0) {
          content.push("项目依赖：");
          const depItems = dependencies.projects.map(
            (dep) => `● ${dep.title}${dep.project_type ? ` (${dep.project_type})` : ""} - ${dep.slug}`
          );
          content.push(depItems.join("\n"));
        }
      }
    }
    const links = [
      project.source_url && `源代码: ${project.source_url}`,
      project.issues_url && `问题追踪: ${project.issues_url}`,
      project.wiki_url && `Wiki: ${project.wiki_url}`,
      project.discord_url && `Discord: ${project.discord_url}`,
      ...project.donation_urls?.map(
        (d) => `赞助 (${d.platform}): ${d.url}`
      ) || []
    ].filter(Boolean);
    if (links.length > 0) content.push(`相关链接：
${links.join("\n")}`);
    if (project.gallery?.length > 0) {
      content.push(`图库：`);
      project.gallery.slice(0, 3).forEach((img) => {
        content.push(import_koishi7.h.image(img.url));
      });
    }
    if (project.body) {
      content.push(`详细介绍：`);
      const bodyParts = processModrinthBody(project.body, config.maxDescLength).slice(0, config.maxParagraphs);
      content.push(...bodyParts.map((part) => {
        const imageMatch = part.match(/^__IMAGE__(.+)__$/);
        if (imageMatch) return import_koishi7.h.image(imageMatch[1]);
        return part;
      }));
      if (processModrinthBody(project.body, config.maxDescLength).length > config.maxParagraphs) {
        content.push("（更多内容请查看完整页面）");
      }
    }
    content.push(`项目地址：${projectUrl}`);
    return { content, url: projectUrl };
  } catch (error) {
    ctx.logger.error("Modrinth 详情获取失败:", error);
    return null;
  }
}
__name(getModrinthProject, "getModrinthProject");
function registerModrinth(ctx, mc, config) {
  mc.subcommand(".modrinth <keyword:string>", "查询 Modrinth 资源").option("type", "-t <type:string> 资源类型").option("version", "-v <version:string> 支持版本").option("loader", "-l <loader:string> 加载器").option("facets", "-f <facets:string> 高级过滤").option("skip", "-k <count:number> 跳过结果数").option("sort", "-sort <sort:string> 排序方式").option("dependencies", "-dep 显示依赖").option("shot", "-s 截图模式").option("download", "-d 下载模式").action(async ({ session, options }, keyword) => {
    if (!keyword) return "请输入关键词";
    try {
      const searchOptions = { sort: options.sort, offset: Math.max(0, options.skip || 0), limit: 1 };
      const facetsArray = [];
      if (options.type) facetsArray.push([`project_type:${options.type}`]);
      if (options.version) facetsArray.push([`versions:${options.version}`]);
      if (options.loader) facetsArray.push([`categories:${options.loader}`]);
      if (options.facets) facetsArray.push(...parseFacets(options.facets));
      if (facetsArray.length) searchOptions["facets"] = JSON.stringify(facetsArray);
      const projects = await searchModrinthProjects(ctx, keyword, searchOptions);
      if (!projects.length) return "未找到匹配的资源";
      if (options.download) return handleDownload(ctx, session, "modrinth", projects[0], config, options);
      const projectInfo = await getModrinthProject(ctx, projects[0].project_id, config, options.dependencies);
      if (!projectInfo) return "获取详情失败";
      const result = await renderOutput(session, projectInfo.content, projectInfo.url, ctx, config, options.shot);
      return config.useForward && result === "" && !options.shot ? void 0 : result;
    } catch (error) {
      ctx.logger.error("Modrinth 查询失败:", error);
      return "查询时出错";
    }
  });
}
__name(registerModrinth, "registerModrinth");

// src/resource/parser.ts
var LINK_PARSERS = [
  {
    /** Modrinth 项目链接匹配规则 */
    regex: /modrinth\.com\/(?:mod|modpack|resourcepack|shader|datapack|plugin)\/([a-zA-Z0-9_-]+)/,
    handler: /* @__PURE__ */ __name((ctx, match, config) => config.modrinthEnabled && getModrinthProject(ctx, match[1], config), "handler")
  },
  {
    /** CurseForge 项目链接匹配规则 */
    regex: /(?:www\.)?curseforge\.com\/(?:minecraft\/(?:mc-mods|modpacks|texture-packs|worlds|bukkit-plugins|customization)\/([a-zA-Z0-9_-]+)(?:\/files\/(\d+))?|projects\/(\d+))/,
    handler: /* @__PURE__ */ __name(async (ctx, match, config) => {
      if (!config.curseforgeEnabled) return null;
      if (match[3]) return getCurseForgeProject(ctx, parseInt(match[3]), config.curseforgeEnabled);
      if (!match[1]) return null;
      const { data } = await ctx.http.get("https://api.curseforge.com/v1/mods/search", {
        headers: { "x-api-key": config.curseforgeEnabled },
        params: { gameId: 432, slug: match[1], pageSize: 1 }
      });
      return data?.[0]?.id ? getCurseForgeProject(ctx, data[0].id, config.curseforgeEnabled) : null;
    }, "handler")
  },
  {
    /** MCMOD 项目链接匹配规则 */
    regex: /(?:www\.)?mcmod\.cn\/(?:(class|modpack|post|item)\/(\d+)\.html|(?:class|modpack|post|item)\/(\d+))/,
    handler: /* @__PURE__ */ __name((ctx, match, config) => {
      if (!config.mcmodEnabled) return null;
      const type = match[1] || ["class", "modpack", "post", "item"].find((t) => match[0].includes(t)) || "class";
      const id = parseInt(match[2] || match[3]);
      if (!id) return null;
      const handlers = { class: getMcmodProject, modpack: getMcmodModpack, post: getMcmodPost, item: getMcmodItem };
      return handlers[type]?.(ctx, { name: `${type} ${id}`, url: match[0], extra: { id, type } }, config);
    }, "handler")
  },
  {
    /** Minecraft Wiki 页面链接匹配规则 */
    regex: /(?:(?:zh|en)\.)?minecraft\.wiki\/(?:w\/|wiki\/)?([^#?&\s]+)/,
    handler: /* @__PURE__ */ __name(async (ctx, match, config) => {
      if (!config.mcwikiEnabled) return null;
      const pageTitle = decodeURIComponent(match[1]).replace(/_/g, " ");
      const { query: { pages } } = await ctx.http.get("https://zh.minecraft.wiki/api.php", {
        params: { action: "query", titles: pageTitle, format: "json" }
      });
      const pageId = pages && Object.keys(pages)[0];
      return pageId !== "-1" ? getMcwikiPage(ctx, parseInt(pageId), config) : null;
    }, "handler")
  }
];
async function parseAndProcess(ctx, session, content, config) {
  for (const { regex, handler } of LINK_PARSERS) {
    const match = content.match(regex);
    if (!match) continue;
    try {
      const result = await handler(ctx, match, config);
      if (result) return await renderOutput(session, result.content, result.url, ctx, config, config.linkParserEnabled === "shot");
    } catch (error) {
      ctx.logger.error(`链接解析失败:`, error.message);
    }
  }
  return null;
}
__name(parseAndProcess, "parseAndProcess");
function registerLinkParser(ctx, config) {
  ctx.middleware(async (session, next) => {
    const isValidMessage = session.type?.includes("message") || session.type === "send";
    const isCommand = session.content?.match(/^[./#!]/) || session.argv?.name;
    if (isValidMessage && session.content?.trim() && !isCommand) {
      try {
        const result = await parseAndProcess(ctx, session, session.content, config);
        if (result) return session.send(result);
      } catch (error) {
        ctx.logger.error("链接解析错误:", error.message);
      }
    }
    return next();
  });
}
__name(registerLinkParser, "registerLinkParser");

// src/resource/search.ts
function preparePlatforms(ctx, options) {
  const validPlatforms = ["modrinth", "curseforge", "mcmod", "mcwiki"];
  if (options.platform) {
    const platforms = options.platform.toLowerCase().split(",").filter((p) => validPlatforms.includes(p));
    return platforms.length ? { platformsToSearch: platforms } : { error: "所指定平台无效" };
  }
  const defaultPlatforms = ["modrinth", "curseforge"];
  return { platformsToSearch: defaultPlatforms };
}
__name(preparePlatforms, "preparePlatforms");
function buildModrinthFacets(options) {
  const facets = [];
  if (options.type) facets.push([`project_type:${options.type}`]);
  if (options.version) facets.push([`versions:${options.version}`]);
  if (options.loader) facets.push([`categories:${options.loader}`]);
  if (options.mrf) {
    try {
      const parsedFacets = JSON.parse(options.mrf);
      if (Array.isArray(parsedFacets)) facets.push(...parsedFacets);
    } catch {
    }
  }
  return facets.length ? JSON.stringify(facets) : null;
}
__name(buildModrinthFacets, "buildModrinthFacets");
function preparePlatformOptions(options, platformStates) {
  return {
    modrinth: {
      facets: options.mrf || buildModrinthFacets(options),
      sort: options.sort,
      version: options.version,
      offset: platformStates.modrinth?.offset || 0,
      limit: 100
    },
    curseforge: {
      categoryId: options.type ? CF_MAPS.TYPE[options.type] : void 0,
      gameVersion: options.version,
      sortOrder: options.cfo,
      pageSize: 50,
      modLoaderType: options.loader ? CF_MAPS.LOADER[options.loader] : void 0,
      sortField: options.sort,
      index: platformStates.curseforge?.offset || 0
    },
    mcmod: { offset: platformStates.mcmod?.offset || 0, type: options.type, mold: options.mcmold ? 1 : 0 },
    mcwiki: { offset: platformStates.mcwiki?.offset || 0, what: options.what }
  };
}
__name(preparePlatformOptions, "preparePlatformOptions");
async function searchPlatform(ctx, platform, keyword, config, platformOptions, platformStates) {
  const handler = PLATFORMS[platform];
  if (!handler?.checkConfig(config)) return { platform, results: [] };
  try {
    const response = await handler.search(ctx, keyword, config, platformOptions || {});
    const results = response?.results || response || [];
    const pagination = response?.pagination || {};
    if (!results?.length) {
      const isExhausted = pagination.exhausted || pagination.totalResults === 0 || pagination.page >= pagination.total && pagination.total > 0;
      if (isExhausted) platformStates[platform] = { ...platformStates[platform], exhausted: true };
      return { platform, results: [] };
    }
    const transformedResults = results.map(handler.transform);
    const currentOffset = platformStates[platform]?.offset || 0;
    let newOffset = currentOffset + transformedResults.length;
    if (platform === "mcwiki" && pagination.nextOffset !== void 0) {
      newOffset = pagination.nextOffset;
    } else if (pagination.offset !== void 0) {
      newOffset = pagination.offset + results.length;
    } else if (pagination.page && pagination.pageSize) {
      newOffset = pagination.page * pagination.pageSize;
    }
    platformStates[platform] = {
      offset: newOffset,
      totalPages: pagination.total || 1,
      totalResults: pagination.totalResults || 0,
      page: pagination.page || Math.floor(newOffset / (pagination.pageSize || 100)) + 1,
      exhausted: pagination.exhausted || (pagination.page !== void 0 && pagination.total ? pagination.page >= pagination.total : false)
    };
    return { platform, results: transformedResults };
  } catch (error) {
    ctx.logger.error(`${handler.name} 搜索失败:`, error);
    return { platform, results: [] };
  }
}
__name(searchPlatform, "searchPlatform");
async function executeSearch(ctx, keyword, options, config, platforms, platformStates, platformResults) {
  const activePlatforms = platforms.filter((p) => !platformStates[p]?.exhausted);
  if (activePlatforms.length === 0) return { success: true };
  const platformOptions = preparePlatformOptions(options, platformStates);
  const searchResults = await Promise.all(
    activePlatforms.map((platform) => searchPlatform(ctx, platform, keyword, config, platformOptions[platform], platformStates))
  );
  let totalResults = 0;
  searchResults.forEach(({ platform, results }) => {
    if (!results || results.length === 0) return;
    totalResults += results.length;
    if (options.type && platform === "modrinth" && ["mod", "modpack", "resourcepack", "shader"].includes(options.type)) {
      results = results.filter((i) => i.extra?.type === options.type);
    }
    platformResults[platform] = (platformResults[platform] || []).concat(results);
    const expectedPageSizes = { curseforge: 50, mcmod: 30, mcwiki: 10 };
    const expectedPageSize = expectedPageSizes[platform] || 100;
    if (results.length < expectedPageSize && !platformStates[platform]?.exhausted) {
      platformStates[platform] = { ...platformStates[platform], exhausted: true };
    }
  });
  const allExhausted = platforms.every((p) => platformStates[p]?.exhausted);
  const noResults = Object.values(platformResults).every((r) => r.length === 0);
  if (allExhausted && noResults) {
    const names = platforms.map((p) => PLATFORMS[p]?.name || p).join("、");
    return {
      success: false,
      message: `${names} 无匹配资源：${options.type ? `类型:${options.type}` : ""}${options.version ? `, 版本:${options.version}` : ""}`
    };
  }
  return { success: true };
}
__name(executeSearch, "executeSearch");
function mergeResults(platformResults) {
  const platforms = Object.keys(platformResults);
  const maxCount = Math.max(...platforms.map((p) => platformResults[p].length), 0);
  const allResults = [];
  for (let i = 0; i < maxCount; i++) {
    for (const platform of platforms) {
      if (i < platformResults[platform].length) allResults.push(platformResults[platform][i]);
    }
  }
  return allResults;
}
__name(mergeResults, "mergeResults");
async function handleUserInput2(input, allResults, currentPage, config, ctx, session, platforms, platformStates, platformResults, options, keyword) {
  if (!input || input?.toLowerCase() === "c") return { done: true, message: input ? "已取消搜索" : "已超时，自动取消搜索" };
  if (input.toLowerCase() === "n") {
    const endIndex = (currentPage + 1) * 10;
    if (endIndex < allResults.length) return { done: false, nextPage: currentPage + 1 };
    const activePlatforms = platforms.filter((p) => !platformStates[p]?.exhausted);
    if (activePlatforms.length > 0) {
      const beforeLoadCount = allResults.length;
      await executeSearch(ctx, keyword, options, config, activePlatforms, platformStates, platformResults);
      const newAllResults = mergeResults(platformResults);
      if (newAllResults.length > beforeLoadCount) return { done: false, nextPage: Math.floor(beforeLoadCount / 10) };
    }
    return { done: true, message: "无更多结果" };
  }
  const choice = parseInt(input);
  if (isNaN(choice) || choice < 1 || choice > allResults.length) {
    return { done: true, message: `请输入 1-${allResults.length} 的数字，或输入n查看下页，输入c取消` };
  }
  try {
    const selected = allResults[choice - 1];
    const platform = Object.values(PLATFORMS).find((p) => p.name === selected.platform);
    if (!platform) return { done: true, message: "获取详情失败：无法找到平台处理器" };
    const detailArg = selected.platform === "MCMOD" ? selected : selected.extra.id;
    const detail = await platform.getDetail(ctx, detailArg, config);
    if (!detail) return { done: true, message: "获取详情失败" };
    return { done: true, message: await renderOutput(session, detail.content, detail.url, ctx, config) };
  } catch (error) {
    ctx.logger.error(`获取详情发生异常:`, error);
    return { done: true, message: "获取详情时发生错误" };
  }
}
__name(handleUserInput2, "handleUserInput");
async function displayResultPage(ctx, session, config, platformResults, currentPage, platforms, platformStates) {
  const allResults = mergeResults(platformResults);
  const startIndex = currentPage * 10;
  const endIndex = startIndex + 10;
  if (endIndex > allResults.length) {
    const activePlatforms = platforms.filter((p) => !platformStates[p]?.exhausted);
    if (activePlatforms.length > 0 && allResults.length < endIndex && startIndex < allResults.length) return { needMoreResults: true };
  }
  const currentResults = allResults.slice(startIndex, endIndex);
  if (currentResults.length === 0) return { message: "无更多结果" };
  let totalPages = Math.ceil(allResults.length / 10);
  if (platforms.length === 1 && platformStates[platforms[0]]?.totalPages > 0) totalPages = platformStates[platforms[0]].totalPages;
  const formattedResults = [
    "请回复序号查看详情，输入n查看下页，输入c取消",
    ...currentResults.map((p, i) => {
      const index = startIndex + i + 1;
      const desc = p.description ? `
  ${p.description.substring(0, 64)}${p.description.length > 64 ? "..." : ""}` : "";
      return `${index}. [${p.platform}] ${p.name}${desc}`;
    }),
    ...totalPages > 1 ? [`第 ${currentPage + 1}/${totalPages} 页`] : []
  ];
  await renderOutput(session, formattedResults, null, ctx, config, false);
  return { success: true };
}
__name(displayResultPage, "displayResultPage");
function registerSearch(ctx, mc, config) {
  mc.subcommand(".search <keyword:string>", "搜索 Minecraft 资源").option("platform", "-p <platform:string> 指定平台").option("sort", "-s <sort:string> 排序方式").option("version", "-v <version:string> 支持版本").option("loader", "-l <loader:string> 加载器").option("skip", "-k <count:number> 跳过结果数").option("type", `-t <type:string> 资源类型(${Object.keys(CF_MAPS.TYPE).join("|")})`).option("mrf", "-mrf <facets:string> [MR]高级过滤").option("cfo", "-cfo <order:string> [CF]升降序").option("what", "-ww <what:string> [Wiki]搜索范围").option("mcmold", "-mm [Mod]复杂搜索").option("download", "-d 下载模式").action(async ({ session, options }, keyword) => {
    if (!keyword) return "需要关键词";
    try {
      const platformResults = {};
      const platformStates = {};
      let currentPage = 0;
      const skip = Math.max(0, options.skip || 0);
      const platformsResult = preparePlatforms(ctx, options);
      if (platformsResult.error) return platformsResult.error;
      const platforms = platformsResult.platformsToSearch;
      platforms.forEach((p) => {
        platformResults[p] = [];
        platformStates[p] = { offset: skip, page: 1, totalPages: 1, totalResults: 0, exhausted: false };
      });
      const initialSearch = await executeSearch(ctx, keyword, options, config, platforms, platformStates, platformResults);
      if (!initialSearch.success) return initialSearch.message;
      let searchAttempts = 0;
      const maxSearchAttempts = 3;
      while (true) {
        const displayResult = await displayResultPage(ctx, session, config, platformResults, currentPage, platforms, platformStates);
        if (displayResult.message) return displayResult.message;
        if (displayResult.needMoreResults) {
          if (searchAttempts >= maxSearchAttempts) {
            const allResults = mergeResults(platformResults);
            if (allResults.length === 0) return "无匹配结果";
            currentPage = Math.max(0, Math.floor((allResults.length - 1) / 10));
            continue;
          }
          const beforeCount = mergeResults(platformResults).length;
          await executeSearch(
            ctx,
            keyword,
            options,
            config,
            platforms.filter((p) => !platformStates[p]?.exhausted),
            platformStates,
            platformResults
          );
          const afterCount = mergeResults(platformResults).length;
          if (afterCount <= beforeCount) {
            searchAttempts++;
          } else {
            searchAttempts = 0;
          }
          continue;
        }
        const input = await session.prompt(60 * 1e3);
        const userAction = await handleUserInput2(
          input,
          mergeResults(platformResults),
          currentPage,
          config,
          ctx,
          session,
          platforms,
          platformStates,
          platformResults,
          options,
          keyword
        );
        if (userAction.done) return userAction.message;
        if (userAction.nextPage !== void 0) currentPage = userAction.nextPage;
      }
    } catch (error) {
      ctx.logger.error("搜索执行失败:", error);
      return "搜索过程中出错";
    }
  });
}
__name(registerSearch, "registerSearch");

// src/index.ts
var name = "mc-tools";
var inject = { optional: ["puppeteer", "database"] };
var usage = `
<div style="border-radius: 10px; border: 1px solid #ddd; padding: 16px; margin-bottom: 20px; box-shadow: 0 2px 5px rgba(0,0,0,0.1);">
  <h2 style="margin-top: 0; color: #4a6ee0;">📌 插件说明</h2>
  <p>📖 <strong>使用文档</strong>：请点击左上角的 <strong>插件主页</strong> 查看插件使用文档</p>
  <p>🔍 <strong>更多插件</strong>：可访问 <a href="https://github.com/YisRime" style="color:#4a6ee0;text-decoration:none;">苡淞的 GitHub</a> 查看本人的所有插件</p>
</div>

<div style="border-radius: 10px; border: 1px solid #ddd; padding: 16px; margin-bottom: 20px; box-shadow: 0 2px 5px rgba(0,0,0,0.1);">
  <h2 style="margin-top: 0; color: #e0574a;">❤️ 支持与反馈</h2>
  <p>🌟 喜欢这个插件？请在 <a href="https://github.com/YisRime" style="color:#e0574a;text-decoration:none;">GitHub</a> 上给我一个 Star！</p>
  <p>🐛 遇到问题？请通过 <strong>Issues</strong> 提交反馈，或加入 QQ 群 <a href="https://qm.qq.com/q/PdLMx9Jowq" style="color:#e0574a;text-decoration:none;"><strong>855571375</strong></a> 进行交流</p>
</div>
`;
var Config = import_koishi8.Schema.intersect([
  import_koishi8.Schema.object({
    linkParserEnabled: import_koishi8.Schema.union([
      import_koishi8.Schema.const("disable").description("禁用"),
      import_koishi8.Schema.const("text").description("启用"),
      import_koishi8.Schema.const("shot").description("启用（截图）")
    ]).description("启用链接解析").default("disable"),
    mcwikiEnabled: import_koishi8.Schema.boolean().description("启用 MC Wiki 查询").default(true),
    modrinthEnabled: import_koishi8.Schema.boolean().description("启用 Modrinth 查询").default(true),
    mcmodEnabled: import_koishi8.Schema.union([
      import_koishi8.Schema.const(false).description("禁用"),
      import_koishi8.Schema.string().description("启用").role("link").default("https://mcmod-api.yis-rime.workers.dev/")
    ]).description("启用 MC 百科 查询").default("https://mcmod-api.yis-rime.workers.dev/"),
    curseforgeEnabled: import_koishi8.Schema.union([
      import_koishi8.Schema.const(false).description("禁用"),
      import_koishi8.Schema.string().description("启用").role("secret")
    ]).description("启用 CurseForge 查询").default(false)
  }).description("查询开关配置"),
  import_koishi8.Schema.object({
    useForward: import_koishi8.Schema.boolean().description("启用合并转发").default(true),
    useScreenshot: import_koishi8.Schema.boolean().description("启用网页截图").default(true),
    useFallback: import_koishi8.Schema.boolean().description("启用发送回退").default(true),
    maxParagraphs: import_koishi8.Schema.number().description("详情段数限制").default(20).min(1).max(100),
    maxDescLength: import_koishi8.Schema.number().description("每段字数限制").default(512).min(128).max(2048)
  }).description("资源查询配置"),
  import_koishi8.Schema.object({
    playerEnabled: import_koishi8.Schema.boolean().description("启用玩家信息查询").default(true),
    verEnabled: import_koishi8.Schema.boolean().description("启用最新版本查询").default(true),
    updInterval: import_koishi8.Schema.number().description("更新检查间隔(分钟)").default(1).min(1).max(1440),
    noticeTargets: import_koishi8.Schema.array(import_koishi8.Schema.object({
      platform: import_koishi8.Schema.string().description("平台 ID"),
      channelId: import_koishi8.Schema.string().description("频道 ID"),
      type: import_koishi8.Schema.union([
        import_koishi8.Schema.const("release").description("仅正式版"),
        import_koishi8.Schema.const("snapshot").description("仅快照版"),
        import_koishi8.Schema.const("both").description("所有版本")
      ]).description("推送类型").default("both")
    })).description("版本更新推送目标").role("table"),
    statusEnabled: import_koishi8.Schema.boolean().description("启用服务状态查询").default(true),
    statusUpdInterval: import_koishi8.Schema.number().description("状态检查间隔(分钟)").default(5).min(1).max(1440),
    statusNoticeTargets: import_koishi8.Schema.array(import_koishi8.Schema.object({
      platform: import_koishi8.Schema.string().description("平台 ID"),
      channelId: import_koishi8.Schema.string().description("频道 ID")
    })).description("状态变更推送目标").role("table")
  }).description("版本&玩家&状态查询配置"),
  import_koishi8.Schema.object({
    infoEnabled: import_koishi8.Schema.boolean().description("启用服务器查询").default(true),
    serverApis: import_koishi8.Schema.array(import_koishi8.Schema.object({
      type: import_koishi8.Schema.union([
        import_koishi8.Schema.const("java").description("Java版"),
        import_koishi8.Schema.const("bedrock").description("基岩版")
      ]).description("API 类型"),
      url: import_koishi8.Schema.string().description("API URL （使用 ${address} 指代地址）")
    })).description("服务器查询 API ").default([
      { type: "java", url: "https://api.mcstatus.io/v2/status/java/${address}" },
      { type: "bedrock", url: "https://api.mcstatus.io/v2/status/bedrock/${address}" },
      { type: "java", url: "https://api.mcsrvstat.us/3/${address}" },
      { type: "bedrock", url: "https://api.mcsrvstat.us/bedrock/3/${address}" },
      { type: "java", url: "https://api.imlazy.ink/mcapi?type=json&host=${address}" },
      { type: "bedrock", url: "https://api.imlazy.ink/mcapi?type=json&host=${address}&be=true" }
    ]).role("table"),
    serverTemplate: import_koishi8.Schema.string().role("textarea").description("服务器信息模板（使用{...:x}指代数据，数字代表数量限制）").default("{icon}\n{motd}\n{edition}{software} {version} | {online}/{max} | {ping}\n{gamemode} {serverid} {eulablock}\n玩家({playercount}):{playerlist:10}\n插件({plugincount}):{pluginlist:10}\n模组({modcount}):{modlist:10}")
  }).description("服务器查询配置"),
  import_koishi8.Schema.object({
    bindEnabled: import_koishi8.Schema.boolean().description("启用白名单管理").default(false),
    serverMaps: import_koishi8.Schema.array(import_koishi8.Schema.object({
      serverId: import_koishi8.Schema.number().description("服务器 ID").required(),
      platform: import_koishi8.Schema.string().description("平台 ID"),
      channelId: import_koishi8.Schema.string().description("频道 ID"),
      serverAddress: import_koishi8.Schema.string().description("服务器地址")
    })).description("服务器映射群组").default([]).role("table"),
    rconServers: import_koishi8.Schema.array(import_koishi8.Schema.object({
      id: import_koishi8.Schema.number().description("服务器 ID").required(),
      rconAddress: import_koishi8.Schema.string().description("地址").default("localhost:25575"),
      rconPassword: import_koishi8.Schema.string().description("密码").role("secret")
    })).description("RCON 配置").default([]).role("table")
  }).description("服务器连接配置")
]);
function apply(ctx, config) {
  const mc = ctx.command("mc", "Minecraft 工具");
  mc.subcommand(".help", "Minecraft 插件帮助").action(() => `MC-Tools 插件帮助
[详细说明](https://github.com/YisRime/koishi-plugin-mc-tools)`);
  config.verEnabled !== false && registerVer(mc);
  config.noticeTargets?.length && regVerCheck(ctx, config);
  config.statusEnabled !== false && registerStatus(mc);
  config.statusNoticeTargets?.length && regStatusCheck(ctx, config);
  config.playerEnabled !== false && registerPlayer(ctx, mc);
  config.infoEnabled !== false && config.serverApis?.length && registerInfo(mc, config);
  if (config.rconServers.length > 0) registerServer(ctx, mc, config);
  if (config.modrinthEnabled) registerModrinth(ctx, mc, config);
  if (typeof config.curseforgeEnabled === "string" && config.curseforgeEnabled) registerCurseForge(ctx, mc, config);
  if (typeof config.mcmodEnabled === "string" && config.mcmodEnabled) registerMcmod(ctx, mc, config);
  if (config.mcwikiEnabled) registerMcwiki(ctx, mc, config);
  if (config.linkParserEnabled !== "disable") registerLinkParser(ctx, config);
  if (config.mcmodEnabled || config.mcwikiEnabled || config.modrinthEnabled || config.curseforgeEnabled) registerSearch(ctx, mc, config);
}
__name(apply, "apply");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Config,
  apply,
  inject,
  name,
  usage
});
