// koishi-plugin-word-core-grammar-basic/lib/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Config: () => Config,
  apply: () => apply,
  inject: () => inject,
  name: () => name
});
module.exports = __toCommonJS(src_exports);
var import_koishi = require("koishi");

// src/tools/index.ts
var numberToChinese = /* @__PURE__ */ __name((num) => {
  const units = ["", "十", "百", "千", "万", "亿", "兆"];
  const chars = ["零", "一", "二", "三", "四", "五", "六", "七", "八", "九"];
  if (num === 0) return "零";
  let str = "";
  let unitPos = 0;
  let zero = true;
  while (num > 0) {
    const digit = num % 10;
    if (digit === 0) {
      if (!zero) {
        zero = true;
        str = chars[digit] + str;
      }
    } else {
      zero = false;
      str = chars[digit] + (unitPos > 0 ? units[unitPos] : "") + str;
    }
    unitPos++;
    num = Math.floor(num / 10);
  }
  str = str.replace(/^一十/, "十");
  return str;
}, "numberToChinese");
var isNumeric = /* @__PURE__ */ __name((str) => {
  return !isNaN(Number(str)) && str.trim() !== "";
}, "isNumeric");
var randomNumber = /* @__PURE__ */ __name((minNumber, maxNumber) => {
  return Math.floor(Math.random() * (maxNumber - minNumber + 1)) + minNumber;
}, "randomNumber");

// src/index.ts
var name = "word-core-grammar-basic";
var Config = import_koishi.Schema.object({
  Leaderboard: import_koishi.Schema.number().default(10).description("排行榜最大显示数量"),
  httpRequest: import_koishi.Schema.boolean().default(true).description("是否启用http请求")
});
var inject = ["word"];
function apply(ctx, config) {
  ctx.word.statement.addStatement("+", async (inData2, session2) => {
    const saveCell2 = inData2.wordData.saveDB;
    let uid2 = inData2.args.length >= 3 ? inData2.args[2] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const item2 = inData2.args[0];
    const number2 = await inData2.internal.getItem(uid2, saveCell2, item2);
    if (typeof number2 != "number") {
      return number2;
    }
    const addNumTemp = inData2.args[1];
    let addNum = 0;
    if (!/^\d+$/.test(addNumTemp) && !/^\d+\~\d+$/.test(addNumTemp) && !/^\d+%$/.test(addNumTemp)) {
      throw `物品 [${item2}] 添加的数量 [${addNum}] 不为数字或标识`;
    }
    switch (true) {
      case /^\d+$/.test(addNumTemp): {
        addNum = Number(addNumTemp);
        break;
      }
      case /^\d+\~\d+$/.test(addNumTemp): {
        const matchData = addNumTemp.match(/^(\d+)\~(\d+)$/);
        addNum = randomNumber(Number(matchData[1]), Number(matchData[2]));
        break;
      }
      case /^\d+%$/.test(addNumTemp): {
        const matchData = addNumTemp.match(/^(\d+)%$/);
        addNum = number2 * Number(matchData[1]) / 100;
        break;
      }
      default: {
        break;
      }
    }
    const now = number2 + Number(addNum);
    await inData2.internal.saveItem(uid2, saveCell2, item2, Math.floor(now));
    return String(Math.floor(addNum));
  });
  ctx.word.statement.addStatement("-", async (inData2, session2) => {
    const saveCell2 = inData2.wordData.saveDB;
    let uid2 = inData2.args.length >= 3 ? inData2.args[2] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const item2 = inData2.args[0];
    const number2 = await inData2.internal.getItem(uid2, saveCell2, item2);
    if (typeof number2 != "number") {
      return number2;
    }
    const rmNumTemp = inData2.args[1];
    let rmNum = 0;
    if (!/^\d+$/.test(rmNumTemp) && !/^\d+\~\d+$/.test(rmNumTemp) && !/^\d+%$/.test(rmNumTemp)) {
      throw `物品 [${item2}] 添加的数量 [${rmNum}] 不为数字或标识`;
    }
    switch (true) {
      case /^\d+$/.test(rmNumTemp): {
        rmNum = Number(rmNumTemp);
        break;
      }
      case /^\d+\~\d+$/.test(rmNumTemp): {
        const matchData = rmNumTemp.match(/^(\d+)\~(\d+)$/);
        rmNum = randomNumber(Number(matchData[1]), Number(matchData[2]));
        break;
      }
      case /^\d+%$/.test(rmNumTemp): {
        const matchData = rmNumTemp.match(/^(\d+)%$/);
        rmNum = number2 * Number(matchData[1]) / 100;
        break;
      }
      default: {
        break;
      }
    }
    const now = number2 - Number(rmNum);
    if (now < 0) {
      return inData2.parPack.kill(`物品 [${item2}] 数量不足`);
    }
    await inData2.internal.saveItem(uid2, saveCell2, item2, Math.floor(now));
    return String(Math.floor(rmNum));
  });
  ctx.word.statement.addStatement("?", async (inData, session) => {
    const saveCell = inData.wordData.saveDB;
    let uid = inData.args.length >= 5 ? inData.args[4] : session.userId;
    if (uid == "that") {
      uid = inData.matchs.id[0];
    }
    const item = inData.args[0];
    let number;
    if (/^\d+$/.test(item)) {
      number = Number(item);
    } else {
      const aaaa = await inData.internal.getItem(uid, saveCell, item);
      if (typeof aaaa != "number") {
        return number;
      }
      number = aaaa;
    }
    number = number ? number : 0;
    const inputNumber = inData.args[2];
    if (!/^\d+$/.test(inputNumber)) {
      return inData.parPack.kill(`数量 [${inputNumber}] 不为数字`);
    }
    const relationship = inData.args[1];
    if (relationship == "==" || relationship == ">" || relationship == "<" || relationship == "!=" || relationship == ">=" || relationship == "<=") {
    } else {
      return inData.parPack.kill(`数量 [${relationship}] 不符合预期`);
    }
    if (eval(`${number}${relationship}${inputNumber}`)) {
      if (inData.args.length >= 4) {
        if (inData.args[3] == "") {
          return "";
        } else {
          return inData.args[3];
        }
      } else {
        return "";
      }
    } else {
      if (inData.args.length >= 4) {
        if (inData.args[3] == "") {
          return inData.parPack.next();
        } else {
          if (inData.args.length >= 6) {
            return inData.args[5];
          } else {
            return "";
          }
        }
      } else {
        return inData.parPack.next();
      }
    }
  }, [0, 1, 1, 1, 0, 1, 0]);
  ctx.word.statement.addStatement("&", async (inData2, session2) => {
    if (!/^\d+$/.test(inData2.args[0])) {
      return inData2.parPack.kill("时间格式错误");
    }
    await (0, import_koishi.sleep)(Number(inData2.args[0]) * 1e3);
    if (inData2.args.length > 1) {
      return inData2.args[1];
    } else {
      return "";
    }
  });
  ctx.word.statement.addStatement("#", async (inData2, session2) => {
    let uid2 = inData2.args.length >= 2 ? inData2.args[1] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const saveCell2 = inData2.wordData.saveDB;
    const item2 = inData2.args[0];
    let number2 = await inData2.internal.getItem(uid2, saveCell2, item2);
    number2 = number2 ? number2 : 0;
    return String(number2);
  });
  ctx.word.statement.addStatement("%", async (inData2, session2) => {
    if (!/^\d+$/.test(inData2.args[0]) || Number(inData2.args[0]) < 0 || Number(inData2.args[0]) > 100) {
      return inData2.parPack.kill("概率格式错误");
    }
    const random = /* @__PURE__ */ __name((minNumber, maxNumber) => {
      return Math.floor(Math.random() * (maxNumber - minNumber + 1)) + minNumber;
    }, "random");
    const randomNumber2 = random(0, 100);
    let msg = "";
    if (inData2.args.length > 1) {
      msg = inData2.args[1];
    }
    if (Number(inData2.args[0]) > randomNumber2) {
      return msg;
    } else {
      if (inData2.args.length > 1) {
        return "";
      } else {
        return inData2.parPack.kill("判定失败");
      }
    }
  }, [0, 1]);
  ctx.word.statement.addStatement("@this", async (inData2, session2) => {
    return session2.username;
  });
  ctx.word.statement.addStatement("#this", async (inData2, session2) => {
    return session2.userId;
  });
  ctx.word.statement.addStatement("@that", async (inData2, session2) => {
    if (!inData2.matchs.hasOwnProperty("name")) {
      return;
    }
    return inData2.matchs.name[0];
  });
  ctx.word.statement.addStatement("#that", async (inData2, session2) => {
    if (!inData2.matchs.hasOwnProperty("id")) {
      return;
    }
    return inData2.matchs.id[0];
  });
  ctx.word.statement.addStatement("@", async (inData2, session2) => {
    const type = inData2.args[0];
    const at = inData2.args[1];
    if (!at) {
      return "创建at时出错：未填写昵称或id";
    }
    if (type == "id") {
      return `<at id="${at}"/>`;
    } else if (type == "name") {
      return `<at name="${at}"/>`;
    } else {
      return "创建at时出错：未填写类型或类型出错";
    }
  });
  ctx.word.statement.addStatement("!", async (inData2, session2) => {
    return "";
  });
  ctx.word.statement.addStatement("cd", async (inData2, session2) => {
    const uid2 = session2.userId;
    const eventName = inData2.args[0];
    const time = inData2.args[1];
    if (!/^\d+$/.test(time)) {
      return inData2.parPack.kill("cd时间输入错误");
    }
    let userEventConfig = await inData2.internal.getUserConfig(uid2, eventName);
    if (!userEventConfig) {
      userEventConfig = Number(time) * 1e3 + Date.now();
      inData2.internal.saveUserConfig(uid2, eventName, userEventConfig);
      if (inData2.args.length > 2) {
        return inData2.args[2];
      } else {
        return "";
      }
    } else {
      if (Date.now() >= userEventConfig) {
        userEventConfig = Number(time) * 1e3 + Date.now();
        inData2.internal.saveUserConfig(uid2, eventName, userEventConfig);
        if (inData2.args.length > 2) {
          return inData2.args[2];
        } else {
          return "";
        }
      } else {
        return inData2.parPack.kill();
      }
    }
  }, [0, 1, 1]);
  if (!ctx.word.trigger.trigger["(数)"]) {
    ctx.word.trigger.addTrigger("inputNumber", "(数)", "(\\d+)+?");
    ctx.word.cache.cacheRefresh();
  }
  ctx.word.statement.addStatement("数", async (inData2, session2) => {
    const inputNumber2 = inData2.args[0];
    if (!/^\d+$/.test(inputNumber2)) {
      return inData2.parPack.kill("获取输入数的输入参数不正确");
    }
    const data = inData2.matchs.inputNumber[Number(inputNumber2) - 1];
    if (!data) {
      return inData2.parPack.kill("未获取到输入的数");
    }
    return data;
  });
  ctx.word.statement.addStatement("算", async (inData, session) => {
    const numArgs1 = inData.args[0];
    const numArgs2 = inData.args[2];
    const Operator = inData.args[1];
    if (!/^\d+$/.test(numArgs1)) {
      return inData.parPack.kill("获取输入数的输入参数不正确");
    }
    if (!/^\d+$/.test(numArgs2)) {
      return inData.parPack.kill("获取输入数的输入参数不正确");
    }
    if (!/[+-/*]/.test(Operator)) {
      return inData.parPack.kill("获取输入数的输入参数不正确");
    }
    const MAX_SAFE = Number.MAX_SAFE_INTEGER;
    const MIN_SAFE = Number.MIN_SAFE_INTEGER;
    if (numArgs1 > MAX_SAFE || numArgs1 < MIN_SAFE) {
      return inData.parPack.kill(`Error: number 1 (${numArgs1}) is out of safe range.`);
    }
    if (numArgs2 > MAX_SAFE || numArgs2 < MIN_SAFE) {
      return inData.parPack.kill(`Error: number 2 (${numArgs2}) is out of safe range.`);
    }
    const result = Math.floor(eval(`${numArgs1}${Operator}${numArgs2}`));
    if (result > MAX_SAFE || result < MIN_SAFE) {
      return inData.parPack.kill(`Error: result (${result}) is out of safe range.`);
    }
    return String(result);
  });
  ctx.word.statement.addStatement("~", async (inData2, session2) => {
    const first = inData2.args[0];
    const second = inData2.args[1];
    if (!/^\d+$/.test(first)) {
      return inData2.parPack.kill("获取输入数的输入参数不正确");
    }
    if (!/^\d+$/.test(second)) {
      return inData2.parPack.kill("获取输入数的输入参数不正确");
    }
    const random = /* @__PURE__ */ __name((minNumber, maxNumber) => {
      return Math.floor(Math.random() * (maxNumber - minNumber + 1)) + minNumber;
    }, "random");
    return String(random(Number(first), Number(second)));
  });
  ctx.word.statement.addStatement("time", async (inData2, session2) => {
    const first = inData2.args[0];
    const isString = inData2.args[1];
    if (!/^\d+$/.test(first)) {
      return inData2.parPack.kill("获取输入数的输入参数不正确");
    }
    const day = /* @__PURE__ */ new Date();
    let time;
    switch (first) {
      case "1": {
        time = day.getFullYear();
        break;
      }
      case "2": {
        time = day.getMonth() + 1;
        break;
      }
      case "3": {
        time = day.getDay();
        if (time == 0) {
          time = 7;
        }
        break;
      }
      case "4": {
        time = day.getDate();
        break;
      }
      case "5": {
        time = day.getHours();
        break;
      }
      case "6": {
        time = day.getMinutes();
        break;
      }
      case "7": {
        time = day.getSeconds();
        break;
      }
      case "8": {
        time = Date.now();
        break;
      }
      case "9": {
        time = Date();
        break;
      }
      default:
        break;
    }
    if (time) {
      if (isString == "1" && first != "9") {
        return numberToChinese(time);
      } else {
        return String(time);
      }
    } else {
      return "";
    }
  });
  ctx.word.statement.addStatement("调", async (inData2, session2) => {
    const whichStart = inData2.args[0];
    if (session2.content == whichStart) {
      return inData2.parPack.kill("禁止调用自身");
    }
    if (!session2.bot) {
      ctx.logger("调用语法中session缺少bot属性");
      return "【当前触发方式不支持调用语法】";
    }
    if (!session2.send) {
      ctx.logger("调用语法中session缺少send属性");
      return "【当前触发方式不支持调用语法】";
    }
    if (!session2.event) {
      ctx.logger("调用语法中session缺少event属性");
      return "【当前触发方式不支持调用语法】";
    }
    const test = session2?.bot.session((0, import_koishi.clone)(session2.event));
    test.content = whichStart;
    let getReturnMsg = "";
    await ctx.word.driver.start(test, async (msg) => {
      if (!msg) {
        return "";
      }
      getReturnMsg = msg;
    });
    return getReturnMsg;
  });
  ctx.word.statement.addStatement("异调", async (inData2, session2) => {
    const whichStart = inData2.args[0];
    if (session2.content == whichStart) {
      return inData2.parPack.kill("禁止调用自身");
    }
    if (!session2.bot) {
      ctx.logger("调用语法中session缺少bot属性");
      return "【当前触发方式不支持调用语法】";
    }
    if (!session2.send) {
      ctx.logger("调用语法中session缺少send属性");
      return "【当前触发方式不支持调用语法】";
    }
    if (!session2.event) {
      ctx.logger("调用语法中session缺少event属性");
      return "【当前触发方式不支持调用语法】";
    }
    const test = session2?.bot.session(session2.event);
    test.content = whichStart;
    await ctx.word.driver.start(test, async (msg) => {
      if (!msg) {
        return "";
      }
      test.send(msg);
    });
    return "";
  });
  ctx.word.statement.addStatement("wi", async (inData2, session2) => {
    if (!session2.prompt) {
      ctx.logger("wi语法中session缺少prompt属性");
      return "当前不支持获取输入语法";
    }
    if (!session2.send) {
      ctx.logger("wi语法中session缺少send属性");
      return "当前不支持获取输入语法";
    }
    session2.send(`请输入${inData2.args[0] ? inData2.args[0] + "的值" : ""}:`);
    const a = await session2?.prompt();
    if (!a) {
      return "";
    }
    return a;
  });
  ctx.word.statement.addStatement("p", async (inData2, session2) => {
    const permission = inData2.args[0];
    let uid2 = session2.userId;
    const have = await ctx.word.permission.isHave(uid2, permission);
    const msg = inData2.args[1] ? inData2.args[1] : "";
    if (msg == "") {
      if (have) {
        return "";
      } else {
        return inData2.parPack.next();
      }
    } else {
      if (have) {
        return msg;
      } else {
        return "";
      }
    }
  });
  ctx.word.statement.addStatement("排行榜", async (inData2, session2) => {
    const itemName = inData2.args[0];
    const saveCell2 = inData2.wordData.saveDB;
    const a = await ctx.word.tools.getDB("wordUserPackData");
    const userList = a.idList;
    const dataList = a.dataList;
    const dataTempList = [];
    const userTempList = [];
    dataList.forEach((item2, index) => {
      if (!item2[saveCell2]) {
        return;
      }
      if (!item2[saveCell2][itemName]) {
        return;
      }
      dataTempList.push(item2[saveCell2][itemName]);
      userTempList.push(userList[index]);
    });
    const outDataList = [].concat(dataTempList).sort((a2, b) => {
      return b - a2;
    });
    const outUserList = [];
    let outMsg = "";
    outDataList.forEach((v, nowIndex) => {
      const index = dataTempList.indexOf(v);
      outUserList.push(userTempList[index]);
      if (nowIndex < config.Leaderboard) {
        outMsg += `${nowIndex + 1}. <at id="${userTempList[index]}"/>  ${v}
`;
      }
      dataTempList.splice(index, 1);
      userTempList.splice(index, 1);
    });
    return outMsg;
  });
  ctx.word.statement.addStatement("歌曲", async (inData2, session2) => {
    const url = inData2.args[0];
    if (!url) {
      return inData2.parPack.kill("无歌曲链接参数");
    }
    return `<audio src="${url}" url="${url}"/>`;
  });
  ctx.word.statement.addStatement("视频", async (inData2, session2) => {
    const url = inData2.args[0];
    if (!url) {
      return inData2.parPack.kill("无歌曲链接参数");
    }
    return `<video src="${url}" url="${url}"/>`;
  });
  ctx.word.statement.addStatement("禁言", async (inData2, session2) => {
    if (!session2.hasOwnProperty("guildId")) {
      return "当前不支持获取输入语法";
    }
    let uid2 = inData2.args.length >= 2 ? inData2.args[1] : session2.userId;
    let long = inData2.args[0];
    if (!/^\d+$/.test(long)) {
      return inData2.parPack.kill("获取禁言时常的输入参数不正确");
    }
    let reason = inData2.args[2];
    if (reason) {
      session2.bot.muteGuildMember(session2.guildId, uid2, long, reason);
    } else {
      session2.bot.muteGuildMember(session2.guildId, uid2, long);
    }
  });
  ctx.word.statement.addStatement("踢", async (inData2, session2) => {
    if (!session2.hasOwnProperty("guildId")) {
      return "当前不支持获取输入语法";
    }
    let uid2 = inData2.args.length >= 0 ? inData2.args[0] : session2.userId;
    let long = inData2.args[1] ? inData2.args[1] : 0;
    if (!/^\d+$/.test(long)) {
      return inData2.parPack.kill("获取禁言时常的输入参数不正确");
    }
    if (long == 0) {
      session2.bot.kickGuildMember(session2.guildId, uid2, false);
    } else {
      session2.bot.kickGuildMember(session2.guildId, uid2, true);
    }
  });
  const httpRequest = /* @__PURE__ */ __name(async (url, method, head, body) => {
    if (method == "post") {
      const config2 = {
        method,
        headers: head,
        data: body
      };
      if (!config2.headers) {
        delete config2.headers;
      }
      if (!config2.data) {
        delete config2.data;
      }
      const data = await ctx.http("post", url, config2);
      return data;
    }
    if (method == "get") {
      const reqUrl = `${url}?${new URLSearchParams(body)}`;
      const data = head ? await ctx.http("get", reqUrl, {
        headers: head
      }) : await ctx.http(reqUrl);
      return data;
    }
  }, "httpRequest");
  const getJson = /* @__PURE__ */ __name((originStr) => {
    const temp = originStr.split("&");
    const json = {};
    temp.forEach((a) => {
      const list = a.split("=");
      const key = list[0];
      const value = list[1];
      json[key] = value;
    });
    return json;
  }, "getJson");
  const dataSplit = /* @__PURE__ */ __name((data, needSplit) => {
    const splitList = needSplit.split(".");
    let returnMsg = data;
    for (let i of splitList) {
      if (returnMsg[i]) {
        returnMsg = returnMsg[i];
      } else {
        return needSplit;
      }
    }
    return returnMsg;
  }, "dataSplit");
  ctx.word.statement.addStatement("get", async (inData2, session2) => {
    if (!config.httpRequest) {
      return;
    }
    const url = `http://${inData2.args[0]}`;
    const headTemp = inData2.args[1];
    const bodyTemp = inData2.args[2];
    let head = null;
    if (headTemp != "") {
      head = getJson(headTemp);
    }
    let body = null;
    if (body != "") {
      body = getJson(bodyTemp);
    }
    const data = await httpRequest(url, "get", head, body);
    const needDataList = inData2.args.slice(3);
    let outMsg = "";
    needDataList.forEach((e) => {
      outMsg += String(dataSplit(structuredClone(data), e));
    });
    return outMsg;
  });
  ctx.word.statement.addStatement("gets", async (inData2, session2) => {
    if (!config.httpRequest) {
      return;
    }
    const url = `https://${inData2.args[0]}`;
    const headTemp = inData2.args[1];
    const bodyTemp = inData2.args[2];
    let head = null;
    if (headTemp != "") {
      head = getJson(headTemp);
    }
    let body = null;
    if (body != "") {
      body = getJson(bodyTemp);
    }
    const data = await httpRequest(url, "get", head, body);
    const needDataList = inData2.args.slice(3);
    let outMsg = "";
    needDataList.forEach((e) => {
      outMsg += String(dataSplit(structuredClone(data), e));
    });
    return outMsg;
  });
  ctx.word.statement.addStatement("posts", async (inData2, session2) => {
    if (!config.httpRequest) {
      return;
    }
    const url = `https://${inData2.args[0]}`;
    const headTemp = inData2.args[1];
    const bodyTemp = inData2.args[2];
    let head = null;
    if (headTemp != "") {
      head = getJson(headTemp);
    }
    let body = null;
    if (body != "") {
      body = getJson(bodyTemp);
    }
    const data = await httpRequest(url, "post", head, body);
    const needDataList = inData2.args.slice(3);
    let outMsg = "";
    needDataList.forEach((e) => {
      outMsg += String(dataSplit(structuredClone(data), e));
    });
    return outMsg;
  });
  ctx.word.statement.addStatement("post", async (inData2, session2) => {
    if (!config.httpRequest) {
      return;
    }
    const url = `http://${inData2.args[0]}`;
    const headTemp = inData2.args[1];
    const bodyTemp = inData2.args[2];
    let head = null;
    if (headTemp != "") {
      head = getJson(headTemp);
    }
    let body = null;
    if (body != "") {
      body = getJson(bodyTemp);
    }
    const data = await httpRequest(url, "post", head, body);
    const needDataList = inData2.args.slice(3);
    let outMsg = "";
    needDataList.forEach((e) => {
      outMsg += String(dataSplit(structuredClone(data), e));
    });
    return outMsg;
  });
  ctx.word.statement.addStatement("http", async (inData2, session2) => {
    if (!config.httpRequest) {
      return;
    }
    const url = inData2.args[0];
    const method = inData2.args[1];
    const headTemp = inData2.args[2];
    const bodyTemp = inData2.args[3];
    if (method != "post" && method != "get") {
      return;
    }
    let head = null;
    if (headTemp != "") {
      head = getJson(headTemp);
    }
    let body = null;
    if (body != "") {
      body = getJson(bodyTemp);
    }
    const data = await httpRequest(url, method, head, body);
    const needDataList = inData2.args.slice(4);
    let outMsg = "";
    needDataList.forEach((e) => {
      outMsg += String(dataSplit(structuredClone(data), e));
    });
    return outMsg;
  });
  ctx.word.statement.addStatement("%%", async (inData2, session2) => {
    const args = inData2.args;
    if (args.length % 2 == 1) {
      return "参数数量为奇数";
    }
    const obj = [];
    for (let i = 0; i < args.length; i += 2) {
      if (!/^\d+$/.test(args[i])) {
        return `${i + 1}个参数不为整数`;
      }
      obj.push({
        value: args[i + 1],
        weight: Number(args[i])
      });
    }
    function weightedRandom(items) {
      const totalWeight = items.reduce((sum, item2) => sum + item2.weight, 0);
      const randomNum = Math.random() * totalWeight;
      let weightSum = 0;
      for (const item2 of items) {
        weightSum += item2.weight;
        if (randomNum < weightSum) {
          return item2.value;
        }
      }
    }
    __name(weightedRandom, "weightedRandom");
    return weightedRandom(obj);
  }, ["while", 0, 1]);
  ctx.word.statement.addStatement("a+", async (inData2, session2) => {
    const listName = inData2.args[0];
    const listItem = inData2.args[1];
    let uid2 = inData2.args.length >= 3 ? inData2.args[2] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const saveCell2 = inData2.wordData.saveDB;
    const getListData = await inData2.internal.getList(uid2, saveCell2, listName);
    if (!Array.isArray(getListData)) {
      return getListData;
    }
    getListData.push(listItem);
    const a = await inData2.internal.saveList(uid2, saveCell2, listName, getListData);
    if (a) {
      return listItem;
    }
  });
  ctx.word.statement.addStatement("a-", async (inData2, session2) => {
    const listName = inData2.args[0];
    const listItem = inData2.args[1];
    let uid2 = inData2.args.length >= 3 ? inData2.args[2] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const saveCell2 = inData2.wordData.saveDB;
    const getListData = await inData2.internal.getList(uid2, saveCell2, listName);
    if (!Array.isArray(getListData)) {
      return getListData;
    }
    const index = getListData.indexOf(listItem);
    if (index < 0) {
      return inData2.parPack.kill("不存在此列表项");
    }
    const item2 = getListData.splice(index, 1);
    const a = await inData2.internal.saveList(uid2, saveCell2, listName, getListData);
    if (a) {
      return item2[0];
    }
  });
  ctx.word.statement.addStatement("ak", async (inData2, session2) => {
    const listName = inData2.args[0];
    let uid2 = inData2.args.length >= 2 ? inData2.args[1] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const saveCell2 = inData2.wordData.saveDB;
    const getListData = [];
    await inData2.internal.saveList(uid2, saveCell2, listName, getListData);
  });
  ctx.word.statement.addStatement("al", async (inData2, session2) => {
    const listName = inData2.args[0];
    let uid2 = inData2.args.length >= 2 ? inData2.args[1] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const saveCell2 = inData2.wordData.saveDB;
    const getListData = await inData2.internal.getList(uid2, saveCell2, listName);
    if (!Array.isArray(getListData)) {
      return getListData;
    }
    return String(getListData.length);
  });
  ctx.word.statement.addStatement("a#", async (inData2, session2) => {
    const listName = inData2.args[0];
    const index = inData2.args[1];
    if (!isNumeric(index)) {
      return inData2.parPack.kill("序号不为数字");
    }
    if (index == "0") {
      return inData2.parPack.kill("序号需要大于1");
    }
    let uid2 = inData2.args.length >= 3 ? inData2.args[2] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const saveCell2 = inData2.wordData.saveDB;
    const getListData = await inData2.internal.getList(uid2, saveCell2, listName);
    if (!Array.isArray(getListData)) {
      return getListData;
    }
    if (getListData.length <= 0) {
      return inData2.parPack.kill("不存在此列表项或列表为空");
    }
    return getListData[Number(index) - 1];
  });
  ctx.word.statement.addStatement("ar", async (inData2, session2) => {
    const listName = inData2.args[0];
    let uid2 = inData2.args.length >= 2 ? inData2.args[1] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const saveCell2 = inData2.wordData.saveDB;
    const getListData = await inData2.internal.getList(uid2, saveCell2, listName);
    if (!Array.isArray(getListData)) {
      return getListData;
    }
    if (getListData.length <= 0) {
      return inData2.parPack.kill("不存在此列表项或列表为空");
    }
    const index = randomNumber(0, getListData.length);
    return getListData[index];
  });
  ctx.word.statement.addStatement("a?", async (inData2, session2) => {
    const listName = inData2.args[0];
    const listItem = inData2.args[1];
    let uid2 = inData2.args.length >= 3 ? inData2.args[2] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const saveCell2 = inData2.wordData.saveDB;
    const getListData = await inData2.internal.getList(uid2, saveCell2, listName);
    if (!Array.isArray(getListData)) {
      return getListData;
    }
    const index = getListData.indexOf(listItem);
    if (index < 0) {
      return "-1";
    }
    return String(index) + 1;
  });
  ctx.word.statement.addStatement("as", async (inData2, session2) => {
    const listName = inData2.args[0];
    const index = inData2.args[1];
    const listItem = inData2.args[2];
    if (!isNumeric(index)) {
      return inData2.parPack.kill("序号不为数字");
    }
    if (index == "0") {
      return inData2.parPack.kill("序号需要大于1");
    }
    let uid2 = inData2.args.length >= 4 ? inData2.args[3] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const saveCell2 = inData2.wordData.saveDB;
    const getListData = await inData2.internal.getList(uid2, saveCell2, listName);
    if (!Array.isArray(getListData)) {
      return getListData;
    }
    getListData[Number(index) - 1] = listItem;
    const a = await inData2.internal.saveList(uid2, saveCell2, listName, getListData);
    if (a) {
      return listItem;
    }
  });
  ctx.word.statement.addStatement("aa", async (inData2, session2) => {
    const listName = inData2.args[0];
    let uid2 = inData2.args.length >= 2 ? inData2.args[1] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const saveCell2 = inData2.wordData.saveDB;
    const getListData = await inData2.internal.getList(uid2, saveCell2, listName);
    if (!Array.isArray(getListData)) {
      return getListData;
    }
    if (getListData.length <= 0) {
      return inData2.parPack.kill("不存在此列表项或列表为空");
    }
    let outMsg = getListData.map((item2, index) => `${index + 1}. ${item2}`).join("");
    return outMsg;
  });
  ctx.word.statement.addStatement("ac", async (inData2, session2) => {
    const listName = inData2.args[0];
    let uid2 = inData2.args.length >= 2 ? inData2.args[1] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const saveCell2 = inData2.wordData.saveDB;
    let getListData = await inData2.internal.getList(uid2, saveCell2, listName);
    if (!Array.isArray(getListData)) {
      return getListData;
    }
    if (getListData.length <= 0) {
      return inData2.parPack.kill("不存在此列表项或列表为空");
    }
    getListData = [...new Set(getListData)];
    const a = await inData2.internal.saveList(uid2, saveCell2, listName, getListData);
    if (!a) {
      return inData2.parPack.kill("列表去重失败");
    }
  });
  ctx.word.statement.addStatement("ah", async (inData2, session2) => {
    const listName = inData2.args[0];
    let uid2 = inData2.args.length >= 4 ? inData2.args[3] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const saveCell2 = inData2.wordData.saveDB;
    let getListData = await inData2.internal.getList(uid2, saveCell2, listName);
    if (!Array.isArray(getListData)) {
      return getListData;
    }
    if (getListData.length <= 0) {
      return inData2.parPack.kill("不存在此列表项或列表为空");
    }
    const index1 = inData2.args[1] - 1;
    const index2 = inData2.args[2] - 1;
    const temp = getListData[index1] + "";
    getListData[index1] = getListData[index2];
    getListData[index2] = temp;
    const a = await inData2.internal.saveList(uid2, saveCell2, listName, getListData);
    if (!a) {
      return inData2.parPack.kill("列表互换失败");
    }
  });
  ctx.word.statement.addStatement("am", async (inData2, session2) => {
    const listName = inData2.args[0];
    let uid2 = inData2.args.length >= 2 ? inData2.args[1] : session2.userId;
    if (uid2 == "that") {
      uid2 = inData2.matchs.id[0];
    }
    const saveCell2 = inData2.wordData.saveDB;
    let getListData = await inData2.internal.getList(uid2, saveCell2, listName);
    if (!Array.isArray(getListData)) {
      return getListData;
    }
    return getListData.join("");
  });
  ctx.word.statement.addStatement("指令", async (inData2, session2) => {
    const args = inData2.args[0];
    let a;
    const off = ctx.once("before-send", (sessionA) => {
      a = sessionA.content;
      return true;
    });
    if (!session2.execute) {
      ctx.logger("调用语法中session缺少execute属性");
      return "【当前触发方式不支持调用语法】";
    }
    if (!session2.send) {
      ctx.logger("调用语法中session缺少send属性");
      return "【当前触发方式不支持调用语法】";
    }
    const send = inData2.args[1];
    if (send == "1") {
      const hList = await session2.execute(args, true);
      off();
      session2.send(hList);
    } else {
      await session2.execute(args);
      return a;
    }
  });
}
__name(apply, "apply");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Config,
  apply,
  inject,
  name
});
