// koishi-plugin-emit-blacklist/lib/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Config: () => Config,
  apply: () => apply,
  name: () => name
});
module.exports = __toCommonJS(src_exports);
var import_koishi = require("koishi");
var name = "emit-blacklist";
var Config = import_koishi.Schema.object({
  enabled: import_koishi.Schema.boolean().default(true).description("是否启用用户黑名单插件。"),
  // 使用 Schema.array(Schema.object({...})) 来定义对象列表
  userList: import_koishi.Schema.array(import_koishi.Schema.object({
    identifier: import_koishi.Schema.string().required().description("需要拉黑的用户，格式为 `平台:用户ID`。"),
    remark: import_koishi.Schema.string().description("备注")
  })).role("table").description("用户黑名单列表。").default([{
    identifier: "onebot:2854196310",
    remark: "Q群管家"
  }])
});
var ORIGINAL_EMIT = Symbol.for("koishi-plugin-emit-blacklist.originalEmit");
function apply(ctx, config) {
  const logger = ctx.logger(name);
  const app = ctx.app;
  if (!config.enabled) {
    if (app[ORIGINAL_EMIT]) {
      app.emit = app[ORIGINAL_EMIT];
      delete app[ORIGINAL_EMIT];
      logger.info("插件已禁用，已恢复原始的 emit 方法。");
    }
    return;
  }
  if (app[ORIGINAL_EMIT]) {
    logger.warn("检测到 emit 方法已被修补，本次加载将跳过。");
    return;
  }
  app[ORIGINAL_EMIT] = app.emit;
  const newEmit = /* @__PURE__ */ __name(function(...allArgs) {
    const original = app[ORIGINAL_EMIT];
    let session = void 0;
    let name2;
    if (typeof allArgs[0] === "string") {
      name2 = allArgs[0];
    } else {
      session = allArgs[0];
      name2 = allArgs[1];
    }
    const targetEvents = ["message", "message-created"];
    if (session?.userId && session?.platform && targetEvents.includes(name2)) {
      const userIdentifier = `${session.platform}:${session.userId}`;
      const isBlacklisted = config.userList.some((user) => user.identifier === userIdentifier);
      if (isBlacklisted) {
        logger.warn(`[黑名单] 在 "${name2}" 事件处拦截到用户: "${userIdentifier}"`);
        return;
      }
    }
    return original.apply(this, allArgs);
  }, "newEmit");
  app.emit = newEmit;
  logger.info("用户黑名单插件已加载，并成功应用 emit 拦截器。");
  ctx.on("dispose", () => {
    if (app[ORIGINAL_EMIT]) {
      app.emit = app[ORIGINAL_EMIT];
      delete app[ORIGINAL_EMIT];
      logger.info("插件已卸载，已成功恢复原始的 emit 方法。");
    }
  });
}
__name(apply, "apply");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Config,
  apply,
  name
});
