// @quanhuzeyu/koishi-plugin-qhzy-sharp/lib/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Config: () => Config,
  SharpService: () => SharpService,
  apply: () => apply,
  filter: () => filter,
  inject: () => inject,
  name: () => name,
  usage: () => usage
});
module.exports = __toCommonJS(src_exports);
var import_koishi = require("koishi");
var import_path = __toESM(require("path"));
var _srcDir = import_path.default.resolve(__dirname);
var name = "qhzy-sharp";
var usage = `该服务通过w-node实现sharp，具体请先了解w-node用法`;
var filter = false;
var inject = ["node"];
var Config = import_koishi.Schema.object({
  nodeBinaryPath: import_koishi.Schema.path({
    filters: ["directory"],
    allowCreate: true
  }).description("sharp 二进制文件路径").default("data/assets/qhzy/sharp"),
  timeout: import_koishi.Schema.number().default(6e4).description("超时时间(ms)"),
  sharpV: import_koishi.Schema.string().description("指定sharp的版本").default("0.33.5")
});
var SharpService = class extends import_koishi.Service {
  static {
    __name(this, "SharpService");
  }
  Sharp;
  tmpDir = import_path.default.resolve(this.ctx.baseDir, "data/assets/qhzy/sharp/tmp");
  sharpV = "0.33.5";
  nodeBinaryPath = "data/assets/qhzy/sharp";
  constructor(ctx, config) {
    super(ctx, "QhzySharp");
    this.config = {
      nodeBinaryPath: "data/assets/qhzy/sharp",
      timeout: 6e4,
      sharpV: "0.33.5",
      ...config
    };
    this.tmpDir = import_path.default.resolve(ctx.baseDir, "data/assets/qhzy/sharp/tmp");
    this.sharpV = this.config.sharpV;
    this.nodeBinaryPath = this.config.nodeBinaryPath;
  }
  async start() {
    const logger = this.ctx.logger;
    await this.ctx.node.safeImport(`sharp`).then((sharp) => {
      this.Sharp = sharp;
    }).catch((err) => {
      logger.error(err);
      logger.error("sharp加载失败，请检查node服务");
    });
  }
};
function apply(ctx) {
  ctx.plugin(SharpService);
}
__name(apply, "apply");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Config,
  SharpService,
  apply,
  filter,
  inject,
  name,
  usage
});
