// yakumo/lib/index.js
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/index.ts
import * as cordis from "cordis";
import globby from "globby";
import yargs from "yargs-parser";
import detect from "detect-indent";
import { manager, spawnAsync } from "./utils.js";
import kleur from "kleur";
import { promises as fs, readFileSync } from "node:fs";
import { deduplicate, isNonNullable, makeArray } from "cosmokit";
export * from "./utils.js";
var cwd = process.cwd();
var content = readFileSync(`${cwd}/package.json`, "utf8");
var meta = JSON.parse(content);
var Context2 = class extends cordis.Context {
  static {
    __name(this, "Context");
  }
  constructor(config) {
    super(config);
    this.provide("yakumo", void 0, true);
    this.plugin(Yakumo, config);
  }
};
var builtin = [
  "list",
  "prepare",
  "publish",
  "run",
  "test",
  "upgrade",
  "version"
];
var Yakumo = class extends cordis.Service {
  constructor(ctx, config) {
    super(ctx, "yakumo", true);
    this.ctx = ctx;
    this.config = config;
    ctx.mixin("yakumo", ["register"]);
    this.cwd = cwd;
    this.manager = manager;
    for (const name in config.pipeline || {}) {
      this.register(name, async (...rest) => {
        const tasks = config.pipeline[name];
        for (const task of tasks) {
          const [name2, ...args] = task.split(/\s+/g);
          await this.execute(name2, ...args, ...rest);
        }
      });
    }
  }
  static {
    __name(this, "Yakumo");
  }
  cwd;
  argv;
  manager;
  workspaces;
  indent = detect(content).indent;
  commands = {};
  register(name, callback, options = {}) {
    this.commands[name] = [callback, options];
  }
  async initialize() {
    const folders = await globby(meta.workspaces || [], {
      cwd,
      onlyDirectories: true,
      expandDirectories: false
    });
    folders.unshift("");
    this.workspaces = Object.fromEntries((await Promise.all(folders.map(async (path) => {
      if (path) path = "/" + path;
      try {
        const content2 = await fs.readFile(`${cwd}${path}/package.json`, "utf8");
        return [path, JSON.parse(content2)];
      } catch {
      }
    }))).filter(isNonNullable));
  }
  resolveIntercept() {
    let result = this.config;
    let intercept = this.ctx[Context2.intercept];
    while (intercept) {
      result = {
        ...result,
        ...intercept.yakumo,
        alias: {
          ...result.alias,
          ...intercept.yakumo?.alias
        },
        exclude: [
          ...result.exclude || [],
          ...intercept.yakumo?.exclude || []
        ]
      };
      intercept = Object.getPrototypeOf(intercept);
    }
    return result;
  }
  locate(name, options = {}) {
    const { alias = {}, exclude } = this.resolveIntercept();
    const defaultFilter = options.filter || ((meta2) => options.includeRoot || !meta2.workspaces);
    const filter = /* @__PURE__ */ __name((meta2, path) => {
      return defaultFilter(meta2, path) && !exclude?.some((pattern) => {
        return new RegExp("^/" + pattern.replace(/\*/g, "[^/]+") + "$").test(path);
      });
    }, "filter");
    if (Array.isArray(name)) {
      if (!name.length) {
        return Object.keys(this.workspaces).filter((folder) => {
          return filter(this.workspaces[folder], folder);
        });
      } else {
        return deduplicate(name.flatMap((name2) => this.locate(name2, options)));
      }
    }
    if (alias[name]) {
      return makeArray(alias[name]).map((path) => {
        if (!this.workspaces[path]) {
          throw new Error(`cannot find workspace ${path} resolved by ${name}`);
        }
        return path;
      });
    }
    for (const key in alias) {
      if (!key.endsWith("*")) continue;
      if (!name.startsWith(key.slice(0, -1))) continue;
      const results = makeArray(alias[key]).map((path) => path.slice(0, -1) + name.slice(key.length - 1)).filter((path) => this.workspaces[path]);
      if (results.length) return results;
    }
    const targets = Object.keys(this.workspaces).filter((folder) => {
      if (!filter(this.workspaces[folder], folder)) return;
      return folder.endsWith("/" + name);
    });
    if (!targets.length) {
      throw new Error(`cannot find workspace "${name}"`);
    } else if (targets.length > 1) {
      throw new Error(`ambiguous workspace "${name}": ${targets.join(", ")}`);
    }
    return targets;
  }
  async save(path) {
    const content2 = JSON.stringify(this.workspaces[path], null, this.indent) + "\n";
    await fs.writeFile(`${cwd}${path}/package.json`, content2);
  }
  async execute(name, ...args) {
    await this.ctx.events.flush();
    if (!this.commands[name]) {
      if (builtin.includes(name)) {
        await this.ctx.get("loader")?.create({
          name: "yakumo/" + name
        });
        return this.execute(name, ...args);
      }
      console.error(kleur.red(`unknown command: ${name}`));
      process.exit(1);
    }
    const [callback, options] = this.commands[name];
    const index = args.indexOf("--");
    const rest = index === -1 ? [] : args.splice(index + 1);
    const argv = yargs(args, options);
    argv["--"] = rest;
    await this.initialize();
    if (!name.startsWith("yakumo:") && name !== "run") {
      await this.execute("run", ...argv._, "--", `yakumo:before:${name}`);
    }
    argv.config = options;
    this.argv = argv;
    await callback(...args);
    if (!name.startsWith("yakumo:") && name !== "run") {
      await this.execute("run", ...argv._.slice(0, index === -1 ? void 0 : index), "--", `yakumo:after:${name}`);
    }
  }
  async start() {
    if (this.ctx.get("loader")?.config.name !== "yakumo") return;
    const [name, ...args] = process.argv.slice(2);
    if (!name) {
      console.log("yakumo");
      process.exit(0);
    }
    this.execute(name, ...args);
  }
  async install() {
    const agent = manager?.name || "npm";
    const code = await spawnAsync([agent, "install"]);
    if (code) process.exit(code);
  }
};
export {
  Context2 as Context,
  cwd,
  Yakumo as default,
  meta
};
