// yakumo/lib/utils.js
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/utils.ts
import ora from "ora";
import prompts from "prompts";
import which from "which-pm-runs";
import spawn from "execa";
import getRegistry from "get-registry";
import semver from "semver";
async function confirm(message) {
  const { value } = await prompts({
    name: "value",
    type: "confirm",
    message
  });
  return value;
}
__name(confirm, "confirm");
function exit(message) {
  const spinner = ora();
  spinner.info(message);
  return process.exit(0);
}
__name(exit, "exit");
function spawnAsync(args, options = {}) {
  const child = spawn(args[0], args.slice(1), options);
  return new Promise((resolve) => {
    child.stderr?.pipe(process.stderr);
    child.stdout?.pipe(process.stdout);
    child.on("close", resolve);
  });
}
__name(spawnAsync, "spawnAsync");
var manager = which();
async function install() {
  const agent = manager?.name || "npm";
  const code = await spawnAsync([agent, "install"]);
  if (code) process.exit(code);
}
__name(install, "install");
var registryTask;
async function fetchRemote(name) {
  const registry = await (registryTask ||= getRegistry());
  const packageUrl = new URL(encodeURIComponent(name).replace(/^%40/, "@"), registry);
  const response = await fetch(packageUrl, {
    headers: {
      "Accept": "application/vnd.npm.install-v1+json; q=1.0, application/json; q=0.8, */*"
    }
  });
  if (!response.ok) throw new Error(`Failed to fetch ${packageUrl}`);
  return response.json();
}
__name(fetchRemote, "fetchRemote");
function selectVersion(data, version) {
  if (data["dist-tags"][version]) {
    return data["dist-tags"][version];
  } else if (data.versions?.[version]) {
    return version;
  } else {
    const versions = Object.keys(data.versions);
    return semver.maxSatisfying(versions, version, { includePrerelease: true, loose: true });
  }
}
__name(selectVersion, "selectVersion");
export {
  confirm,
  exit,
  fetchRemote,
  install,
  manager,
  selectVersion,
  spawnAsync
};
