// @koishijs/scripts/lib/clone.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = default_1;
const child_process_1 = require("child_process");
const which_pm_runs_1 = __importDefault(require("which-pm-runs"));
const prompts_1 = __importDefault(require("prompts"));
async function getRepo() {
    const { name } = await (0, prompts_1.default)({
        type: 'text',
        name: 'name',
        message: 'repository name:',
    });
    return name.trim();
}
async function getName() {
    const { name } = await (0, prompts_1.default)({
        type: 'text',
        name: 'name',
        message: 'target directory:',
    });
    return name.trim();
}
function default_1(cli) {
    cli.command('clone [repo] [name]', 'clone a plugin')
        .action(async (repo, name, options) => {
        let cap;
        repo ||= await getRepo();
        if ((cap = /^(?:https:\/\/github\.com\/)?([\w-]+)\/([\w-]+)(?:\.git)?$/.exec(repo))) {
            name ||= cap[2].replace('koishi-plugin-', '');
            if (!repo.startsWith('https:')) {
                repo = 'https://github.com/' + repo;
            }
            if (!repo.endsWith('.git')) {
                repo = repo + '.git';
            }
        }
        name ||= await getName();
        (0, child_process_1.execSync)(['git', 'clone', repo, 'external/' + name].join(' '), { stdio: 'inherit' });
        const agent = (0, which_pm_runs_1.default)()?.name || 'npm';
        (0, child_process_1.execSync)([agent, 'exec', 'yakumo', 'prepare'].join(' '), { stdio: 'inherit' });
        const args = agent === 'yarn' ? [] : ['install'];
        (0, child_process_1.execSync)([agent, ...args].join(' '), { stdio: 'inherit' });
    });
}
