// @koishijs/scripts/lib/setup.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = default_1;
const fs_extra_1 = require("fs-extra");
const child_process_1 = require("child_process");
const path_1 = require("path");
const _1 = require(".");
const kleur_1 = require("kleur");
const which_pm_runs_1 = __importDefault(require("which-pm-runs"));
const prompts_1 = __importDefault(require("prompts"));
function supports(command) {
    try {
        (0, child_process_1.execSync)(command, { stdio: 'ignore' });
        return true;
    }
    catch {
        return false;
    }
}
class Initiator {
    options;
    name;
    desc;
    fullname;
    monorepo;
    target;
    source = (0, path_1.resolve)(__dirname, '../template');
    constructor(options) {
        this.options = options;
    }
    async start(name) {
        await this.init(name);
        const agent = (0, which_pm_runs_1.default)()?.name || 'npm';
        const args = agent === 'yarn' ? [] : ['install'];
        (0, child_process_1.execSync)([agent, ...args].join(' '), { stdio: 'inherit' });
    }
    async init(name) {
        name ||= await this.getName();
        const oldName = name;
        name = name.toLowerCase().replace(/_/g, '-');
        if (!/^(?:@[a-z0-9-]+\/)?[a-z0-9-]+$/.test(name)) {
            console.log((0, kleur_1.red)('error'), 'plugin name contains invalid character');
            process.exit(1);
        }
        if (oldName !== name) {
            console.log((0, kleur_1.blue)('info'), `plugin name will be converted to "${name}"`);
            const { confirm } = await (0, prompts_1.default)({
                type: 'confirm',
                name: 'confirm',
                message: 'OK?',
            });
            if (!confirm)
                process.exit(0);
        }
        if (name.includes('koishi-plugin-')) {
            this.fullname = name;
            this.name = name.replace('koishi-plugin-', '').replace(/^@.+\//, '');
            console.log((0, kleur_1.blue)('info'), 'prefix "koishi-plugin-" can be omitted');
        }
        else {
            this.name = name.replace(/^@.+\//, '');
            this.fullname = name.replace(/^(.+\/)?/, '$1koishi-plugin-');
        }
        this.desc = await this.getDesc();
        if (this.options.monorepo) {
            this.monorepo = (0, path_1.resolve)(_1.cwd, 'external', this.name);
            this.target = (0, path_1.resolve)(_1.cwd, 'external', this.name, 'packages', this.name);
        }
        else {
            this.target = (0, path_1.resolve)(_1.cwd, 'external', this.name);
        }
        await this.write();
    }
    async getName() {
        const { name } = await (0, prompts_1.default)({
            type: 'text',
            name: 'name',
            message: 'plugin name:',
        });
        return name.trim();
    }
    async getDesc() {
        const { desc } = await (0, prompts_1.default)({
            type: 'text',
            name: 'desc',
            message: 'description:',
        });
        return desc;
    }
    async write() {
        await (0, fs_extra_1.mkdir)(this.target, { recursive: true });
        await Promise.all([
            this.writeManifest(),
            this.writeTsConfig(),
            this.writeIndex(),
            this.writeReadme(),
            this.writeClient(),
        ]);
        await this.initGit();
    }
    async writeManifest() {
        if (this.monorepo) {
            const source = await (0, fs_extra_1.readJson)(this.source + '/package.root.json', 'utf8');
            source.devDependencies = _1.meta.devDependencies;
            await (0, fs_extra_1.writeFile)(this.monorepo + '/package.json', JSON.stringify({
                name: `@root/${this.name}`,
                ...source,
            }, null, 2) + '\n');
        }
        const source = await (0, fs_extra_1.readJson)(this.source + '/package.json', 'utf8');
        if (this.options.console) {
            source.devDependencies['@koishijs/client'] = _1.meta.devDependencies['@koishijs/client'];
            source.peerDependencies['@koishijs/plugin-console'] = _1.meta.dependencies['@koishijs/plugin-console'];
        }
        source.peerDependencies['koishi'] = _1.meta.dependencies['koishi'];
        await (0, fs_extra_1.writeFile)(this.target + '/package.json', JSON.stringify({
            name: this.fullname,
            description: this.desc,
            ...source,
        }, null, 2) + '\n');
    }
    async writeTsConfig() {
        const source = await (0, fs_extra_1.readJson)(this.source + '/tsconfig.base.json', 'utf8');
        if (this.monorepo) {
            await (0, fs_extra_1.writeFile)(this.monorepo + '/tsconfig.base.json', JSON.stringify(source, null, 2));
            await (0, fs_extra_1.writeFile)(this.monorepo + '/tsconfig.json', JSON.stringify({
                extends: './tsconfig.base',
                compilerOptions: {
                    baseUrl: '.',
                    paths: {
                        [`koishi-plugin-${this.name}-*`]: ['packages/*/src'],
                        [`koishi-plugin-*`]: ['packages/*/src'],
                    },
                },
            }, null, 2) + '\n');
            await (0, fs_extra_1.writeFile)(this.target + '/tsconfig.json', JSON.stringify({
                extends: '../../tsconfig.base',
                compilerOptions: {
                    outDir: 'lib',
                    rootDir: 'src',
                },
                include: [
                    'src',
                ],
            }, null, 2) + '\n');
        }
        else {
            await (0, fs_extra_1.writeFile)(this.target + '/tsconfig.json', JSON.stringify({
                ...source,
                include: [
                    'src',
                ],
            }, null, 2) + '\n');
        }
    }
    async writeIndex() {
        await (0, fs_extra_1.mkdir)(this.target + '/src');
        const filename = `/src/index.${this.options.console ? 'console' : 'default'}.ts`;
        const source = await (0, fs_extra_1.readFile)(this.source + filename, 'utf8');
        await (0, fs_extra_1.writeFile)(this.target + '/src/index.ts', source
            .replace(/\{\{name\}\}/g, this.name.replace(/^@\w+\//, '')));
    }
    async writeReadme() {
        const source = await (0, fs_extra_1.readFile)(this.source + '/readme.md', 'utf8');
        await (0, fs_extra_1.writeFile)(this.target + '/readme.md', source
            .replace(/\{\{name\}\}/g, this.fullname)
            .replace(/\{\{desc\}\}/g, this.desc));
    }
    async writeClient() {
        if (!this.options.console)
            return;
        await (0, fs_extra_1.mkdir)(this.target + '/client');
        await Promise.all([
            (0, fs_extra_1.copyFile)(this.source + '/client/index.ts', this.target + '/client/index.ts'),
            (0, fs_extra_1.copyFile)(this.source + '/client/page.vue', this.target + '/client/page.vue'),
            (0, fs_extra_1.copyFile)(this.source + '/client/tsconfig.json', this.target + '/client/tsconfig.json'),
        ]);
    }
    async initGit() {
        if (!this.options.git || !supports('git --version'))
            return;
        await Promise.all([
            (0, fs_extra_1.copyFile)(this.source + '/_editorconfig', this.target + '/.editorconfig'),
            (0, fs_extra_1.copyFile)(this.source + '/_gitattributes', this.target + '/.gitattributes'),
            (0, fs_extra_1.copyFile)(this.source + '/_gitignore', this.target + '/.gitignore'),
        ]);
        (0, child_process_1.execSync)('git init', { cwd: this.target, stdio: 'ignore' });
        (0, child_process_1.execSync)('git add .', { cwd: this.target, stdio: 'ignore' });
        (0, child_process_1.execSync)('git commit -m "initial commit"', { cwd: this.target, stdio: 'ignore' });
    }
}
function default_1(cli) {
    cli.command('setup [name]', 'initialize a new plugin')
        .alias('create')
        .alias('init')
        .alias('new')
        .option('-m, --monorepo', 'setup for monorepo')
        .option('-c, --console', 'with console extension')
        .option('-G, --no-git', 'skip git initialization')
        .action(async (name, options) => {
        new Initiator(options).start(name);
    });
}
