// yakumo-tsc/lib/index.js
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/index.ts
import * as fs from "node:fs/promises";
import { join } from "path";
import { cwd } from "yakumo";
import { compile, load } from "tsconfig-utils";
import * as atsc from "atsc";
import * as dtsc from "dtsc";
import { extname } from "node:path";
async function prepareBuild(nodes) {
  if (!nodes.length) return;
  await fs.writeFile(cwd + "/tsconfig.temp.json", JSON.stringify({
    files: [],
    references: nodes.map((node) => ({ path: "." + node.path }))
  }, null, 2));
}
__name(prepareBuild, "prepareBuild");
function getFiles(meta, outDir) {
  function addExport(pattern) {
    if (!pattern) return;
    if (pattern.startsWith("./")) pattern = pattern.slice(2);
    if (!pattern.startsWith(outDir + "/")) return;
    const ext = extname(pattern);
    pattern = pattern.slice(outDir.length + 1, -ext.length);
    if (pattern.endsWith(".d")) {
      pattern = pattern.slice(0, -2);
    }
    files.add(pattern);
  }
  __name(addExport, "addExport");
  function addConditionalExport(pattern) {
    if (!pattern) return;
    if (typeof pattern === "string") return addExport(pattern);
    if ("types" in pattern) return addConditionalExport(pattern.types);
    if ("typings" in pattern) return addConditionalExport(pattern.typings);
    for (const key in pattern) {
      addConditionalExport(pattern[key]);
    }
  }
  __name(addConditionalExport, "addConditionalExport");
  const files = /* @__PURE__ */ new Set();
  addExport(meta.main);
  addExport(meta.module);
  addConditionalExport(meta.exports);
  return [...files];
}
__name(getFiles, "getFiles");
async function bundleNodes(nodes) {
  for (const node of nodes) {
    await dtsc.build(join(cwd, node.path));
  }
}
__name(bundleNodes, "bundleNodes");
var inject = ["yakumo"];
function apply(ctx) {
  ctx.register("tsc", async () => {
    const { argv } = ctx.yakumo;
    const paths = ctx.yakumo.locate(ctx.yakumo.argv._);
    if (argv.clean) {
      const tasks = paths.map(async (path) => {
        const fullpath = join(cwd, path);
        const tsconfig = await load(fullpath);
        await Promise.allSettled([
          fs.rm(join(fullpath, tsconfig?.compilerOptions?.outDir || "lib"), { recursive: true }),
          fs.rm(join(fullpath, "tsconfig.tsbuildinfo"))
        ]);
      });
      tasks.push(fs.rm(join(cwd, "tsconfig.temp.json")));
      await Promise.allSettled(tasks);
      return;
    }
    const nodes = {};
    for (const path of paths) {
      const meta = ctx.yakumo.workspaces[path];
      if (!meta.main && !meta.exports) continue;
      const fullpath = join(cwd, path);
      try {
        const config = await load(fullpath);
        if (config.compilerOptions?.noEmit) continue;
        const files = getFiles(meta, config.compilerOptions?.outDir || "lib");
        const bundle2 = config.compilerOptions?.outFile ? true : config.compilerOptions?.outDir ? false : files.length === 1 && !!meta.exports;
        nodes[meta.name] = { config, bundle: bundle2, path, meta, prev: [], next: /* @__PURE__ */ new Set() };
      } catch (error) {
        if (error.code !== "ENOENT") throw error;
      }
    }
    for (const name in nodes) {
      const { meta } = nodes[name];
      const deps = {
        ...meta.dependencies,
        ...meta.devDependencies
      };
      for (const dep in deps) {
        if (!nodes[dep] || meta.yakumo?.tsc?.ignore?.includes(dep)) continue;
        nodes[name].prev.push(dep);
        nodes[dep].next.add(name);
      }
    }
    let bundle = false;
    const layers = [];
    while (Object.keys(nodes).length) {
      const layer = [];
      bundle = !bundle;
      let flag = true;
      while (flag) {
        flag = false;
        for (const name of Object.keys(nodes)) {
          const node = nodes[name];
          if (node.next.size || node.bundle === bundle) continue;
          flag = true;
          layer.unshift(node);
          delete nodes[name];
          node.prev.forEach((prev) => {
            nodes[prev].next.delete(name);
          });
        }
      }
      if (layers.length && !layer.length) {
        console.log(nodes);
        throw new Error("circular dependency detected");
      }
      layers.unshift(layer);
    }
    if (bundle) layers.unshift([]);
    for (let i = 0; i < layers.length; i += 2) {
      const bundleTargets = layers[i];
      const buildTargets = layers[i + 1];
      const tasks = buildTargets.map((node) => atsc.build(join(cwd, node.path)));
      await Promise.all([
        prepareBuild(buildTargets),
        bundleNodes(bundleTargets)
      ]);
      if (buildTargets.length) {
        const code = await compile(["-b", "tsconfig.temp.json", "--listEmittedFiles"]);
        if (code) process.exit(code);
      }
      await Promise.all(tasks);
    }
    await fs.rm(join(cwd, "tsconfig.temp.json")).catch(() => {
    });
  }, {
    boolean: ["clean"]
  });
}
__name(apply, "apply");
export {
  apply,
  inject
};
