// @koishijs/wordle/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defineVariation = exports.Wordle = void 0;
const koishi_1 = require("koishi");
var Wordle;
(function (Wordle) {
    let GameState;
    (function (GameState) {
        GameState["Active"] = "active";
    })(GameState = Wordle.GameState || (Wordle.GameState = {}));
})(Wordle || (exports.Wordle = Wordle = {}));
function defineVariation(variation) {
    var _a, _b;
    var _c;
    let command;
    const sessionState = new Map();
    const handleInput = (_a = variation.handleInput) !== null && _a !== void 0 ? _a : async function (input, state, session, ctx) {
        const unitResults = [];
        if (input.length !== state.currentWord.length) {
            return { unitResults, type: 'bad-length' };
        }
        // check if word is in the word list
        if (variation.validWords) {
            if (!variation.validWords.some((word) => word.join('') === input)) {
                return { unitResults, type: 'invalid' };
            }
        }
        if (variation.validateWord) {
            if (!(await variation.validateWord(input.split(''), session, ctx))) {
                return { unitResults, type: 'invalid' };
            }
        }
        const currentWord = [...state.currentWord];
        for (let i = 0; i < input.length; i++) {
            const char = input[i];
            if (char === currentWord[i]) {
                unitResults.push({ type: 'correct', char });
                currentWord[i] = '';
            }
            else if (currentWord.some((char) => char === input[i])) {
                unitResults.push({ type: 'bad-position', char });
            }
            else {
                unitResults.push({ type: 'incorrect', char });
            }
        }
        if (unitResults.every((result) => result.type === 'correct')) {
            return { unitResults, type: 'correct' };
        }
        else {
            return { unitResults, type: 'incorrect' };
        }
    };
    function normalizeInject(inject) {
        var _a;
        if (typeof inject === 'undefined') {
            return {
                required: ['canvas'],
                optional: [],
            };
        }
        else if (Array.isArray(inject)) {
            return {
                required: Array.from(new Set(inject)).concat(['canvas']),
                optional: [],
            };
        }
        else {
            return {
                required: Array.from(new Set(((_a = inject.required) !== null && _a !== void 0 ? _a : []).concat(['canvas']))),
                optional: inject.optional,
            };
        }
    }
    return _c = class {
            constructor(ctx, config) {
                var _a;
                this.ctx = ctx;
                this.config = config;
                this._variation = variation;
                // re-export koishi fields
                this.name = variation.name;
                // define locales
                ctx.i18n.define('zh-CN', require('./locales/zh-CN'));
                // register global command messages for this variation
                ctx.i18n.define('zh-CN', {
                    commands: {
                        [typeof variation.command === 'string'
                            ? variation.command
                            : variation.command.name]: require('./locales/commands.zh-CN'),
                    },
                });
                if (variation.locales) {
                    Object.entries(variation.locales).forEach(([locale, data]) => {
                        ctx.i18n.define(locale, data);
                    });
                }
                command = typeof variation.command === 'string' ? ctx.command(variation.command) : variation.command;
                (_a = variation.init) === null || _a === void 0 ? void 0 : _a.call(variation, command, this);
                command.option('quit', '-q');
                command.action(async (argv, word) => {
                    var _a, _b, _d, _e, _f, _g, _h, _j;
                    const { session } = argv;
                    const state = sessionState.get(`${session.guildId}.${session.channelId}`);
                    if (argv.options.quit) {
                        if ((state === null || state === void 0 ? void 0 : state.state) !== Wordle.GameState.Active) {
                            return session === null || session === void 0 ? void 0 : session.text('wordle.messages.not-started', [command.name]);
                        }
                        sessionState.delete(`${session.guildId}.${session.channelId}`);
                        (_a = variation.onGameEnd) === null || _a === void 0 ? void 0 : _a.call(variation, argv, ctx);
                        return session === null || session === void 0 ? void 0 : session.text('wordle.messages.game-ended', [command.name, state.currentWord.join('')]);
                    }
                    if ((state === null || state === void 0 ? void 0 : state.state) !== Wordle.GameState.Active && word) {
                        return session === null || session === void 0 ? void 0 : session.text('wordle.messages.not-started', [command.name]);
                    }
                    if ((state === null || state === void 0 ? void 0 : state.state) === Wordle.GameState.Active) {
                        if (word) {
                            const result = await handleInput(word, state, argv, this);
                            const text = [];
                            switch (result.type) {
                                case 'bad-length':
                                    text.push(koishi_1.h.text(session === null || session === void 0 ? void 0 : session.text('wordle.messages.bad-length')));
                                    break;
                                case 'invalid':
                                    text.push(koishi_1.h.text(session === null || session === void 0 ? void 0 : session.text('wordle.messages.invalid')));
                                    break;
                                case 'correct':
                                    text.push(koishi_1.h.text(session === null || session === void 0 ? void 0 : session.text('wordle.messages.correct')));
                                // Note: here we do not use `break` here as we need to render the result as well
                                // eslint-disable-next-line no-fallthrough
                                case 'incorrect':
                                    text.push(await this.render(result.unitResults, (_b = state.guessedWords) !== null && _b !== void 0 ? _b : [], session));
                                    break;
                            }
                            await session.send(text);
                            if (result.type === 'correct') {
                                // game ended
                                sessionState.delete(`${session.guildId}.${session.channelId}`);
                                (_d = variation.onGameEnd) === null || _d === void 0 ? void 0 : _d.call(variation, argv, ctx);
                            }
                            else if ((_e = state.guessedCount >= variation.guessCount) !== null && _e !== void 0 ? _e : 6) {
                                await session.send(session === null || session === void 0 ? void 0 : session.text('wordle.messages.game-over', [command.name, state.currentWord.join('')]));
                                (_f = variation.onGameEnd) === null || _f === void 0 ? void 0 : _f.call(variation, argv, ctx);
                                sessionState.delete(`${session.guildId}.${session.channelId}`);
                            }
                            else if (result.type === 'incorrect') {
                                sessionState.set(`${session.guildId}.${session.channelId}`, {
                                    state: Wordle.GameState.Active,
                                    currentWord: state.currentWord,
                                    guessedWords: [...((_g = state.guessedWords) !== null && _g !== void 0 ? _g : []), result],
                                    guessedCount: state.guessedCount + 1,
                                });
                            }
                        }
                        else {
                            return session === null || session === void 0 ? void 0 : session.text('wordle.messages.no-input', [command.name]);
                        }
                    }
                    else {
                        // start a new game
                        (_h = variation.onGameStart) === null || _h === void 0 ? void 0 : _h.call(variation, argv, this);
                        const currentWord = await this.getCurrentWord(argv, ctx);
                        sessionState.set(`${session.guildId}.${session.channelId}`, { state: Wordle.GameState.Active, currentWord });
                        await session.send(session === null || session === void 0 ? void 0 : session.text('wordle.messages.game-started', [command.name, (_j = variation.guessCount) !== null && _j !== void 0 ? _j : 6]));
                    }
                });
            }
            getCurrentWord(argv, ctx) {
                return variation.getCurrentWord(argv, this);
            }
            async render(word, guessedWords, session) {
                var _a, _b;
                const width = word.length * 60 + 5 * (word.length + 1);
                const height = ((_a = variation.guessCount) !== null && _a !== void 0 ? _a : 6) * 60 + 5 * (((_b = variation.guessCount) !== null && _b !== void 0 ? _b : 6) + 1) + 68 + 20 * 2;
                return await session.app.canvas.render(width, height, (ctx) => {
                    // Set background color to white
                    ctx.fillStyle = '#fff';
                    ctx.fillRect(0, 0, width, height);
                    // Draw centered header
                    ctx.font = 'bold 68px sans-serif';
                    ctx.textAlign = 'center';
                    ctx.textBaseline = 'middle';
                    ctx.fillStyle = '#000';
                    ctx.fillText(command.name.replace(/^./, (p) => p.toLocaleUpperCase()), width / 2, 68 / 2 + 20);
                    let y = 68 + 20 * 2;
                    ctx.font = 'bold 36px sans-serif';
                    // Draw every line of word
                    for (const w of [...guessedWords, { unitResults: word, type: '' }]) {
                        let x = 5;
                        for (const letter of w.unitResults) {
                            ctx.fillStyle =
                                letter.type === 'correct' ? '#6aaa64' : letter.type === 'bad-position' ? '#c9b458' : '#787c7e';
                            ctx.fillRect(x, y, 60, 60);
                            ctx.fillStyle = '#fff';
                            ctx.textAlign = 'center';
                            ctx.textBaseline = 'middle';
                            ctx.fillText(letter.char.toLocaleUpperCase(), x + 60 / 2, y + 60 / 2 + 2);
                            x += 60 + 5;
                        }
                        y += 60 + 5;
                    }
                });
            }
        },
        _c.Config = variation.Config,
        _c.inject = normalizeInject((_b = variation.inject) !== null && _b !== void 0 ? _b : variation.using),
        _c.reusable = variation.reusable,
        _c.reactive = variation.reactive,
        _c;
}
exports.defineVariation = defineVariation;
