// @quanhuzeyu/koishi-plugin-node/lib/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Config: () => Config,
  QhzyNode: () => QhzyNode,
  apply: () => apply,
  name: () => name
});
module.exports = __toCommonJS(src_exports);
var import_node_path = __toESM(require("node:path"));
var import_node_fs = __toESM(require("node:fs"));
var import_koishi = require("koishi");
var name = "QhzyNode";
var Config = import_koishi.Schema.object({
  storeDir: import_koishi.Schema.string().default("data/@QuanhuZeYu/node"),
  proxyPrefix: import_koishi.Schema.string().description("代理前缀").default("")
});
var QhzyNode = class extends import_koishi.Service {
  static {
    __name(this, "QhzyNode");
  }
  constructor(ctx, config) {
    super(ctx, "node");
    this.config = {
      storeDir: "data/@QuanhuZeYu/node",
      proxyPrefix: "",
      ...config
    };
  }
  async start() {
    const logger = this.ctx.logger("QhzyNode");
    logger.info("node 服务启动成功");
  }
  /**
   * 执行跨平台命令的函数
   * @param command 要执行的命令
   * @param args 执行命令的参数
   * @param cwd 执行命令的工作目录
   * @returns 执行结果的 Promise
   */
  async runCommand(command, args = [], cwd) {
    const execa = await import("execa");
    try {
      const { stdout } = await execa.execa(command, args, { cwd });
      return stdout;
    } catch (error) {
      throw new Error(`Error executing command: ${error.stderr || error.message}`);
    }
  }
  async safeImport(packageName) {
    const logger = this.ctx.logger("QhzyNode");
    const proxy = this.config.proxyPrefix;
    const storeDir = import_node_path.default.resolve(this.ctx.baseDir, this.config.storeDir);
    const packDir = import_node_path.default.resolve(storeDir, packageName);
    logger.info(`导入包路径: [ ${storeDir} ]`);
    if (!import_node_fs.default.existsSync(storeDir)) {
      logger.info(`创建目录: [ ${storeDir} ]`);
      try {
        import_node_fs.default.mkdirSync(storeDir, { recursive: true });
      } catch (e) {
        logger.error(`创建目录失败: [ ${storeDir} ], 错误信息: ${e.message}`);
        throw e;
      }
    }
    if (!import_node_fs.default.existsSync(packDir)) {
      logger.info(`正在安装包: [ ${packageName} ]`);
      try {
        await this.runCommand(`${proxy}npm`, ["init", "-y"], storeDir);
        await this.runCommand(`${proxy}npm`, ["install", packageName], storeDir);
        logger.info(`包安装完成: [ ${packageName} ]`);
      } catch (e) {
        logger.error(`安装包时出错: [ ${packageName} ], 错误信息: ${e.message}`);
        if (import_node_fs.default.existsSync(packDir)) {
          import_node_fs.default.rmSync(packDir, { recursive: true, force: true });
        }
        throw e;
      }
    } else {
      logger.info(`找到现有包: [ ${packageName} ]`);
    }
    try {
      return require(import_node_path.default.resolve(storeDir, "node_modules", packageName));
    } catch (e) {
      logger.error(`加载模块失败: [ ${packageName} ], 错误信息: ${e.message}`);
      try {
        logger.info(`清理目录: [ ${packDir} ] 并重试安装`);
        import_node_fs.default.rmSync(packDir, { recursive: true, force: true });
      } catch (err) {
        logger.error(`清理目录失败: [ ${packDir} ], 错误信息: ${err.message}`);
        throw err;
      }
      return this.safeImport(packageName);
    }
  }
};
function apply(ctx) {
  ctx.plugin(QhzyNode);
}
__name(apply, "apply");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Config,
  QhzyNode,
  apply,
  name
});
