// koishi-plugin-arcaea/lib/source/aua.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArcaeaUnlimitedApi = void 0;
const index_1 = require("./index");
const koishi_1 = require("koishi");
class ArcaeaUnlimitedApi extends koishi_1.Service {
    constructor(ctx, config) {
        super(ctx, 'arcaea');
        this.ctx = ctx;
        this.config = config;
        this.client = ctx.http.extend({
            endpoint: config.endpoint,
            headers: {
                Authorization: 'Bearer ' + config.access_token
            }
        });
    }
    parseError(e) {
        if (e.response) {
            try {
                return (0, index_1.ArcaeaResponseStatus)(e.response.data);
            }
            catch (e) {
                return {
                    status: -114514,
                    message: "Internal error:" + e.toString() + ", arcaea server response:" + JSON.stringify(e.response.body)
                };
            }
        }
        else {
            return {
                status: -114513,
                message: "Internal error or parser error:" + e.toString()
            };
        }
    }
    async sendRequest(path, method, parameters, body) {
        try {
            const response = (await this.client.axios({
                method,
                url: path,
                params: parameters,
                data: body,
                responseType: 'json'
            }));
            if (response.data.status < 0)
                return this.parseError({ response });
            return (0, index_1.ArcaeaResponse)(koishi_1.Schema.any())(response.data['content']);
        }
        catch (e) {
            return this.parseError(e);
        }
    }
    async sendUserRequest(path, userIdentifier, parameters) {
        return await this.sendRequest(path, 'GET', {
            ...parameters,
            user: userIdentifier
        });
    }
    async getUserInfo(userIdentifier, recent) {
        const response = await this.sendUserRequest('/user/info', userIdentifier, {});
        console.info(response);
        if (response.status < 0)
            throw new index_1.ArcaeaInternalError(`${response.message}(${response.status})` ?? "unknown error(-114515)");
        try {
            return (0, index_1.ArcaeaUserInfo)(response);
        }
        catch (e) {
            throw new index_1.ArcaeaInternalError(e.toString());
        }
    }
    async getUserBest30(userIdentifier) {
        const response = await this.sendUserRequest('/user/best', userIdentifier, {});
        if (!response.status)
            throw new index_1.ArcaeaInternalError(response.message ?? "unknown error");
        try {
            return (0, index_1.ArcaeaBest30Records)(response);
        }
        catch (e) {
            throw new index_1.ArcaeaInternalError(e.toString());
        }
    }
}
exports.ArcaeaUnlimitedApi = ArcaeaUnlimitedApi;
ArcaeaUnlimitedApi.Config = koishi_1.Schema.object({
    type: koishi_1.Schema.const('aua'),
    endpoint: koishi_1.Schema.string().required(),
    access_token: koishi_1.Schema.string().role('secret')
});
