// koishi-plugin-arcaea/lib/source/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArcaeaInternalError = exports.ArcaeaResponse = exports.ArcaeaBest30Records = exports.ArcaeaUserInfo = exports.ArcaeaPlayRecord = exports.ArcaeaSuccessfulResponse = exports.ArcaeaResponseStatus = void 0;
const koishi_1 = require("koishi");
exports.ArcaeaResponseStatus = koishi_1.Schema.object({
    status: koishi_1.Schema.number().required()
});
exports.ArcaeaSuccessfulResponse = koishi_1.Schema.object({
    status: koishi_1.Schema.const(0)
});
exports.ArcaeaPlayRecord = koishi_1.Schema.object({
    song_id: koishi_1.Schema.string().required(),
    difficulty: koishi_1.Schema.number().required(),
    score: koishi_1.Schema.number().required(),
    shiny_perfect_count: koishi_1.Schema.number().required(),
    perfect_count: koishi_1.Schema.number().required(),
    near_count: koishi_1.Schema.number().required(),
    miss_count: koishi_1.Schema.number().required(),
    clear_type: koishi_1.Schema.number(),
    best_clear_type: koishi_1.Schema.number(),
    health: koishi_1.Schema.number(),
    time_played: koishi_1.Schema.transform(koishi_1.Schema.number(), t => new Date(t)).required(),
    modifier: koishi_1.Schema.number().required(),
    rating: koishi_1.Schema.number().required()
});
exports.ArcaeaUserInfo = koishi_1.Schema.object({
    account_info: koishi_1.Schema.object({
        user_id: koishi_1.Schema.number(),
        code: koishi_1.Schema.transform(koishi_1.Schema.string(), t => parseInt(t)),
        name: koishi_1.Schema.string().required(),
        character: koishi_1.Schema.number(),
        join_date: koishi_1.Schema.transform(koishi_1.Schema.number(), t => new Date(t)),
        is_skill_sealed: koishi_1.Schema.boolean(),
        is_char_uncapped: koishi_1.Schema.boolean(),
        is_char_uncapped_override: koishi_1.Schema.boolean(),
        is_mutual: koishi_1.Schema.boolean(),
        rating: koishi_1.Schema.number()
    }),
    recent_score: koishi_1.Schema.array(exports.ArcaeaPlayRecord)
});
exports.ArcaeaBest30Records = koishi_1.Schema.object({
    best30_avg: koishi_1.Schema.number().required(),
    recent10_avg: koishi_1.Schema.number().required(),
    best30_list: koishi_1.Schema.array(exports.ArcaeaPlayRecord).required()
});
const ArcaeaResponse = (T) => koishi_1.Schema.union([
    exports.ArcaeaResponseStatus,
    koishi_1.Schema.intersect([
        exports.ArcaeaSuccessfulResponse,
        T
    ])
]);
exports.ArcaeaResponse = ArcaeaResponse;
class ArcaeaInternalError extends Error {
}
exports.ArcaeaInternalError = ArcaeaInternalError;
