// koishi-plugin-sayo-roll/lib/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Config: () => Config,
  apply: () => apply,
  inject: () => inject,
  leftQuotes: () => leftQuotes,
  matchRegExps: () => matchRegExps,
  name: () => name,
  parseArgs: () => parseArgs,
  replaceMy: () => replaceMy,
  rightQuotes: () => rightQuotes
});
module.exports = __toCommonJS(index_exports);
var import_koishi = require("koishi");

// src/locales/zh-CN.yml
var zh_CN_default = { commands: { "sayo-roll": { description: "随机选择", examples: "<p>\n  sayo-roll\n  <br /> &gt;{' '}\n  <i18n path='commands.sayo-roll.messages.roll-integer'>(随机 1~100 整数)</i18n>\n</p>\n<p>\n  sayo-roll 10\n  <br /> &gt;{' '}\n  <i18n path='commands.sayo-roll.messages.roll-integer'>(随机 1~10 整数)</i18n>\n</p>\n<p>\n  sayo-roll 我是不是耳聋\n  <br /> &gt;{' '}\n  <i18n path='commands.sayo-roll.messages.i-think'>你 (是 / 不是) 耳聋</i18n>\n</p>\n<p>\n  sayo-roll 我今天吃没吃\n  <br /> &gt;{' '}\n  <i18n path='commands.sayo-roll.messages.i-think'>你今天 (吃了 / 没吃)</i18n>\n</p>\n<p>\n  sayo-roll 今晚刷pp还是摆烂\n  <br /> &gt;{' '}\n  <i18n path='commands.sayo-roll.messages.surely'>(今晚刷pp / 摆烂)</i18n>\n</p>\n<p>\n  sayo-roll 吃饭 睡觉 打游戏\n  <br /> &gt; {' '}\n  <i18n path='commands.sayo-roll.messages.choice'>(吃饭 / 睡觉 / 打游戏)</i18n>\n</p>\n<p>\n  sayo-roll 我今晚刷到pp的概率\n  <br /> &gt;{' '}\n  <i18n path='commands.sayo-roll.messages.probability'>\n    <>你今晚刷到pp</>\n    <>(0.01 ~ 100)</>\n  </i18n>\n</p>", messages: { "invalid-arg": "看不懂", "number-too-small": "数字太小了！", "block-word": "你在说什么！！！", "roll-integer": "你的数字是 {0}", "i-think": "我觉得{0}", surely: "当然是{0}喽", probability: "{0} 的概率为：{1}%", choice: "我选择{0}" } } }, _config: { blockWords: "屏蔽词列表" } };

// src/index.ts
var name = "sayo-roll";
var inject = [];
var defaultBlockWords = [
  "打胶",
  "傻逼",
  "做爱",
  "sb",
  "打炮",
  "打飞机",
  "打枪",
  "自慰",
  "鸡巴",
  "鸡吧",
  "鸡把",
  "鸡鸡",
  "自杀",
  "去世",
  "紫砂",
  "屌",
  "破处",
  "处女"
];
var Config = import_koishi.Schema.intersect([
  import_koishi.Schema.object({
    blockWords: import_koishi.Schema.array(import_koishi.Schema.string()).role("table").default(defaultBlockWords)
  })
]).i18n({
  "zh-CN": zh_CN_default._config,
  zh: zh_CN_default._config
});
function matchRegExps(str, regexps) {
  for (const [r, f] of regexps) {
    const m = r.exec(str);
    if (!m) continue;
    return f(m);
  }
}
__name(matchRegExps, "matchRegExps");
var leftQuotes = [`"`, `'`, `“`, `‘`];
var rightQuotes = [`"`, `'`, `”`, `’`];
function parseArgs(text) {
  const args = [];
  let arg = "";
  let currentRight = null;
  for (const char of text) {
    if (char === "\\") {
      arg += char;
    } else if (arg.slice(-1) === "\\") {
      arg = arg.slice(0, -1) + char;
    } else if (!currentRight && leftQuotes.includes(char)) {
      currentRight = rightQuotes[leftQuotes.indexOf(char)];
    } else if (char === currentRight || !currentRight && char.match(/\s/)) {
      currentRight = null;
      if (arg) {
        args.push(arg);
        arg = "";
      }
    } else {
      arg += char;
    }
  }
  if (currentRight) throw new TypeError("unmatched quote");
  if (arg) args.push(arg);
  return args;
}
__name(parseArgs, "parseArgs");
function replaceMy(text) {
  return text.replace("我", "你");
}
__name(replaceMy, "replaceMy");
function apply(ctx, config) {
  ctx.i18n.define("zh-CN", zh_CN_default);
  ctx.i18n.define("zh", zh_CN_default);
  const blockWordLower = config.blockWords.map((v) => v.toLowerCase());
  ctx.command(`${name} <args:text>`).action(({ session }, arg) => {
    if (!session) return;
    if (!arg) {
      return session.text(".roll-integer", [import_koishi.Random.int(1, 100)]);
    }
    const argLower = arg.toLowerCase();
    for (const word of blockWordLower) {
      if (argLower.includes(word)) return session.text(".block-word");
    }
    const matched = matchRegExps(arg, [
      [
        /^(?<int>\d+)$/,
        (match) => {
          const max = parseInt(match.groups.int);
          if (max <= 1) return session.text(".number-too-small");
          return session.text(".roll-integer", [import_koishi.Random.int(1, max + 1)]);
        }
      ],
      [
        // 我是不是耳聋 => { prefix: '我', word: '是', type: '不', suffix: '耳聋' }
        // 我今天吃没吃 => { prefix: '我今天', word: '吃', type: '没', suffix: '' }
        /^(?<prefix>.*)(?<word>.+)(?<type>不|没)(\k<word>)(?<suffix>.*?)([呀呢啊])?([？\?！!]*)$/,
        (match) => {
          const prefix = replaceMy(match.groups.prefix);
          const suffix = replaceMy(match.groups.suffix);
          const { word, type } = match.groups;
          const result = import_koishi.Random.bool(0.5);
          return session.text(".i-think", [
            type === "不" ? (
              // 我是不是耳聋 => 你是耳聋 / 你不是耳聋
              `${prefix}${result ? "" : "不"}${word}${suffix}`
            ) : (
              // 我今天吃没吃 => 你今天吃了 / 你今天没吃
              `${prefix}${result ? "" : "没"}${word}${result ? "了" : ""}${suffix}`
            )
          ]);
        }
      ],
      [
        /^(?<prefix>.+?)还是(?<suffix>.+?)([呀呢啊])?([？\?]*)$/,
        (match) => session.text(".surely", [
          import_koishi.Random.pick([
            replaceMy(match.groups.prefix),
            replaceMy(match.groups.suffix)
          ])
        ])
      ],
      [
        /^(?<name>.+?)的?概率([是|为])?([？\?]*)$/,
        (match) => session.text(".probability", [
          replaceMy(match.groups.name),
          (Math.random() * 100).toFixed(2)
        ])
      ]
    ]);
    if (matched) return matched;
    let args;
    try {
      args = parseArgs(arg).map(replaceMy);
    } catch (e) {
      ctx.logger.debug(e);
      return session.text(".invalid-arg");
    }
    if (args.length < 2) {
      return session.text(".invalid-arg");
    }
    return session.text(".choice", [import_koishi.Random.pick(args)]);
  });
}
__name(apply, "apply");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Config,
  apply,
  inject,
  leftQuotes,
  matchRegExps,
  name,
  parseArgs,
  replaceMy,
  rightQuotes
});
