// koishi-plugin-anti-space/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apply = exports.Config = exports.usage = exports.name = void 0;
const koishi_1 = require("koishi");
exports.name = 'anti-space';
exports.usage = `去掉koishi指令与第一个参数之间的空格
例如：rryth koishi可以变为rrythkoishi
注意本插件只对指令生效，如chess插件中的“落子”不是指令，不会被本插件处理
新增nocommandlist可免输入指令，当且仅当参数足够特别时可以使用本配置项。
例如emojimix的参数是两个emoji，可以直接用正则匹配，从而省略指令的输入。
效果：原来需要输入emojimix emoji1emoji2才能触发，现在只需要输入emoji1emoji2即可触发。\n
更新：所有at参数无需单独配置，已预置`;
exports.Config = koishi_1.Schema.object({
    ignoreat: koishi_1.Schema.boolean().description('是否忽略@与引用').default(true),
    commandslist: koishi_1.Schema.dict(String).description('左面填指令名，右面填用于匹配参数的正则表达式，留空则默认全匹配，下面以rryth为例')
        .default({
        'rryth': '/.*/',
    }),
    nocommandlist: koishi_1.Schema.dict(String).description('左面填要忽略的指令名，右面填用于匹配参数的正则表达式，一定不要留空！！下面以emojimix为例')
        .default({
        'emojimix': '/^(\p{Emoji_Presentation}\uFE0F?){2}$/u',
    }),
});
function createRegexFromString(str) {
    // 解析字符串，获取正则表达式模式和标志
    const match = /^\/(.*)\/([a-z]*)$/.exec(str);
    if (match) {
        const pattern = match[1];
        const flags = match[2];
        return new RegExp(pattern, flags);
    }
    else {
        // 如果字符串不是有效的正则表达式，则抛出错误
        throw new Error(`Invalid regular expression: ${str}`);
    }
}
function findAllAtTags(str) {
    const regex = /<.*?>/g;
    const matches = [];
    let match;
    while ((match = regex.exec(str)) !== null) {
        matches.push(match[0]);
    }
    return matches;
}
function apply(ctx, config) {
    const prefixs = ctx.root.config.prefix;
    const prefix = prefixs ? prefixs[0] : '';
    ctx.middleware(async (session, next) => {
        let mid = session.content.trim();
        if (config.ignoreat) {
            while (mid.startsWith(`<`)) {
                mid = mid.slice(mid.indexOf('>') + 1).trim();
            }
        }
        if (mid.startsWith(prefix)) {
            mid = mid.trim().slice(prefix.length, mid.length);
        }
        if (mid.includes('<')) {
            const regs = findAllAtTags(mid);
            const regList = regs.join(' ');
            const command = mid.slice(0, mid.indexOf('<')).trim();
            return await session.execute(`${command} ${regList}`);
        }
        const commandList = Object.entries(config.commandslist);
        const [command, regex] = commandList.find(([cmd]) => mid.startsWith(cmd)) || [];
        if (command && regex) {
            const reg = mid.slice(command.length);
            const test1 = (!regex) ? '/.*/' : regex;
            const test = createRegexFromString(test1);
            if (!reg.startsWith(' ') && reg !== '' && test.test(reg)) {
                return await session.execute(`${command} ${reg}`);
            }
        }
        return next();
    });
    ctx.before('send', async (session) => {
        if (session.content.startsWith('<execute>help ')) {
            session.content = '';
            session.execute('help');
        }
    }, true);
    ctx.middleware(async (session, next) => {
        let mid = session.content.trim();
        if (config.ignoreat) {
            while (mid.startsWith(`<`)) {
                mid = mid.slice(mid.indexOf('>') + 1).trim();
            }
        }
        const noCommandList = Object.entries(config.nocommandlist);
        for (const [command, regex] of noCommandList) {
            if (!regex) {
                throw new Error('省略指令的参数正则不可留空！');
            }
            const test = createRegexFromString(regex);
            if (test.test(mid)) {
                return await session.execute(`${command.replace(prefix, '')} ${mid}`);
            }
        }
        return next();
    });
}
exports.apply = apply;
