// yakumo/lib/plugins/publish.js
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/plugins/publish.ts
import { cwd, exit, spawnAsync } from "../index.js";
import { gt, prerelease } from "semver";
import { isNonNullable } from "cosmokit";
import { join } from "path";
import { fetchRemote, selectVersion } from "../utils.js";
import ora from "ora";
import prompts from "prompts";
import assert from "assert";
async function getVersion(name, isNext2 = false) {
  const remote = await fetchRemote(name).catch(() => null);
  if (!remote) return "0.0.0";
  if (isNext2) {
    return selectVersion(remote, "next") || selectVersion(remote, "latest") || "0.0.0";
  } else {
    return selectVersion(remote, "latest") || "0.0.0";
  }
}
__name(getVersion, "getVersion");
function isNext(version) {
  const parts = prerelease(version);
  if (!parts) return false;
  return parts[0] !== "rc";
}
__name(isNext, "isNext");
function isBerry(manager) {
  return manager?.name === "yarn" && !manager.version.startsWith("1.");
}
__name(isBerry, "isBerry");
async function publish(manager, path, meta, args, argv) {
  if (!isBerry(manager)) {
    args = [manager?.name || "npm", "publish", join(cwd, path), "--color", ...args];
    return await spawnAsync(args, { stdio: argv.debug ? "inherit" : "ignore" });
  }
  return await spawnAsync(["yarn", "workspace", meta.name, "npm", "publish", ...args], { stdio: argv.debug ? "inherit" : "ignore" });
}
__name(publish, "publish");
async function parallel(list, fn) {
  await Promise.all(list.map(fn));
}
__name(parallel, "parallel");
async function serial(list, fn) {
  for (const item of list) await fn(item);
}
__name(serial, "serial");
var inject = ["yakumo"];
function apply(ctx) {
  ctx.register("publish", async () => {
    const { argv } = ctx.yakumo;
    const spinner = ora();
    let paths = ctx.yakumo.locate(argv._, {
      filter: /* @__PURE__ */ __name((meta) => {
        return argv._.length ? !meta.workspaces : !meta.private;
      }, "filter")
    });
    if (argv._.length) {
      const pending = paths.filter((path) => ctx.yakumo.workspaces[path].private);
      if (pending.length) {
        const paths2 = pending.map((path) => ctx.yakumo.workspaces[path].name).join(", ");
        const { value } = await prompts({
          name: "value",
          type: "confirm",
          message: `workspace ${paths2} ${pending.length > 1 ? "are" : "is"} private, switch to public?`
        });
        if (!value) exit("operation cancelled.");
        await Promise.all(pending.map(async (path) => {
          delete ctx.yakumo.workspaces[path].private;
          await ctx.yakumo.save(path);
        }));
      }
    }
    let progress = 0, skipped = 0;
    spinner.start(`Loading workspaces (0/${paths.length})`);
    paths = (await Promise.all(paths.map(async (path) => {
      const meta = ctx.yakumo.workspaces[path];
      spinner.text = `Loading workspaces (${++progress}/${paths.length})`;
      const version = await getVersion(meta.name, isNext(meta.version));
      if (gt(meta.version, version)) return path;
      skipped += 1;
    }))).filter(isNonNullable);
    spinner.succeed();
    const total = paths.length;
    if (!argv.debug && total > 0) spinner.start(`Publishing packages (0/${total})`);
    let completed = 0, failed = 0;
    await (argv.debug ? serial : parallel)(paths, async (path) => {
      const meta = ctx.yakumo.workspaces[path];
      try {
        await ctx.parallel("publish/before", path, meta);
        const args = [
          "--tag",
          argv.tag ?? (isNext(meta.version) ? "next" : "latest"),
          "--access",
          argv.access ?? "public"
        ];
        if (argv.registry) args.push("--registry", argv.registry);
        if (argv.otp) args.push("--otp", argv.otp);
        const code = await publish(ctx.yakumo.manager, path, meta, args, argv);
        assert(!code);
        fetch("https://registry-direct.npmmirror.com/" + meta.name + "/sync?sync_upstream=true", {
          method: "PUT"
        }).catch(() => {
        });
        await ctx.parallel("publish/after", path, meta);
      } catch (e) {
        failed++;
      } finally {
        if (!argv.debug) spinner.text = `Publishing packages (${++completed}/${total})`;
      }
    });
    const skippedText = skipped ? `, ${skipped} skipped` : "";
    if (failed) {
      spinner.fail(`Published ${total - failed} packages, ${failed} failed${skippedText}.`);
    } else {
      spinner.succeed(`Published ${total} packages${skippedText}.`);
    }
  }, {
    boolean: ["debug"]
  });
}
__name(apply, "apply");
export {
  apply,
  inject
};
