// yakumo/lib/plugins/test.js
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/plugins/test.ts
import { spawn } from "child_process";
import parse from "yargs-parser";
import unparse from "yargs-unparser";
import globby from "globby";
var inject = ["yakumo"];
function apply(ctx) {
  ctx.register("test", async () => {
    function getFiles(names) {
      if (!names.length) return ["**/tests/*.spec.ts"];
      return names.flatMap((name) => {
        const [folder] = name.split("/", 1);
        name = name.slice(folder.length + 1) || "*";
        return ctx.yakumo.locate(folder, { includeRoot: true }).map((path) => {
          return `${path}/tests/${name}.spec.ts`.slice(1);
        });
      });
    }
    __name(getFiles, "getFiles");
    const argv = parse(process.argv.slice(3));
    const files = await globby(getFiles(argv._), {
      cwd: ctx.yakumo.cwd,
      onlyFiles: true
    });
    argv._ = [];
    const child = spawn(process.execPath, [
      ...unparse(argv),
      "--test",
      ...files
    ], {
      stdio: "inherit",
      cwd: ctx.yakumo.cwd
    });
    child.on("exit", (code, signal) => {
      process.on("exit", () => {
        if (signal) {
          process.kill(process.pid, signal);
        } else {
          process.exit(code ?? void 0);
        }
      });
    });
    process.on("SIGINT", () => {
      child.kill("SIGINT");
    });
  });
}
__name(apply, "apply");
export {
  apply,
  inject
};
