// yakumo/lib/plugins/version.js
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/plugins/version.ts
import { writeFile } from "fs/promises";
import { readFileSync } from "fs";
import { gt, SemVer } from "semver";
import kleur from "kleur";
import { confirm, cwd } from "../index.js";
var bumpTypes = ["major", "minor", "patch", "version", "reset", "local"];
var Package = class {
  constructor(path) {
    this.path = path;
    const content = readFileSync(`${cwd}/${path}/package.json`, "utf8");
    this.meta = JSON.parse(content);
    this.version = this.meta.version;
  }
  static {
    __name(this, "Package");
  }
  meta;
  version;
  dirty;
  bump(flag, options, args) {
    if (this.meta.private) return;
    if (flag === "local") return this.meta.version;
    let version = new SemVer(this.meta.version);
    const reset = flag === "reset";
    if (args.stable) {
      version.prerelease = [];
    } else if (args.prerelease) {
      if (version.prerelease.length) {
        version.prerelease = [{
          alpha: "beta",
          beta: "rc"
        }[version.prerelease[0]], 0];
      } else {
        flag ??= "major";
        if (flag === "major") {
          version = new SemVer(`${version.major + 1}.0.0-alpha.0`);
        } else if (flag === "minor") {
          version = new SemVer(`${version.major}.${version.minor + 1}.0-alpha.0`);
        } else if (flag === "patch") {
          version = new SemVer(`${version.major}.${version.minor}.${version.patch + 1}-alpha.0`);
        }
      }
    } else if (!flag || reset) {
      if (version.prerelease.length) {
        const prerelease = version.prerelease.slice();
        prerelease[1] += reset ? -1 : 1;
        version.prerelease = prerelease;
      } else {
        version.patch += reset ? -1 : 1;
      }
      if (reset) {
        this.dirty = true;
        return this.version = version.format();
      }
    } else if (flag === "version") {
      if (this.version !== options.version) {
        this.dirty = true;
      }
      this.version = options.version;
      return options.version;
    } else {
      if (version.prerelease.length) {
        version.prerelease = [];
      } else {
        version[flag] += 1;
        if (flag !== "patch") version.patch = 0;
        if (flag === "major") version.minor = 0;
      }
    }
    const formatted = version.format();
    if (gt(formatted, this.version)) {
      this.dirty = true;
      this.version = formatted;
      return formatted;
    }
  }
  save(indent) {
    this.meta.version = this.version;
    const content = JSON.stringify(this.meta, null, indent) + "\n";
    return writeFile(`${cwd}/${this.path}/package.json`, content);
  }
};
var Graph = class {
  constructor(project) {
    this.project = project;
    for (const path in project.workspaces) {
      this.nodes[path] = new Package(path);
    }
  }
  static {
    __name(this, "Graph");
  }
  nodes = {};
  each(callback) {
    const results = [];
    for (const path in this.nodes) {
      results.push(callback(this.nodes[path], path));
    }
    return results;
  }
  bump(node, flag, args) {
    const version = node.bump(flag, this.project.argv, args);
    if (!version) return;
    const dependents = /* @__PURE__ */ new Set();
    this.each((target) => {
      const { name } = node.meta;
      if (target.meta.name === name) return;
      const npmLinkPrefix = `npm:${name}@`;
      for (const type of ["devDependencies", "peerDependencies", "dependencies", "optionalDependencies"]) {
        const deps = target.meta[type] || {};
        for (const key in deps) {
          let update2 = function(prefix) {
            const range = value.slice(prefix.length);
            if (range.includes(":")) return;
            const modifier = /^[\^~]?/.exec(range)[0];
            if (range === modifier + version) return;
            target.meta[type][key] = prefix + modifier + version;
            target.dirty = true;
            if (type !== "devDependencies") {
              dependents.add(target);
            }
          };
          var update = update2;
          __name(update2, "update");
          const value = deps[key];
          if (key === name) {
            update2("");
          } else if (value.startsWith(npmLinkPrefix)) {
            update2(npmLinkPrefix);
          }
        }
      }
    });
    if (!this.project.argv.recursive) return;
    dependents.forEach((dep) => this.bump(dep, flag, args));
  }
  async save() {
    let hasUpdate = false;
    await Promise.all(this.each((node) => {
      if (!node.dirty) return;
      hasUpdate = true;
      if (node.version === node.meta.version) {
        console.log(`- ${node.meta.name}: dependency updated`);
      } else {
        console.log(`- ${node.meta.name}: ${kleur.cyan(node.meta.version)} => ${kleur.green(node.version)}`);
      }
      return node.save(this.project.indent);
    }));
    if (!hasUpdate) {
      console.log("Everything is up-to-date.");
    }
  }
};
var inject = ["yakumo"];
function apply(ctx) {
  ctx.register("version", async () => {
    if (!ctx.yakumo.argv._.length) {
      const yes = await confirm("You did not specify any packages to bump. Do you want to bump all the packages?");
      if (!yes) return;
    }
    const flags = bumpTypes.filter((type) => type in ctx.yakumo.argv);
    if (flags.length > 1) {
      console.log(kleur.red("You can only specify one bump type."));
      return;
    }
    const flag = flags[0];
    if (flag === "version") {
      new SemVer(ctx.yakumo.argv.version);
    }
    const graph = new Graph(ctx.yakumo);
    const paths = ctx.yakumo.locate(ctx.yakumo.argv._);
    for (const path of paths) {
      graph.bump(graph.nodes[path], flag, ctx.yakumo.argv);
    }
    await graph.save();
  }, {
    alias: {
      major: ["1"],
      minor: ["2"],
      patch: ["3"],
      reset: ["0"],
      prerelease: ["p"],
      stable: ["P"],
      version: ["v"],
      recursive: ["r"],
      local: ["l"]
    },
    boolean: ["major", "minor", "patch", "reset", "local", "prerelease", "stable", "recursive"]
  });
}
__name(apply, "apply");
export {
  apply,
  inject
};
