// koishi-plugin-w-node/lib/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  default: () => src_default,
  name: () => name
});
module.exports = __toCommonJS(src_exports);
var import_node_os = __toESM(require("node:os"));
var import_node_path = __toESM(require("node:path"));
var import_promises = __toESM(require("node:fs/promises"));
var import_get_registry = __toESM(require("get-registry"));
var import_koishi = require("koishi");
var name = "w-node";
var exists = /* @__PURE__ */ __name(async (path2) => {
  try {
    await import_promises.default.stat(path2);
    return true;
  } catch {
    return false;
  }
}, "exists");
var NodeService = class extends import_koishi.Service {
  constructor(ctx, config) {
    super(ctx, "node");
    this.config = config;
    ctx.command("node", "Node 服务");
    ctx.command("node.list", "列出安装的包").action(async () => {
      const dir = this.config.packagePath;
      if (!exists(dir)) return "包目录不存在";
      const subs = await Promise.all((await import_promises.default.readdir(dir)).map(async (sub) => {
        const packageJsonPath = import_node_path.default.resolve(dir, sub, "node_modules", this.unescapePackageName(sub), "package.json");
        const packageJson = await import_promises.default.readFile(packageJsonPath, "utf-8");
        const { name: name2, version } = JSON.parse(packageJson);
        return `${name2}@${version}`;
      }));
      return `共安装了 ${subs.length} 个包：${subs.join(", ")}`;
    });
    ctx.command("node.add <package:string> [version:string]", "安装包", { authority: 4 }).action(async (_, packageName2, version) => {
      try {
        await this.install(packageName2, version);
        return "安装成功";
      } catch (err) {
        return `安装失败：${err}`;
      }
    });
    ctx.command("node.remove <package:string>", "移除包", { authority: 4 }).action(async (_, packageName2) => {
      const cwd = import_node_path.default.resolve(this.config.packagePath, this.escapePackageName(packageName2));
      if (await exists(cwd)) {
        await import_promises.default.rm(cwd, { recursive: true, force: true });
        return "移除成功";
      }
      return `包 ${packageName2} 不存在`;
    });
    ctx.command("node.debug.evalwith <package:string> <code:text>", "引入包并运行 JavaScript", { authority: 4 }).option("varname", "-v <varname:string> 引入包的变量名").action(async (argv, packageName, code) => {
      const varName = argv.options.varname || packageName.replace(/^@.*\//, "").replace(/(^(?=\d))/, "_");
      try {
        const _package = await this.safeImport(packageName);
        const result = await eval(`const ${varName} = _package; ${code}`);
        return JSON.stringify(result);
      } catch (error) {
        return String(error);
      }
    });
  }
  static {
    __name(this, "NodeService");
  }
  async start() {
    this.execaPackage = await import("execa");
    const wrapStdoutStderr = /* @__PURE__ */ __name((type) => {
      const log = this.logger[type === "stdout" ? "info" : "error"];
      return function* (line) {
        log(" ".repeat("Running $ ".length) + line);
      };
    }, "wrapStdoutStderr");
    const wrapExeca = /* @__PURE__ */ __name((execa) => (...p) => {
      if (p.length > 1 && p[0] instanceof Array) this.logger.info("Running $ %s", String.raw(...p));
      const result2 = execa({
        stdout: wrapStdoutStderr("stdout"),
        stderr: wrapStdoutStderr("stderr")
      })(...p);
      if (typeof result2 === "function") return wrapExeca(result2);
      return result2;
    }, "wrapExeca");
    this.execa = wrapExeca(this.execaPackage.execa);
    if (!this.config.registry) {
      this.config.registry = await (0, import_get_registry.default)();
      this.ctx.scope.update(this.config);
    }
  }
  logger = this.ctx.logger("w-node");
  execaPackage;
  execa;
  escapePackageName = /* @__PURE__ */ __name((packageName2) => packageName2.replace("@", "at__").replace("/", "__slash__"), "escapePackageName");
  unescapePackageName = /* @__PURE__ */ __name((escapedPackageName) => escapedPackageName.replace("at__", "@").replace("__slash__", "/"), "unescapePackageName");
  async install(packageName2, version) {
    const cwd = import_node_path.default.resolve(this.config.packagePath, this.escapePackageName(packageName2));
    this.logger.info(`Making directory '${cwd}'.`);
    await import_promises.default.mkdir(cwd, { recursive: true });
    await import_promises.default.writeFile(import_node_path.default.resolve(cwd, "package.json"), "{}");
    const packageStr = `${packageName2}@${version || "latest"}`;
    this.logger.info(`Installing '${packageName2}'...`);
    await this.execa({ cwd })`npm add ${packageName2} --color always --registry ${this.config.registry}`;
    this.logger.info(`Installed package '${packageName2}'.`);
  }
  async safeImport(packageName2, { maxRetry = 3, forceInstall = false } = {}) {
    const cwd = import_node_path.default.resolve(this.config.packagePath, this.escapePackageName(packageName2));
    const pkgd = import_node_path.default.resolve(cwd, "node_modules", packageName2);
    if (forceInstall || !await exists(pkgd)) await this.install(packageName2);
    else this.logger.info(`Hit cached package '${packageName2}'.`);
    try {
      return require(pkgd);
    } catch (err) {
      this.logger.error("Failed to require package: %o", err);
      if (maxRetry > 0) {
        return this.safeImport(packageName2, { maxRetry: maxRetry - 1, forceInstall: true });
      }
      throw err;
    }
  }
};
((NodeService2) => {
  NodeService2.Config = import_koishi.z.object({
    packagePath: import_koishi.z.string().default(import_node_path.default.resolve(import_node_os.default.tmpdir(), "w-node")).description("存放 npm 包的位置"),
    registry: import_koishi.z.string().default("").description("npm 源地址，默认与当前项目相同")
  });
})(NodeService || (NodeService = {}));
var src_default = NodeService;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  name
});
