// koishi-plugin-menu/lib/index.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Config: () => Config,
  apply: () => apply,
  inject: () => inject,
  logger: () => logger,
  name: () => name,
  usage: () => usage
});
module.exports = __toCommonJS(src_exports);
var import_koishi = require("koishi");

// src/extract.ts
var Extract = class {
  constructor(ctx) {
    this.ctx = ctx;
  }
  static {
    __name(this, "Extract");
  }
  /**
   * 获取会话语言代码
   * @param session 会话对象
   * @returns 语言代码
   */
  locale(session) {
    const locales = [...session?.locales || [], ...session?.channel?.locales || [], ...session?.guild?.locales || [], ...session?.user?.locales || []];
    return this.ctx.i18n?.fallback(locales)?.[0] || "zh-CN";
  }
  /**
   * 获取所有可用指令
   * @param session 会话对象
   * @param locale 语言代码
   * @returns 指令列表
   */
  async all(session, locale = "") {
    if (!this.ctx.$commander) return [];
    this.setLocale(session, locale);
    const userAuth = session.user?.authority || 0;
    const commandList = this.ctx.$commander._commandList?.filter((cmd) => !cmd?.parent && cmd?.ctx?.filter?.(session) && userAuth >= this.getAuthority(cmd)) || [];
    const commands = (await Promise.all(commandList.map((cmd) => this.build(cmd, session, userAuth)))).filter(Boolean);
    return commands.sort((a, b) => a.name[0].name.localeCompare(b.name[0].name));
  }
  /**
   * 获取单个指令信息
   * @param session 会话对象
   * @param cmdName 指令名称
   * @param locale 语言代码
   * @returns 指令对象或null
   */
  async single(session, cmdName, locale = "") {
    if (!cmdName) return null;
    this.setLocale(session, locale);
    const command = this.find(cmdName, session);
    return command && !Array.isArray(command) ? this.build(command, session, session.user?.authority || 0) : null;
  }
  /**
   * 获取相关指令列表
   * @param session 会话对象
   * @param cmdName 指令名称
   * @param locale 语言代码
   * @returns 相关指令列表
   */
  async related(session, cmdName, locale = "") {
    const target = await this.single(session, cmdName, locale);
    if (!target) return [];
    const commands = [target];
    if (cmdName.includes(".")) {
      const parent = await this.single(session, cmdName.split(".")[0], locale);
      if (parent && !commands.some((cmd) => cmd.name[0].name === parent.name[0].name)) commands.unshift(parent);
    }
    return commands;
  }
  /**
   * 过滤指令列表
   * @param commands 指令列表
   * @param showHidden 是否显示隐藏项
   * @param isDetailView 是否为详情视图
   * @returns 过滤后的指令列表
   */
  filter(commands, showHidden = false, isDetailView = false) {
    return commands.map((cmd) => ({
      ...cmd,
      options: cmd.options.filter((opt) => showHidden || !opt.hidden),
      subs: cmd.subs?.filter((sub) => showHidden || !sub.hidden)
    })).filter((cmd) => isDetailView || showHidden || !cmd.hidden);
  }
  /**
   * 设置会话语言
   * @param session 会话对象
   * @param locale 语言代码
   */
  setLocale(session, locale) {
    if (locale) session.locales = [locale, ...session?.locales || []];
  }
  /**
   * 查找指令
   * @param target 目标指令名
   * @param session 会话对象
   * @returns 指令对象或快捷方式列表
   */
  find(target, session) {
    if (!target || !this.ctx.$commander) return null;
    const command = this.ctx.$commander.resolve(target, session);
    if (command?.ctx?.filter?.(session)) {
      const { aliases } = this.getEffectiveData(command);
      return aliases[target]?.filter !== false ? command : null;
    }
    const shortcuts = this.ctx.i18n?.find?.("commands.(name).shortcuts.(variant)", target)?.map((item) => ({ ...item, command: this.ctx.$commander?.resolve(item?.data?.name, session) }))?.filter((item) => item?.command?.match?.(session)) || [];
    const perfect = shortcuts.filter((item) => item?.similarity === 1);
    return perfect.length ? perfect[0]?.command : shortcuts.length ? shortcuts : null;
  }
  /**
   * 处理指令名称配置
   * @param command 指令对象
   * @returns 名称配置数组
   */
  processNameConfig(command) {
    if (!command) return [];
    const { aliases } = this.getEffectiveData(command);
    const aliasEntries = Object.entries(aliases || {});
    if (!aliasEntries.length) return [{ name: command.name || "", enabled: true, isDefault: true }];
    const nameItems = [];
    let hasDefault = false;
    aliasEntries.forEach(([alias, config]) => {
      const enabled = config?.filter !== false;
      const isDefault = !hasDefault && enabled;
      if (isDefault) hasDefault = true;
      nameItems.push({ name: alias, enabled, isDefault });
    });
    return nameItems.sort((a, b) => {
      if (a.isDefault) return -1;
      if (b.isDefault) return 1;
      return a.name.localeCompare(b.name);
    });
  }
  /**
   * 获取有效数据
   * @param command 指令对象
   * @returns 有效的配置数据
   */
  getEffectiveData(command) {
    if (!command) return { config: {}, options: {}, aliases: {}, texts: null };
    const snapshot = this.ctx.get("commands")?.snapshots?.[command.name];
    if (!snapshot) {
      return {
        config: command.config || {},
        options: command._options || {},
        aliases: command._aliases || {},
        texts: null
      };
    }
    const mergedOptions = { ...snapshot.initial.options };
    Object.entries(snapshot.override.options || {}).forEach(([key, override]) => {
      mergedOptions[key] = override && typeof override === "object" ? { ...mergedOptions[key], ...override } : override;
    });
    return {
      config: { ...snapshot.initial.config || {}, ...snapshot.override.config || {} },
      options: mergedOptions,
      aliases: snapshot.override.aliases || command._aliases || {},
      texts: snapshot.override.texts || null
    };
  }
  /**
   * 获取指令权限等级
   * @param command 指令对象
   * @returns 权限等级
   */
  getAuthority(command) {
    return this.getEffectiveData(command).config?.authority || command.config?.authority || 0;
  }
  /**
   * 获取有效文本内容
   * @param texts 文本配置
   * @param path 文本路径
   * @param session 会话对象
   * @param fallback 回退函数
   * @returns 文本内容
   */
  getEffectiveText(texts, path, session, fallback) {
    if (!texts?.[path]) return fallback();
    const overrideText = texts[path];
    if (typeof overrideText === "string") return overrideText;
    if (typeof overrideText === "object" && overrideText) {
      const locale = session.locales?.[0] || "zh-CN";
      return overrideText[locale] || overrideText["zh-CN"] || overrideText["en-US"] || Object.values(overrideText).find((v) => typeof v === "string") || fallback();
    }
    return fallback();
  }
  /**
   * 构建指令对象
   * @param command 原始指令对象
   * @param session 会话对象
   * @param userAuth 用户权限
   * @returns 构建的指令对象或null
   */
  async build(command, session, userAuth) {
    if (!command) return null;
    const { config, options, texts } = this.getEffectiveData(command);
    const cmdAuth = config?.authority || 0;
    if (userAuth < cmdAuth) return null;
    const getText = /* @__PURE__ */ __name((path) => session?.text?.(path) || "", "getText");
    const nameConfig = this.processNameConfig(command);
    const [description, usage2, examples] = await Promise.all([
      Promise.resolve(this.getEffectiveText(
        texts,
        "description",
        session,
        () => this.cleanText(getText([`commands.${command.name}.description`, ""]))
      )),
      this.getUsageText(command, session, command.name, texts),
      Promise.resolve(this.getExamplesText(command, getText, command.name))
    ]);
    return {
      group: session?.resolve?.(config?.group) || "其它",
      name: nameConfig,
      authority: cmdAuth,
      hidden: session?.resolve?.(config?.hidden) || false,
      desc: description,
      usage: this.cleanText(usage2),
      examples,
      options: this.buildOptions(options, session, userAuth, getText, command.name, texts),
      subs: await this.buildSubCommands(command, session, userAuth)
    };
  }
  /**
   * 清理文本内容
   * @param data 原始数据
   * @returns 清理后的文本
   */
  cleanText(data) {
    if (typeof data === "string") return data;
    if (Array.isArray(data)) return data.map((item) => typeof item === "string" ? item : item?.attrs?.content || "").filter(Boolean).join(" ").trim();
    return "";
  }
  /**
   * 获取使用方法文本
   * @param command 指令对象
   * @param session 会话对象
   * @param name 指令名称
   * @param texts 文本配置
   * @returns 使用方法文本
   */
  async getUsageText(command, session, name2, texts) {
    const overrideUsage = this.getEffectiveText(texts, "usage", session, () => "");
    if (overrideUsage) return overrideUsage;
    try {
      return command?._usage ? typeof command._usage === "string" ? command._usage : await command._usage(session) : session?.text?.([`commands.${name2}.usage`, ""]) || "";
    } catch {
      return session?.text?.([`commands.${name2}.usage`, ""]) || "";
    }
  }
  /**
   * 获取示例文本
   * @param command 指令对象
   * @param getText 获取文本函数
   * @param name 指令名称
   * @returns 示例文本
   */
  getExamplesText(command, getText, name2) {
    return command?._examples?.length ? command._examples.join("\n\n") : getText([`commands.${name2}.examples`, ""]).split("\n").filter((line) => line.trim()).join("\n\n");
  }
  /**
   * 构建选项列表
   * @param effectiveOptions 有效选项配置
   * @param session 会话对象
   * @param userAuth 用户权限
   * @param getText 获取文本函数
   * @param cmdName 指令名称
   * @param texts 文本配置
   * @returns 选项列表
   */
  buildOptions(effectiveOptions, session, userAuth, getText, cmdName, texts) {
    if (!effectiveOptions || typeof effectiveOptions !== "object") return [];
    const options = [];
    const seenOptions = /* @__PURE__ */ new Set();
    const addOption = /* @__PURE__ */ __name((option, name2) => {
      if (!option || !name2 || userAuth < (option?.authority || 0) || seenOptions.has(name2)) return;
      const desc = this.getEffectiveText(texts?.options, name2, session, () => getText(option?.descPath ?? [`commands.${cmdName}.options.${name2}`, ""]));
      if (desc || option?.syntax) {
        seenOptions.add(name2);
        options.push({
          name: name2,
          desc: desc || "",
          syntax: option?.syntax || "",
          hidden: session?.resolve?.(option?.hidden) || false,
          authority: option?.authority || 0
        });
      }
    }, "addOption");
    Object.entries(effectiveOptions).forEach(([key, opt]) => {
      if (!opt) return;
      if (!("value" in opt)) addOption(opt, opt?.name || key);
      if (opt?.variants) Object.keys(opt.variants).forEach((variantKey) => {
        addOption(opt.variants[variantKey], `${opt?.name || key}.${variantKey}`);
      });
    });
    return options;
  }
  /**
   * 构建子指令列表
   * @param command 指令对象
   * @param session 会话对象
   * @param userAuth 用户权限
   * @returns 子指令列表或undefined
   */
  async buildSubCommands(command, session, userAuth) {
    if (!command?.children?.length) return void 0;
    try {
      const uniqueChildren = command.children.filter(
        (sub, index, arr) => sub?.name && arr.findIndex((s) => s?.name === sub.name) === index
      );
      const subs = (await Promise.all(
        uniqueChildren.filter((sub) => sub?.ctx?.filter?.(session) && userAuth >= this.getAuthority(sub)).map((sub) => this.build(sub, session, userAuth))
      )).filter(Boolean);
      return subs.length ? subs : void 0;
    } catch {
      return void 0;
    }
  }
};

// src/utils.ts
var import_path = require("path");
var import_crypto = require("crypto");
var import_fs = require("fs");
var FileStore = class {
  static {
    __name(this, "FileStore");
  }
  baseDir;
  cacheDir;
  /**
   * 创建文件存储实例
   * @param rootDir 根目录路径
   */
  constructor(rootDir) {
    this.baseDir = (0, import_path.join)(rootDir, "data/menu");
    this.cacheDir = (0, import_path.join)(this.baseDir, "cache");
  }
  /**
   * 构建文件路径
   * @param type 文件类型
   * @param locale 可选的语言代码
   * @returns 完整文件路径
   */
  makePath(type, locale) {
    const fileName = type === "commands" && locale ? `commands-${locale}.json` : `${type}.json`;
    return (0, import_path.join)(this.baseDir, fileName);
  }
  /**
   * 检查文件是否存在
   * @param type 文件类型
   * @param locale 可选的语言代码
   * @returns 文件是否存在
   */
  async exists(type, locale) {
    try {
      await import_fs.promises.access(this.makePath(type, locale));
      return true;
    } catch {
      return false;
    }
  }
  /**
   * 保存数据到文件
   * @param type 文件类型
   * @param data 要保存的数据
   * @param locale 可选的语言代码
   */
  async save(type, data, locale) {
    const path = this.makePath(type, locale);
    await import_fs.promises.mkdir((0, import_path.dirname)(path), { recursive: true });
    await import_fs.promises.writeFile(path, JSON.stringify(data, null, 2), "utf8");
  }
  /**
   * 从文件读取数据
   * @param type 文件类型
   * @param locale 可选的语言代码
   * @returns 读取的数据或null
   */
  async load(type, locale) {
    try {
      const content = await import_fs.promises.readFile(this.makePath(type, locale), "utf8");
      return JSON.parse(content);
    } catch {
      return null;
    }
  }
  /**
   * 解析资源路径
   * @param path 输入路径（URL或文件名）
   * @returns 处理后的路径
   */
  resolve(path) {
    return !path || /^https?:\/\//.test(path) ? path || "" : `file:///${(0, import_path.join)(this.baseDir, path).replace(/\\/g, "/")}`;
  }
  /**
   * 检查本地资源文件是否存在
   * @param filename 文件名
   * @returns 是否存在
   */
  async check(filename) {
    if (!filename || /^https?:\/\//.test(filename)) return true;
    try {
      await import_fs.promises.access((0, import_path.join)(this.baseDir, filename));
      return true;
    } catch {
      return false;
    }
  }
  /**
   * 生成缓存键
   * @param commands 指令列表
   * @param config 配置对象
   * @param cmdName 可选的指令名称
   * @returns 缓存键字符串
   */
  generateCacheKey(commands, config, cmdName) {
    if (!commands?.length || !config) return "invalid";
    const prefix = cmdName?.replace(/[^a-zA-Z0-9\-_\.]/g, "_") || "menu";
    const hash = (0, import_crypto.createHash)("md5").update(`${commands.length}:${this.hashConfig(config)}:${this.hashCommands(commands)}`).digest("hex").substring(0, 12);
    return `${prefix}_${hash}`;
  }
  /**
   * 生成配置哈希
   * @param config 配置对象
   * @returns 配置哈希字符串
   */
  hashConfig(config) {
    if (!config || typeof config !== "object") return "empty";
    const keys = ["padding", "radius", "fontSize", "titleSize", "primary", "secondary", "bgColor", "textColor", "glassBlur"];
    return (0, import_crypto.createHash)("md5").update(keys.map((k) => `${k}:${config[k] ?? ""}`).join("|")).digest("hex").substring(0, 8);
  }
  /**
   * 生成指令列表哈希
   * @param commands 指令列表
   * @returns 指令哈希字符串
   */
  hashCommands(commands) {
    if (!Array.isArray(commands) || !commands.length) return "empty";
    const signature = commands.filter((cmd) => cmd && typeof cmd === "object").map((cmd) => `${cmd.name?.name || "unknown"}:${cmd.desc?.length || 0}:${cmd.options?.length || 0}:${cmd.subs?.length || 0}`).sort().join("|");
    return (0, import_crypto.createHash)("md5").update(signature || "empty").digest("hex").substring(0, 8);
  }
  /**
   * 获取缓存文件内容
   * @param key 缓存键
   * @returns 缓存文件的Buffer数据，如果不存在则返回null
   */
  async getCache(key, ext = "png") {
    try {
      return await import_fs.promises.readFile((0, import_path.join)(this.cacheDir, `${key}.${ext}`));
    } catch {
      return null;
    }
  }
  /**
   * 保存数据到缓存文件
   * @param key 缓存键
   * @param buffer 要保存的Buffer数据
   */
  async saveCache(key, buffer, ext = "png") {
    const path = (0, import_path.join)(this.cacheDir, `${key}.${ext}`);
    await import_fs.promises.mkdir((0, import_path.dirname)(path), { recursive: true });
    await import_fs.promises.writeFile(path, buffer);
  }
  /**
   * 清除缓存文件
   * @param cmdName 可选的指令名称，用于清除特定指令的缓存
   */
  async clearCache(cmdName) {
    try {
      const files = await import_fs.promises.readdir(this.cacheDir).catch(() => []);
      const pattern = cmdName?.replace(/[^a-zA-Z0-9\-_\.]/g, "_");
      await Promise.all(files.filter((f) => (f.endsWith(".png") || f.endsWith(".jpg")) && (!pattern || f.startsWith(pattern + "_"))).map((f) => import_fs.promises.unlink((0, import_path.join)(this.cacheDir, f)).catch(() => {
      })));
    } catch {
    }
  }
};

// src/render.ts
var Render = class {
  static {
    __name(this, "Render");
  }
  /**
   * 构建完整的HTML页面
   * @param config 渲染配置
   * @param commands 指令列表
   * @param cmdName 可选的指令名称
   * @returns 完整的HTML字符串
   */
  build(config, commands, currentPage = 1, pageSize = 8, cmdName) {
    const css = this.buildCSS(config);
    const body = this.buildContent(config, commands, currentPage, pageSize, cmdName);
    return `<!DOCTYPE html><html><head><meta charset="UTF-8"><style>${css}${config.customCss ? "\n" + config.customCss : ""}</style></head><body>${body}</body></html>`;
  }
  /**
   * 生成CSS样式
   * @param config 渲染配置
   * @returns CSS样式字符串
   */
  buildCSS(config) {
    const { fontUrl, bgImage, bgSize, primary, secondary, bgColor, textColor, radius, padding, fontSize, titleSize, glassBlur } = config;
    let bgSizeValue;
    if (bgSize === "stretch") {
      bgSizeValue = "100% 100%";
    } else {
      bgSizeValue = bgSize || "cover";
    }
    const bodyBg = bgImage ? `url('${bgImage}') center/${bgSizeValue}, linear-gradient(135deg, ${primary.replace(/[\d.]+\)$/, "0.1)")}, ${secondary.replace(/[\d.]+\)$/, "0.05)")})` : `linear-gradient(135deg, ${primary.replace(/[\d.]+\)$/, "0.15)")} 0%, ${secondary.replace(/[\d.]+\)$/, "0.1)")} 50%, ${bgColor} 100%)`;
    const containerBg = bgImage ? "transparent" : glassBlur > 0 ? bgColor.replace(/[\d.]+\)$/, "0.85)") : bgColor.replace(/[\d.]+\)$/, "0.95)");
    const cardBg = glassBlur > 0 ? bgColor.replace(/[\d.]+\)$/, "0.8)") : bgColor.replace(/[\d.]+\)$/, "0.95)");
    const contentBg = glassBlur > 0 ? bgColor.replace(/[\d.]+\)$/, "0.7)") : bgColor.replace(/[\d.]+\)$/, "0.95)");
    const glassEffect = glassBlur > 0 ? `backdrop-filter:blur(${glassBlur}px);-webkit-backdrop-filter:blur(${glassBlur}px);` : "";
    return `${fontUrl ? `@import url('${fontUrl}');` : ""}
:root{--primary:${primary};--secondary:${secondary};--bg:${bgColor};--text:${textColor};--radius:${radius}px;--padding:${padding}px;--font-size:${fontSize}px;--title-size:${fontSize * titleSize}px}
*{box-sizing:border-box}
body{font:var(--font-size)/1.6 -apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,sans-serif;color:var(--text);background:${bodyBg};padding:var(--padding);display:flex;align-items:flex-start;justify-content:center;min-height:100vh}
.container{background:${containerBg};${bgImage && glassBlur > 0 ? glassEffect : ""}border-radius:calc(var(--radius)*1.5);padding:calc(var(--padding)*1.2);${bgImage ? "" : `box-shadow:0 20px 40px ${primary.replace(/[\d.]+\)$/, "0.15)")},0 8px 32px ${primary.replace(/[\d.]+\)$/, "0.1)")};border:1px solid ${primary.replace(/[\d.]+\)$/, "0.2)")};`}display:grid;grid-template-columns:repeat(auto-fit,minmax(185px,1fr));gap:calc(var(--padding)*0.8);width:100%;max-width:925px}
.container.detail-view{display:flex;flex-direction:column;max-width:800px;width:auto;min-width:400px}
.header,.group-title{grid-column:1/-1;text-align:center;font-weight:700;background:linear-gradient(135deg,var(--primary),var(--secondary));color:white;border-radius:var(--radius);position:relative;overflow:hidden}
.detail-view .header{grid-column:unset}
.header::before,.group-title::before{content:'';position:absolute;top:0;left:0;right:0;bottom:0;background:linear-gradient(135deg,transparent 0%,${primary.replace(/[\d.]+\)$/, "0.1)")} 50%,transparent 100%);pointer-events:none}
.header{font-size:var(--title-size);padding:calc(var(--padding)*1.5)}
.group-title{font-size:calc(var(--font-size)*1.25);padding:calc(var(--padding)*0.8) calc(var(--padding)*1.2);margin-top:calc(var(--padding)*0.6)}
.group-title:first-child{margin-top:0}
.command-card,.detail-card{background:${cardBg};${glassEffect}border-radius:var(--radius);border:1px solid ${primary.replace(/[\d.]+\)$/, "0.2)")};position:relative;overflow:hidden;transition:all 0.3s cubic-bezier(0.4,0,0.2,1)}
.command-card:hover,.detail-card:hover{transform:translateY(-2px);box-shadow:0 12px 24px ${primary.replace(/[\d.]+\)$/, "0.15)")},0 4px 16px ${primary.replace(/[\d.]+\)$/, "0.1)")};border-color:${primary.replace(/[\d.]+\)$/, "0.3)")}}
.command-card.main-command{grid-column:1/-1}
.detail-view .command-card.main-command{grid-column:unset;margin-bottom:calc(var(--padding)*0.8)}
.detail-view .detail-card{margin-bottom:calc(var(--padding)*0.8);width:100%;min-width:300px;max-width:100%}
.detail-view .detail-card:last-child{margin-bottom:0}
.command-card::before,.detail-card::before{content:'';position:absolute;top:0;left:0;width:4px;height:100%;background:linear-gradient(180deg,var(--primary),var(--secondary))}
.detail-card::before{width:3px}
.command-card-content,.detail-card-content{padding:calc(var(--padding)*1.1);margin-left:4px;background:${contentBg};${glassEffect}}
.detail-card-content{margin-left:3px;padding:var(--padding)}
.command-name,.detail-title{font-weight:700;margin-bottom:calc(var(--padding)*0.5);background:linear-gradient(90deg,var(--primary),var(--secondary));-webkit-background-clip:text;background-clip:text;color:transparent;display:inline-block}
.command-name{font-size:calc(var(--font-size)*1.1)}
.detail-title{font-size:calc(var(--font-size)*1.05)}
.command-desc,.detail-content{color:${textColor.replace(/[\d.]+\)$/, "0.8)")};line-height:1.5}
.command-desc{font-size:calc(var(--font-size)*0.9)}
.detail-content{font-size:calc(var(--font-size)*0.88);white-space:pre-wrap}
.command-card:not(.main-command) .command-desc{display:-webkit-box;-webkit-line-clamp:3;-webkit-box-orient:vertical;overflow:hidden}
.footer{grid-column:1/-1;text-align:center;font-size:calc(var(--font-size)*0.85);background:${cardBg};${glassEffect}color:${textColor.replace(/[\d.]+\)$/, "0.7)")};padding:calc(var(--padding)*0.8);border:1px solid ${primary.replace(/[\d.]+\)$/, "0.1)")};border-radius:var(--radius)}
.detail-view .footer{grid-column:unset;margin-top:calc(var(--padding)*0.8)}
@media(min-width:1200px){.container:not(.detail-view){grid-template-columns:repeat(auto-fit,minmax(211px,1fr));max-width:1056px}}
@media(max-width:1199px)and(min-width:900px){.container:not(.detail-view){grid-template-columns:repeat(auto-fit,minmax(198px,1fr))}}
@media(max-width:899px)and(min-width:600px){.container:not(.detail-view){grid-template-columns:repeat(auto-fit,minmax(165px,1fr));gap:calc(var(--padding)*0.6)}}
@media(max-width:599px){body{padding:calc(var(--padding)*0.6)}.container:not(.detail-view){grid-template-columns:1fr;padding:calc(var(--padding)*0.8);gap:calc(var(--padding)*0.5)}.container.detail-view{min-width:unset;padding:calc(var(--padding)*0.8)}.header{font-size:calc(var(--title-size)*0.9)}.group-title{font-size:calc(var(--font-size)*1.1)}.command-card-content,.detail-card-content{padding:calc(var(--padding)*0.9)}.detail-view .detail-card{min-width:unset}}`;
  }
  /**
   * 构建页面内容
   * @param config 渲染配置
   * @param commands 指令列表
   * @param cmdName 可选的指令名称
   * @returns HTML内容字符串
   */
  buildContent(config, commands, currentPage, pageSize, cmdName) {
    const body = cmdName ? this.buildDetail(commands, cmdName) : this.buildList(commands, currentPage, pageSize);
    let footerContent = config.footer || "";
    if (!cmdName) {
      const groups = Array.from(new Set(commands.map((c) => c.group || "其它")));
      const totalPages = Math.ceil(groups.length / pageSize);
      if (totalPages > 1) {
        footerContent += `<div style="margin-top:8px;opacity:0.8">第 ${currentPage} / ${totalPages} 页 (共 ${groups.length} 个分组)</div>`;
      }
    }
    const header = config.header ? `<div class="header">${config.header}</div>` : "";
    const footer = footerContent ? `<div class="footer">${footerContent}</div>` : "";
    const containerClass = cmdName ? "container detail-view" : "container";
    return `<div class="${containerClass}">${header}${body}${footer}</div>`;
  }
  /**
   * 构建指令列表视图
   * @param commands 指令列表
   * @returns HTML字符串
   */
  buildList(commands, currentPage, pageSize) {
    const groupMap = /* @__PURE__ */ new Map();
    commands.forEach((cmd) => {
      const group = cmd.group || "其它";
      if (!groupMap.has(group)) groupMap.set(group, []);
      groupMap.get(group).push(cmd);
    });
    const allGroups = Array.from(groupMap.entries());
    const totalPages = Math.ceil(allGroups.length / pageSize);
    const page = Math.max(1, Math.min(currentPage, totalPages));
    const pagedGroups = allGroups.slice((page - 1) * pageSize, page * pageSize);
    const elements = [];
    pagedGroups.forEach(([name2, cmds]) => {
      elements.push(`<div class="group-title">${name2} (${cmds.length})</div>`);
      cmds.forEach((cmd) => {
        const primaryName = cmd.name.find((n) => n.isDefault)?.name || cmd.name[0]?.name || "";
        elements.push(`<div class="command-card"><div class="command-card-content"><div class="command-name">${primaryName}</div><div class="command-desc">${cmd.desc || "无描述"}</div></div></div>`);
      });
    });
    if (elements.length === 0) return '<div class="group-title">此页没有指令</div>';
    return elements.join("");
  }
  /**
   * 构建指令详情视图
   * @param commands 指令列表
   * @param cmdName 指令名称
   * @returns HTML字符串
   */
  buildDetail(commands, cmdName) {
    const cmd = commands.find((c) => c.name.some((n) => n.name === cmdName && n.enabled)) || commands[0];
    if (!cmd) return `<div class="command-card main-command"><div class="command-card-content"><div class="command-name">指令未找到</div><div class="command-desc">未找到指令 "${cmdName}"</div></div></div>`;
    const elements = [];
    const primaryName = cmd.name.find((n) => n.isDefault)?.name || cmd.name[0]?.name || "";
    elements.push(`<div class="command-card main-command"><div class="command-card-content"><div class="command-name">${primaryName}</div><div class="command-desc">${cmd.desc || "无描述"}</div></div></div>`);
    const aliases = cmd.name.filter((n) => !n.isDefault && n.enabled).map((n) => n.name);
    if (aliases.length > 0) {
      const aliasesText = aliases.join(", ");
      elements.push(this.buildDetailCard("指令别名", aliasesText));
    }
    if (cmd.usage) elements.push(this.buildDetailCard("使用方法", cmd.usage));
    if (cmd.options?.length) {
      const content = cmd.options.map((o) => o.desc ? `${o.syntax || o.name}
  ${o.desc}` : o.syntax || o.name).join("\n");
      elements.push(this.buildDetailCard(`选项参数 (${cmd.options.length})`, content));
    }
    if (cmd.examples) elements.push(this.buildDetailCard("使用示例", cmd.examples));
    if (cmd.subs?.length) {
      const content = cmd.subs.map((s) => `${s.name.find((n) => n.isDefault)?.name || s.name[0]?.name || ""} - ${s.desc || "无描述"}`).join("\n");
      elements.push(this.buildDetailCard(`子指令 (${cmd.subs.length})`, content));
    }
    return elements.join("");
  }
  /**
   * 构建详情卡片
   * @param title 卡片标题
   * @param content 卡片内容
   * @returns HTML字符串
   */
  buildDetailCard(title, content) {
    const escapedContent = content.replace(/</g, "&lt;").replace(/>/g, "&gt;");
    return `<div class="detail-card"><div class="detail-card-content"><div class="detail-title">${title}</div><div class="detail-content">${escapedContent}</div></div></div>`;
  }
};

// src/index.ts
var name = "menu";
var inject = ["puppeteer"];
var logger = new import_koishi.Logger("menu");
var usage = `
<div style="border-radius: 10px; border: 1px solid #ddd; padding: 16px; margin-bottom: 20px; box-shadow: 0 2px 5px rgba(0,0,0,0.1);">
  <h2 style="margin-top: 0; color: #4a6ee0;">📌 插件说明</h2>
  <p>📖 <strong>使用文档</strong>：请点击左上角的 <strong>插件主页</strong> 查看插件使用文档</p>
  <p>🔍 <strong>更多插件</strong>：可访问 <a href="https://github.com/YisRime" style="color:#4a6ee0;text-decoration:none;">苡淞的 GitHub</a> 查看本人的所有插件</p>
</div>

<div style="border-radius: 10px; border: 1px solid #ddd; padding: 16px; margin-bottom: 20px; box-shadow: 0 2px 5px rgba(0,0,0,0.1);">
  <h2 style="margin-top: 0; color: #e0574a;">❤️ 支持与反馈</h2>
  <p>🌟 喜欢这个插件？请在 <a href="https://github.com/YisRime" style="color:#e0574a;text-decoration:none;">GitHub</a> 上给我一个 Star！</p>
  <p>🐛 遇到问题？请通过 <strong>Issues</strong> 提交反馈，或加入 QQ 群 <a href="https://qm.qq.com/q/PdLMx9Jowq" style="color:#e0574a;text-decoration:none;"><strong>855571375</strong></a> 进行交流</p>
</div>
`;
var Config = import_koishi.Schema.intersect([
  import_koishi.Schema.object({
    source: import_koishi.Schema.union([
      import_koishi.Schema.const("file").description("本地配置"),
      import_koishi.Schema.const("inline").description("内存读取")
    ]).description("指令数据源").default("inline"),
    enableCache: import_koishi.Schema.boolean().description("启用图片缓存").default(true),
    imageFormat: import_koishi.Schema.union([
      import_koishi.Schema.const("png").description("PNG"),
      import_koishi.Schema.const("jpeg").description("JPEG")
    ]).description("图片格式").default("png"),
    pageSize: import_koishi.Schema.number().description("分组数量").min(1).default(8)
  }).description("数据源配置"),
  import_koishi.Schema.object({
    padding: import_koishi.Schema.number().description("边距(px)").min(0).default(16),
    radius: import_koishi.Schema.number().description("圆角(px)").min(0).default(12),
    glassBlur: import_koishi.Schema.number().description("毛玻璃强度(px)").min(0).default(0),
    fontSize: import_koishi.Schema.number().description("字体(px)").min(1).default(24),
    titleSize: import_koishi.Schema.number().description("标题倍数").min(1).default(2)
  }).description("样式配置"),
  import_koishi.Schema.object({
    fontlink: import_koishi.Schema.string().description("字体链接"),
    bgimg: import_koishi.Schema.string().description("背景图片"),
    bgSize: import_koishi.Schema.union([
      import_koishi.Schema.const("cover").description("覆盖"),
      import_koishi.Schema.const("contain").description("包含"),
      import_koishi.Schema.const("stretch").description("拉伸"),
      import_koishi.Schema.const("auto").description("自动")
    ]).description("覆盖方式").default("cover"),
    header: import_koishi.Schema.string().role("textarea").description("页头HTML"),
    footer: import_koishi.Schema.string().role("textarea").description("页脚HTML"),
    customCss: import_koishi.Schema.string().role("textarea").description("自定义CSS")
  }).description("页面配置"),
  import_koishi.Schema.object({
    primary: import_koishi.Schema.string().description("主色调").role("color").default("rgba(139, 92, 246, 1)"),
    secondary: import_koishi.Schema.string().description("副色调").role("color").default("rgba(56, 189, 248, 1)"),
    bgColor: import_koishi.Schema.string().description("背景色").role("color").default("rgba(255, 255, 255, 1)"),
    textColor: import_koishi.Schema.string().description("文本色").role("color").default("rgba(100, 116, 139, 1)")
  }).description("颜色配置")
]);
function apply(ctx, config) {
  const files = new FileStore(ctx.baseDir);
  const render = new Render();
  const extract = new Extract(ctx);
  ctx.schema.extend("command", import_koishi.Schema.object({
    hidden: import_koishi.Schema.computed(import_koishi.Schema.boolean()).description("在菜单中隐藏指令").default(false),
    group: import_koishi.Schema.computed(import_koishi.Schema.string()).description("指令分组名称").default("其它")
  }), 900);
  ctx.schema.extend("command-option", import_koishi.Schema.object({
    hidden: import_koishi.Schema.computed(import_koishi.Schema.boolean()).description("在菜单中隐藏选项").default(false)
  }), 900);
  ctx.before("command/execute", (argv) => {
    const { command, session } = argv;
    if (command["_actions"].length || !session.app.$commander.get("menu")) return;
    return session.execute({ name: "menu", args: [command.name] });
  });
  ctx.command("menu [cmd:string]", "显示指令菜单").userFields(["authority"]).option("hidden", "-H  显示所有指令和选项").option("clear", "-c  清理缓存并重新生成").option("page", "-p <page:number>  分页页码").action(async ({ session, options }, cmd) => {
    try {
      if (config.enableCache && options.clear) await files.clearCache(cmd);
      const locale = extract.locale(session);
      const commands = await getCommands(cmd, session, locale, options.hidden);
      if (cmd && commands.length === 0) return `找不到指令 ${cmd}`;
      if (commands.length === 0) return "暂无可用指令";
      const renderConfig = {
        ...config,
        fontUrl: config.fontlink ? files.resolve(config.fontlink) : void 0,
        bgImage: config.bgimg ? files.resolve(config.bgimg) : void 0
      };
      const cacheKey = files.generateCacheKey(commands, { ...renderConfig, page: options.page, format: config.imageFormat }, cmd);
      const ext = config.imageFormat === "jpeg" ? "jpg" : "png";
      if (config.enableCache && !options.clear) {
        const cached = await files.getCache(cacheKey, ext);
        if (cached) return import_koishi.h.image(cached, `image/${config.imageFormat}`);
      }
      const html = render.build(renderConfig, commands, options.page, config.pageSize, cmd);
      const buffer = await toImage(html, config.imageFormat);
      if (config.enableCache) await files.saveCache(cacheKey, buffer, ext);
      return import_koishi.h.image(buffer, `image/${config.imageFormat}`);
    } catch (error) {
      logger.error("渲染失败:", error);
      return "渲染菜单失败";
    }
  });
  const toImage = /* @__PURE__ */ __name(async (html, format) => {
    const page = await ctx.puppeteer.page();
    try {
      await page.setContent(html);
      const element = await page.$(".container");
      return await element.screenshot({
        type: format,
        quality: format === "jpeg" ? 70 : void 0,
        omitBackground: format === "png"
      });
    } finally {
      await page.close();
    }
  }, "toImage");
  async function getCommands(cmdName, session, locale, showHidden = false) {
    try {
      if (config.source === "inline") {
        const commands = cmdName ? await extract.related(session, cmdName, locale) : await extract.all(session, locale);
        return extract.filter(commands, showHidden, !!cmdName);
      }
      let allCommands = await files.load("commands", locale);
      if (!allCommands) {
        allCommands = await extract.all(session, locale);
        if (allCommands?.length) await files.save("commands", allCommands, locale);
      }
      if (!cmdName) return extract.filter(allCommands || [], showHidden, false);
      let found = findCommand(allCommands || [], cmdName);
      if (!found) {
        found = await extract.single(session, cmdName, locale);
        if (found) {
          const exists = (allCommands || []).some((c) => c?.name === found.name);
          if (!exists) {
            allCommands = allCommands || [];
            allCommands.push(found);
            await files.save("commands", allCommands, locale);
          }
        }
      }
      return found ? extract.filter([found], showHidden, true) : [];
    } catch (error) {
      logger.error("获取指令数据失败:", error);
      return [];
    }
  }
  __name(getCommands, "getCommands");
  function findCommand(commands, nameOrAlias) {
    if (!Array.isArray(commands) || !nameOrAlias?.trim()) return null;
    for (const cmd of commands) {
      if (!cmd || typeof cmd !== "object" || !cmd.name) continue;
      if (Array.isArray(cmd.name)) {
        for (const nameItem of cmd.name) if (nameItem?.name === nameOrAlias && nameItem?.enabled) return cmd;
      }
      if (cmd.subs && Array.isArray(cmd.subs)) {
        for (const sub of cmd.subs) {
          if (!sub || typeof sub !== "object" || !sub.name) continue;
          if (Array.isArray(sub.name)) {
            for (const nameItem of sub.name) if (nameItem?.name === nameOrAlias && nameItem?.enabled) return sub;
          }
        }
      }
    }
    return null;
  }
  __name(findCommand, "findCommand");
}
__name(apply, "apply");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Config,
  apply,
  inject,
  logger,
  name,
  usage
});
