// koishi-plugin-word-core/lib/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __export = (target, all) => {
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Config: () => Config2,
  apply: () => apply2,
  inject: () => inject,
  logger: () => logger,
  name: () => name,
  usage: () => usage,
  word: () => word
});
module.exports = __toCommonJS(src_exports);
var import_koishi5 = require("koishi");

// src/core/word.ts
var import_koishi2 = require("koishi");

// src/core/service/tools/tools.ts
var import_koishi = require("koishi");
var readDBFunction = /* @__PURE__ */ __name(async (ctx, dbName, key) => {
  const readObjTemp = await ctx.database.get(dbName, key);
  if (readObjTemp.length <= 0) {
    return {};
  }
  return readObjTemp[0].data;
}, "readDBFunction");
var writeDBFunction = /* @__PURE__ */ __name(async (ctx, dbName, key, newData) => {
  const readObjTemp = await ctx.database.get(dbName, key);
  if (readObjTemp.length <= 0) {
    await ctx.database.create(dbName, { id: key, data: newData });
    return true;
  } else {
    await ctx.database.set(dbName, key, { data: newData });
    return true;
  }
}, "writeDBFunction");
var getDBFunction = /* @__PURE__ */ __name(async (ctx, dbName) => {
  const dbData = await ctx.database.get(dbName, { id: { $regex: /^[\s\S]+$/ } });
  const data = {
    idList: dbData.map((v) => {
      return v.id;
    }),
    dataList: dbData.map((v) => {
      const temp = v.data;
      return temp;
    })
  };
  return data;
}, "getDBFunction");
var removeDBFunction = /* @__PURE__ */ __name(async (ctx, dbName, key) => {
  ctx.database.remove(dbName, key);
}, "removeDBFunction");
var random = new import_koishi.Random(() => Math.random());
var randomNumber = /* @__PURE__ */ __name((minNumber, maxNumber) => {
  if (minNumber > maxNumber) {
    return Math.floor(Math.random() * (minNumber - maxNumber + 1)) + maxNumber;
  } else {
    return Math.floor(Math.random() * (maxNumber - minNumber + 1)) + minNumber;
  }
}, "randomNumber");

// src/core/service/user/user.ts
var _User = class _User {
  constructor(readDBTools, writeDBTools) {
    this.configTempData = {};
    this.readDBTools = readDBTools;
    this.writeDBTools = writeDBTools;
  }
  // private tempData: Record<string, Record<string, Record<string, number>>> = {};
  /**
   * 获取用户背包数据
   * @param uid 
   * @returns 背包数据
   */
  async getData(uid) {
    const data = await this.readDBTools("wordUserPackData", uid);
    const a = data;
    return a;
  }
  /**
   * 保存数据为用户数据
   * @param uid 用户id
   * @param data 完整用户数据
   * @returns 
   */
  async updateData(uid, data) {
    const backData = await this.writeDBTools("wordUserPackData", uid, data);
    return backData;
  }
  /**
   * 获取用户背包数据
   * @param uid 用户id
   * @param cell 存储格
   * @param itemName 物品名称
   * @returns 数量
   */
  async getItem(uid, cell, itemName) {
    const data = await this.getData(uid);
    if (!data[cell]) {
      return null;
    }
    if (!data[cell][itemName]) {
      return null;
    }
    return data[cell][itemName];
  }
  /**
   * 强制保存物品数量(一般不使用)
   * @param uid 用户id
   * @param cell 存储格
   * @param itemName 物品名称
   * @param itemData 数量
   * @returns 
   */
  async updateItem(uid, cell, itemName, itemData) {
    const data = await this.getData(uid);
    if (!data[cell]) {
      data[cell] = {};
    }
    data[cell][itemName] = itemData;
    return await this.updateData(uid, data);
  }
  /**
   * 获取用户编辑指针
   * @param uid 用户id
   * @returns 
   */
  async getEditWord(uid) {
    const userConfig = await this.readDBTools("wordUserConfig", uid);
    if (!userConfig.nowEdit) {
      userConfig.nowEdit = ["default"];
    }
    return userConfig.nowEdit[0];
  }
  /**
   * 修改编辑目标
   * @param uid 用户id
   * @param newDB 编辑目标
   * @returns 
   */
  async setEditWord(uid, newDB) {
    const userConfig = await this.readDBTools("wordUserConfig", uid);
    userConfig.nowEdit = [newDB];
    return await this.writeDBTools("wordUserConfig", uid, userConfig);
  }
  /**
   * 在配置中读取设置
   * @param uid 
   * @returns 
   */
  async getConfig(uid) {
    let data;
    if (this.configTempData.hasOwnProperty(uid)) {
      data = this.configTempData[uid];
    } else {
      data = await this.readDBTools("wordUserTemp", uid);
    }
    const a = data;
    this.configTempData[uid] = a;
    return this.configTempData[uid];
  }
  // 对某一项进行键值对设置
  async setConfig(uid, key, value) {
    const configData = await this.getConfig(uid);
    configData[key] = value;
    this.configTempData[uid] = configData;
  }
  // 强制对某一项进行键值对设置
  async setConfigForce(uid, key, value) {
    const configData = await this.getConfig(uid);
    configData[key] = value;
    this.configTempData[uid] = configData;
    const ok = await this.saveUserConfig(uid, configData);
    if (ok) {
      delete this.configTempData[uid];
    } else {
      return;
    }
  }
  // 保存设置
  async saveConfig() {
    const errList = {};
    for (const data in this.configTempData) {
      const bl = await this.saveUserConfig(data, this.configTempData[data]);
      if (!bl) {
        errList[data] = this.configTempData[data];
      }
    }
    this.configTempData = errList;
    if (JSON.stringify(this.configTempData) == "{}") {
      return true;
    } else {
      return false;
    }
  }
  // 存储某用户的数据
  async saveUserConfig(uid, data) {
    const backData = await this.writeDBTools("wordUserTemp", uid, data);
    return backData;
  }
};
__name(_User, "User");
var User = _User;

// src/core/extend/trigger.ts
var trigger = {
  "(@)": { reg: ['\\s*<at\\s+(?:id=".+"\\s+?)?name="([^"]*)"\\s*\\/>\\s*'], id: "name" },
  "(#)": { reg: ['\\s*<at\\s+id="(.+?)"\\s*(?:name="[^"]*"\\s*)?\\/>\\s*', '\\s*<at id=\\"([\\s\\S]+?)\\"\\/>\\s*'], id: "id" }
};
var addTrigger = /* @__PURE__ */ __name((triggerName, replaceStr, matchReg) => {
  if (!trigger[replaceStr]) {
    trigger[replaceStr] = {
      reg: [],
      id: triggerName
    };
  }
  trigger[replaceStr].reg.push(matchReg);
  return trigger;
}, "addTrigger");
var rmTrigger = /* @__PURE__ */ __name((replaceStr, matchReg) => {
  if (!trigger[replaceStr]) {
    return trigger;
  }
  if (!trigger[replaceStr].reg.includes(matchReg)) {
    return trigger;
  }
  const index = trigger[replaceStr].reg.indexOf(matchReg);
  trigger[replaceStr].reg = trigger[replaceStr].reg.splice(index, 1);
  return trigger;
}, "rmTrigger");

// src/core/service/editor/cache.ts
var wordCache = {
  hasKey: {},
  idList: [],
  grammarKeys: [],
  normalKeys: []
};
function getCache() {
  return wordCache;
}
__name(getCache, "getCache");
function isRegexSyntax(str) {
  const regexSyntax = /[.*+?^${}()|[\]\\]/;
  if (!regexSyntax.test(str)) {
    return false;
  }
  try {
    new RegExp(str);
    return true;
  } catch (e) {
    return false;
  }
}
__name(isRegexSyntax, "isRegexSyntax");
var cacheRefresh = /* @__PURE__ */ __name(async (getDBTools) => {
  const triggerKeys = Object.keys(trigger);
  wordCache = {
    hasKey: {},
    idList: [],
    grammarKeys: [],
    normalKeys: []
  };
  const { dataList, idList } = await getDBTools("wordData");
  const dataListTemp = dataList;
  wordCache.idList = idList;
  dataListTemp.forEach((item) => {
    const itemTemp = item.data;
    const itemName = item.name;
    Object.keys(itemTemp).forEach((v) => {
      if (!wordCache.hasKey[v]) {
        wordCache.hasKey[v] = [];
      }
      if (!wordCache.hasKey[v].includes(itemName)) wordCache.hasKey[v].push(itemName);
      for (let a of triggerKeys) {
        if (v.includes(a) || isRegexSyntax(v)) {
          if (!wordCache.grammarKeys.includes(v)) {
            wordCache.grammarKeys.push(v);
          }
          return;
        }
      }
      if (!wordCache.normalKeys.includes(v)) {
        wordCache.normalKeys.push(v);
      }
    });
  });
  return wordCache;
}, "cacheRefresh");
var addCache = /* @__PURE__ */ __name((q, wordName) => {
  if (Object.keys(wordCache.hasKey).includes(q)) {
    wordCache.hasKey[q].push(wordName);
  } else {
    wordCache.hasKey[q] = [wordName];
  }
  const triggerKeys = Object.keys(trigger);
  for (let a of triggerKeys) {
    if (q.includes(a) || isRegexSyntax(q)) {
      if (!wordCache.grammarKeys.includes(q)) {
        wordCache.grammarKeys.push(q);
      }
      return;
    }
  }
  if (!wordCache.normalKeys.includes(q)) {
    wordCache.normalKeys.push(q);
  }
}, "addCache");
var rmCache = /* @__PURE__ */ __name((q, wordName) => {
  if (Object.keys(wordCache.hasKey).includes(q)) {
    const index = wordCache.hasKey[q].indexOf(wordName);
    wordCache.hasKey[q].splice(index, 1);
    if (wordCache.hasKey[q].length <= 0) {
      delete wordCache.hasKey[q];
    }
  }
  wordCache.normalKeys = wordCache.normalKeys.filter((item) => item !== q);
  wordCache.grammarKeys = wordCache.grammarKeys.filter((item) => item !== q);
}, "rmCache");

// src/core/service/editor/editor.ts
var import_form_data = __toESM(require("form-data"));
var import_axios = __toESM(require("axios"));
var _Editor = class _Editor {
  constructor(tools, cache) {
    this.res = {};
    this.readDB = tools.readDB;
    this.writeDB = tools.writeDB;
    this.getDB = tools.getDB;
    this.removeDB = tools.removeDB;
    this.addCache = cache.addCache;
    this.rmCache = cache.rmCache;
    this.refreshCache = cache.cacheRefresh;
    this.getCache = cache.getCache;
    this.tools = tools;
  }
  /**
   * 读取词库
   * @param name 目标词库
   * @returns 结果
   */
  async readWord(name2) {
    const readData = await this.readDB("wordData", name2);
    let out;
    if (Object.keys(readData).length === 0) {
      out = {
        name: name2,
        saveDB: "default",
        author: [],
        data: {}
      };
    } else {
      out = readData;
    }
    return out;
  }
  /**
   * 更新词库
   * @param name 目标词库
   * @param data 词库数据
   * @returns 结果
   */
  async updateWord(name2, data) {
    return await this.writeDB("wordData", name2, data);
  }
  // 写入数据到回收站
  async writeRecycleBin(name2, data) {
    return await this.writeDB("recycleBinList", name2, data);
  }
  // 读一个回收站词库内容
  async readRecycleBin(name2) {
    return await this.readDB("recycleBinList", name2);
  }
  /**
   * 读取回收站列表
   * @returns idList
   */
  async getRecycleBinList() {
    const list = await this.getDB("recycleBinList");
    return list.idList;
  }
  /**
   * 读取普通词库列表
   * @returns idList
   */
  async getWordList() {
    const list = await this.getDB("wordData");
    return list.idList;
  }
  /**
   * 删除词库
   * @param name 需要删除的词库名
   * @returns 结果
   */
  async removeWord(name2) {
    const wordList = await this.getWordList();
    if (!wordList.includes(name2)) {
      return "词库列表不存在此词库";
    }
    const data = await this.readWord(name2);
    this.writeRecycleBin(name2, data);
    this.removeDB("wordData", name2);
    return "ok";
  }
  /**
   * 恢复词库
   * @param name 被恢复的词库名
   * @returns 返回结果
   */
  async restoreWord(name2) {
    const recycleBinList = await this.getRecycleBinList();
    if (!recycleBinList.includes(name2)) {
      return "回收站不存在此词库";
    }
    const data = await this.readRecycleBin(name2);
    this.updateWord(name2, data);
    await this.removeDB("recycleBinList", name2);
    return "ok";
  }
  /**
   * 设置属性
   * @param name 目标词库
   * @param key author|data|saveDB之一
   * @param value 设置为的值
   * @returns 结果
   */
  async setWordKeyValue(name2, key, value) {
    if (!(await this.getWordList()).includes(name2)) {
      return "无此词库";
    }
    const wordData = await this.readWord(name2);
    if (key == "author") {
      wordData.author = value;
    }
    if (key == "data") {
      wordData.data = value;
    }
    if (key == "saveDB") {
      wordData.saveDB = value;
    }
    return await this.updateWord(name2, wordData);
  }
  /**
   * 查询属性
   * @param name 目标词库 
   * @param key author|data|saveDB之一
   * @returns 结果
   */
  async readWordKeyValue(name2, key) {
    if (!(await this.getWordList()).includes(name2)) {
      return "无此词库";
    }
    const wordData = await this.readWord(name2);
    return wordData[key];
  }
  /**
   * 设置存储格
   * @param name 目标词库名
   * @param cell 存储格名
   * @param uid 操作者uid
   */
  async setSaveCell(name2, cell, uid) {
    const author = await this.isAuthor(name2, uid);
    if (!author) {
      return "你不是作者";
    }
    return await this.setWordKeyValue(name2, "saveDB", cell);
  }
  /**
   * 重置存储格
   * @param name 目标词库名
   * @param uid 操作者uid
   */
  async resetSaveCell(name2, uid) {
    const author = await this.isAuthor(name2, uid);
    if (!author) {
      return "你不是作者";
    }
    return await this.setWordKeyValue(name2, "saveDB", "default");
  }
  /**
   * 读取存储格子
   * @param name 目标词库名
   * @param uid 操作者uid
   */
  async readSaveCell(name2, uid) {
    return await this.readWordKeyValue(name2, "saveDB");
  }
  /**
   * 更新作者
   * @param name 目标词库名
   * @param authorList 更新作者列表
   */
  async updateAuthor(name2, authorList) {
    await this.setWordKeyValue(name2, "author", authorList);
  }
  /**
   * 读取作者
   * @param name 目标词库名
   * @returns 
   */
  async getAuthor(name2) {
    return await this.readWordKeyValue(name2, "author");
  }
  /**
   * 增加作者
   * @param name 目标词库名
   * @param uid 操作者uid
   * @param authorID 新增uid
   * @returns 
   */
  async addAuthor(name2, uid, authorID) {
    const author = await this.getAuthor(name2);
    if (!await this.isAuthor(name2, uid)) {
      return "您不是作者，无权操作";
    }
    if (!author.includes(authorID)) {
      author.push(authorID);
      await this.updateAuthor(name2, author);
      return "添加完成";
    } else {
      return "此作者已存在";
    }
  }
  /**
   * 减少作者
   * @param name 目标词库名
   * @param uid 操作者uid
   * @param authorID 去除uid
   * @returns 
   */
  async removeAuthor(name2, uid, authorID) {
    let author = await this.getAuthor(name2);
    if (!await this.isAuthor(name2, uid)) {
      return "您不是作者，无权操作";
    }
    if (!author.includes(authorID)) {
      return "此作者不存在";
    } else {
      author.splice(author.indexOf(name2), 1);
      await this.updateAuthor(name2, author);
      return "已删除作者";
    }
  }
  /**
   * 是否有作者权限
   * @param name 目标词库名
   * @param authorID 是否含有此uid
   * @returns 
   */
  async isAuthor(name2, authorID) {
    const author = await this.getAuthor(name2);
    if (!author.includes(authorID)) {
      return false;
    } else {
      return true;
    }
  }
  /**
   * 添加词条
   * @param name 目标词库
   * @param uid 操作者uid
   * @param q 触发词
   * @param a 回答
   * @returns 
   */
  async addWordItem(name2, uid, q, a) {
    const list = await this.getWordList();
    if (!list.includes(name2)) {
      const dataTemp = {};
      dataTemp[q] = [a];
      this.addCache(q, name2);
      if (await this.updateWord(name2, {
        name: name2,
        author: [uid],
        data: dataTemp,
        saveDB: "default"
      })) {
        return 1;
      }
    }
    const wordData = await this.readWord(name2);
    if (!await this.isAuthor(name2, uid)) {
      return "您不是作者，无权操作";
    }
    if (!Object.keys(wordData.data).includes(q)) {
      wordData.data[q] = [a];
      this.addCache(q, name2);
    } else {
      wordData.data[q].push(a);
    }
    const index = wordData.data[q].length;
    await this.updateWord(name2, wordData);
    return index;
  }
  /**
   * 删除词条
   * @param name 目标词库
   * @param uid 操作者uid
   * @param q 触发词
   * @param a 删除序号(从1开始，为'all'时删除此触发)
   * @returns 
   */
  async rmWordItem(name2, uid, q, a) {
    const wordData = await this.readWord(name2);
    if (!await this.isAuthor(name2, uid)) {
      return "您不是作者，无权操作";
    }
    if (!Object.keys(wordData.data).includes(q)) {
      return "此触发词不存在";
    } else {
      if (a == "all") {
        delete wordData.data[q];
        this.rmCache(q, name2);
      } else {
        if (a > wordData.data[q].length) {
          return "超过触发词的回答上限";
        }
        wordData.data[q].splice(a - 1, 1);
        if (wordData.data[q].length == 0) {
          delete wordData.data[q];
          this.rmCache(q, name2);
        }
      }
    }
    await this.updateWord(name2, wordData);
    return "over";
  }
  /**
   * 查找词条（查询某触发词所在的词库）
   * @param q 触发词 
   * @returns 
   */
  async getQuestion(q) {
    const hasKey = this.getCache().hasKey;
    const questionList = Object.keys(hasKey);
    const matchedString = questionList.find((regText) => {
      const reg = new RegExp(`^${regText}$`);
      return reg.test(q);
    });
    if (!matchedString) {
      return [];
    } else {
      return hasKey[q];
    }
  }
  // 获取云端插件序列
  async getCloudWordList() {
    const fetchDataTemp = await fetch(`${this.tools.url}/getList`);
    const fetchData = await fetchDataTemp.json();
    this.res = fetchData;
  }
  // 云端下载
  async getCloudWord(name2) {
    if (JSON.stringify(this.res) == "{}") {
      await this.getCloudWordList();
    }
    const pluginData = this.res[name2];
    if (!pluginData) {
      return "获取的插件格式异常";
    }
    const authorId = pluginData.authorId;
    const pluginName = pluginData.name;
    try {
      const url = `${this.tools.url}/getPlugin/${authorId}/${pluginName}.json`;
      const fetchDataTemp = await fetch(url);
      const fetchData = await fetchDataTemp.json();
      const a = await this.getWordList();
      const getDBName = fetchData.name;
      if (a.includes(getDBName)) {
        return "词库已存在，无法安装";
      }
      this.updateWord(getDBName, fetchData);
      this.getCache();
      return "ok";
    } catch (err) {
      console.log(err);
      return "获取的插件格式异常";
    }
  }
  // 云端上传
  async updateCloudWord(pluginData) {
    const formData = new import_form_data.default();
    const a = await this.readWord(pluginData.dbName);
    if (a.author.length == 0) {
      return;
    }
    const file = Buffer.from(JSON.stringify(a));
    formData.append("file", file, { filename: `${pluginData.name}.json` });
    formData.append("tag", JSON.stringify(pluginData.tag));
    Object.keys(pluginData).forEach((v) => {
      const a2 = v;
      if (v == "file" || v == "tag") {
        return;
      }
      {
        formData.append(a2, pluginData[a2]);
      }
    });
    await import_axios.default.post(`${this.tools.url}/newPlugin`, formData);
  }
};
__name(_Editor, "Editor");
var Editor = _Editor;

// src/core/service/permission/permission.ts
var _Permission = class _Permission {
  constructor(readDB, writeDB) {
    this.readDB = readDB;
    this.writeDB = writeDB;
  }
  /**
   * 为某人添加权限
   * @param uid 被操作者uid
   * @param newPermission 权限树
   * @returns 
   */
  async add(uid, newPermission) {
    const permissionData = await this.readDB("wordUserConfig", uid);
    if (!permissionData.permission) {
      permissionData.permission = [];
    }
    if (permissionData.permission.includes(newPermission)) {
      return "已存在此权限";
    }
    permissionData.permission.push(newPermission);
    return await this.writeDB("wordUserConfig", uid, permissionData);
  }
  /**
   * 删除某人的权限
   * @param uid 被操作者uid
   * @param permission 权限树
   * @returns 
   */
  async rm(uid, permission) {
    const permissionData = await this.readDB("wordUserConfig", uid);
    if (!permissionData.permission) {
      permissionData.permission = [];
    }
    const index = permissionData.permission.indexOf(permission);
    if (index <= -1) {
      return "不存在此权限";
    }
    permissionData.permission.splice(index, 1);
    return await this.writeDB("wordUserConfig", uid, permissionData);
  }
  /**
   * 判断某人是否拥有某权限
   * @param uid 查询目标uid
   * @param permission 权限树
   * @returns 
   */
  async isHave(uid, permission) {
    const permissionData = await this.readDB("wordUserConfig", uid);
    if (!permissionData.permission) {
      permissionData.permission = [];
    }
    if (permissionData.permission.includes(permission)) {
      return true;
    }
    const arr = permission.split(".");
    for (const node of permissionData.permission) {
      const nodeList = node.split(".");
      for (let index = 0; index < nodeList.length; index++) {
        if (nodeList[index] === "*") {
          return true;
        }
        if (nodeList[index] !== arr[index]) {
          break;
        }
      }
    }
    return false;
  }
  /**
   * 查看某人的全部权限
   * @param uid 被操作者uid
   * @returns 权限树
   */
  async all(uid) {
    const permissionData = await this.readDB("wordUserConfig", uid);
    if (!permissionData.permission) {
      permissionData.permission = [];
    }
    return permissionData.permission;
  }
  // 增加权限组？
  // 减少权限组？
};
__name(_Permission, "Permission");
var Permission = _Permission;

// src/core/extend/statement.ts
var statement = {};
var ifStatement = {};
var addStatement = /* @__PURE__ */ __name((ruleTrigger, callback, rule) => {
  if (!statement[ruleTrigger] || !ifStatement[ruleTrigger]) {
    if (rule) {
      ifStatement[ruleTrigger] = {
        func: callback,
        rule
      };
    } else {
      statement[ruleTrigger] = {
        func: callback
      };
    }
    return statement;
  } else {
    console.log(`语法包:${ruleTrigger}已被注册`);
    return statement;
  }
}, "addStatement");
var rmStatement = /* @__PURE__ */ __name((ruleTrigger) => {
  delete statement[ruleTrigger];
  return statement;
}, "rmStatement");

// src/core/service/config/config.ts
var _Config = class _Config {
  constructor(tools, cache) {
    this.readDB = tools.readDB;
    this.writeDB = tools.writeDB;
    this.getDB = tools.getDB;
    this.removeDB = tools.removeDB;
    this.addCache = cache.addCache;
    this.rmCache = cache.rmCache;
    this.refreshCache = cache.cacheRefresh;
  }
  /**
   * 获取配置
   * @param key 配置的键
   * @returns 
   */
  async getConfig(key) {
    const readData = await this.readDB("wordCoreConfig", key);
    let out;
    if (Object.keys(readData).length === 0) {
      out = {};
    } else {
      out = readData;
    }
    return out;
  }
  /**
   * 更新(新建)配置
   * @param key 配置的键
   * @param value 配置的值
   * @returns 
   */
  async updateConfig(key, value) {
    return await this.writeDB("wordCoreConfig", key, value);
  }
};
__name(_Config, "Config");
var Config = _Config;

// src/core/service/index.ts
var _wordService = class _wordService {
  constructor(ctx) {
    this.Tools = {};
    this.User = {};
    this.Cache = {};
    this.Editor = {};
    this.Permission = {};
    this.trigger = {};
    this.statement = {};
    this.config = {};
    this.ctx = ctx;
    this.ctx.inject(["database"], async (ctx2) => {
      this.Tools.readDB = (dbName, key) => {
        return readDBFunction(ctx2, dbName, key);
      };
      this.Tools.writeDB = (dbName, key, data) => {
        return writeDBFunction(ctx2, dbName, key, data);
      };
      this.Tools.getDB = (dbName) => {
        return getDBFunction(ctx2, dbName);
      };
      this.Tools.removeDB = async (dbName, key) => {
        return removeDBFunction(ctx2, dbName, key);
      };
      this.Tools.randomNumber = randomNumber;
    });
    while (!this.Tools.readDB) {
    }
    const user = new User(this.Tools.readDB, this.Tools.writeDB);
    this.User.getData = (uid) => {
      return user.getData(uid);
    };
    this.User.updateData = (uid, data) => {
      return user.updateData(uid, data);
    };
    this.User.getItem = (uid, cell, item) => {
      return user.getItem(uid, cell, item);
    };
    this.User.getEditWord = (uid) => {
      return user.getEditWord(uid);
    };
    this.User.setEditWord = (uid, newDB) => {
      return user.setEditWord(uid, newDB);
    };
    this.User.updateItem = (uid, cell, itemName, amount) => {
      return user.updateItem(uid, cell, itemName, amount);
    };
    this.User.getConfig = (uid) => {
      return user.getConfig(uid);
    };
    this.User.setConfig = (uid, key, value) => {
      return user.setConfig(uid, key, value);
    };
    this.User.setConfigForce = (uid, key, value) => {
      return user.setConfigForce(uid, key, value);
    };
    this.User.saveConfig = () => {
      return user.saveConfig();
    };
    this.Cache.getCache = getCache;
    this.Cache.cacheRefresh = () => {
      return cacheRefresh(this.Tools.getDB);
    };
    this.Cache.nowCache = wordCache;
    this.Cache.rmCache = rmCache;
    this.Cache.addCache = addCache;
    this.Permission = new Permission(this.Tools.readDB, this.Tools.writeDB);
    this.Editor = new Editor(this.Tools, this.Cache);
    this.trigger = {
      trigger,
      addTrigger,
      rmTrigger
    };
    this.statement = {
      statement,
      ifStatement,
      rmStatement,
      addStatement
    };
    let config = new Config(this.Tools, this.Cache);
    this.config.getConfig = (key) => {
      return config.getConfig(key);
    };
    this.config.updateConfig = (key, value) => {
      return config.updateConfig(key, value);
    };
  }
};
__name(_wordService, "wordService");
var wordService = _wordService;

// src/core/driver/src/index.ts
var saveItemDataTemp = {};
var next = /* @__PURE__ */ __name((data) => {
  return { status: "next", data };
}, "next");
var end = /* @__PURE__ */ __name((data) => {
  return { status: "end", data };
}, "end");
var kill = /* @__PURE__ */ __name((data) => {
  return { status: "kill", data };
}, "kill");
var parPack = {
  next,
  end,
  kill
};
var funcPackKeys = Object.keys(statement);
var ifFuncPackKeys = Object.keys(ifStatement);
var parsStart = /* @__PURE__ */ __name(async (questionList, wordData, word2, session, matchList) => {
  funcPackKeys = Object.keys(statement);
  ifFuncPackKeys = Object.keys(ifStatement);
  const tree = getTree(questionList);
  if (!questionList) {
    return null;
  }
  const msg = [];
  let userData = {};
  let oldUserData = {};
  for (let needParMsg of tree) {
    let over;
    if (Array.isArray(needParMsg)) {
      over = await parseTrees(questionList, word2, needParMsg, session, wordData, !matchList ? {} : matchList, JSON.parse(JSON.stringify(userData)));
      if (over) {
        msg.push(over.message);
        oldUserData = userData;
        userData = over.data;
      } else {
        userData = oldUserData;
      }
    } else {
      msg.push(needParMsg);
    }
  }
  const msgOut = msg.join("");
  if (msgOut) {
    return { message: msgOut, data: userData };
  } else {
    if (userData.hasOwnProperty("item")) {
      return { message: "", data: userData };
    } else {
      return null;
    }
  }
}, "parsStart");
var getTree = /* @__PURE__ */ __name((str) => {
  let parseStr = str;
  const par = /* @__PURE__ */ __name(() => {
    let tempArr = [];
    let index = 0;
    while (parseStr.length > 0) {
      const v = parseStr[0];
      parseStr = parseStr.slice(1);
      if (v == "(") {
        if (!tempArr[index]) {
          tempArr[index] = [];
        }
        const length = tempArr[index].length;
        if (typeof tempArr[index][length - 1] == "string") {
          if (tempArr[index][length - 1].endsWith("\\")) {
            tempArr[index][length - 1] = tempArr[index][length - 1].slice(0, -1);
            tempArr[index][length - 1] += v;
          } else {
            tempArr[index].push(par());
          }
        } else {
          tempArr[index].push(par());
        }
      } else if (v == ")") {
        if (!tempArr[index]) {
          tempArr[index] = [];
        }
        const length = tempArr[index].length;
        if (typeof tempArr[index][length - 1] == "string") {
          if (tempArr[index][length - 1]?.endsWith("\\")) {
            tempArr[index][length - 1] = tempArr[index][length - 1].slice(0, -1);
            tempArr[index][length - 1] += v;
          } else {
            return tempArr;
          }
        } else {
          return tempArr;
        }
      } else if (v == ":") {
        if (!tempArr[index]) {
          tempArr[index] = [""];
        }
        const length = tempArr[index].length;
        if (Array.isArray(tempArr[index][length - 1])) {
          index++;
        } else if (!funcPackKeys.includes(tempArr[0][0]) && !ifFuncPackKeys.includes(tempArr[0][0]) || (tempArr[index][length - 1].endsWith("http") || tempArr[index][length - 1].endsWith("https") || tempArr[index][length - 1].endsWith("\\"))) {
          const lastList = tempArr[index][length - 1];
          if (Array.isArray(lastList)) {
            tempArr[index].push(v);
          } else if (tempArr[index][length - 1].endsWith("\\")) {
            tempArr[index][length - 1] = tempArr[index][length - 1].slice(0, -1);
            tempArr[index][length - 1] += v;
          } else {
            tempArr[index][length - 1] += v;
          }
        } else {
          index++;
        }
      } else {
        if (!tempArr[index]) {
          tempArr[index] = [""];
        }
        const length = tempArr[index].length;
        if (Array.isArray(tempArr[index][length - 1])) {
          tempArr[index].push(v);
        } else {
          tempArr[index][length - 1] += v;
        }
      }
    }
    return tempArr[0];
  }, "par");
  const aTemp = par();
  const a = aTemp.length > 0 ? aTemp : aTemp[0];
  const par2 = /* @__PURE__ */ __name((arr) => {
    const c = [];
    for (let d of arr) {
      if (Array.isArray(d)) {
        if (d.length == 1 && !Array.isArray(d[0])) {
          c.push(d[0]);
        } else {
          c.push(par2(d));
        }
      } else {
        c.push(d);
      }
    }
    return c;
  }, "par2");
  const b = par2(a);
  return b;
}, "getTree");
var parseTrees = /* @__PURE__ */ __name(async (questionList, word2, inData, session, wordData, matchList, inputUserData) => {
  const par = /* @__PURE__ */ __name(async (functonArray, data) => {
    let userDataTemp = data;
    let isIF = ifFuncPackKeys.includes(functonArray[0]);
    let rule = isIF ? ifStatement[functonArray[0]].rule : [0];
    if (rule[0] == "while") {
      const a = rule.slice(1);
      const b = functonArray.slice(1);
      rule = Array.from({ length: b.length }, (_, i) => a[i % a.length]);
    }
    for (let i = 0; i < functonArray.length; i++) {
      const canPar = isIF ? Array.isArray(functonArray[i]) && rule[i] == 1 : Array.isArray(functonArray[i]);
      if (canPar) {
        const a = await par(functonArray[i], JSON.parse(JSON.stringify(userDataTemp)));
        if (a) {
          functonArray[i] = a.message;
          userDataTemp = a.data;
        } else {
          saveItemDataTemp[questionList ? questionList : ""] = { item: {}, userConfig: {} };
          userDataTemp = { item: {}, userConfig: {} };
        }
      }
    }
    const getPar = /* @__PURE__ */ __name(async () => {
      const which = functonArray[0];
      const arg = functonArray.slice(1);
      const matchs = matchList;
      const canPar = funcPackKeys.includes(which) && isIF == false || ifFuncPackKeys.includes(which) && isIF == true;
      if (canPar) {
        const overPar = await parStatement(which, {
          args: arg,
          matchs,
          wordData,
          parPack,
          internal: {
            // 缓存功能
            saveItem: /* @__PURE__ */ __name(async (uid, saveDB, itemName, number) => {
              if (!userDataTemp.item) {
                userDataTemp.item = {};
              }
              if (!userDataTemp.item[uid]) {
                userDataTemp.item[uid] = {};
              }
              if (!userDataTemp.item[uid][saveDB]) {
                userDataTemp.item[uid][saveDB] = {};
              }
              userDataTemp.item[uid][saveDB][itemName] = number;
              saveItemDataTemp[questionList ? questionList : ""] = userDataTemp;
              return true;
            }, "saveItem"),
            getItem: /* @__PURE__ */ __name(async (uid, saveDB, itemName) => {
              const num = await word2.user.getItem(uid, saveDB, itemName);
              if (Array.isArray(num) && num) {
                return kill("该物品是列表型");
              }
              if (!userDataTemp.item) {
                userDataTemp.item = {};
              }
              if (!userDataTemp.item[uid]) {
                userDataTemp.item[uid] = {};
              }
              if (!userDataTemp.item[uid][saveDB]) {
                userDataTemp.item[uid][saveDB] = {};
              }
              if (!userDataTemp.item[uid][saveDB][itemName] && userDataTemp.item[uid][saveDB][itemName] != 0) {
                userDataTemp.item[uid][saveDB][itemName] = num ? num : 0;
              }
              saveItemDataTemp[questionList ? questionList : ""] = userDataTemp;
              userDataTemp.item[uid][saveDB][itemName] = userDataTemp.item[uid][saveDB][itemName];
              return userDataTemp.item[uid][saveDB][itemName];
            }, "getItem"),
            getUserConfig: /* @__PURE__ */ __name(async (uid, key) => {
              const userConfig = await word2.user.getConfig(uid);
              if (!userDataTemp.userConfig) {
                userDataTemp.userConfig = {};
              }
              if (!userDataTemp.userConfig[uid]) {
                userDataTemp.userConfig[uid] = {};
              }
              userDataTemp.userConfig[uid] = userConfig;
              saveItemDataTemp[questionList ? questionList : ""] = userDataTemp;
              return userDataTemp.userConfig[uid][key];
            }, "getUserConfig"),
            saveUserConfig: /* @__PURE__ */ __name(async (uid, key, value) => {
              if (!userDataTemp.userConfig) {
                userDataTemp.userConfig = {};
              }
              if (!userDataTemp.userConfig[uid]) {
                userDataTemp.userConfig[uid] = await word2.user.getConfig(uid);
              }
              userDataTemp.userConfig[uid][key] = value;
              saveItemDataTemp[questionList ? questionList : ""] = userDataTemp;
            }, "saveUserConfig"),
            removeUserConfig: /* @__PURE__ */ __name(async (uid, key) => {
              if (!userDataTemp.userConfig[uid]) {
                userDataTemp.userConfig[uid] = await word2.user.getConfig(uid);
              }
              if (!userDataTemp.userConfig[uid][key]) {
                return;
              }
              delete userDataTemp.userConfig[uid][key];
              saveItemDataTemp[questionList ? questionList : ""] = userDataTemp;
            }, "removeUserConfig"),
            getList: /* @__PURE__ */ __name(async (uid, saveDB, listName) => {
              const num = await word2.user.getItem(uid, saveDB, listName);
              if (!Array.isArray(num) && num) {
                return kill("该物品是数量型");
              }
              if (!userDataTemp.item) {
                userDataTemp.item = {};
              }
              if (!userDataTemp.item[uid]) {
                userDataTemp.item[uid] = {};
              }
              if (!userDataTemp.item[uid][saveDB]) {
                userDataTemp.item[uid][saveDB] = {};
              }
              if (!userDataTemp.item[uid][saveDB][listName] && !(Array.isArray(userDataTemp.item[uid][saveDB][listName]) && userDataTemp.item[uid][saveDB][listName].length === 0)) {
                userDataTemp.item[uid][saveDB][listName] = num ? num : [];
              }
              saveItemDataTemp[questionList ? questionList : ""] = userDataTemp;
              userDataTemp.item[uid][saveDB][listName] = userDataTemp.item[uid][saveDB][listName];
              return userDataTemp.item[uid][saveDB][listName];
            }, "getList"),
            saveList: /* @__PURE__ */ __name(async (uid, saveDB, itemName, list) => {
              if (!userDataTemp.item) {
                userDataTemp.item = {};
              }
              if (!userDataTemp.item[uid]) {
                userDataTemp.item[uid] = {};
              }
              if (!userDataTemp.item[uid][saveDB]) {
                userDataTemp.item[uid][saveDB] = {};
              }
              userDataTemp.item[uid][saveDB][itemName] = list;
              saveItemDataTemp[questionList ? questionList : ""] = userDataTemp;
              return true;
            }, "saveList")
          }
        }, session, isIF);
        if (Array.isArray(overPar)) {
          const a = await par(overPar, JSON.parse(JSON.stringify(userDataTemp)));
          if (a) {
            return { message: a.message, data: a.data };
          } else {
            saveItemDataTemp[questionList ? questionList : ""] = { item: {}, userConfig: {} };
            userDataTemp = { item: {}, userConfig: {} };
            return { message: "", data: userDataTemp };
          }
        } else if (overPar || overPar == "") {
          return { message: overPar, data: userDataTemp };
        } else {
          return null;
        }
      } else {
        return { message: functonArray.join(""), data: userDataTemp };
      }
    }, "getPar");
    const outData = getPar();
    return outData;
  }, "par");
  const aaa = await par(inData, JSON.parse(JSON.stringify(inputUserData)));
  return aaa;
}, "parseTrees");
var parStatement = /* @__PURE__ */ __name(async (which, toInData, session, isIF) => {
  let str;
  if (isIF) {
    str = await ifStatement[which].func(toInData, session);
  } else {
    str = await statement[which].func(toInData, session);
  }
  if (Array.isArray(str)) {
    return str;
  } else if (typeof str == "object") {
    const status = str.status;
    if (status == "end" || status == "next" || status == "kill") {
      const errorMsg = `${status}${str.data ? ":" + str.data : ""}`;
      throw new Error(errorMsg);
    }
    if (status == "killthis") {
      return null;
    }
  } else {
    return str ? str : "";
  }
  return "";
}, "parStatement");

// src/core/driver/index.ts
function parseTriggrtString(str, template, trigger2) {
  const matchList = {};
  const templatePattern = /\(([@#数])\)/g;
  const parts = [];
  let lastIndex = 0;
  let match;
  while ((match = templatePattern.exec(template)) !== null) {
    const symbol = match[1];
    const fullSymbol = `(${symbol})`;
    const triggerItem = trigger2[fullSymbol];
    if (!triggerItem) {
      throw new Error(`Missing trigger config for ${fullSymbol}`);
    }
    if (match.index > lastIndex) {
      parts.push({
        type: "text",
        value: template.slice(lastIndex, match.index)
      });
    }
    parts.push({
      type: "token",
      key: symbol,
      trigger: triggerItem
    });
    lastIndex = match.index + match[0].length;
  }
  if (lastIndex < template.length) {
    parts.push({
      type: "text",
      value: template.slice(lastIndex)
    });
  }
  let cursor = 0;
  for (const part of parts) {
    if (part.type === "text") {
      if (!str.startsWith(part.value, cursor)) {
        return null;
      }
      cursor += part.value.length;
    } else if (part.type === "token") {
      const { reg, id } = part.trigger;
      let matched = false;
      for (const regStr of reg) {
        const regex = new RegExp(regStr, "y");
        regex.lastIndex = cursor;
        const result = regex.exec(str);
        if (result && result[1] !== void 0) {
          const value = result[1];
          if (!matchList[id]) matchList[id] = [];
          matchList[id].push(value);
          cursor = regex.lastIndex;
          matched = true;
          break;
        }
      }
      if (!matched) {
        return null;
      }
    }
  }
  if (cursor !== str.length) return null;
  return matchList;
}
__name(parseTriggrtString, "parseTriggrtString");
var _wordDriver = class _wordDriver {
  constructor(word2, ctx) {
    this.ctx = ctx;
    this.word = word2;
  }
  /**
   * 开始解析当前的对话对象
   * @param session 当前对话对象
   * @returns 
   */
  async start(session, callback) {
    if (!session.content) {
      return;
    }
    let q = session.content;
    const wordCache2 = this.word.cache.getCache();
    let matchList = {};
    let list = wordCache2.hasKey[q];
    const triggerList = Object.keys(this.word.trigger.trigger);
    if (!wordCache2.normalKeys.includes(q)) {
      const testMsgGrammar = /* @__PURE__ */ __name(() => {
        let item = "";
        for (let a of wordCache2.grammarKeys) {
          let b = a;
          for (const key of triggerList) {
            for (const reg of this.word.trigger.trigger[key].reg) {
              item = b;
              const escapedKey = key.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
              item = item.replace(new RegExp(escapedKey, "g"), `${reg}`);
              const msgReg = new RegExp(`^${item}$`);
              if (msgReg.test(q)) {
                list = wordCache2.hasKey[a];
                return a;
              }
              b = item;
            }
          }
        }
      }, "testMsgGrammar");
      const grammarMssg = testMsgGrammar();
      if (!list) {
        return;
      }
      if (list.length <= 0) {
        return;
      }
      if (!grammarMssg) {
        return;
      }
      matchList = parseTriggrtString(q, grammarMssg, this.word.trigger.trigger) || {};
      q = grammarMssg;
    }
    if (!list) {
      return;
    }
    if (list.length <= 0) {
      return;
    }
    let primitiveList = list;
    const killWordListTemp = await this.word.config.getConfig("filtration");
    const channelId = session.channelId || "";
    const killWordList = killWordListTemp[channelId] ? killWordListTemp[channelId] : [];
    let overPrimitiveList = [];
    primitiveList.forEach((e) => {
      if (killWordList.includes(e)) {
        return;
      } else {
        overPrimitiveList.push(e);
      }
    });
    const witchWordDB = this.word.tools.randomNumber(0, overPrimitiveList.length - 1);
    const parsedList = [];
    let witchWord = 0;
    let needPar = "";
    const parOne = /* @__PURE__ */ __name(async () => {
      if (!q) {
        return;
      }
      const item = overPrimitiveList[witchWordDB];
      const wordData = await this.word.editor.readWord(item);
      const questionList = wordData.data[q];
      if (!questionList) {
        return;
      }
      do {
        if (parsedList.length >= questionList.length) {
          return;
        }
        witchWord = this.word.tools.randomNumber(0, questionList.length - 1);
      } while (parsedList.includes(witchWord));
      parsedList.push(witchWord);
      needPar = questionList[witchWord];
      const contentList = needPar.split("(换)");
      for (let i of contentList) {
        try {
          let abc = await this.parMsgPrivate(i, wordData, session, matchList);
          if (!abc) {
            continue;
          }
          const a = abc.message;
          const needSaveObj = abc.data;
          const needSave = needSaveObj.item;
          if (needSave) {
            for (let uid in needSave) {
              const saveDBList = needSave[uid];
              for (let saveDB in saveDBList) {
                const itemNameList = saveDBList[saveDB];
                for (let itemName in itemNameList) {
                  const num = itemNameList[itemName];
                  await this.word.user.updateItem(uid, saveDB, itemName, num);
                }
              }
            }
          }
          const needSaveConfig = needSaveObj.userConfig;
          if (needSaveConfig) {
            for (let uid in needSaveConfig) {
              for (let key in needSaveConfig[uid]) {
                await this.word.user.setConfig(uid, key, needSaveConfig[uid][key]);
              }
            }
          }
          const ok = await this.word.user.saveConfig();
          if (ok) {
            callback(a);
            continue;
          } else {
            callback(" [word-core] 数据保存失败");
            continue;
          }
        } catch (err) {
          console.log(err);
          const errorType = err.message;
          if (!errorType) {
            return err;
          }
          const msg = errorType.split(":")[1];
          if (errorType.startsWith("kill")) {
            if (msg) {
              callback(msg);
              continue;
            }
            continue;
          }
          if (errorType.startsWith("end")) {
            const needSaveObj = saveItemDataTemp[q];
            const needSave = needSaveObj.item;
            if (needSave) {
              for (let uid in needSave) {
                const saveDBList = needSave[uid];
                for (let saveDB in saveDBList) {
                  const itemNameList = saveDBList[saveDB];
                  for (let itemName in itemNameList) {
                    const num = itemNameList[itemName];
                    await this.word.user.updateItem(uid, saveDB, itemName, num);
                  }
                }
              }
            }
            const needSaveConfig = needSaveObj.userConfig;
            if (needSaveConfig) {
              for (let uid in needSaveConfig) {
                for (let key in needSaveConfig[uid]) {
                  await this.word.user.setConfig(uid, key, needSaveConfig[uid][key]);
                }
              }
            }
            const ok = await this.word.user.saveConfig();
            if (ok) {
              if (msg) {
                callback(msg);
                continue;
              }
              continue;
            } else {
              if (msg) {
                callback(" [word-core] 数据保存失败，并且" + msg);
                continue;
              }
              callback(" [word-core] 数据保存失败");
              continue;
            }
          }
          if (errorType.startsWith("next")) {
            return await parOne();
          }
        }
      }
    }, "parOne");
    await parOne();
  }
  /**
   * 词库本体解析文本
   * @param msg 需要解析的文本
   * @param wordSaveData 文本所在的词库配置，{data: xxx, author: xxx, saveDB: xxx, name: xxx}
   * @param session koishi的上下文Session，或是session的关键元素
   * @param matchList 输入匹配列表，可空
   * @returns 返回解析结果
   */
  async parMsgPrivate(msg, wordSaveData, session, matchList) {
    return await parsStart(msg, wordSaveData, this.word, session, matchList);
  }
  /**
     * 词库解析文本
     * @param msg 需要解析的文本
     * @param wordParConfig 文本所在的词库配置，{data?: xxx, author?: xxx, saveDB: xxx, name?: xxx}
     * @param session koishi的上下文Session，或是session的关键元素
     * @param matchList 输入匹配列表，可空
     * @returns 返回解析结果
     */
  async parMsg(msg, wordParConfig, session, matchList) {
    try {
      let abc = await this.parMsgPrivate(msg, wordParConfig, session, matchList);
      if (!abc) {
        return;
      }
      const a = abc.message;
      const needSaveObj = abc.data;
      const needSave = needSaveObj.item;
      if (needSave) {
        for (let uid in needSave) {
          const saveDBList = needSave[uid];
          for (let saveDB in saveDBList) {
            const itemNameList = saveDBList[saveDB];
            for (let itemName in itemNameList) {
              const num = itemNameList[itemName];
              await this.word.user.updateItem(uid, saveDB, itemName, num);
            }
          }
        }
      }
      const needSaveConfig = needSaveObj.userConfig;
      if (needSaveConfig) {
        for (let uid in needSaveConfig) {
          for (let key in needSaveConfig[uid]) {
            await this.word.user.setConfig(uid, key, needSaveConfig[uid][key]);
          }
        }
      }
      const ok = await this.word.user.saveConfig();
      if (ok) {
        return a;
      } else {
        return " [word-core] 数据保存失败";
      }
    } catch (err) {
      console.log(err);
      const errorType = err.message;
      if (!errorType) {
        return err;
      }
      const msg2 = errorType.split(":")[1];
      if (errorType.startsWith("kill")) {
        if (msg2) {
          return msg2;
        }
        return;
      }
      if (errorType.startsWith("next")) {
        return "";
      }
    }
  }
};
__name(_wordDriver, "wordDriver");
var wordDriver = _wordDriver;

// src/core/word.ts
var _word = class _word extends import_koishi2.Service {
  constructor(ctx, config) {
    super(ctx, "word", true);
    const WordDriver = new wordDriver(this, ctx);
    this.driver = WordDriver;
    const wordServiceTemp = new wordService(ctx);
    this.tools = wordServiceTemp.Tools;
    this.tools.url = config.searchEndpoint;
    this.cache = wordServiceTemp.Cache;
    this.editor = wordServiceTemp.Editor;
    this.user = wordServiceTemp.User;
    this.permission = wordServiceTemp.Permission;
    this.trigger = wordServiceTemp.trigger;
    this.statement = wordServiceTemp.statement;
    this.cache.cacheRefresh();
    this.config = wordServiceTemp.config;
  }
};
__name(_word, "word");
var word = _word;
((word2) => {
  word2.Config = import_koishi2.Schema.object({
    searchEndpoint: import_koishi2.Schema.string().description("词库插件市场后端地址").default("https://wplugin.reifuu.icu")
  });
})(word || (word = {}));

// src/index.ts
var import_path = require("path");

// src/core/index.ts
var core_exports = {};
__export(core_exports, {
  apply: () => apply,
  using: () => using
});
var using = ["database"];
var dbInit = /* @__PURE__ */ __name((ctx) => {
  ctx.model.extend("wordUserPackData", {
    id: "string",
    data: "json"
  }, {
    primary: "id"
  });
  ctx.model.extend("wordData", {
    id: "string",
    data: "json"
  }, {
    primary: "id"
  });
  ctx.model.extend("wordUserConfig", {
    id: "string",
    data: "json"
  }, {
    primary: "id"
  });
  ctx.model.extend("recycleBinList", {
    id: "string",
    data: "json"
  }, {
    primary: "id"
  });
  ctx.model.extend("wordUserTemp", {
    id: "string",
    data: "json"
  }, {
    primary: "id"
  });
  ctx.model.extend("wordUserTemp", {
    id: "string",
    data: "json"
  }, {
    primary: "id"
  });
  ctx.model.extend("wordCoreConfig", {
    id: "string",
    data: "json"
  }, {
    primary: "id"
  });
}, "dbInit");
var apply = /* @__PURE__ */ __name(async (ctx) => {
  dbInit(ctx);
}, "apply");

// src/sand-adapter/bot.ts
var import_koishi4 = require("koishi");

// src/sand-adapter/message.ts
var import_koishi3 = require("koishi");
var _SandboxMessenger = class _SandboxMessenger extends import_koishi3.MessageEncoder {
  constructor() {
    super(...arguments);
    this.buffer = "";
    this.rules = Object.fromEntries(["image", "img", "audio", "video", "file"].map((type) => {
      return [type, async (attrs) => {
        const src = attrs.src || attrs.url;
        const type1 = type === "image" ? "img" : type;
        if (src.startsWith("file:")) {
          return (0, import_koishi3.h)(type1, { ...attrs, src: `${this.bot.ctx.server.selfUrl}/sandbox/${src}` });
        }
        return (0, import_koishi3.h)(type1, { ...attrs, src });
      }];
    }));
  }
  async flush() {
    if (!this.buffer.trim()) return;
    const content = await import_koishi3.h.transformAsync(this.buffer.trim(), this.rules);
    const session = this.bot.session(this.session.event);
    session.messageId = import_koishi3.Random.id();
    this.bot.ctx.console.broadcast(
      "word-sanbox-request",
      content
    );
    if (session.event.message) {
      this.results.push(session.event.message);
    }
    this.buffer = "";
  }
  async visit(element) {
    const { type, attrs, children } = element;
    switch (type) {
      case "figure":
      case "message": {
        await this.flush();
        await this.render(children);
        await this.flush();
        break;
      }
      case "at": {
        let msg = "";
        if (attrs.name) {
          msg = attrs.name;
        }
        if (attrs.id) {
          msg = attrs.id;
        }
        this.buffer += `@${msg}`;
        break;
      }
      default: {
        this.buffer += element.toString();
        break;
      }
    }
  }
};
__name(_SandboxMessenger, "SandboxMessenger");
var SandboxMessenger = _SandboxMessenger;

// src/sand-adapter/bot.ts
var _wordBot = class _wordBot extends import_koishi4.Bot {
  constructor(ctx, config) {
    super(ctx, config, "word-sandbox");
    this.platform = "word-sandbox";
    this.selfId = "word-core";
    this.user = {
      id: "word-core",
      name: "word-core",
      nick: "Word Core"
    };
  }
  // 这里接下来引用的是
  // https://github.com/koishijs/webui/blob/main/plugins/sandbox/src/bot.ts#L13C10-L13C25
  createSession(msg) {
    const session = this.session({
      user: { id: "word-user", name: "word-user" }
    });
    session.type = "message";
    session.userId = "word-user";
    session.username = "word-user";
    session.channelId = "word-sandbox";
    session.isDirect = true;
    session.content = msg;
    this.dispatch(session);
  }
  async sendMessage(channelId, content, guildId, options) {
    await new SandboxMessenger(this, channelId).send(content);
    return [];
  }
  async sendPrivateMessage(userId, content, guildId, options) {
    await new SandboxMessenger(this, "private:word-sandbox").send(content);
    return [];
  }
};
__name(_wordBot, "wordBot");
var wordBot = _wordBot;
var bot_default = wordBot;

// src/index.ts
var name = "word-core";
var inject = {
  // required: ['database', 'console', 'notifier']
  required: ["database", "console"]
};
var Config2 = import_koishi5.Schema.object({
  EnableAt: import_koishi5.Schema.boolean().default(false).description("是否开启仅接收at机器人的消息"),
  masterID: import_koishi5.Schema.array(String).description("管理员的唯一标识").default(["word-user"]),
  searchEndpoint: import_koishi5.Schema.string().description("词库插件市场后端地址").default("https://wplugin.reifuu.icu"),
  telemetry: import_koishi5.Schema.boolean().description("告诉作者我正在使用词库~\n\n（求求大家开启这个选项！这对作者真的很重要！）").default(true)
});
var logger = new import_koishi5.Logger("Word-core");
var usage = `# 快速入门

这个插件是一个用于编辑小游戏，或者让机器人根据关键词回复的插件。

快速入门请查看这里：https://forum.koishi.xyz/t/topic/7663

# 备注
1. 词库的教程文档在这里：https://docs.reifuu.icu/
2. 如果有问题的话欢迎联系我，QQ：1946831552！
3. 词库的商店暂时无法主动提交，如果你有想提交的词库的话！
4. 请将wordData数据库内对应的词库的文本发送到我的邮箱中！请告诉我你的名字和提交的词库类型！
5. 邮箱：1946831552@qq.com

# 欢迎加入我们的群聊一起交流！QQ群：312762918`;
var apply2 = /* @__PURE__ */ __name(async (ctx, config) => {
  ctx.plugin(core_exports);
  ctx.plugin(word, { searchEndpoint: config.searchEndpoint });
  ctx.inject(["word"], async (ctx2) => {
    ctx2.command("word", "词库核心！");
    ctx2.command("word", "词库核心！").subcommand(".add <question:string> <answer:string>", "为一个触发词添加回复").usage("添加一个词库").example("word.add 你好 你也好").action(async ({ session }, question, answer) => {
      if (!session) {
        return;
      }
      if (!question) {
        return `<at name="${session.username}" /> 你没有设置触发词`;
      }
      if (!answer) {
        return `<at name="${session.username}" /> 你没有设置回答`;
      }
      const uid = session.userId || session.uid;
      const nowWordDB = await ctx2.word.user.getEditWord(uid);
      const hasPermission = await ctx2.word.permission.isHave(uid, `word.edit.${nowWordDB}`);
      if (!hasPermission && !config.masterID.includes(uid)) {
        return `<at name="${session.username}" /> 你没有词库【${nowWordDB}】的编辑权限`;
      }
      const a = await ctx2.word.editor.addWordItem(nowWordDB, uid, question, answer);
      if (typeof a === "number") {
        return `<at name="${session.username}" /> 添加到【${nowWordDB}】词库成功，序号为【${a}】`;
      } else {
        return a;
      }
    });
    ctx2.command("word", "词库核心！").subcommand(".rm <question:string> <listNumber:text>", "删除触发词的回复").usage("序号为数字或者all").example("word.rm 你好 all").example("word.rm 你好 1").action(async ({ session }, question, whichTemp) => {
      if (!session) {
        return;
      }
      if (!question) {
        return `<at name="${session.username}" /> 你没有设置触发词`;
      }
      if (!/^\d+$|^all$/.test(whichTemp)) {
        return `<at name="${session.username}" /> 你没有设置需要被删除的序号或序号不正确`;
      }
      const uid = session.userId || session.uid;
      const which = whichTemp === "all" ? "all" : Number(whichTemp);
      const nowWordDB = await ctx2.word.user.getEditWord(uid);
      const hasPermission = await ctx2.word.permission.isHave(uid, `word.edit.${nowWordDB}`);
      if (!hasPermission && !config.masterID.includes(uid) && nowWordDB != "default") {
        return `<at name="${session.username}" /> 你没有词库【${nowWordDB}】的编辑权限`;
      }
      const a = await ctx2.word.editor.rmWordItem(nowWordDB, uid, question, which);
      if (a === "over") {
        return `<at name="${session.username}" /> 删除触发词成功`;
      } else {
        return `<at name="${session.username}" /> ${a}`;
      }
    });
    ctx2.command("word", "词库核心！").subcommand(".setedit <dbname:text>", "选择库进行编辑").usage("当setedit后不加参数则代表选择为默认库").example("word.setedit 测试").action(async ({ session }, test) => {
      if (!session) {
        return;
      }
      let newDB = test;
      if (!test) {
        newDB = "default";
      }
      const uid = session.userId || session.uid;
      const a = await ctx2.word.user.setEditWord(uid, newDB);
      if (a) {
        return `<at name="${session.username}" /> 设置成功`;
      } else {
        return `<at name="${session.username}" /> 设置失败`;
      }
    });
    ctx2.command("word", "词库核心！").subcommand(".readedit", "查看当前正在编辑的词库").example("word.readedit").action(async ({ session }) => {
      if (!session) {
        return;
      }
      const uid = session.userId || session.uid;
      const a = await ctx2.word.user.getEditWord(uid);
      return `<at name="${session.username}" /> 你正在编辑【${a}】`;
    });
    ctx2.command("word", "词库核心！").subcommand(".find <question:text>", "寻找某个触发词所在的词库").example("word.find 你好").action(async ({ session }, question) => {
      if (!session) {
        return;
      }
      const a = await ctx2.word.editor.getQuestion(question);
      let outMsg = "";
      a.forEach((value, index) => {
        outMsg = outMsg + `
${index + 1}. ${value}`;
      });
      return `<at name="${session.username}" /> 此关键词存在以下词库：` + import_koishi5.h.text(outMsg);
    });
    ctx2.command("word", "词库核心！").subcommand(".get <question:text>", "查看当前词库某触发词的所有回答").example("word.get 测试").action(async ({ session }, question) => {
      if (!session) {
        return;
      }
      if (!question) {
        return `<at name="${session.username}" /> 你没有输入需要查询的关键词`;
      }
      const uid = session.userId || session.uid;
      const nowDB = await ctx2.word.user.getEditWord(uid);
      const a = await ctx2.word.editor.readWord(nowDB);
      if (!a.data[question]) {
        return `<at name="${session.username}" /> 当前编辑词库没有此触发词`;
      }
      let outMsg = `<at name="${session.username}" /> 此关键词含有以下回答：`;
      a.data[question].forEach((value, index) => {
        outMsg = outMsg + `
${index + 1}. ${value}`;
      });
      return import_koishi5.h.text(outMsg);
    });
    ctx2.command("word", "词库核心！").subcommand(".getDB <dbName:text>", "查看当前/某词库的所有触发词").example("word.getDB 测试").example("word.getDB").action(async ({ session }, dbName) => {
      if (!session) {
        return;
      }
      const uid = session.userId || session.uid;
      dbName = dbName ? dbName : await ctx2.word.user.getEditWord(uid);
      const a = await ctx2.word.editor.readWord(dbName);
      const questionList = Object.keys(a.data);
      let outMsg = `<at name="${session.username}" /> 你当前编辑的库含有以下触发词：`;
      questionList.forEach((value, index) => {
        outMsg = outMsg + `
${index + 1}. ${value}`;
      });
      return import_koishi5.h.text(outMsg);
    });
    ctx2.command("word", "词库核心！").subcommand(".setsave <cell:text>", "设置当前词库的存储格子").example("word.setsave 存储格1").action(async ({ session }, cell) => {
      if (!session) {
        return;
      }
      if (!cell) {
        return `<at name="${session.username}" /> 你没有输入存储格子名称`;
      }
      const uid = session.userId || session.uid;
      const nowWordDB = await ctx2.word.user.getEditWord(uid);
      const hasPermission = await ctx2.word.permission.isHave(uid, `word.edit.${nowWordDB}`);
      if (!hasPermission && !config.masterID.includes(uid)) {
        return `<at name="${session.username}" /> 你没有词库【${nowWordDB}】的编辑权限`;
      }
      const a = await ctx2.word.editor.setSaveCell(nowWordDB, cell, uid);
      if (typeof a === "boolean") {
        return `<at name="${session.username}" /> 修改成功`;
      } else {
        return `<at name="${session.username}" /> ${a}`;
      }
    });
    ctx2.command("word", "词库核心！").subcommand(".resetsave", "重置当前词库的存储格子").example("word.resetsave").action(async ({ session }) => {
      if (!session) {
        return;
      }
      const uid = session.userId || session.uid;
      const nowWordDB = await ctx2.word.user.getEditWord(uid);
      const hasPermission = await ctx2.word.permission.isHave(uid, `word.edit.${nowWordDB}`);
      if (!hasPermission && !config.masterID.includes(uid)) {
        return `<at name="${session.username}" /> 你没有词库【${nowWordDB}】的编辑权限`;
      }
      const a = await ctx2.word.editor.setSaveCell(nowWordDB, "default", uid);
      if (typeof a === "boolean") {
        return `<at name="${session.username}" /> 修改成功`;
      } else {
        return `<at name="${session.username}" /> ${a}`;
      }
    });
    ctx2.command("word", "词库核心！").subcommand(".getsave", "查看当前词库的存储格子").example("word.getsave").action(async ({ session }) => {
      if (!session) {
        return;
      }
      const uid = session.userId || session.uid;
      const nowWordDB = await ctx2.word.user.getEditWord(uid);
      const a = await ctx2.word.editor.readSaveCell(nowWordDB, uid);
      return `<at name="${session.username}" /> 当前词库的存储格子为【${a}】`;
    });
    ctx2.command("word", "词库核心！").subcommand(".addp <uid:string> <permission:text>", "增加权限").usage([
      "权限节点说明：",
      "编辑词库：word.edit.词库名",
      "编辑所有词库：word.edit.*",
      "添加权限：word.admin.add",
      "删除权限：word.admin.rm",
      "管理员级权限：word.admin.*"
    ].join("\n")).example("word.addp 5b0fe8a3b1ff2 word.edit.*").action(async ({ session }, uid, permission) => {
      if (!session) {
        return;
      }
      const mid = session.userId || session.uid;
      const hasPermission = await ctx2.word.permission.isHave(mid, "word.admin.add");
      if (!hasPermission && !config.masterID.includes(mid)) {
        return `<at name="${session.username}" /> 你没有词库的【添加权限】权限`;
      }
      const a = await ctx2.word.permission.add(uid, permission);
      if (typeof a === "boolean") {
        return a ? `<at name="${session.username}" /> 添加成功` : `<at name="${session.username}" /> 添加失败`;
      } else {
        return `<at name="${session.username}" /> ${a}`;
      }
    });
    ctx2.command("word", "词库核心！").subcommand(".rmp <uid:string> <permission:text>", "取消权限").usage([
      "权限节点说明：",
      "编辑词库：word.edit.词库名",
      "编辑所有词库：word.edit.*",
      "添加权限：word.admin.add",
      "删除权限：word.admin.rm",
      "管理员级权限：word.admin.*"
    ].join("\n")).example("word.rmp 6503fb7b50308 word.edit.*").action(async ({ session }, uid, permission) => {
      if (!session) {
        return;
      }
      const mid = session.userId || session.uid;
      const hasPermission = await ctx2.word.permission.isHave(mid, "word.admin.rm");
      if (!hasPermission && !config.masterID.includes(mid)) {
        return `<at name="${session.username}" /> 你没有词库的【删除权限】权限`;
      }
      const a = await ctx2.word.permission.rm(uid, permission);
      if (typeof a === "boolean") {
        return a ? `<at name="${session.username}" /> 添加成功` : `<at name="${session.username}" /> 添加失败`;
      } else {
        return `<at name="${session.username}" /> ${a}`;
      }
    });
    ctx2.command("word", "词库核心！").subcommand(".addauthor <uid:string>", "设置某uid为作者").example("word.addauthor 6503fb7b50308").action(async ({ session }, uid) => {
      if (!session) {
        return;
      }
      const mid = session.userId || session.uid;
      const nowWordDB = await ctx2.word.user.getEditWord(mid);
      const hasPermission = await ctx2.word.permission.isHave(uid, `word.edit.${nowWordDB}`);
      if (!hasPermission && !config.masterID.includes(mid)) {
        return `<at name="${session.username}" /> 你没有词库【${nowWordDB}】的编辑权限`;
      }
      const a = await ctx2.word.editor.addAuthor(nowWordDB, mid, uid);
      return `<at name="${session.username}" /> ${a}`;
    });
    ctx2.command("word", "词库核心！").subcommand(".rmauthor <uid:string>", "删除某uid的作者权限").example("word.rmauthor 6503fb7b50308").action(async ({ session }, uid) => {
      if (!session) {
        return;
      }
      const mid = session.userId || session.uid;
      const nowWordDB = await ctx2.word.user.getEditWord(mid);
      const hasPermission = await ctx2.word.permission.isHave(uid, `word.edit.${nowWordDB}`);
      if (!hasPermission && !config.masterID.includes(mid)) {
        return `<at name="${session.username}" /> 你没有词库【${nowWordDB}】的编辑权限`;
      }
      const a = await ctx2.word.editor.removeAuthor(nowWordDB, mid, uid);
      return `<at name="${session.username}" /> ${a}`;
    });
    ctx2.command("word", "词库核心！").subcommand(".getauthor", "查看当前词库的作者").example("word.getauthor").action(async ({ session }, uid) => {
      if (!session) {
        return;
      }
      const mid = session.userId || session.uid;
      const nowWordDB = await ctx2.word.user.getEditWord(mid);
      const a = await ctx2.word.editor.getAuthor(nowWordDB);
      return `<at name="${session.username}" /> 当前词库的作者有：

${a.join("\n")}`;
    });
    ctx2.command("word", "词库核心！").subcommand(".id", "查看自己的id及名字").example("word.id").action(({ session }) => {
      if (!session) {
        return "发生异常";
      }
      return `您的名字是：【${session.username}】，您的id是：【${session.userId}】`;
    });
    ctx2.command("word", "词库核心！").subcommand(".skip <dbname:text>", "禁止某词库在本频道运行").example("word.skip default").action(async ({ session }, dbName) => {
      if (!session) {
        return "发生异常";
      }
      if (!dbName) {
        return `<at name="${session.username}" /> 你没有输入词库名称`;
      }
      const skipData = await ctx2.word.config.getConfig("filtration");
      if (!session.channelId) {
        return `<at name="${session.username}" /> 你没有在频道中使用此命令`;
      }
      if (!skipData[session.channelId]) {
        skipData[session.channelId] = [];
      }
      skipData[session.channelId].push(dbName);
      await ctx2.word.config.updateConfig("filtration", skipData);
      return `<at name="${session.username}" /> 修改成功`;
    });
    ctx2.command("word", "词库核心！").subcommand(".unskip <dbname:text>", "允许某词库在本频道运行").example("word.unskip default").action(async ({ session }, dbName) => {
      if (!session) {
        return "发生异常";
      }
      if (!dbName) {
        return `<at name="${session.username}" /> 你没有输入词库名称`;
      }
      const skipData = await ctx2.word.config.getConfig("filtration");
      if (!session.channelId) {
        return `<at name="${session.username}" /> 你没有在频道中使用此命令`;
      }
      if (!skipData[session.channelId]) {
        skipData[session.channelId] = [];
      }
      const findIndex = skipData[session.channelId].indexOf(dbName);
      skipData[session.channelId].splice(findIndex, 1);
      if (skipData[session.channelId].length == 0) {
        delete skipData[session.channelId];
      }
      await ctx2.word.config.updateConfig("filtration", skipData);
      return `<at name="${session.username}" /> 修改成功`;
    });
    ctx2.command("word", "词库核心！").subcommand(".skip.list", "查看本频道词库黑名单").example("word.skip.list").action(async ({ session }) => {
      if (!session) {
        return "发生异常";
      }
      const skipData = await ctx2.word.config.getConfig("filtration");
      if (!session.channelId) {
        return `<at name="${session.username}" /> 你没有在频道中使用此命令`;
      }
      if (!skipData[session.channelId]) {
        return `<at name="${session.username}" /> 当前群内未设置禁用的词库`;
      }
      const skipList = skipData[session.channelId];
      return `<at name="${session.username}" /> 当前群内有以下词库被禁用：

 ${skipList.join("，")}`;
    });
    ctx2.command("word", "词库核心！").subcommand(".alldb", "查看所有词库").example("word.alldb").action(async ({ session }) => {
      if (!session) {
        return "发生异常";
      }
      const { idList } = ctx2.word.cache.getCache();
      return `<at name="${session.username}" /> 当前存在以下词库：

 ${idList.join("\n")}`;
    });
    ctx2.command("word", "词库核心！").subcommand(".rmdb <dbName:string>", "删除某个词库").example("word.rmdb default").action(async ({ session }, dbName) => {
      if (!session) {
        return;
      }
      const mid = session.userId || session.uid;
      const nowWordDB = await ctx2.word.user.getEditWord(mid);
      const hasPermission = await ctx2.word.permission.isHave(mid, `word.edit.${nowWordDB}`);
      if (!hasPermission && !config.masterID.includes(mid)) {
        return `<at name="${session.username}" /> 你没有词库【${nowWordDB}】的编辑权限`;
      }
      const a = await ctx2.word.editor.removeWord(dbName);
      return `<at name="${session.username}" /> ${a}`;
    });
    ctx2.command("word", "词库核心！").subcommand(".recycledb", "查看回收站").example("word.recycledb").action(async ({ session }) => {
      if (!session) {
        return;
      }
      const mid = session.userId || session.uid;
      const nowWordDB = await ctx2.word.user.getEditWord(mid);
      const hasPermission = await ctx2.word.permission.isHave(mid, `word.edit.${nowWordDB}`);
      if (!hasPermission && !config.masterID.includes(mid)) {
        return `<at name="${session.username}" /> 你没有词库【${nowWordDB}】的编辑权限`;
      }
      const a = await ctx2.word.editor.getRecycleBinList();
      return `<at name="${session.username}" /> 回收站包含以下词库：
${a.join("\n")}`;
    });
    ctx2.command("word", "词库核心！").subcommand(".restoredb <dbName:string>", "从回收站回收某个词库").example("word.restoredb default").action(async ({ session }, dbName) => {
      if (!session) {
        return;
      }
      const mid = session.userId || session.uid;
      const nowWordDB = await ctx2.word.user.getEditWord(mid);
      const hasPermission = await ctx2.word.permission.isHave(mid, `word.edit.${nowWordDB}`);
      if (!hasPermission && !config.masterID.includes(mid)) {
        return `<at name="${session.username}" /> 你没有词库【${nowWordDB}】的编辑权限`;
      }
      const a = await ctx2.word.editor.restoreWord(dbName);
      return `<at name="${session.username}" /> ${a}`;
    });
    ctx2.on("message", async (session) => {
      if (!session.content) {
        return;
      }
      const uid = session.userId || session.uid;
      const botId = session.bot.user?.id || session.bot.selfId || session.bot.userId || session.bot.sid;
      if (uid == botId) {
        return;
      }
      const forkSession = session.bot.session((0, import_koishi5.clone)(session.event));
      if (!forkSession.content) {
        return;
      }
      const firstElement = import_koishi5.h.parse(forkSession.content);
      if (firstElement[0].type === "at") {
        const atId = firstElement[0].attrs.id;
        if (atId == botId) {
          forkSession.content = forkSession.content.replace(firstElement[0].toString(), "").trim();
        } else {
          if (config.EnableAt) {
            return;
          }
        }
      } else {
        if (config.EnableAt) {
          return;
        }
      }
      await ctx2.word.driver.start(forkSession, (str) => {
        if (!str) {
          return;
        }
        forkSession.send(str);
      });
    });
    ctx2.console.addEntry({
      dev: (0, import_path.resolve)(__dirname, "../client/index.ts"),
      prod: (0, import_path.resolve)(__dirname, "../dist")
    });
    setInterval(async () => {
      try {
        if (config.telemetry) {
          await fetch(`https://xc.null.red:8043/api/online/heartbeat?t=word_core&_=${Date.now()}`);
        }
      } catch (err) {
        return "词库提交当前状态异常...作者木有办法收到你的鼓励..";
      }
    }, 2 * 60 * 1e3);
    ctx2.console.addListener("getWord", async (name2) => {
      const a = await ctx2.word.editor.getCloudWord(name2);
      await ctx2.word.cache.cacheRefresh();
      return a;
    });
    ctx2.console.addListener("rmWord", async (name2) => {
      const a = await ctx2.word.editor.removeWord(name2);
      return a;
    });
    ctx2.console.addListener("getPluginServerUrl", () => {
      return ctx2.word.tools.url;
    });
    ctx2.console.addListener("getWordList", async () => {
      const { idList } = ctx2.word.cache.getCache();
      return idList;
    });
  });
  const getWebuiBot = /* @__PURE__ */ __name(() => {
    return new bot_default(ctx, {});
  }, "getWebuiBot");
  ctx.console.addListener("sandbox-send", async (msg) => {
    const bot = getWebuiBot();
    bot.createSession(msg);
  });
}, "apply");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Config,
  apply,
  inject,
  logger,
  name,
  usage,
  word
});
