// koishi-plugin-user-blacklist/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Config = void 0;
exports.apply = apply;
const koishi_1 = require("koishi");
// 定义插件配置Schema
exports.Config = koishi_1.Schema.object({
    enabled: koishi_1.Schema.boolean().default(true).description('是否启用黑名单功能'),
    blacklistedUsers: koishi_1.Schema.array(koishi_1.Schema.string())
        .role('textarea')
        .default([])
        .description('黑名单用户列表（每行一个user_openid）'),
    replyText: koishi_1.Schema.string()
        .role('textarea')
        .default('您已被加入黑名单，无法使用机器人功能。')
        .description('回复黑名单用户的文本内容（留空则不发送文本）'),
    replyImageUrl: koishi_1.Schema.string()
        .default('')
        .description('回复黑名单用户的图片URL（留空则不发送图片）'),
    sendBoth: koishi_1.Schema.boolean()
        .default(true)
        .description('当同时配置了文本和图片时，是否两者都发送')
});
// 导出插件函数
function apply(ctx, config) {
    // 如果插件未启用，直接返回
    if (!config.enabled)
        return;
    // 使用中间件拦截消息
    ctx.middleware(async (session, next) => {
        // 检查发送者是否在黑名单中
        if (isUserBlacklisted(session, config.blacklistedUsers)) {
            // 用户在黑名单中，发送配置的回复
            await replyToBlacklistedUser(session, config);
            // 不继续处理消息（阻止黑名单用户使用任何功能）
            return;
        }
        // 用户不在黑名单中，继续处理消息
        return next();
    }, true); // 设置为true表示这个中间件优先级高，会在其他中间件之前执行
    // 日志记录
    ctx.logger('blacklist').info('黑名单插件已加载');
}
// 检查用户是否在黑名单中
function isUserBlacklisted(session, blacklistedUsers) {
    // 获取用户的openid
    const userOpenid = session.userId;
    // 检查用户是否在黑名单中
    return blacklistedUsers.includes(userOpenid);
}
// 回复黑名单用户
async function replyToBlacklistedUser(session, config) {
    try {
        const hasText = config.replyText && config.replyText.trim() !== '';
        const hasImage = config.replyImageUrl && config.replyImageUrl.trim() !== '';
        // 根据配置发送回复
        if (hasText && hasImage && config.sendBoth) {
            // 同时发送文本和图片
            await session.send(config.replyText);
            await session.send(`<image url="${config.replyImageUrl}"/>`);
        }
        else if (hasText) {
            // 只发送文本
            await session.send(config.replyText);
        }
        else if (hasImage) {
            // 只发送图片
            await session.send(`<image url="${config.replyImageUrl}"/>`);
        }
        // 记录日志
        session.app.logger('blacklist').info(`已拦截黑名单用户 ${session.userId} 的消息`);
    }
    catch (error) {
        // 错误处理
        session.app.logger('blacklist').error(`回复黑名单用户时出错: ${error}`);
    }
}
