// koishi-plugin-arcaea/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArcaeaUnlimitedApi = exports.apply = exports.Config = exports.name = void 0;
const koishi_1 = require("koishi");
const aua_1 = require("./source/aua");
Object.defineProperty(exports, "ArcaeaUnlimitedApi", { enumerable: true, get: function () { return aua_1.ArcaeaUnlimitedApi; } });
const source_1 = require("./source");
const text_1 = require("./render/text");
exports.name = 'arcaea';
exports.Config = koishi_1.Schema.object({
    api: koishi_1.Schema.intersect([
        koishi_1.Schema.object({
            type: koishi_1.Schema.union([
                koishi_1.Schema.const('aua').description("Arcaea Unlimited API(BotArcAPI)").required(),
                koishi_1.Schema.const('estertion')
            ]).description('Estertion Arcaea API(施工中)').required()
        }),
        koishi_1.Schema.union([
            aua_1.ArcaeaUnlimitedApi.Config,
            koishi_1.Schema.object({ type: koishi_1.Schema.const('estertion') })
        ])
    ])
});
function apply(ctx, config) {
    if (!config.api)
        return;
    switch (config.api.type) {
        case 'aua':
            ctx.plugin(aua_1.ArcaeaUnlimitedApi, config.api);
            break;
    }
    ctx.using(['arcaea'], (ctx) => {
        console.info("Arcaea Loaded");
        ctx.command('arcaea.recent [user]', 'Arcaea用户信息查询')
            .action(async ({ session }, user) => {
            try {
                const response = await ctx.arcaea.getUserInfo(user);
                return (0, text_1.arcaeaRecentRender)(session, response);
            }
            catch (e) {
                if (e instanceof source_1.ArcaeaInternalError) {
                    return e.toString();
                }
            }
        });
    });
}
exports.apply = apply;
