// dumble/lib/index.js
var __defProp = Object.defineProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });

// src/index.ts
import { dirname, extname, isAbsolute, join, relative, resolve } from "node:path";
import { isBuiltin } from "node:module";
import { build } from "esbuild";
import * as fs from "node:fs/promises";
import * as yaml from "js-yaml";
import kleur from "kleur";
import globby from "globby";
var DependencyType = ["dependencies", "devDependencies", "peerDependencies", "optionalDependencies"];
var ignored = [
  'This call to "require" will not be bundled because the argument is not a string literal',
  'Indirect calls to "require" will not be bundled',
  'should be marked as external for use with "require.resolve"'
];
function display(prefix) {
  return ({ location, text }) => {
    if (ignored.some((message) => text.includes(message))) return;
    if (!location) return console.log(prefix, text);
    const { file, line, column } = location;
    console.log(kleur.cyan(`${file}:${line}:${column}:`), prefix, text);
  };
}
__name(display, "display");
var displayError = display(kleur.red("error:"));
var displayWarning = display(kleur.yellow("warning:"));
function bundle(options) {
  const base = process.cwd();
  for (const [key, value] of Object.entries(options.entryPoints)) {
    const source = relative(base, value);
    const target = relative(base, resolve(options.outdir, key + options.outExtension[".js"]));
    console.log("esbuild:", source, "->", target);
  }
  return build(options).then(({ warnings }) => {
    warnings.forEach(displayWarning);
  }, ({ warnings, errors }) => {
    errors.forEach(displayError);
    warnings.forEach(displayWarning);
  });
}
__name(bundle, "bundle");
var externalPlugin = /* @__PURE__ */ __name(({ manifest, exports }) => ({
  name: "external library",
  setup(build2) {
    const { entryPoints, platform, format } = build2.initialOptions;
    const currentEntry = Object.values(entryPoints)[0];
    build2.onResolve({ filter: /^[@\w].+$/ }, (args) => {
      if (isAbsolute(args.path)) return null;
      if (isBuiltin(args.path)) return { external: true };
      const name = args.path.startsWith("@") ? args.path.split("/", 2).join("/") : args.path.split("/", 1)[0];
      if (name === manifest.name) return { external: true };
      const types = new Set(DependencyType.filter((type) => manifest[type]?.[name]));
      if (types.size === 0) {
        throw new Error(`Missing dependency: ${name} from ${args.importer}`);
      }
      types.delete("devDependencies");
      return { external: types.size > 0 };
    });
    build2.onResolve({ filter: /^\./, namespace: "file" }, async (args) => {
      const { path } = await build2.resolve(args.path, {
        namespace: "internal",
        importer: args.importer,
        resolveDir: args.resolveDir,
        kind: args.kind
      });
      const ext = extname(args.path);
      if (!path && ext.endsWith("js")) {
        const base = join(args.resolveDir, args.path.slice(0, -ext.length));
        const entry = exports[base + ".d" + ext.replace(/js$/, "ts")];
        if (entry) return { path: entry.types, external: true };
      }
      if (currentEntry === path || !exports[path]) return null;
      if (format === "cjs") return { external: true };
      const outFile = exports[path][platform] || exports[path].default;
      if (!outFile) return null;
      const outDir = dirname(exports[currentEntry][platform]);
      let relpath = relative(outDir, outFile).replace(/\\/g, "/");
      if (!relpath.startsWith(".")) relpath = "./" + relpath;
      return { path: relpath, external: true };
    });
    build2.onLoad({ filter: /\.d\.ts$/ }, async (args) => {
      const contents = await fs.readFile(args.path);
      return { loader: "copy", contents };
    });
  }
}), "externalPlugin");
var yamlPlugin = /* @__PURE__ */ __name((options = {}) => ({
  name: "yaml",
  setup(build2) {
    build2.initialOptions.resolveExtensions.push(".yml", ".yaml");
    build2.onLoad({ filter: /\.ya?ml$/ }, async ({ path }) => {
      const source = await fs.readFile(path, "utf8");
      return {
        loader: "json",
        contents: JSON.stringify(yaml.load(source, options))
      };
    });
  }
}), "yamlPlugin");
var hashbangPlugin = /* @__PURE__ */ __name((binaries) => ({
  name: "hashbang",
  setup(build2) {
    build2.onLoad({ filter: /\.tsx?$/ }, async ({ path }) => {
      if (!binaries.includes(path)) return null;
      let contents = await fs.readFile(path, "utf8");
      if (!contents.startsWith("#!")) {
        contents = "#!/usr/bin/env node\n" + contents;
      }
      return { contents, loader: extname(path).slice(1) };
    });
  }
}), "hashbangPlugin");
async function dumble(cwd, manifest, tsconfig, config = {}) {
  const { rootDir = "", outFile, noEmit, emitDeclarationOnly, sourceMap } = tsconfig.compilerOptions;
  if (!noEmit && !emitDeclarationOnly) return;
  const outDir = tsconfig.compilerOptions.outDir ?? dirname(outFile);
  const define = Object.fromEntries(Object.entries(config.env ?? {}).map(([key, value]) => {
    return [`process.env.${key}`, JSON.stringify(value)];
  }));
  const outdir = resolve(cwd, outDir);
  const outbase = resolve(cwd, rootDir);
  const matrix = [];
  const exports = /* @__PURE__ */ Object.create(null);
  const outFiles = /* @__PURE__ */ new Set();
  const binaries = [];
  const resolveCache = /* @__PURE__ */ Object.create(null);
  async function resolvePattern(pattern) {
    if (!pattern.startsWith(outDir + "/")) {
      pattern = pattern.replace("*", "**");
      const targets = await globby(pattern, { cwd });
      for (const target of targets) {
        if (!relative(rootDir, target).startsWith("../")) continue;
        const filename = join(cwd, target);
        exports[filename] = { default: filename };
      }
      return;
    }
    const outExt = extname(pattern);
    pattern = pattern.slice(outDir.length + 1, -outExt.length).replace("*", "**") + ".{ts,tsx}";
    return [outExt, await globby(pattern, { cwd: outbase })];
  }
  __name(resolvePattern, "resolvePattern");
  async function addExport(pattern, preset2, prefix = "", isBinary = false) {
    if (!pattern) return;
    if (pattern.startsWith("./")) pattern = pattern.slice(2);
    const result = await (resolveCache[pattern] ??= resolvePattern(pattern));
    if (!result) return;
    const [outExt, targets] = result;
    preset2 = { ...preset2 };
    if (outExt === ".cjs") {
      preset2.format = "cjs";
    } else if (outExt === ".mjs") {
      preset2.format = "esm";
    }
    for (const target of targets) {
      const srcFile = join(cwd, rootDir, target);
      if (isBinary) binaries.push(srcFile);
      const srcExt = extname(target);
      const entry = target.slice(0, -srcExt.length);
      const outFile2 = join(outdir, entry + outExt);
      if (outFiles.has(outFile2)) return;
      outFiles.add(outFile2);
      if (!preset2.platform) {
        ;
        (exports[srcFile] ||= {}).types = `${manifest.name}/${prefix}`;
      } else {
        ;
        (exports[srcFile] ||= {})[preset2.platform] = outFile2;
      }
      matrix.push({
        absWorkingDir: cwd,
        outdir,
        outbase,
        target: tsconfig.compilerOptions?.target,
        outExtension: { ".js": outExt },
        entryPoints: { [entry]: srcFile },
        bundle: true,
        minify: config.minify,
        sourcemap: sourceMap,
        sourcesContent: false,
        keepNames: true,
        charset: "utf8",
        logLevel: "silent",
        resolveExtensions: [".tsx", ".ts", ".jsx", ".js", ".css", ".json"],
        tsconfig: cwd + "/tsconfig.json",
        plugins: [
          yamlPlugin(),
          externalPlugin({ cwd, manifest, tsconfig, exports }),
          hashbangPlugin(binaries)
        ],
        define,
        ...preset2
      });
    }
  }
  __name(addExport, "addExport");
  const tasks = [];
  function addConditionalExport(pattern, preset2, prefix = "") {
    if (typeof pattern === "string") {
      tasks.push(addExport(pattern, preset2, prefix));
      return;
    }
    for (const key in pattern) {
      if (key === "require") {
        addConditionalExport(pattern[key], { ...preset2, format: "cjs" }, prefix);
      } else if (key === "import") {
        addConditionalExport(pattern[key], { ...preset2, format: "esm" }, prefix);
      } else if (["browser", "node"].includes(key)) {
        addConditionalExport(pattern[key], { ...preset2, platform: key }, prefix);
      } else if (["types", "typings"].includes(key)) {
        addConditionalExport(pattern[key], { ...preset2, platform: void 0 }, prefix);
      } else {
        addConditionalExport(pattern[key], preset2, key.startsWith(".") ? join(prefix, key) : prefix);
      }
    }
  }
  __name(addConditionalExport, "addConditionalExport");
  const preset = {
    platform: "node",
    format: manifest.type === "module" ? "esm" : "cjs"
  };
  tasks.push(addExport(manifest.main, preset));
  tasks.push(addExport(manifest.module, { ...preset, format: "esm" }));
  addConditionalExport(manifest.exports, preset);
  if (!manifest.exports) {
    tasks.push(addExport("package.json", preset, null));
  }
  if (typeof manifest.bin === "string") {
    tasks.push(addExport(manifest.bin, preset, null, true));
  } else if (manifest.bin) {
    for (const key in manifest.bin) {
      tasks.push(addExport(manifest.bin[key], preset, null, true));
    }
  }
  await Promise.all(tasks);
  const build2 = config.build ?? ((options, callback) => callback(options));
  await Promise.all(matrix.map(async (options) => {
    try {
      await build2(options, bundle);
    } catch (error) {
      console.error(error);
    }
  }));
}
__name(dumble, "dumble");
var index_default = dumble;
export {
  DependencyType,
  index_default as default
};
