// @koishijs/plugin-hmr/lib/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/locales/zh-CN.yml
var require_zh_CN = __commonJS({
  "src/locales/zh-CN.yml"(exports, module2) {
    module2.exports = { base: "用户显示路径的根目录，默认为当前工作路径。", root: "要监听的文件或目录列表，相对于 `base` 路径。", ignored: "要忽略的文件或目录。支持 [Glob Patterns](https://github.com/micromatch/micromatch) 语法。", debounce: "延迟触发更新的等待时间。" };
  }
});

// src/index.ts
var src_exports = {};
__export(src_exports, {
  default: () => src_default
});
module.exports = __toCommonJS(src_exports);
var import_koishi = require("koishi");
var import_chokidar = require("chokidar");
var import_path = require("path");
var import_module = require("module");
var import_loader = require("@koishijs/loader");

// src/error.ts
var import_code_frame = require("@babel/code-frame");
var import_fs = require("fs");
function isBuildFailure(e) {
  return Array.isArray(e?.errors) && e.errors.every((error) => error.text);
}
__name(isBuildFailure, "isBuildFailure");
function handleError(e, logger) {
  if (!isBuildFailure(e)) {
    logger.warn(e);
    return;
  }
  for (const error of e.errors) {
    if (!error.location) {
      logger.warn(error.text);
      continue;
    }
    try {
      const { file, line, column } = error.location;
      const source = (0, import_fs.readFileSync)(file, "utf8");
      const formatted = (0, import_code_frame.codeFrameColumns)(source, {
        start: { line, column }
      }, {
        highlightCode: true,
        message: error.text
      });
      logger.warn(`File: ${file}:${line}:${column}
` + formatted);
    } catch (e2) {
      logger.warn(e2);
    }
  }
}
__name(handleError, "handleError");

// src/index.ts
function loadDependencies(filename, ignored) {
  const dependencies = /* @__PURE__ */ new Set();
  function traverse({ filename: filename2, children }) {
    if (ignored.has(filename2) || dependencies.has(filename2) || filename2.includes("/node_modules/"))
      return;
    dependencies.add(filename2);
    children.forEach(traverse);
  }
  __name(traverse, "traverse");
  traverse(require.cache[filename]);
  return dependencies;
}
__name(loadDependencies, "loadDependencies");
var Watcher = class {
  constructor(ctx, config) {
    this.ctx = ctx;
    this.config = config;
    this.base = (0, import_path.resolve)(ctx.baseDir, config.base || "");
    this.logger = ctx.logger("hmr");
    ctx.provide("watcher", this);
    ctx.on("ready", () => this.start());
    ctx.on("dispose", () => this.stop());
  }
  static {
    __name(this, "Watcher");
  }
  static inject = ["loader"];
  base;
  watcher;
  require = (0, import_module.createRequire)(require.resolve("@koishijs/loader/package.json"));
  /**
   * changes from externals E will always trigger a full reload
   *
   * - root R -> external E -> none of plugin Q
   */
  externals;
  /**
   * files X that should be reloaded
   *
   * - including all stashed files S
   * - some plugin P -> file X -> some change C
   */
  accepted;
  /**
   * files X that should not be reloaded
   *
   * - including all externals E
   * - some change C -> file X -> none of change D
   */
  declined;
  /** stashed changes */
  stashed = /* @__PURE__ */ new Set();
  logger;
  relative(filename) {
    if (!this.base)
      return filename;
    return (0, import_path.relative)(this.base, filename);
  }
  start() {
    const { loader } = this.ctx;
    const { root, ignored } = this.config;
    this.watcher = (0, import_chokidar.watch)(root, {
      ...this.config,
      cwd: this.base,
      ignored: (0, import_koishi.makeArray)(ignored)
    });
    this.externals = loadDependencies(require.resolve("koishi"), new Set(Object.values(loader.cache)));
    const triggerLocalReload = this.ctx.debounce(() => this.triggerLocalReload(), this.config.debounce);
    this.watcher.on("change", async (path) => {
      const filename = (0, import_path.resolve)(this.base, path);
      const isEntry = filename === loader.filename || loader.envFiles.includes(filename);
      if (loader.suspend && isEntry) {
        loader.suspend = false;
        return;
      }
      this.logger.debug("change detected:", path);
      if (isEntry) {
        if (require.cache[filename]) {
          this.ctx.loader.fullReload();
        } else {
          const config = await loader.readConfig();
          this.ctx.root.state.update(config);
          this.ctx.emit("config");
        }
      } else {
        if (this.externals.has(filename)) {
          this.ctx.loader.fullReload();
        } else if (require.cache[filename]) {
          this.stashed.add(filename);
          triggerLocalReload();
        }
      }
    });
  }
  stop() {
    return this.watcher.close();
  }
  analyzeChanges() {
    const pending = [];
    this.accepted = new Set(this.stashed);
    this.declined = new Set(this.externals);
    this.stashed.forEach((filename) => {
      const { children } = require.cache[filename];
      for (const { filename: filename2 } of children) {
        if (this.accepted.has(filename2) || this.declined.has(filename2) || filename2.includes("/node_modules/"))
          continue;
        pending.push(filename2);
      }
    });
    while (pending.length) {
      let index = 0, hasUpdate = false;
      while (index < pending.length) {
        const filename = pending[index];
        const { children } = require.cache[filename];
        let isDeclined = true, isAccepted = false;
        for (const { filename: filename2 } of children) {
          if (this.declined.has(filename2) || filename2.includes("/node_modules/"))
            continue;
          if (this.accepted.has(filename2)) {
            isAccepted = true;
            break;
          } else {
            isDeclined = false;
            if (!pending.includes(filename2)) {
              hasUpdate = true;
              pending.push(filename2);
            }
          }
        }
        if (isAccepted || isDeclined) {
          hasUpdate = true;
          pending.splice(index, 1);
          if (isAccepted) {
            this.accepted.add(filename);
          } else {
            this.declined.add(filename);
          }
        } else {
          index++;
        }
      }
      if (!hasUpdate)
        break;
    }
    for (const filename of pending) {
      this.declined.add(filename);
    }
  }
  triggerLocalReload() {
    this.analyzeChanges();
    const pending = /* @__PURE__ */ new Map();
    const reloads = /* @__PURE__ */ new Map();
    for (const filename of Object.values(this.ctx.loader.cache)) {
      const module2 = require.cache[filename];
      const plugin = (0, import_loader.unwrapExports)(module2.exports);
      if (!plugin || this.declined.has(filename))
        continue;
      const runtime = this.ctx.registry.get(plugin);
      pending.set(filename, [plugin, runtime]);
      this.declined.add(filename);
    }
    for (const [filename, [plugin, runtime]] of pending) {
      this.declined.delete(filename);
      const dependencies = [...loadDependencies(filename, this.declined)];
      this.declined.add(filename);
      if (!dependencies.some((dep) => this.accepted.has(dep)))
        continue;
      dependencies.forEach((dep) => this.accepted.add(dep));
      if (runtime) {
        let isMarked = false;
        const visited = /* @__PURE__ */ new Set();
        const queued = [runtime];
        while (queued.length) {
          const runtime2 = queued.shift();
          if (visited.has(runtime2))
            continue;
          visited.add(runtime2);
          if (reloads.has(plugin)) {
            isMarked = true;
            break;
          }
          for (const state of runtime2.children) {
            queued.push(state.runtime);
          }
        }
        if (!isMarked) {
          const children = /* @__PURE__ */ new Map();
          reloads.set(plugin, { filename, children });
          for (const state of runtime.children) {
            children.set(state, this.ctx.loader.getRefName(state));
          }
        }
      } else {
        reloads.set(plugin, { filename, children: /* @__PURE__ */ new Map() });
      }
    }
    const backup = {};
    for (const filename of this.accepted) {
      backup[filename] = require.cache[filename];
      delete require.cache[filename];
    }
    function rollback() {
      for (const filename in backup) {
        require.cache[filename] = backup[filename];
      }
    }
    __name(rollback, "rollback");
    const attempts = {};
    try {
      for (const [, { filename }] of reloads) {
        attempts[filename] = (0, import_loader.unwrapExports)(this.require(filename));
      }
    } catch (e) {
      handleError(e, this.logger);
      return rollback();
    }
    this.ctx.emit("hmr/reload", reloads);
    try {
      for (const [plugin, { filename, children }] of reloads) {
        const path = this.relative(filename);
        try {
          this.ctx.registry.delete(plugin);
        } catch (err) {
          this.logger.warn("failed to dispose plugin at %c\n" + (0, import_koishi.coerce)(err), path);
        }
        this.ctx.loader.replace(plugin, attempts[filename]);
        try {
          for (const [state, name] of children) {
            const fork = state.parent.plugin(attempts[filename], state.config);
            fork.key = state.key;
            if (name)
              state.parent.scope[import_loader.Loader.kRecord][name] = fork;
          }
          this.logger.info("reload plugin at %c", path);
        } catch (err) {
          this.logger.warn("failed to reload plugin at %c\n" + (0, import_koishi.coerce)(err), path);
          throw err;
        }
      }
    } catch {
      rollback();
      for (const [plugin, { filename, children }] of reloads) {
        try {
          this.ctx.registry.delete(attempts[filename]);
          for (const [state, name] of children) {
            const fork = state.parent.plugin(plugin, state.config);
            fork.key = state.key;
            if (name)
              state.parent.scope[import_loader.Loader.kRecord][name] = fork;
          }
        } catch (err) {
          this.logger.warn(err);
        }
      }
      return;
    }
    this.stashed = /* @__PURE__ */ new Set();
  }
};
((Watcher2) => {
  Watcher2.Config = import_koishi.Schema.object({
    base: import_koishi.Schema.string(),
    root: import_koishi.Schema.union([
      import_koishi.Schema.array(String).role("table"),
      import_koishi.Schema.transform(String, (value) => [value])
    ]).default(["."]),
    ignored: import_koishi.Schema.union([
      import_koishi.Schema.array(String).role("table"),
      import_koishi.Schema.transform(String, (value) => [value])
    ]).default([
      "**/node_modules/**",
      "**/.git/**",
      "**/logs/**"
    ]),
    debounce: import_koishi.Schema.natural().role("ms").default(100)
  }).i18n({
    "zh-CN": require_zh_CN()
  });
})(Watcher || (Watcher = {}));
var src_default = Watcher;
