// koishi-plugin-koishi/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.apply = exports.Config = exports.name = void 0;
const koishi_1 = require("koishi");
const execa_1 = __importDefault(require("execa"));
const plugin_console_1 = require("@koishijs/plugin-console");
const which_pm_runs_1 = __importDefault(require("which-pm-runs"));
const path_1 = require("path");
const promises_1 = require("fs/promises");
const yaml_1 = __importDefault(require("yaml"));
const get_port_1 = __importDefault(require("get-port"));
exports.name = 'koishi';
exports.Config = koishi_1.Schema.object({});
var isReady = 0;
var log = new koishi_1.Logger("2koishi");
class CustomProvider1 extends plugin_console_1.DataService {
    constructor(ctx) {
        super(ctx, 'get-koishi-plugin-koishi-status');
    }
    async get() {
        return isReady;
    }
}
function spawnAsync(args, options = {}) {
    const child = (0, execa_1.default)(args[0], args.slice(1), options);
    return new Promise((resolve) => {
        // child.stderr?.pipe(process.stderr)
        // child.stdout?.pipe(process.stdout)
        child.stderr.on("data", (data) => {
            log.error(data.toString());
        });
        child.stdout.on("data", (data) => {
            log.info(data.toString());
        });
        child.on('close', resolve);
    });
}
function spawnKoishi(args, options = {}) {
    const child = (0, execa_1.default)(args[0], args.slice(1), options);
    return new Promise((resolve) => {
        // child.stderr?.pipe(process.stderr)
        // child.stdout?.pipe(process.stdout)
        child.stderr.on("data", (data) => {
            log.error(data.toString());
        });
        child.stdout.on("data", (data) => {
            log.info(data.toString());
        });
        resolve(child);
    });
}
function getManager(manager) {
    if (!manager)
        return ['npm'];
    if (manager.name !== 'yarn' || manager.version.startsWith('1.'))
        return [manager.name];
    return ['yarn', 'npm'];
}
async function apply(ctx) {
    var koishiYAML = yaml_1.default.parse((await (0, promises_1.readFile)("./koishi.yml")).toString());
    var thisPort = koishiYAML.port;
    var thisUiPath = ctx.console.config.uiPath || "/";
    console.log(thisPort, thisUiPath);
    ctx.plugin(CustomProvider1);
    ctx.using(['console'], (ctx) => {
        ctx.console.addEntry({
            dev: (0, path_1.resolve)(__dirname, '../client/index.ts'),
            prod: (0, path_1.resolve)(__dirname, '../dist'),
        });
    });
    // write your plugin here
    // console.log(getManager(which()));
    const manager = getManager((0, which_pm_runs_1.default)())[0];
    // console.log(__dirname);
    // console.log(resolve(__dirname,"../bp"));
    const bp_path = (0, path_1.resolve)(__dirname, "../bp");
    var child = null;
    var port = await (0, get_port_1.default)();
    spawnAsync([manager], { cwd: bp_path }).then(async (code) => {
        var bpConfig = yaml_1.default.parse((await (0, promises_1.readFile)((0, path_1.resolve)(bp_path, "koishi.yml"))).toString());
        bpConfig.port = port;
        await (0, promises_1.writeFile)((0, path_1.resolve)(bp_path, "koishi.yml"), yaml_1.default.stringify(bpConfig));
        if (code == 0)
            isReady = 1;
        else
            isReady = 2;
        // var child = await spawnKoishi([manager,"start"], { cwd: bp_path })
        child = (0, execa_1.default)(manager, ["start"], { cwd: bp_path });
        child.stderr.on("data", (data) => {
            // log.error(data.toString())
        });
        child.stdout.on("data", (data) => {
            // log.info(data.toString())
        });
        child.on("close", () => {
            log.info("process close!");
        });
    });
    ctx.on("dispose", () => {
        child.kill("SIGINT");
    });
    ctx.console.addListener('koishi-get-port', () => {
        return port;
    });
    // console.log(global.endpoint)
}
exports.apply = apply;
