// koishi-plugin-starfx-bot/lib/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __name = (target, value) => __defProp(target, "name", { value, configurable: true });
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __export = (target, all) => {
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/locales/zh-CN.yml
var require_zh_CN = __commonJS({
  "src/locales/zh-CN.yml"(exports2, module2) {
    module2.exports = { commands: { roll: { messages: { noParam: "你问啥?", invalid: "你要不看看你说的是啥", "too-many": "太、太多了啦!", noodles: "扔出了{num}个{noodles}面的骰子，点数如下:{points}", possibility: "{param}是{possibility}%", choose: "当然是{option}啦!" } }, "room-number": { messages: { roomNumber: "当前的房间号是: {room}", changeRoom: "已将当前房间号 {oldRoom} 修改为 {newRoom}", newRoom: "已添加房间号: {room}", noRoom: "当前没有保存的房间", delRoom: "已删除房间: {room}", invalid: "无效输入" } }, forward: { messages: { delete: "已删除转发群聊: {gid}", noTarget: "未配置转发群组", success: "已转发消息到: {target}", noMessage: "请引用要转发的消息", setOK: "已配置转发: {target}", forwardContent: "[转发消息]<br/> {content}" } }, "waiting-play": { messages: { addReady: "已添加，当前待机{num}人: <br/>{list}", delReady: "已删除，当前待机{num}人: <br/>{list}", delFailed: "未删除，当前待机{num}人: <br/>{list}", clearReady: "待机列表已清空", listReady: "当前待机{num}人: <br/>{list}", invalid: "输入无效" } } }, middleware: { messages: { atNotReply: "艾特我又不说话", atNotReplyOther: "艾特他又不说话", notBot: "我才不是机器人！", loveMessage: "我喜欢你", iLoveU: "我也喜欢你" } } };
  }
});

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Config: () => Config,
  apply: () => apply,
  assetsDir: () => assetsDir,
  baseDir: () => baseDir,
  inject: () => inject,
  name: () => name,
  repeatContextMap: () => repeatContextMap,
  starfxLogger: () => starfxLogger,
  usage: () => usage
});
module.exports = __toCommonJS(src_exports);
var fs2 = __toESM(require("node:fs"));
var import_node_path4 = __toESM(require("node:path"));
var import_koishi5 = require("koishi");
var import_mime_types = __toESM(require("mime-types"));

// package.json
var package_default = {
  name: "koishi-plugin-starfx-bot",
  description: "StarFreedomX机器人的小功能，自用",
  contributors: [
    "StarFreedomX <starfreedomx@outlook.com>"
  ],
  version: "0.26.5",
  main: "lib/index.js",
  typings: "lib/index.d.ts",
  files: [
    "lib",
    "dist",
    "assets"
  ],
  license: "MIT",
  homepage: "https://github.com/StarFreedomX/starfx-bot",
  repository: {
    type: "git",
    url: "git+https://github.com/StarFreedomX/starfx-bot.git"
  },
  keywords: [
    "chatbot",
    "koishi",
    "plugin"
  ],
  koishi: {
    description: {
      zh: "starfx的机器人小插件，目前支持: 明日方舟封印功能 闲鱼“卖掉了”功能 艾特我/他又不说话 我才不是机器人！ 我也喜欢你！ BanG Dream!边框绘制 群语录 roll 待机人数记录 房间号记录 推特查看原图 复读 echo 撤回 查汇率 黑白名单配置"
    },
    service: {
      optional: [
        "skia",
        "QhzySharp"
      ]
    },
    preview: true,
    hidden: false
  },
  peerDependencies: {
    koishi: "^4.18.7"
  },
  dependencies: {
    axios: "^1.9.0",
    "chart.js": "^4.5.1",
    "chartjs-adapter-dayjs-3": "^1.2.3",
    cheerio: "^1.0.0",
    dayjs: "^1.11.19",
    "http-proxy-agent": "^7.0.2",
    "https-proxy-agent": "^7.0.6",
    "mime-types": "^3.0.1",
    "rss-parser": "^3.13.0"
  },
  devDependencies: {
    "@biomejs/biome": "2.3.7",
    "@ltxhhz/koishi-plugin-skia-canvas": "^0.0.10",
    "@quanhuzeyu/koishi-plugin-qhzy-sharp": "^1.2.1",
    "@quanhuzeyu/sharp-for-koishi": "^0.0.7",
    "@types/mime-types": "^3.0.1"
  }
};

// src/plugins/currencySearch.ts
var import_node_path = __toESM(require("node:path"));
var import_chart = require("chart.js");
var cheerio = __toESM(require("cheerio"));
var import_koishi = require("koishi");
import_chart.Chart.register(...import_chart.registerables);
var mvpAPIKey;
var currencyCodes = [];
var currencyCodesMap = {};
async function intervalGetExchangeRate(_ctx, _cfg, _session, _searchString, _exchangeRatePath) {
}
__name(intervalGetExchangeRate, "intervalGetExchangeRate");
async function parseNaturalCurrency(searchString) {
  const index = await getCurrencyCodesMap();
  const s = searchString.replace(/\s+/g, "");
  const match = s.match(/(.+?)(兑|对)(.+)/);
  if (match) {
    const from = index[match[1]];
    const to = index[match[3]];
    if (from && to) {
      return from + to;
    }
  }
  if (index[s]) {
    return index[s];
  }
  return null;
}
__name(parseNaturalCurrency, "parseNaturalCurrency");
async function getExchangeRate(ctx, cfg, session, searchString, raw, retryTimes = 1) {
  const apiKey = retryTimes ? await getMvpAPIKey() : await updateMvpAPIKey();
  try {
    let guids = [];
    if (searchString) {
      const parsed = await parseNaturalCurrency(searchString);
      if (parsed) {
        searchString = parsed;
      }
    }
    if (raw && /^([a-z0-9]{6})([,\s]([a-z0-9]{6}))*$/.test(raw)) {
      guids = raw.split(",");
    } else if (/^([a-zA-Z]{3}|[a-zA-Z]{6})([,\s]([a-zA-Z]{3}|[a-zA-Z]{6}))*$/.test(
      searchString
    )) {
      const currencies = await getCurrencyCodes();
      const codes = searchString.split(/[,\s]+/);
      const invalidCodes = [];
      const search6Strings = [];
      codes.forEach((code) => {
        const first = code.slice(0, 3).toUpperCase();
        const second = code.length === 3 ? "CNY" : code.slice(3, 6).toUpperCase();
        search6Strings.push(code.length === 3 ? first + second : code);
        if (!currencies.some((c) => c.code === first)) invalidCodes.push(first);
        if (code.length === 6 && !currencies.some((c) => c.code === second))
          invalidCodes.push(second);
      });
      if (invalidCodes.length > 0) {
        throw new Error(`输入的货币代码不合法: ${invalidCodes.join(", ")}`);
      }
      guids = await getExchangeGuids(search6Strings.join(","), apiKey);
    }
    if (!guids.length || guids.every((g) => !g)) {
      throw new Error(`输入无效: ${searchString}`);
    }
    const result = [];
    for (const guid of guids) {
      if (!guid) throw new Error("GUID not found");
      const nowPriceArr = await getQuotes(guid, apiKey);
      const monthPriceArr = await getMSNPrices(guid, apiKey, "1M1H");
      const day5PriceArr = await getMSNPrices(guid, apiKey, "5D");
      const nowPrice = nowPriceArr?.[0];
      const oneMonthPrice = monthPriceArr?.[0];
      const fiveDayPrice = day5PriceArr?.[0];
      if (!nowPrice || !oneMonthPrice || !fiveDayPrice) {
        throw new Error("Failed to fetch price");
      }
      const monthChartBuffer = await drawCurrencyChartSkia(
        ctx,
        oneMonthPrice.prices,
        oneMonthPrice.timeStamps,
        `1 ${nowPrice.fromCurrency} to 1 ${nowPrice.currency}`
      );
      const day5ChartBuffer = await drawCurrencyChartSkia(
        ctx,
        fiveDayPrice.prices,
        fiveDayPrice.timeStamps,
        `1 ${nowPrice.fromCurrency} to 1 ${nowPrice.currency}`
      );
      const mergeImgSrc = "data:image/png;base64," + (await mergeBuffersVertical(ctx, day5ChartBuffer, monthChartBuffer)).toString("base64");
      result.push({
        name: nowPrice.symbolName,
        fromCurrency: nowPrice.fromCurrency,
        currency: nowPrice.currency,
        nowPrice: nowPrice.price,
        // oneMonthChart: h.image(monthImgSrc),
        // fiveDayChart: h.image(day5ImgSrc),
        currencyChart: import_koishi.h.image(mergeImgSrc)
      });
    }
    for (const item of result) {
      await session.send([
        import_koishi.h.text(`
${(/* @__PURE__ */ new Date()).toLocaleString()}
${item.name}
(${item.fromCurrency}/${item.currency})
当前价格: ${item.nowPrice}
近5天/近1个月价格走势:
`),
        item.currencyChart
        // item.fiveDayChart,
        // h.text(`
        // 近30天价格走势:
        // `),
        // item.oneMonthChart
      ]);
    }
  } catch (err) {
    if (retryTimes === 0) {
      const message = err instanceof Error ? err.message : typeof err === "string" ? err : JSON.stringify(err);
      await session.send(`查询失败：${message}`);
    } else {
      await getExchangeRate(
        ctx,
        cfg,
        session,
        searchString,
        raw,
        retryTimes - 1
      );
    }
  }
}
__name(getExchangeRate, "getExchangeRate");
async function mergeBuffersVertical(ctx2, buf1, buf2) {
  const { Canvas, loadImage } = ctx2.skia;
  const img1 = await loadImage(buf1);
  const img2 = await loadImage(buf2);
  const width = img1.width;
  const height = img1.height + img2.height;
  const canvas = new Canvas(width, height);
  const ctx = canvas.getContext("2d");
  ctx.drawImage(img1, 0, 0);
  ctx.drawImage(img2, 0, img1.height);
  return await canvas.toBuffer("png");
}
__name(mergeBuffersVertical, "mergeBuffersVertical");
async function updateMvpAPIKey() {
  try {
    const htmlResp = await fetch(
      "https://www.msn.com/zh-cn/money/tools/currencyconverter"
    );
    if (!htmlResp.ok)
      throw new Error(`Failed to fetch page: ${htmlResp.status}`);
    const html = await htmlResp.text();
    const $ = cheerio.load(html);
    const head = $("head");
    const clientSettingsRaw = head.attr("data-client-settings");
    if (!clientSettingsRaw) throw new Error("未找到 data-client-settings 属性");
    const clientSettings = JSON.parse(
      clientSettingsRaw.replace(/&quot;/g, '"')
    );
    const version = clientSettings?.bundleInfo?.v;
    if (!version) throw new Error("未找到 bundleInfo.v");
    const targetScope = encodeURIComponent(
      JSON.stringify({
        audienceMode: "adult",
        browser: {
          browserType: "edgeChromium",
          version: "142",
          ismobile: "false"
        },
        deviceFormFactor: "desktop",
        domain: "www.msn.com",
        locale: {
          content: { language: "zh", market: "cn" },
          display: { language: "zh", market: "cn" }
        },
        os: "windows",
        modes: { audienceMode: "adult" },
        platform: "web",
        pageType: "finance::financetools::financecurrencyconverter",
        pageExperiments: ["prg-cmc-river"]
      })
    );
    const apiUrl = `https://assets.msn.com/resolver/api/resolve/v3/config/?expType=AppConfig&apptype=finance&v=${version}&targetScope=${targetScope}`;
    const apiResp = await fetch(apiUrl);
    if (!apiResp.ok)
      throw new Error(`Failed to fetch config API: ${apiResp.status}`);
    const json = await apiResp.json();
    const newMvpAPIKey = json?.configs?.["shared/msn-ns/CommonAutoSuggest/default"]?.properties?.mvpAPIkey ?? null;
    mvpAPIKey = newMvpAPIKey;
    return newMvpAPIKey;
  } catch (err) {
    console.error(err);
    return null;
  }
}
__name(updateMvpAPIKey, "updateMvpAPIKey");
async function getMvpAPIKey() {
  return mvpAPIKey ? mvpAPIKey : updateMvpAPIKey();
}
__name(getMvpAPIKey, "getMvpAPIKey");
async function getExchangeGuids(symbols, apikey) {
  try {
    const url = `https://assets.msn.cn/service/Finance/IdMap?apikey=${apikey}&MStarIds=${encodeURIComponent(symbols)}`;
    const resp = await fetch(url);
    if (!resp.ok) throw new Error(`请求失败: ${resp.status}`);
    let data = await resp.json();
    data = data.filter((d) => d.status === 200);
    const inputSymbols = symbols.split(",");
    const guidMap = new Map(data.map((item) => [item.mStarId, item.guid]));
    return inputSymbols.map((sym) => guidMap.get(sym) || "");
  } catch (err) {
    console.error(err);
    return [];
  }
}
__name(getExchangeGuids, "getExchangeGuids");
async function getQuotes(ids, apiKey) {
  try {
    const url = `https://assets.msn.com/service/Finance/Quotes?apikey=${encodeURIComponent(apiKey)}&cm=zh-cn&it=edgeid&ids=${encodeURIComponent(ids)}&wrapodata=false`;
    const resp = await fetch(url);
    if (!resp.ok) throw new Error(`请求失败: ${resp.status}`);
    const data = await resp.json();
    const arrayData = Array.isArray(data) ? data : [data];
    let quotes = [];
    if (arrayData.length > 0) {
      if (Array.isArray(arrayData[0])) {
        quotes = arrayData.flat(1);
      } else {
        quotes = arrayData;
      }
    }
    const inputIds = ids.split(",");
    const quoteMap = new Map(
      quotes.map((q) => [
        q.instrumentId,
        {
          price: q.price,
          symbolName: q.localizedAttributes?.["zh-cn"]?.symbolName || q.localizedAttributes?.["zh-cn"]?.displayName || q.symbol,
          fromCurrency: q.fromCurrency || q.displayName,
          currency: q.currency
        }
      ])
    );
    return inputIds.map(
      (id) => quoteMap.get(id) ?? {
        price: 0,
        symbolName: "",
        fromCurrency: "",
        currency: ""
      }
    );
  } catch (err) {
    console.error(err);
    return [];
  }
}
__name(getQuotes, "getQuotes");
async function getMSNPrices(ids, apiKey, type = "1M1H") {
  const url = `https://assets.msn.com/service/Finance/QuoteSummary?apikey=${apiKey}&ids=${ids}&intents=Charts,Exchanges&type=${type}&wrapodata=false`;
  const res = await fetch(url);
  if (!res.ok) throw new Error(`HTTP error! status: ${res.status}`);
  let data = await res.json();
  if (Array.isArray(data) && data.length && Array.isArray(data[0])) {
    data = data.flat(1);
  }
  if (!Array.isArray(data)) {
    throw new Error("Unexpected API response format");
  }
  return data.map(
    (item) => {
      const series = item.chart?.series;
      return {
        prices: series?.prices || [],
        timeStamps: series?.timeStamps || []
      };
    }
  );
}
__name(getMSNPrices, "getMSNPrices");
async function drawCurrencyChartSkia(ctx, prices, timeStamps, title = "兑换汇率", width = 1200, height = 600) {
  if (!prices.length || !timeStamps.length || prices.length !== timeStamps.length)
    return;
  const dataPoints = timeStamps.map((t, i) => ({
    x: new Date(t),
    y: prices[i]
  }));
  const { Canvas } = ctx.skia;
  const canvas = new Canvas(width, height);
  const whiteBackground = {
    id: "whiteBackground",
    beforeDraw(chart) {
      const { ctx: ctx2, width: width2, height: height2 } = chart;
      ctx2.save();
      ctx2.fillStyle = "#ffffff";
      ctx2.fillRect(0, 0, width2, height2);
      ctx2.restore();
    }
  };
  new import_chart.Chart(canvas, {
    type: "line",
    data: {
      datasets: [
        {
          label: title,
          data: dataPoints,
          showLine: true,
          // ⭐ 不连线，只画点
          pointRadius: 0,
          // 小点
          pointBorderWidth: 0,
          // ⭐ 去掉外边框 → 变成实心点
          borderWidth: 2,
          borderColor: "red",
          backgroundColor: "rgba(0,0,255,0.1)",
          tension: 0,
          segment: {
            borderDash: /* @__PURE__ */ __name((ctx2) => {
              const p0 = ctx2.p0.parsed.x;
              const p1 = ctx2.p1.parsed.x;
              const diffHours = (p1 - p0) / (1e3 * 60 * 60);
              return diffHours > 2 ? [6, 4] : [];
            }, "borderDash")
          }
        }
      ]
    },
    options: {
      responsive: false,
      scales: {
        x: {
          type: "time",
          time: {
            parser: "YYYY-MM-DD HH:mm:ss",
            tooltipFormat: "YYYY/MM/DD HH:mm",
            displayFormats: {
              hour: "HH:mm",
              day: "MM-DD"
            },
            unit: "day"
          },
          title: { display: true, text: "日期" }
        },
        y: {
          title: { display: true, text: "价格" }
        }
      },
      layout: {
        padding: { top: 20, bottom: 20, left: 20, right: 40 }
      }
    },
    plugins: [whiteBackground]
  });
  return await canvas.toBuffer("png");
}
__name(drawCurrencyChartSkia, "drawCurrencyChartSkia");
async function getCurrencyCodes() {
  if (currencyCodes?.length) return currencyCodes;
  currencyCodes = require(import_node_path.default.resolve(assetsDir, "./currency.json"));
  return currencyCodes;
}
__name(getCurrencyCodes, "getCurrencyCodes");
async function getCurrencyCodesMap() {
  if (Object.keys(currencyCodesMap)?.length) return currencyCodesMap;
  currencyCodesMap = await buildCurrencyIndex();
  return currencyCodesMap;
}
__name(getCurrencyCodesMap, "getCurrencyCodesMap");
async function buildCurrencyIndex() {
  const map = {};
  (await getCurrencyCodes()).forEach((c) => {
    map[c.country] = c.code;
    map[c.currencyCn] = c.code;
  });
  map.日元 = "JPY";
  map.人民币 = "CNY";
  map.港币 = "HKD";
  return map;
}
__name(buildCurrencyIndex, "buildCurrencyIndex");

// src/plugins/drawHead.ts
var fsp = __toESM(require("node:fs/promises"));
var import_node_path3 = __toESM(require("node:path"));
var import_koishi3 = require("koishi");

// src/utils.ts
var import_node_fs = __toESM(require("node:fs"));
var import_node_path2 = __toESM(require("node:path"));
var import_koishi2 = require("koishi");
var import_chartjs_adapter_dayjs_3 = require("chartjs-adapter-dayjs-3");
async function addRecord(ctx, gid, avatarUrl) {
  const recordDir = `${assetsDir}/record/${gid}`;
  const avatarBuffer = await ctx.http.get(avatarUrl, {
    responseType: "arraybuffer"
  });
  saveImage(avatarBuffer, recordDir);
  return "投稿收到啦";
}
__name(addRecord, "addRecord");
async function getRecord(cfg, gid, tag) {
  const links = structuredClone(cfg.recordLink);
  links[gid] = { linkGroup: gid, linkWeight: 100 };
  const selectGid = getRandomLinkGroup(links).replaceAll(":", "_");
  const recordDir = import_node_path2.default.join(assetsDir, "record", selectGid);
  const tagConfigPath = import_node_path2.default.join(assetsDir, "tagConfig", `${selectGid}.json`);
  if (!import_node_fs.default.existsSync(recordDir)) return null;
  const files = import_node_fs.default.readdirSync(recordDir).filter((file) => /\.(png|jpe?g|webp|gif)$/i.test(file));
  if (!files.length) return null;
  const tagConfigJson = import_node_fs.default.existsSync(tagConfigPath) ? JSON.parse(import_node_fs.default.readFileSync(tagConfigPath, "utf8") || "{}") : {};
  const weighted = files.map((file) => {
    const name2 = import_node_path2.default.parse(file).name;
    const tags = tagConfigJson[name2] || [];
    const weight = tag && tags.includes(tag) ? cfg.tagWeight : 1;
    return { file, weight };
  });
  const totalWeight = weighted.reduce((acc, cur) => acc + cur.weight, 0);
  let rand = Math.random() * totalWeight;
  for (const item of weighted) {
    rand -= item.weight;
    if (rand <= 0) {
      return import_node_path2.default.join(recordDir, item.file);
    }
  }
  return null;
}
__name(getRecord, "getRecord");
function getRandomLinkGroup(record) {
  const entries = Object.values(record);
  const totalWeight = entries.reduce((sum, item) => sum + item.linkWeight, 0);
  let r = Math.random() * totalWeight;
  for (const item of entries) {
    r -= item.linkWeight;
    if (r <= 0) {
      return item.linkGroup;
    }
  }
  return entries[entries.length - 1].linkGroup;
}
__name(getRandomLinkGroup, "getRandomLinkGroup");
function getTodayPrefix() {
  const now = /* @__PURE__ */ new Date();
  const yyyy = now.getFullYear();
  const mm = String(now.getMonth() + 1).padStart(2, "0");
  const dd = String(now.getDate()).padStart(2, "0");
  return `${yyyy}${mm}${dd}`;
}
__name(getTodayPrefix, "getTodayPrefix");
function getNextSequenceNumber(directory, prefix, suffix = "jpg") {
  const files = import_node_fs.default.existsSync(directory) ? import_node_fs.default.readdirSync(directory) : [];
  const regex = new RegExp(`^${prefix}-(\\d+)\\.${suffix}$`);
  let maxNum = 0;
  files.forEach((file) => {
    const match = file.match(regex);
    if (match) {
      const num = parseInt(match[1], 10);
      if (num > maxNum) {
        maxNum = num;
      }
    }
  });
  return maxNum + 1;
}
__name(getNextSequenceNumber, "getNextSequenceNumber");
function saveImage(arrayBuffer, directory, filename) {
  if (!filename) {
    const prefix = getTodayPrefix();
    const seq = getNextSequenceNumber(directory, prefix);
    filename = `${prefix}-${seq}.jpg`;
  }
  const filepath = import_node_path2.default.join(directory, filename);
  if (!import_node_fs.default.existsSync(directory)) {
    import_node_fs.default.mkdirSync(directory, { recursive: true });
  }
  const buffer = Buffer.from(arrayBuffer);
  import_node_fs.default.writeFileSync(filepath, buffer);
  console.log(`已保存图片：${filepath}`);
}
__name(saveImage, "saveImage");
async function getImageSrc(session, param, option) {
  const number = option?.number ?? true, img = option?.img ?? true, at = option?.at ?? true, noParam = option?.noParam ?? true, quote = option?.quote ?? true;
  if (number && param?.length && /^\d+$/.test(param)) {
    return `https://q1.qlogo.cn/g?b=qq&nk=${param}&s=640`;
  } else if (noParam && !param?.length) {
    return `https://q1.qlogo.cn/g?b=qq&nk=${session.userId}&s=640`;
  }
  if (quote) {
    const quoteElementArray = session?.quote?.elements;
    if (quoteElementArray?.length) {
      for (const element of quoteElementArray) {
        if (img && element?.type === "img") {
          return element?.attrs?.src;
        } else if (at && element?.type === "at" && element?.attrs?.id && element.attrs.id !== session.selfId) {
          return `https://q1.qlogo.cn/g?b=qq&nk=${element?.attrs?.id}&s=640`;
        }
      }
    }
  }
  const elementArray = session.elements;
  for (const element of elementArray) {
    if (img && element?.type === "img") {
      return element?.attrs?.src;
    } else if (at && element?.type === "at" && element?.attrs?.id) {
      return `https://q1.qlogo.cn/g?b=qq&nk=${element.attrs.id}&s=640`;
    }
  }
  return "";
}
__name(getImageSrc, "getImageSrc");
async function getImageFromUrl(ctx, url) {
  if (!url) throw new Error("URL must be provided");
  const sharp = ctx.QhzySharp.Sharp;
  try {
    let input;
    if (/^https?:\/\//i.test(url)) {
      const config = {
        responseType: "arraybuffer"
      };
      input = await ctx.http.get(url, config);
    } else {
      input = url;
    }
    return sharp(input).png();
  } catch (err) {
    console.error(`Error processing image from ${url}:`, err);
    throw new Error("Get image failed");
  }
}
__name(getImageFromUrl, "getImageFromUrl");
async function atNotSayReply(cfg, session, elements) {
  const trimElements = elements.filter(
    (e) => !(e.type === "text" && /^\s*$/.test(e.attrs.content))
  );
  if ((cfg.atNotSay || cfg.atNotSayOther) && trimElements.length === 1 && trimElements[0].type === "at") {
    const isAtSelf = trimElements[0].attrs.id === session.selfId;
    if (isAtSelf && cfg.atNotSay && import_koishi2.Random.bool(cfg.atNotSayProperty)) {
      await session.send(session.text("middleware.messages.atNotReply"));
    } else if (!isAtSelf && cfg.atNotSayOther && import_koishi2.Random.bool(cfg.atNotSayOtherProperty)) {
      await session.send(session.text("middleware.messages.atNotReplyOther"));
    }
  }
}
__name(atNotSayReply, "atNotSayReply");
async function replyBot(cfg, session, elements) {
  if (cfg.replyBot !== "关闭") {
    const bots = ["bot", "机器人", "Bot", "BOT", "机器人！", "机器人!", "人机"];
    const texts = elements?.filter((e) => e.type === "text").map((e) => e?.attrs?.content?.trim());
    const ats = elements?.filter((e) => e.type === "at").map((e) => e?.attrs?.id);
    const mentionedBot = texts?.some((t) => bots.includes(t));
    const atMe = ats?.includes(session.selfId);
    if (elements?.length === 1 && mentionedBot && cfg.replyBot === "无需at" || elements?.length === 2 && mentionedBot && atMe) {
      await session.send(session.text("middleware.messages.notBot"));
    }
  }
}
__name(replyBot, "replyBot");
async function iLoveYou(cfg, session, elements) {
  if (cfg.iLoveYou) {
    const hasAtMe = elements?.some(
      (e) => e.type === "at" && e?.attrs?.id === session.selfId
    );
    const hasLoveMessage = elements?.some(
      (e) => e.type === "text" && e?.attrs?.content?.trim() === session.text("middleware.messages.loveMessage")
    );
    if (elements?.length === 2 && hasAtMe && hasLoveMessage) {
      await session.send(
        session.text("middleware.messages.iLoveU", {
          at: import_koishi2.h.at(session.userId),
          quote: import_koishi2.h.quote(session.messageId)
        })
      );
    }
  }
}
__name(iLoveYou, "iLoveYou");
function parseFeatureControl(array) {
  try {
    return Object.fromEntries(
      array.map(
        ({ functionName, whitelist, groups }) => functionName?.length ? [
          functionName,
          {
            whitelist: !!whitelist,
            groups: groups?.split(",")?.map(Number).filter(Boolean) || []
          }
        ] : void 0
      )
    ) || {};
  } catch (e) {
    starfxLogger.warn("[功能控制] 解析失败", e);
    return {};
  }
}
__name(parseFeatureControl, "parseFeatureControl");
function detectControl(controlJson, guildId, funName) {
  const rule = controlJson?.[funName];
  if (!rule || rule.whitelist === void 0 || !Array.isArray(rule.groups)) {
    return true;
  }
  const inList = rule.groups.includes(Number(guildId));
  return rule.whitelist ? inList : !inList;
}
__name(detectControl, "detectControl");
function handleRoll(session) {
  const elements = session.elements;
  const parts = [];
  for (const element of elements) {
    if (element?.type === "text") {
      let str = element.attrs.content;
      let placeholder = "__TEMP__";
      while (str.includes(placeholder)) {
        placeholder += "_X";
      }
      str = str.replace(/我/g, placeholder).replace(/你/g, "我").replace(new RegExp(placeholder, "g"), "你");
      parts.push(...str.split(/(?:\s+)+/).filter(Boolean));
    } else {
      parts.push(element);
    }
  }
  console.log(parts);
  parts.shift();
  if (!parts) return session.text(".noParam");
  const last = session.elements[session.elements.length - 1];
  if (last?.type === "text" && last?.attrs?.content?.endsWith("概率") && last?.attrs?.content?.length > 3) {
    return session.text(".possibility", {
      param: parts,
      possibility: Math.floor(Math.random() * 1e4 + 1) / 100
    });
  }
  const items = parts.join(" ").split("r").filter(Boolean);
  if (items.length === 2) {
    const [num, noodles] = items.map(Number);
    return getPoints(session, num, noodles);
  }
  const newParts = [];
  parts.forEach((element) => {
    if (typeof element === "string") {
      newParts.push(...element.split(/(?:、|还是|，|,)+/).filter(Boolean));
    } else {
      newParts.push(element);
    }
  });
  if (newParts.length > 1) {
    return session.text(".choose", {
      option: import_koishi2.Random.pick(newParts)
    });
  }
  return session.text(".noParam");
}
__name(handleRoll, "handleRoll");
function getPoints(session, num, noodles) {
  if (!Number.isInteger(num) || !Number.isInteger(noodles) || num < 0 || noodles > 0)
    return session.text(".invalid");
  if (num > 20 || noodles > 1e8) return session.text(".too-many");
  const points = Array(num).fill(0).map(() => Math.floor(Math.random() * noodles + 1));
  return session.text(".noodles", {
    num,
    noodles,
    points: points.join(", ")
  });
}
__name(getPoints, "getPoints");
async function undo(_cfg, session) {
  if (session?.quote?.id && session.quote.user.id === session.selfId && Date.now() - session.quote.timestamp < 2 * 60 * 1e3 - 5 * 1e3) {
    await session.bot.deleteMessage(
      session.channelId || session.guildId,
      session.quote.id
    );
  }
}
__name(undo, "undo");
function safeQuote(str, useQuote) {
  let unquoted = str.trim();
  if (unquoted.startsWith('"')) {
    unquoted = unquoted.slice(1);
  }
  if (unquoted.endsWith('"')) {
    unquoted = unquoted.slice(0, -1);
  }
  if (useQuote) {
    return `"${unquoted}"`;
  } else {
    return unquoted;
  }
}
__name(safeQuote, "safeQuote");
function writeMap(map, dest) {
  const dir = import_node_path2.default.dirname(dest);
  if (!import_node_fs.default.existsSync(dir)) {
    import_node_fs.default.mkdirSync(dir, { recursive: true });
  }
  import_node_fs.default.writeFileSync(dest, JSON.stringify([...map], null, 2), "utf-8");
}
__name(writeMap, "writeMap");
function readMap(url) {
  if (!import_node_fs.default.existsSync(url)) {
    const map = /* @__PURE__ */ new Map();
    writeMap(map, url);
  }
  const raw = import_node_fs.default.readFileSync(url, "utf-8");
  const mapArray = JSON.parse(raw);
  return new Map(mapArray);
}
__name(readMap, "readMap");
function ready(session, cfg, param, readyMap) {
  const nowReadyMap = cfg.saveReadyAsFile ? readMap(cfg.saveReadyAsFile) : readyMap;
  let strArr = nowReadyMap.get(session.gid) ?? [];
  let returnMessage = session.text(".invalid");
  if (param === "+" || param === "+1") {
    strArr.push(session.username);
    returnMessage = session.text(".addReady", {
      num: strArr.length,
      list: strArr.join("\n")
    });
  } else if (param === "-" || param === "-1") {
    const newStrArr = strArr.filter((item) => item !== session.username);
    returnMessage = newStrArr.length !== strArr.length ? session.text(".delReady", {
      num: newStrArr.length,
      list: newStrArr.join("\n")
    }) : session.text(".delFailed", {
      num: newStrArr.length,
      list: newStrArr.join("\n")
    });
    strArr = newStrArr;
  } else if (param === "0") {
    strArr.length = 0;
    returnMessage = session.text(".clearReady");
  } else if (param === "" || !param) {
    returnMessage = session.text(".listReady", {
      num: strArr.length,
      list: strArr.join("\n")
    });
  }
  nowReadyMap.set(session.gid, strArr);
  writeMap(nowReadyMap, cfg.saveReadyAsFile);
  return returnMessage;
}
__name(ready, "ready");

// src/plugins/drawHead.ts
async function handleBanGDreamConfig(options) {
  const types = {
    cool: ["cool", "blue", "蓝", "蓝色"],
    powerful: ["powerful", "red", "红", "红色"],
    pure: ["pure", "green", "绿", "绿色"],
    happy: ["happy", "orange", "橙", "橙色"]
  };
  const bands = {
    ppp: [
      "ppp",
      "poppin'Party",
      "破琵琶",
      "步品破茶",
      "poppin",
      "popipa",
      "poppinparty",
      "ポピパ"
    ],
    ag: ["ag", "afterglow", "夕阳红", "悪蓋愚狼"],
    pp: [
      "pp",
      "pastel＊palettes",
      "pastel*palettes",
      "pastelPalettes",
      "怕死怕累",
      "pastel",
      "palettes",
      "pasupare",
      "パスパレ",
      "破巣照破烈斗"
    ],
    hhw: [
      "hhw",
      "ハロー、ハッピーワールド！",
      "hello,happyworld!",
      "hellohappyworld",
      "ハロハピ",
      "hello，happyworld！",
      "harohapi",
      "破狼法被威悪怒",
      "儿歌团",
      "好好玩"
    ],
    r: [
      "r",
      "roselia",
      "露世里恶",
      "萝",
      "露世裏悪",
      "ロゼリア",
      "r组",
      "相声团",
      "相声组"
    ],
    ras: [
      "ras",
      "raiseasuilen",
      "raise",
      "suilen",
      "ラス",
      "零図悪酔恋",
      "睡莲",
      "麗厨唖睡蓮",
      "睡蓮"
    ],
    mnk: [
      "mnk",
      "モニカ",
      "蝶团",
      "蝶",
      "morfonica",
      "毛二力",
      "monika",
      "monica"
    ],
    go: [
      "go",
      "mygo!!!!!",
      "mygo！！！！！",
      "mygo",
      "我去！！！！！",
      "我去!!!!!",
      "我去",
      "卖狗"
    ]
  };
  const trains = {
    color_star: ["花后", "1", "彩", "true"],
    normal_star: ["花前", "0", "false"]
  };
  const drawConfig = {
    color: "",
    band: "",
    starType: "",
    starNum: 0,
    border: ""
  };
  if (options?.color) {
    for (const [type, aliases] of Object.entries(types)) {
      if (aliases.includes(options.color.toLowerCase())) {
        drawConfig.color = type;
        break;
      }
    }
  }
  if (options?.band) {
    for (const [band, aliases] of Object.entries(bands)) {
      if (aliases.includes(options.band.toLowerCase())) {
        drawConfig.band = band;
        break;
      }
    }
  }
  const starNum = options?.starNum ? parseInt(options.starNum, 10) : 0;
  drawConfig.starNum = starNum > 0 && starNum < 10 ? starNum : 0;
  if (options?.train) {
    for (const [train, aliases] of Object.entries(trains)) {
      if (aliases.includes(options.train.toLowerCase())) {
        drawConfig.starType = train;
        break;
      }
    }
  }
  return drawConfig;
}
__name(handleBanGDreamConfig, "handleBanGDreamConfig");
async function drawLock(ctx, baseImage) {
  let image;
  try {
    image = await getImageFromUrl(ctx, baseImage);
  } catch (error) {
    if (error.message === "Get image failed") {
      return "发生错误";
    } else if (error.message === "URL must be provided") {
      return "输入无效";
    } else {
      console.error(error.message);
    }
  }
  const sharp = ctx.QhzySharp.Sharp;
  const imageMetadata = await image.metadata();
  const lockUrl = `${baseDir}/data/starfx-bot/assets/lock.png`;
  const size1 = Math.min(imageMetadata.width, imageMetadata.height);
  image.resize({ width: size1, height: size1, fit: "cover" });
  const overlay = sharp(lockUrl).png();
  overlay.resize({ width: size1 });
  image.composite([{ input: await overlay.toBuffer() }]);
  return import_koishi3.h.image(await image.png().toBuffer(), "image/png");
}
__name(drawLock, "drawLock");
async function drawSold(ctx, baseImage) {
  let image;
  try {
    image = await getImageFromUrl(ctx, baseImage);
  } catch (error) {
    if (error.message === "Get image failed") {
      return "发生错误";
    } else if (error.message === "URL must be provided") {
      return "输入无效";
    } else {
      console.error(error.message);
    }
  }
  const sharp = ctx.QhzySharp.Sharp;
  const imageMetadata = await image.metadata();
  const size1 = Math.min(imageMetadata.width, imageMetadata.height);
  image.resize({ width: size1, height: size1, fit: "cover" });
  const middle = sharp({
    create: {
      width: size1,
      height: size1,
      channels: 4,
      background: { r: 255, g: 255, b: 255, alpha: 0.4 }
    }
  }).png();
  const soldUrl = `${assetsDir}/sold.png`;
  const overlay = sharp(soldUrl).png();
  const overlaySize = Math.round(size1 * 182 / 240);
  overlay.resize({
    width: overlaySize,
    height: overlaySize
  });
  const topLeft = Math.round(overlaySize * 29 / 182);
  image.composite([
    { input: await middle.toBuffer() },
    {
      input: await overlay.toBuffer(),
      top: topLeft,
      left: topLeft
    }
  ]);
  return import_koishi3.h.image(await image.png().toBuffer(), "image/png");
}
__name(drawSold, "drawSold");
async function drawBanGDream(ctx, avatar, inputOptions) {
  if (!avatar) return "";
  const colors = ["cool", "pure", "happy", "powerful"];
  const bands = ["ppp", "ag", "pp", "r", "hhw", "ras", "mnk", "go"];
  const starTypes = ["normal_star", "color_star"];
  const starNums = [1, 2, 3, 4, 5];
  const options = {
    color: inputOptions?.color || colors[Math.floor(Math.random() * colors.length)],
    band: inputOptions?.band || bands[Math.floor(Math.random() * bands.length)],
    starNum: inputOptions?.starNum || starNums[Math.floor(Math.random() * starNums.length)],
    starType: inputOptions?.starType || "",
    border: inputOptions?.border || ""
  };
  options.starType ||= options.starNum < 3 ? starTypes[0] : starTypes[Math.floor(Math.random() * starTypes.length)];
  options.border ||= `card-${starNums.includes(options.starNum) ? options.starNum : 5}${options.starNum === 1 ? `-${options.color}` : ""}`;
  try {
    const zoom = 2;
    const [avatarSharp, colorBuffer, bandBuffer, starBuffer, borderBuffer] = await Promise.all([
      getImageFromUrl(ctx, avatar),
      fsp.readFile(
        import_node_path3.default.join(assetsDir, "bangborder", `${options.color}.png`)
      ),
      fsp.readFile(import_node_path3.default.join(assetsDir, "bangborder", `${options.band}.png`)),
      fsp.readFile(
        import_node_path3.default.join(assetsDir, "bangborder", `${options.starType}.png`)
      ),
      fsp.readFile(
        import_node_path3.default.join(assetsDir, "bangborder", `${options.border}.png`)
      )
    ]);
    const sharp = ctx.QhzySharp.Sharp;
    let image = avatarSharp.resize(500 * zoom, 500 * zoom, { fit: "cover" });
    const borderImage = await sharp(borderBuffer).resize(500 * zoom, 500 * zoom, { fit: "cover" }).toBuffer();
    const colorImage = await sharp(colorBuffer).resize(130 * zoom, 130 * zoom, { fit: "cover" }).toBuffer();
    let bandSharp = sharp(bandBuffer);
    const bandMeta = await bandSharp.metadata();
    if (bandMeta.width !== void 0 && bandMeta.height !== void 0) {
      if (bandMeta.width > bandMeta.height) {
        bandSharp = bandSharp.resize({ width: Math.round(120 * zoom) });
      } else {
        bandSharp = bandSharp.resize({ height: Math.round(120 * zoom) });
      }
    } else {
      starfxLogger.warn(
        "Sharp metadata missing width or height. Skipping resize logic."
      );
      return "无法识别图片Metadata";
    }
    const bandImage = await bandSharp.toBuffer();
    const starImage = await sharp(starBuffer).resize({ width: Math.round(90 * zoom) }).toBuffer();
    const starComposites = Array.from({ length: options.starNum }, (_, i) => ({
      input: starImage,
      left: Math.round(10 * zoom),
      top: Math.round(410 * zoom - i * 60 * zoom)
    }));
    image = image.composite([
      // composite border
      { input: borderImage, left: 0, top: 0 },
      // composite color
      {
        input: colorImage,
        left: Math.round(500 * zoom - 130 * zoom - 3 * zoom),
        top: 6
      },
      // composite band
      {
        input: bandImage,
        left: Math.round(15 * zoom),
        top: Math.round(15 * zoom)
      },
      // composite stars
      ...starComposites
    ]);
    const buffer = await image.png().toBuffer();
    return `data:image/png;base64,${buffer.toString("base64")}`;
  } catch (err) {
    console.error(err);
    return "";
  }
}
__name(drawBanGDream, "drawBanGDream");

// src/plugins/getOriginImg.ts
var cheerio2 = __toESM(require("cheerio"));
var import_koishi4 = require("koishi");
var import_rss_parser = __toESM(require("rss-parser"));
async function getXUrl(urls) {
  const regex = /https:\/\/x\.com\/([^/]+)\/status\/(\d+)/g;
  let match;
  const results = [];
  match = regex.exec(urls);
  do {
    if (match === null) {
      break;
    }
    const [fullUrl] = match;
    results.push(fullUrl);
    match = regex.exec(urls);
  } while (match !== null);
  return results;
}
__name(getXUrl, "getXUrl");
var parser = new import_rss_parser.default({
  customFields: {
    item: ["description", "link"]
  }
});
async function getXNum(session) {
  return session.content.trim().split(" ").slice(1).filter((item) => !Number.isNaN(+item) && item).map((str) => Number(str) - 1);
}
__name(getXNum, "getXNum");
async function getXImage(rssUrl, xUrls) {
  const xUrlsArray = Array.isArray(xUrls) ? xUrls : [xUrls];
  const feed = await parser.parseURL(rssUrl);
  const allImageUrls = [];
  for (const xUrl of xUrlsArray) {
    const item = feed.items.find((i) => i.link === xUrl);
    if (item) {
      const $ = cheerio2.load(item.description);
      $("img").each((_, el) => {
        const src = $(el).attr("src");
        if (src) allImageUrls.push(src);
      });
    }
  }
  return allImageUrls;
}
__name(getXImage, "getXImage");
function chunk(arr, size) {
  const res = [];
  for (let i = 0; i < arr.length; i += size) {
    res.push(arr.slice(i, i + size));
  }
  return res;
}
__name(chunk, "chunk");
async function sendImages(ctx, session, imageUrls) {
  const chunks = chunk(imageUrls, 10);
  for (const group of chunks) {
    const messages = await Promise.all(
      group.map(async (url) => import_koishi4.h.image(await getXImageBase64(ctx, url)))
    );
    if (messages.length > 0) {
      const message = messages.join("");
      await session.send(message);
    } else {
      await session.send(
        "未找到图片，请引用包含图片且处于RSS列表中的的推特链接"
      );
    }
  }
}
__name(sendImages, "sendImages");
async function getXImageBase64(ctx, url) {
  const res = await ctx.http.get(url, { responseType: "arraybuffer" });
  const base64 = Buffer.from(res).toString("base64");
  return `data:image/png;base64,${base64}`;
}
__name(getXImageBase64, "getXImageBase64");

// src/index.ts
var name = "starfx-bot";
var inject = {
  optional: ["skia", "QhzySharp"]
};
var baseDir;
var assetsDir;
var starfxLogger = new import_koishi5.Logger("starfx-bot");
var usage = `
<h1>StarFreedomX的自用插件</h1>
<h2>可选功能依赖：</h2>
<h3><a href="/market?keyword=skia-canvas">skia</a></h3>
    <li>查汇率</li>
<h3><a href="/market?keyword=@quanhuzeyu+sharp">QhzySharp</a></h3>
    <li>卖掉了</li>
    <li>封印</li>
    <li>bdbd</li>
  `;
var repeatContextMap = /* @__PURE__ */ new Map();
var Config = import_koishi5.Schema.intersect([
  import_koishi5.Schema.object({
    openLock: import_koishi5.Schema.boolean().default(true).description("开启明日方舟封印功能"),
    openSold: import_koishi5.Schema.boolean().default(true).description('开启闲鱼"卖掉了"功能'),
    bangdreamBorder: import_koishi5.Schema.boolean().default(true).description("开启BanG Dream!边框功能")
  }).description("绘图功能"),
  import_koishi5.Schema.object({
    record: import_koishi5.Schema.boolean().default(true).description("开启群语录功能"),
    tagWeight: import_koishi5.Schema.number().default(5).min(1).description("tag匹配时的权重，越高权重越大"),
    recordLink: import_koishi5.Schema.dict(
      import_koishi5.Schema.object({
        linkGroup: import_koishi5.Schema.string(),
        linkWeight: import_koishi5.Schema.number()
      })
    ).role("table").description(
      "群组链接，使得群可以调用被链接群的语录，<br>可以配置权重，作为键的群自身的权重为100(注意格式为平台名:群组名)"
    ),
    saveArchive: import_koishi5.Schema.boolean().default(false).description("开启入典功能").hidden()
  }).description("语录记录功能"),
  import_koishi5.Schema.object({
    roll: import_koishi5.Schema.boolean().default(true).description("开启roll随机数功能"),
    undo: import_koishi5.Schema.boolean().default(true).description("机器人撤回消息功能(只测试了qq的onebot适配器)"),
    echo: import_koishi5.Schema.boolean().default(true).description("echo回声洞功能"),
    echoBanner: import_koishi5.Schema.array(String).role("table").description("echo屏蔽词，对文本生效"),
    ready: import_koishi5.Schema.boolean().default(false).description("待机人数记录功能"),
    saveReadyAsFile: import_koishi5.Schema.string().description(
      "写入待机人数的本地地址，留空则不写入"
    ),
    roomNumber: import_koishi5.Schema.boolean().default(false).description("主跑房间号记录功能"),
    saveRoomAsFile: import_koishi5.Schema.string().description(
      "写入房间号的本地地址，留空则不写入"
    ),
    forward: import_koishi5.Schema.boolean().default(true).description("消息转发功能"),
    searchExchangeRate: import_koishi5.Schema.boolean().default(false).description("查汇率功能"),
    intervalGetExchangeRate: import_koishi5.Schema.boolean().default(false).description("汇率定时推送功能").hidden()
  }).description("指令小功能"),
  import_koishi5.Schema.object({
    atNotSay: import_koishi5.Schema.boolean().default(true).description("开启‘艾特我又不说话’功能"),
    atNotSayProperty: import_koishi5.Schema.number().role("slider").min(0).max(1).step(0.01).default(0.5).description("'艾特我又不说话'回复概率"),
    atNotSayOther: import_koishi5.Schema.boolean().default(true).description("开启‘艾特他又不说话’功能"),
    atNotSayOtherProperty: import_koishi5.Schema.number().role("slider").min(0).max(1).step(0.01).default(0.5).description("'艾特他又不说话'回复概率"),
    iLoveYou: import_koishi5.Schema.boolean().default(true).description("开启‘我喜欢你’功能"),
    replyBot: import_koishi5.Schema.union(["关闭", "无需at", "必须at"]).default("无需at").description("回复‘我才不是机器人！’功能")
  }).description("特定回应功能"),
  import_koishi5.Schema.object({
    sendLocalImage: import_koishi5.Schema.dict(
      import_koishi5.Schema.object({
        hiddenInHelp: import_koishi5.Schema.boolean(),
        imgPath: import_koishi5.Schema.string()
      })
    ).role("table").description(
      "特定指令发送本地图片功能，其中键是指令名称，imgPath是图片文件的绝对路径"
    )
  }),
  import_koishi5.Schema.object({
    openRepeat: import_koishi5.Schema.boolean().default(true).description("开启复读功能"),
    minRepeatTimes: import_koishi5.Schema.number().default(2).description("最少重复次数"),
    repeatPossibility: import_koishi5.Schema.number().role("slider").min(0).max(1).step(0.01).default(0.3).description("复读发生概率")
  }).description("复读功能"),
  import_koishi5.Schema.object({
    myId: import_koishi5.Schema.boolean().default(false).description("查询gid uid cid")
  }).description("我的信息查询"),
  import_koishi5.Schema.object({
    filePathToBase64: import_koishi5.Schema.boolean().default(false).description(
      "在消息发送前检查是否有file://,如果有那么转换为base64再发送"
    ),
    originImg: import_koishi5.Schema.boolean().default(false).description("根据链接获取原图开关")
  }).description("自用功能"),
  import_koishi5.Schema.union([
    import_koishi5.Schema.object({
      originImg: import_koishi5.Schema.const(true).required(),
      originImgRSSUrl: import_koishi5.Schema.string().required().description("推特列表rss地址")
    }),
    import_koishi5.Schema.object({})
  ]),
  import_koishi5.Schema.object({
    featureControl: import_koishi5.Schema.array(
      import_koishi5.Schema.object({
        functionName: import_koishi5.Schema.string(),
        whitelist: import_koishi5.Schema.boolean(),
        groups: import_koishi5.Schema.string()
      })
    ).role("table").description(`黑/白名单配置，群组间用英文半角逗号分隔，<br>
可配置功能键及用法详见 [项目地址](https://github.com/StarFreedomX/starfx-bot)或[npm发布页](https://www.npmjs.com/package/koishi-plugin-bangdream-ccg)`)
  }).description("高级配置")
]);
function apply(ctx, cfg) {
  ctx.i18n.define("zh-CN", require_zh_CN());
  baseDir = ctx.baseDir;
  assetsDir = `${ctx.baseDir}/data/starfx-bot/assets`;
  initAssets();
  let featureControl = parseFeatureControl(cfg.featureControl);
  if (cfg.openLock) {
    ctx.command("封印 [param]").action(async ({ session }, param) => {
      if (ctx.QhzySharp && detectControl(featureControl, session.guildId, "lock"))
        await session.send(
          await drawLock(ctx, await getImageSrc(session, param))
        );
    });
  }
  if (cfg.openSold) {
    ctx.command("卖掉了 [param]").action(async ({ session }, param) => {
      if (ctx.QhzySharp && detectControl(featureControl, session.guildId, "sold"))
        await session.send(
          await drawSold(ctx, await getImageSrc(session, param))
        );
    });
  }
  if (cfg.roll) {
    ctx.command("roll").action(async ({ session }) => {
      if (detectControl(featureControl, session.guildId, "roll")) {
        return handleRoll(session);
      }
    });
  }
  if (cfg.echo) {
    ctx.command("echo <params>").option("time", "-t <time: number> 指定时间(min)").action(async ({ session, options }, params) => {
      if (detectControl(featureControl, session.guildId, "echo")) {
        const elements = session.elements;
        const getEchoMessage = /* @__PURE__ */ __name(() => {
          try {
            while (elements[0].type === "at" || elements[0].type === "text" && !elements[0].attrs?.content.trim())
              elements.shift();
            elements[0].attrs.content = elements[0].attrs?.content.trim().split(/\s/).slice(1).join(" ");
            elements.forEach((ele) => {
              ele.attrs.content = ele.attrs?.content.trim().split(/\s/).filter(
                (v, i, a) => v !== "-t" && a[i - 1] !== "-t"
              ).join(" ");
            });
            if (elements.length === 1 && !elements[0].attrs.content?.length) {
              if (cfg.echoBanner?.some(
                (banText) => session.quote?.content?.includes(banText)
              ))
                return "包含屏蔽词，打断echo";
              return session.quote?.elements;
            }
            if (cfg.echoBanner?.some(
              (banText) => session.content?.includes(banText)
            ))
              return "包含屏蔽词，打断echo";
            return elements;
          } catch (_e) {
            return params;
          }
        }, "getEchoMessage");
        const echoMessages = getEchoMessage();
        if (!options?.time && options.time > 0) {
          return echoMessages;
        } else {
          setTimeout(
            async () => {
              await ctx.broadcast([session.gid], echoMessages);
            },
            options.time * 60 * 1e3
          );
        }
      }
    });
  }
  if (cfg.bangdreamBorder) {
    ctx.command("bdbd [param]").option("starNum", "-n <starNum: number>").option("color", "-c <color: string>").option("train", "-t <train: string>").option("band", "-b <band: string>").action(async ({ session, options }, param) => {
      if (ctx.QhzySharp && detectControl(featureControl, session.guildId, "bdbd")) {
        const drawConfig = await handleBanGDreamConfig(options);
        const imgSrc = await getImageSrc(session, param);
        if (!imgSrc?.length) return "输入无效";
        const imageBase64 = await drawBanGDream(
          ctx,
          imgSrc,
          drawConfig
        );
        if (!imageBase64?.length) return "输入无效";
        await session.send(import_koishi5.h.image(imageBase64));
      }
    });
  }
  if (cfg.record) {
    ctx.command("投稿 [param]").action(async ({ session }, param) => {
      if (detectControl(featureControl, session.guildId, "record") && detectControl(featureControl, session.guildId, "record-push")) {
        const imageSrc = await getImageSrc(session, param, {
          img: true,
          at: false,
          quote: true,
          noParam: false,
          number: false
        });
        if (!imageSrc) {
          return "请发送带图片的指令消息或引用图片消息进行投稿";
        }
        return await addRecord(
          ctx,
          session.gid.replaceAll(":", "_"),
          imageSrc
        );
      }
    });
    ctx.command("语录 [tag:string]").action(async ({ session }, tag) => {
      if (detectControl(featureControl, session.guildId, "record") && detectControl(featureControl, session.guildId, "record-get")) {
        const filepath = await getRecord(
          cfg,
          session.gid.replaceAll(":", "_"),
          tag
        );
        starfxLogger.info(`send record: ${filepath}`);
        if (!filepath) return "暂无语录呢";
        await session.send(import_koishi5.h.image(filepath));
      }
    });
  }
  for (const key in cfg.sendLocalImage) {
    ctx.command(key, { hidden: cfg.sendLocalImage[key].hiddenInHelp }).action(async ({ session }) => {
      if (detectControl(
        featureControl,
        session.guildId,
        "sendLocalImage"
      ) && detectControl(featureControl, session.guildId, key))
        return import_koishi5.h.image(
          safeQuote(cfg.sendLocalImage[key].imgPath, false)
        );
    });
  }
  if (cfg.roomNumber) {
    const roomNumMap = /* @__PURE__ */ new Map();
    ctx.command("room-number [param: string]").usage("记录房间号").action(async ({ session }, param) => {
      const nowRoomNumMap = cfg.saveRoomAsFile ? readMap(cfg.saveRoomAsFile) : roomNumMap;
      const room = nowRoomNumMap.get(session.gid);
      if (!param) {
        return room ? session.text(".roomNumber", {
          room
        }) : session.text(".noRoom");
      } else {
        let returnMessage = session.text(".invalid");
        if (/^[0-9]{5,6}$/.test(param)) {
          const had = nowRoomNumMap.get(session.gid);
          nowRoomNumMap.set(session.gid, param);
          returnMessage = had ? session.text(".changeRoom", { oldRoom: room, newRoom: param }) : session.text(".newRoom", { room: param });
        } else if (String(param) === "0") {
          nowRoomNumMap.delete(session.gid);
          returnMessage = session.text(".delRoom", { room });
        }
        writeMap(nowRoomNumMap, cfg.saveRoomAsFile);
        return returnMessage;
      }
    });
  }
  if (cfg.ready) {
    const readyMap = /* @__PURE__ */ new Map();
    ctx.command("waiting-play [param:text]", { strictOptions: true }).usage("待机").action(async ({ session }, param) => {
      return ready(session, cfg, param, readyMap);
    });
  }
  if (cfg.saveArchive) {
    ctx.command("入典").action(async ({ session }) => {
      if (!session.quote) return "请引用合并转发聊天记录进行入典";
    });
  }
  if (cfg.undo) {
    ctx.command("undo").alias("撤回").usage("撤回消息").action(async ({ session }) => {
      if (detectControl(featureControl, session.guildId, "undo"))
        await undo(cfg, session);
    });
  }
  if (cfg.forward) {
    ctx.command("forward").option("group", "-g <group:string>").option("platform", "-p <platform:string>").usage("转发消息").action(async ({ session, options }) => {
      if (detectControl(featureControl, session.guildId, "forward")) {
        const mapPath = import_node_path4.default.join(assetsDir, "forward.json");
        const groupMap = readMap(mapPath);
        if (options.group) {
          if (["0", "clear", "del"].includes(options.group)) {
            const gid = groupMap.get(session.gid);
            groupMap.delete(session.gid);
            writeMap(groupMap, mapPath);
            return session.text(".delete", { gid });
          }
          const target2 = `${options.platform || session.platform}:${options.group}`;
          groupMap.set(session.gid, target2);
          writeMap(groupMap, mapPath);
          if (!session.quote?.content?.length)
            return session.text(".setOK", { target: target2 });
        }
        const target = groupMap.get(session.gid);
        if (!target) return session.text(".noTarget");
        if (!session.quote?.content?.length)
          return session.text(".noMessage");
        const forwardContent = session.text(".forwardContent", {
          content: session.quote.content
        });
        await ctx.broadcast([target], forwardContent);
        return session.text(".success", { target });
      }
    });
  }
  if (cfg.originImg) {
    ctx.command("获取X原图").alias("推特原图").usage("获取推特原图").action(async ({ session }) => {
      if (detectControl(featureControl, session.guildId, "originImg")) {
        let [xUrls, xIndex] = await Promise.all([
          getXUrl(session?.quote?.content),
          getXNum(session)
        ]);
        xIndex = xIndex.length ? xIndex : xUrls.map((_, i) => i);
        const filteredUrls = xIndex.filter((i) => i >= 0 && i < xUrls.length).map((i) => xUrls[i]);
        const imageUrls = await getXImage(
          cfg.originImgRSSUrl,
          filteredUrls
        );
        await sendImages(ctx, session, imageUrls);
      }
    });
  }
  if (cfg.myId) {
    ctx.command("my-gid").action(
      ({ session }) => detectControl(featureControl, session.guildId, "myId") ? session.gid : ""
    );
    ctx.command("my-uid").action(
      ({ session }) => detectControl(featureControl, session.guildId, "myId") ? session.uid : ""
    );
    ctx.command("my-cid").action(
      ({ session }) => detectControl(featureControl, session.guildId, "myId") ? session.cid : ""
    );
  }
  if (cfg.searchExchangeRate) {
    ctx.command("查汇率 <exchangeParam:text>").usage("查询当前汇率").example("查汇率 JPY : 查询日元兑换人民币的汇率(3位字母)").example("查汇率 JPYCNY : 查询日元兑换人民币的汇率(6位字母)").example("查汇率 -r avdzk2 : 查询日元兑换人民币的汇率(msn代码avdzk2)").example(
      "查汇率 -r auvwoc : 查询黄金的价格(msn代码auvwoc, 很怪吧我也不知道为什么是这个)"
    ).option("raw", "-r <raw:string>").action(async ({ session, options }, exchangeParam) => {
      if (ctx.skia && detectControl(featureControl, session.guildId, "exchangeRate")) {
        return await getExchangeRate(
          ctx,
          cfg,
          session,
          exchangeParam,
          options?.raw
        );
      }
    });
  }
  if (cfg.intervalGetExchangeRate) {
    ctx.command("开启汇率推送 [exchangeParam:string]").action(async ({ session }, exchangeParam) => {
      if (ctx.skia && detectControl(featureControl, session.guildId, "exchangeRate")) {
        const exchangeRatePath = import_node_path4.default.join(assetsDir, "exchangeRate.json");
        return await intervalGetExchangeRate(
          ctx,
          cfg,
          session,
          exchangeParam,
          exchangeRatePath
        );
      }
    });
  }
  if (cfg.filePathToBase64) {
    let isLocalPath = function(src) {
      const LOCAL_PATH_REGEX = /^(\/|\\|file:\/\/\/|\.|[A-Za-z]:\\|\/home\/|\/root\/|\.\.\/|\.\/)/;
      return LOCAL_PATH_REGEX.test(src);
    }, convertUriToLocalPath = function(uri) {
      if (uri.startsWith("file://")) {
        try {
          const url = new URL(uri);
          let filePath = url.pathname;
          if (process.platform === "win32") {
            if (filePath.match(/^\/[A-Za-z]:\//)) {
              filePath = filePath.substring(1);
            } else {
            }
          }
          return import_node_path4.default.normalize(filePath);
        } catch (e) {
          starfxLogger.error("URL解析失败:", e);
          return uri;
        }
      }
      if (import_node_path4.default.isAbsolute(uri) || uri.startsWith(".") || uri.startsWith("..")) {
        return import_node_path4.default.normalize(uri);
      }
      return uri;
    }, toBase64String = function(src) {
      try {
        const data = fs2.readFileSync(src);
        return data.toString("base64");
      } catch (err) {
        starfxLogger.error(
          `[Error] 无法读取本地文件 (${src}) 并转换为 Base64:`,
          err
        );
        return void 0;
      }
    }, guessTypeFromElement = function(type) {
      switch (type) {
        case "img":
        case "image":
          return "image/png";
        case "record":
          return "audio/mpeg";
        case "video":
          return "video/mp4";
        default:
          return void 0;
      }
    };
    __name(isLocalPath, "isLocalPath");
    __name(convertUriToLocalPath, "convertUriToLocalPath");
    __name(toBase64String, "toBase64String");
    __name(guessTypeFromElement, "guessTypeFromElement");
    ctx.before("send", (session) => {
      for (const element of session.elements) {
        const src = element.attrs?.src;
        if (!src || !isLocalPath(src)) continue;
        const filePath = convertUriToLocalPath(src);
        const mimeType = import_mime_types.default.lookup(filePath) || guessTypeFromElement(element.type) || "application/octet-stream";
        const base64 = toBase64String(filePath);
        if (base64) element.attrs.src = `data:${mimeType};base64,${base64}`;
      }
    });
  }
  ctx.middleware(async (session, next) => {
    const elements = session.elements;
    if (cfg.openRepeat && detectControl(featureControl, session.guildId, "repeat")) {
      const content = session.content;
      const ctxArr = repeatContextMap.get(session.gid);
      if (!ctxArr || ctxArr[0] !== content) {
        repeatContextMap.set(session.gid, [content, 1]);
      } else {
        if (ctxArr[1] !== -1 && ++ctxArr[1] >= cfg.minRepeatTimes && import_koishi5.Random.bool(cfg.repeatPossibility)) {
          ctxArr[1] = -1;
          await session.send(content);
        }
      }
    }
    if (cfg.atNotSay && detectControl(featureControl, session.guildId, "atNotSay"))
      await atNotSayReply(cfg, session, elements);
    if (cfg.replyBot && detectControl(featureControl, session.guildId, "replyBot"))
      await replyBot(cfg, session, elements);
    if (cfg.iLoveYou && detectControl(featureControl, session.guildId, "iLoveYou"))
      await iLoveYou(cfg, session, elements);
    return next();
  });
  if (process.env.NODE_ENV === "development") {
    ctx.command("test [params]").action(async ({ session }) => {
      await session.send("test");
    });
    ctx.middleware(async (session, next) => {
      await session.send("");
      return next();
    });
  }
  function initAssets() {
    const defaultAssetsDir = import_node_path4.default.join(__dirname, "../assets");
    assetsDir = import_node_path4.default.join(ctx.baseDir, "data/starfx-bot/assets");
    if (!fs2.existsSync(assetsDir)) fs2.mkdirSync(assetsDir, { recursive: true });
    const versionFile = import_node_path4.default.join(assetsDir, "plugin_version.json");
    let localVersion = "0";
    if (fs2.existsSync(versionFile)) {
      try {
        localVersion = JSON.parse(fs2.readFileSync(versionFile, "utf-8")).version || "0";
      } catch {
      }
    }
    const pluginVersion = package_default.version;
    if (pluginVersion > localVersion) {
      try {
        if (fs2.existsSync(defaultAssetsDir)) {
          fs2.cpSync(defaultAssetsDir, assetsDir, {
            recursive: true,
            force: true
          });
        }
        fs2.writeFileSync(
          versionFile,
          JSON.stringify({ version: pluginVersion })
        );
      } catch (err) {
        console.error("initAssets copy failed:", err);
      }
    }
  }
  __name(initAssets, "initAssets");
}
__name(apply, "apply");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Config,
  apply,
  assetsDir,
  baseDir,
  inject,
  name,
  repeatContextMap,
  starfxLogger,
  usage
});
