// koishi-plugin-reactle/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const wordle_1 = require("@koishijs/wordle");
const koishi_1 = require("koishi");
const wordlist_json_1 = __importDefault(require("./data/wordlist.json"));
exports.default = (0, wordle_1.defineVariation)({
    name: 'koishi-plugin-reactle',
    command: 'reactle',
    Config: koishi_1.Schema.intersect([]).i18n({
        'zh-CN': require('./locales/schema.zh-CN'),
    }),
    locales: {
        'zh-CN': require('./locales/zh-CN'),
    },
    validWords: wordlist_json_1.default.map((word) => word.split('')),
    init(command) {
        command.option('random', '-r');
    },
    async getCurrentWord({ options }) {
        if (options.random) {
            return wordlist_json_1.default[Math.floor(Math.random() * wordlist_json_1.default.length)].split('');
        }
        // 1 January 2022 Game Epoch
        const daysSinceLaunch = (Number(new Date()) - Number(new Date(2022, 0))) / 1000 / 60 / 60 / 24;
        return wordlist_json_1.default[Math.floor(daysSinceLaunch) % wordlist_json_1.default.length].split('');
    },
});
